---
title: "Empty Droplets"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/QualityControl/EmptyDrops.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in data

```{r data, message=FALSE}
# Load libraries
library(Matrix)
library(scran)
library(plot3D)
library(Rtsne)
library(ggplot2)
library(DropletUtils)
library(scater)
library(reshape2)
library(viridis)
library(RColorBrewer)
source("../../Functions/auxiliary.R")

# Read in data - Single cell experiment object per batch
sce.do17821 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17821/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17821)$Sample <- rep("P10", ncol(sce.do17821))
colData(sce.do17821)$Library <- rep("do17821", ncol(sce.do17821))

# P15
sce.do18195 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do18195/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do18195)$Sample <- rep("P15", ncol(sce.do18195))
colData(sce.do18195)$Library <- rep("do18195", ncol(sce.do18195))

# P20
sce.do17824 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17824/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17824)$Sample <- rep("P20", ncol(sce.do17824))
colData(sce.do17824)$Library <- rep("do17824", ncol(sce.do17824))

# P25
sce.do18196 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do18196/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do18196)$Sample <- rep("P25", ncol(sce.do18196))
colData(sce.do18196)$Library <- rep("do18196", ncol(sce.do18196))

# P30
sce.do17825 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17825/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17825)$Sample <- rep("P30", ncol(sce.do17825))
colData(sce.do17825)$Library <- rep("do17825", ncol(sce.do17825))

# P35
sce.do17827 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17827/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17827)$Sample <- rep("P35", ncol(sce.do17827))
colData(sce.do17827)$Library <- rep("do17827", ncol(sce.do17827))

# B6
sce.do17815 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17815/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17815)$Sample <- rep("B6", ncol(sce.do17815))
colData(sce.do17815)$Library <- rep("do17815", ncol(sce.do17815))

sce.do17816 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17816/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17816)$Sample <- rep("B6", ncol(sce.do17816))
colData(sce.do17816)$Library <- rep("do17816", ncol(sce.do17816))

# Tc0
sce.do15984 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do15984/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do15984)$Sample <- rep("Tc0", ncol(sce.do15984))
colData(sce.do15984)$Library <- rep("do15984", ncol(sce.do15984))

sce.do17622 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17622/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17622)$Sample <- rep("Tc0", ncol(sce.do17622))
colData(sce.do17622)$Library <- rep("do17622", ncol(sce.do17622))

sce.do18199 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do18199/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do18199)$Sample <- rep("Tc0", ncol(sce.do18199))
colData(sce.do18199)$Library <- rep("do18199", ncol(sce.do18199))

# Tc1
sce.do15983 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do15983/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do15983)$Sample <- rep("Tc1", ncol(sce.do15983))
colData(sce.do15983)$Library <- rep("do15983", ncol(sce.do15983))

sce.do17623 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17623/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17623)$Sample <- rep("Tc1", ncol(sce.do17623))
colData(sce.do17623)$Library <- rep("do17623", ncol(sce.do17623))

sce.do18197 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do18197/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do18197)$Sample <- rep("Tc1", ncol(sce.do18197))
colData(sce.do18197)$Library <- rep("do18197", ncol(sce.do18197))

sce.do18198 <- read10xCounts("../../../../Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do18198/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do18198)$Sample <- rep("Tc1", ncol(sce.do18198))
colData(sce.do18198)$Library <- rep("do18198", ncol(sce.do18198))
```

# Identify empty droplets

```{r emptyDrops}
# P10
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17821)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17821.empty <- sce.do17821[,!is.cell]
sce.do17821 <- sce.do17821[,is.cell]
#saveRDS(sce.do17821.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17821_empty.rds")
rm(sce.do17821.empty)

# P15
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do18195)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do18195.empty <- sce.do18195[,!is.cell]
sce.do18195 <- sce.do18195[,is.cell]
#saveRDS(sce.do17828.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17828_empty.rds")
rm(sce.do17828.empty)

# P20
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17824)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17824.empty <- sce.do17824[,!is.cell]
sce.do17824 <- sce.do17824[,is.cell]
#saveRDS(sce.do17824.empty, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17824_empty.rds")
rm(sce.do17824.empty)

# P25
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do18196)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do18196.empty <- sce.do18196[,!is.cell]
sce.do18196 <- sce.do18196[,is.cell]
#saveRDS(sce.do18196.empty, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do18196_empty.rds")
rm(sce.do18196.empty)

# P30
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17825)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17825.empty <- sce.do17825[,!is.cell]
sce.do17825 <- sce.do17825[,is.cell]
#saveRDS(sce.do17825.empty, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17825_empty.rds")
rm(sce.do17825.empty)

# P35
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17827)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17827.empty <- sce.do17827[,!is.cell]
sce.do17827 <- sce.do17827[,is.cell]
#saveRDS(sce.do17827.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17827_empty.rds")
rm(sce.do17827.empty)

# B6
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17815)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17815.empty <- sce.do17815[,!is.cell]
sce.do17815 <- sce.do17815[,is.cell]
#saveRDS(sce.do17815.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17815_empty.rds")
rm(sce.do17815.empty)

set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17816)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17816.empty <- sce.do17816[,!is.cell]
sce.do17816 <- sce.do17816[,is.cell]
#saveRDS(sce.do17816.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17816_empty.rds")
rm(sce.do17816.empty)

# TC0
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do15984)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do15984.empty <- sce.do15984[,!is.cell]
sce.do15984 <- sce.do15984[,is.cell]
#saveRDS(sce.do15984.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do15984_empty.rds")
rm(sce.do15984.empty)

set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17622)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17622.empty <- sce.do17622[,!is.cell]
sce.do17622 <- sce.do17622[,is.cell]
#saveRDS(sce.do17622.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17622_empty.rds")
rm(sce.do17622.empty)

set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do18199)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do18199.empty <- sce.do18199[,!is.cell]
sce.do18199 <- sce.do18199[,is.cell]
#saveRDS(sce.do18199.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17622_empty.rds")
rm(sce.do18199.empty)

# Tc1
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do15983)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do15983.empty <- sce.do15983[,!is.cell]
sce.do15983 <- sce.do15983[,is.cell]
#saveRDS(sce.do15983.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do15983_empty.rds")
rm(sce.do15983.empty)

set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17623)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17623.empty <- sce.do17623[,!is.cell]
sce.do17623 <- sce.do17623[,is.cell]
#saveRDS(sce.do17623.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17623_empty.rds")
rm(sce.do17623.empty)

set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do18197)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do18197.empty <- sce.do18197[,!is.cell]
sce.do18197 <- sce.do18197[,is.cell]
#saveRDS(sce.do15983.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do15983_empty.rds")
rm(sce.do18197.empty)

set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do18198)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do18198.empty <- sce.do18198[,!is.cell]
sce.do18198 <- sce.do18198[,is.cell]
#saveRDS(sce.do17623.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17623_empty.rds")
rm(sce.do18198.empty)

save.image("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops_temp.RData")
load("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops_temp.RData")
```

# Filtering

```{r filtering}
# Merge datasets
sce.all.1 <- cbind(sce.do17815, sce.do17816, sce.do15984, sce.do17622, sce.do18199,
                 sce.do15983, sce.do17623)
rm("sce.do17815")
rm("sce.do17816")
rm("sce.do15984")
rm("sce.do17622")
rm("sce.do18199")
rm("sce.do15983")
rm("sce.do17623")
sce.all.2 <- cbind(sce.do18197, sce.do18198, sce.do17821,
                 sce.do18195, sce.do17824, sce.do18196, sce.do17825, sce.do17827)
rm("sce.do18197")
rm("sce.do18198")
rm("sce.do17821")
rm("sce.do18195")
rm("sce.do17824")
rm("sce.do18196")
rm("sce.do17825")
rm("sce.do17827")
sce.all <- cbind(sce.all.1, sce.all.2)
rm("sce.all.1")
rm("sce.all.2")

# Save statistics on these libraries
cur_stats <- melt(table(colData(sce.all)$Sample, colData(sce.all)$Library))
cur_stats <- cur_stats[cur_stats$value > 0,]
cur_stats <- cur_stats[order(cur_stats$Var1),]
stats.df <- data.frame(row.names = cur_stats$Var2,
                       Sample = cur_stats$Var1,
                       Library = cur_stats$Var2,
                       CellRanger_filter = cur_stats$value)

# Save dataframe
write.table(stats.df, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Tables/CellRangerFiltering_emptyDrops_stats.txt", 
            sep = "\t", row.names = TRUE, col.names = TRUE, quote = FALSE)

sce.all <- calculateQCMetrics(sce.all)

plot(colData(sce.all)$total_counts, colData(sce.all)$total_features, log = "x", pch = 16, 
     xlab = "Total counts", ylab = "Number of genes", 
     col = ifelse(colData(sce.all)$total_features > 500, "black", "red"))
sce.all <- sce.all[,colData(sce.all)$total_features > 500]

# Remove cells with high mitochondrial reads
mt <- read.table("../../Data/Mouse_genes.txt", sep = "\t", header = TRUE,
                 stringsAsFactors = FALSE)
mt <- mt[mt$Chromosome.scaffold.name == "MT",]
plot(Matrix::colSums(counts(sce.all)[mt$Gene.stable.ID,])/colData(sce.all)$total_counts, 
     pch = 16, ylab = "% mitochondrial reads")
sce.all <- sce.all[,Matrix::colSums(counts(sce.all)[mt$Gene.stable.ID,])/colData(sce.all)$total_counts < 0.1]

# Remove genes that are not expressed
sce.all <- sce.all[Matrix::rowSums(counts(sce.all)) > 0,]

saveRDS(sce.all, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
sce.all <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
```

# Normalization

```{r normalization}
clusters <- quickCluster(sce.all, method = "igraph", irlba.args = c("work" = 100), 
                         max.size = 2000)

sce.all <- computeSumFactors(sce.all, clusters=clusters)
sce.all <- sce.all[,sizeFactors(sce.all) > 0]

sce.all <- normalize(sce.all, return_log = TRUE)


saveRDS(sce.all, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")

sce.all <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
```

# Batch correction

```{r}
# Split datasets into batches
sce.single <- split.sce(sce = sce.all, groups = unique(colData(sce.all)$Library), 
                        colData.name = "Library")

# Batch correction
corrected <- batch.correction(sce.single)

# Save batch corrected matrix in sce object
metadata(sce.all)$corrected <- corrected

saveRDS(sce.all, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
```

# Mark the cells that were labeled before

```{r}
sce.old <- readRDS("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all_clusters.rds")

annotatedclust <- rep("NewCell", ncol(sce.all))
m <- match(colData(sce.old)$Barcode, colData(sce.all)$Barcode)
annotatedclust[m[!is.na(m)]] <- as.character(colData(sce.old)$AnnotatedClusters)[!is.na(m)]

colData(sce.all)$AnnotatedClusters <- factor(annotatedclust,
      levels = c(levels(colData(sce.old)$AnnotatedClusters), "NewCell"))

col_vector <- metadata(sce.old)$color_vector
col_vector <- c(col_vector, "NewCell" = "black")

metadata(sce.all)$color_vector <- col_vector

rm(sce.old)
saveRDS(sce.all, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
```

# Compute tSNE

```{r}
set.seed(111)
tsne <- Rtsne(t(metadata(sce.all)$corrected), perplexity = 60)
reducedDims(sce.all)$TSNE <- tsne$Y

saveRDS(sce.all, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
```

# Visualize results

```{r}
df1 <- data.frame(tsne1 =reducedDims(sce.all)$TSNE[
                      colData(sce.all)$AnnotatedClusters != "NewCell",1],
                  tsne2 = reducedDims(sce.all)$TSNE[
                    colData(sce.all)$AnnotatedClusters != "NewCell",2],
                  batch = colData(sce.all)$AnnotatedClusters[
                    colData(sce.all)$AnnotatedClusters != "NewCell"
                  ],
                  Dazl = logcounts(sce.all)[rowData(sce.all)$Symbol == "Dazl",
                  colData(sce.all)$AnnotatedClusters != "NewCell"],
                  Prm1 = logcounts(sce.all)[rowData(sce.all)$Symbol == "Prm1",
                  colData(sce.all)$AnnotatedClusters != "NewCell"])
df2 <- data.frame(tsne1 =reducedDims(sce.all)$TSNE[
                      colData(sce.all)$AnnotatedClusters == "NewCell",1],
                  tsne2 = reducedDims(sce.all)$TSNE[
                    colData(sce.all)$AnnotatedClusters == "NewCell",2],
                  batch = colData(sce.all)$AnnotatedClusters[
                    colData(sce.all)$AnnotatedClusters == "NewCell"
                  ],
                  Dazl = logcounts(sce.all)[rowData(sce.all)$Symbol == "Dazl",
                  colData(sce.all)$AnnotatedClusters == "NewCell"],
                  Prm1 = logcounts(sce.all)[rowData(sce.all)$Symbol == "Prm1",
                  colData(sce.all)$AnnotatedClusters == "NewCell"])
  
emptyDrops <- ggplot() +
  geom_point(data = df2, aes(tsne1, tsne2, colour = batch)) +
  geom_point(data = df1, aes(tsne1, tsne2, colour = batch)) +
  scale_color_manual(values = metadata(sce.all)$color_vector)
ggsave("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S3/emptyDrops.pdf", emptyDrops, 
       width = 12, height = 8)

emptyDrops.Dazl <- ggplot() +
  geom_point(data = df2, aes(tsne1, tsne2, colour = Dazl)) +
  geom_point(data = df1, aes(tsne1, tsne2, colour = Dazl)) +
  scale_color_viridis()
ggsave("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S3/emptyDrops_Dazl.pdf", emptyDrops.Dazl,
       width = 12, height = 8)

emptyDrops.Prm1 <- ggplot() +
  geom_point(data = df2, aes(tsne1, tsne2, colour = Prm1)) +
  geom_point(data = df1, aes(tsne1, tsne2, colour = Prm1)) +
  scale_color_viridis()
ggsave("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S3/emptyDrops_Prm1.pdf", emptyDrops.Prm1,
       width = 12, height = 8)

# Only P10 sample
emptyDrops.P10 <- ggplot(data.frame(tsne1 =reducedDims(sce.all)$TSNE[
                      grepl("P10", colData(sce.all)$Sample),1],
                  tsne2 = reducedDims(sce.all)$TSNE[
                    grepl("P10", colData(sce.all)$Sample),2],
                  batch = colData(sce.all)$AnnotatedClusters[
                    grepl("P10", colData(sce.all)$Sample)
                  ])) + geom_point(aes(tsne1, tsne2, colour = batch)) +
    scale_color_manual(values = metadata(sce.all)$color_vector)
ggsave("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S3/emptyDrops_P10.pdf", emptyDrops.P10,
       width = 12, height = 8)
```

