---
title: "Empty Droplets"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/QualityControl/EmptyDrops.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in data

```{r data, message=FALSE}
# Load libraries
library(Matrix)
library(scran)
library(plot3D)
library(Rtsne)
library(ggplot2)
library(DropletUtils)
library(scater)
library(reshape2)
library(viridis)
library(RColorBrewer)
library(irlba)
library(openxlsx)
source("../../Functions/auxiliary.R")

# Read in data - Single cell experiment object per batch
# P5
sce.do26386 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do26386/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do26386)$Sample <- rep("P5", ncol(sce.do26386))
colData(sce.do26386)$Library <- rep("do26386", ncol(sce.do26386))

sce.do26387 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do26387/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do26387)$Sample <- rep("P5", ncol(sce.do26387))
colData(sce.do26387)$Library <- rep("do26387", ncol(sce.do26387))

# P10
sce.do17821 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17821/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17821)$Sample <- rep("P10", ncol(sce.do17821))
colData(sce.do17821)$Library <- rep("do17821", ncol(sce.do17821))

# P15
sce.do18195 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do18195/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do18195)$Sample <- rep("P15", ncol(sce.do18195))
colData(sce.do18195)$Library <- rep("do18195", ncol(sce.do18195))

# P20
sce.do17824 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17824/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17824)$Sample <- rep("P20", ncol(sce.do17824))
colData(sce.do17824)$Library <- rep("do17824", ncol(sce.do17824))

# P25
sce.do18196 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do18196/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do18196)$Sample <- rep("P25", ncol(sce.do18196))
colData(sce.do18196)$Library <- rep("do18196", ncol(sce.do18196))

# P30
sce.do17825 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17825/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17825)$Sample <- rep("P30", ncol(sce.do17825))
colData(sce.do17825)$Library <- rep("do17825", ncol(sce.do17825))

# P35
sce.do17827 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17827/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17827)$Sample <- rep("P35", ncol(sce.do17827))
colData(sce.do17827)$Library <- rep("do17827", ncol(sce.do17827))

# B6
sce.do17815 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17815/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17815)$Sample <- rep("B6", ncol(sce.do17815))
colData(sce.do17815)$Library <- rep("do17815", ncol(sce.do17815))

sce.do17816 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17816/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17816)$Sample <- rep("B6", ncol(sce.do17816))
colData(sce.do17816)$Library <- rep("do17816", ncol(sce.do17816))
```

# Identify empty droplets

```{r emptyDrops}
# P5
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do26386)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do26386.empty <- sce.do26386[,!is.cell]
sce.do26386 <- sce.do26386[,is.cell]
rm(sce.do26386.empty)

set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do26387)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do26387.empty <- sce.do26387[,!is.cell]
sce.do26387 <- sce.do26387[,is.cell]
rm(sce.do26387.empty)

# P10
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17821)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17821.empty <- sce.do17821[,!is.cell]
sce.do17821 <- sce.do17821[,is.cell]
#saveRDS(sce.do17821.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17821_empty.rds")
rm(sce.do17821.empty)

# P15
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do18195)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do18195.empty <- sce.do18195[,!is.cell]
sce.do18195 <- sce.do18195[,is.cell]
#saveRDS(sce.do17828.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17828_empty.rds")
rm(sce.do17828.empty)

# P20
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17824)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17824.empty <- sce.do17824[,!is.cell]
sce.do17824 <- sce.do17824[,is.cell]
#saveRDS(sce.do17824.empty, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17824_empty.rds")
rm(sce.do17824.empty)

# P25
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do18196)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do18196.empty <- sce.do18196[,!is.cell]
sce.do18196 <- sce.do18196[,is.cell]
#saveRDS(sce.do18196.empty, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do18196_empty.rds")
rm(sce.do18196.empty)

# P30
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17825)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17825.empty <- sce.do17825[,!is.cell]
sce.do17825 <- sce.do17825[,is.cell]
#saveRDS(sce.do17825.empty, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17825_empty.rds")
rm(sce.do17825.empty)

# P35
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17827)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17827.empty <- sce.do17827[,!is.cell]
sce.do17827 <- sce.do17827[,is.cell]
#saveRDS(sce.do17827.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17827_empty.rds")
rm(sce.do17827.empty)

# B6
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17815)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17815.empty <- sce.do17815[,!is.cell]
sce.do17815 <- sce.do17815[,is.cell]
#saveRDS(sce.do17815.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17815_empty.rds")
rm(sce.do17815.empty)

set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17816)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17816.empty <- sce.do17816[,!is.cell]
sce.do17816 <- sce.do17816[,is.cell]
#saveRDS(sce.do17816.empty, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/EmptyDrops/sce_do17816_empty.rds")
rm(sce.do17816.empty)

save.image("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops_temp.RData")
load("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops_temp.RData")
```

# Filtering

```{r filtering}
# Merge datasets
sce.all.1 <- cbind(sce.do17815, sce.do17816, sce.do26386, sce.do26387)
sce.all.2 <- cbind(sce.do17821, sce.do18195)
sce.all.3 <- cbind(sce.do17824, sce.do18196, sce.do17825, sce.do17827)
rm("sce.do17815")
rm("sce.do17816")
rm("sce.do26386")
rm("sce.do26387")
rm("sce.do17821")
rm("sce.do18195")
rm("sce.do17824")
rm("sce.do18196")
rm("sce.do17825")
rm("sce.do17827")
sce.all <- cbind(sce.all.1, sce.all.2, sce.all.3)
rm("sce.all.1")
rm("sce.all.2")
rm("sce.all.3")

# Save statistics on these libraries
cur_stats <- melt(table(colData(sce.all)$Sample, colData(sce.all)$Library))
cur_stats <- cur_stats[cur_stats$value > 0,]
cur_stats <- cur_stats[order(cur_stats$Var1),]
stats.df <- data.frame(row.names = cur_stats$Var2,
                       Sample = cur_stats$Var1,
                       Library = cur_stats$Var2,
                       CellRanger_filter = cur_stats$value)

# Save dataframe
write.table(stats.df, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Tables/CellRangerFiltering_emptyDrops_stats.txt", 
            sep = "\t", row.names = TRUE, col.names = TRUE, quote = FALSE)

sce.all <- calculateQCMetrics(sce.all)

plot(colData(sce.all)$total_counts, colData(sce.all)$total_features_by_counts, log = "x", pch = 16, 
     xlab = "Total counts", ylab = "Number of genes", 
     col = ifelse(colData(sce.all)$total_features_by_counts > 500, "black", "red"))
sce.all <- sce.all[,colData(sce.all)$total_features_by_counts > 500]

# Remove cells with high mitochondrial reads
mt <- read.table("../../Data/Mouse_genes.txt", sep = "\t", header = TRUE,
                 stringsAsFactors = FALSE)
mt <- mt[mt$Chromosome.scaffold.name == "MT",]
plot(Matrix::colSums(counts(sce.all)[mt$Gene.stable.ID,])/colData(sce.all)$total_counts, 
     pch = 16, ylab = "% mitochondrial reads")
sce.all <- sce.all[,Matrix::colSums(counts(sce.all)[mt$Gene.stable.ID,])/colData(sce.all)$total_counts < 0.1]

# Remove genes that are not expressed
sce.all <- sce.all[Matrix::rowSums(counts(sce.all)) > 0,]

# Save stats on filtered cells
cur_stats <- melt(table(colData(sce.all)$Sample, colData(sce.all)$Library))
cur_stats <- cur_stats[cur_stats$value > 0,]
cur_stats <- cur_stats[order(cur_stats$Var1),]
stats.df <- data.frame(row.names = cur_stats$Var2,
                       Sample = cur_stats$Var1,
                       Library = cur_stats$Var2,
                       CellRanger_filter = cur_stats$value)

# Save dataframe
write.table(stats.df, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Tables/CellRangerFiltering_emptyDrops_stats_afterQC.txt", 
            sep = "\t", row.names = TRUE, col.names = TRUE, quote = FALSE)


saveRDS(sce.all, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
#sce.all <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
```

# Normalization

```{r normalization}
clusters <- quickCluster(sce.all, method = "igraph", irlba.args = c("work" = 100), 
                         max.size = 2000)

sce.all <- computeSumFactors(sce.all, clusters=clusters)
sce.all <- sce.all[,sizeFactors(sce.all) > 0]

sce.all <- normalize(sce.all, return_log = TRUE)


saveRDS(sce.all, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")

#sce.all <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
```

# Batch correction

```{r}
# Split datasets into batches
sce.single <- split.sce(sce = sce.all, groups = unique(colData(sce.all)$Library), 
                        colData.name = "Library")

# Batch correction
corrected <- batch.correction(sce.single)

# Save batch corrected matrix in sce object
metadata(sce.all)$corrected <- corrected

saveRDS(sce.all, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
```

# Mark the cells that were labeled before

```{r}
sce.old <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")

annotatedclust <- rep("NewCell", ncol(sce.all))
m <- match(paste(colData(sce.old)$Barcode, colData(sce.old)$Library), 
             paste(colData(sce.all)$Barcode, colData(sce.all)$Library))
annotatedclust[m[!is.na(m)]] <- as.character(colData(sce.old)$AnnotatedClusters)[!is.na(m)]

colData(sce.all)$AnnotatedClusters <- factor(annotatedclust,
      levels = c(levels(colData(sce.old)$AnnotatedClusters), "NewCell"))

col_vector <- metadata(sce.old)$color_vector
col_vector <- c(col_vector, "NewCell" = "black")

metadata(sce.all)$color_vector <- col_vector

rm(sce.old)
saveRDS(sce.all, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
```

# Compute tSNE

```{r}
sce.all <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
set.seed(12345)
pca <- prcomp_irlba(t(metadata(sce.all)$corrected), n = 50)
set.seed(12345)
tsne <- Rtsne(pca$x, pca = FALSE, perplexity = 600)

reducedDims(sce.all)$TSNE <- tsne$Y

saveRDS(sce.all, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
```

# Visualize results

```{r}
df1 <- data.frame(tsne1 =reducedDims(sce.all)$TSNE[
                      colData(sce.all)$AnnotatedClusters != "NewCell",1],
                  tsne2 = reducedDims(sce.all)$TSNE[
                    colData(sce.all)$AnnotatedClusters != "NewCell",2],
                  cluster = colData(sce.all)$AnnotatedClusters[
                    colData(sce.all)$AnnotatedClusters != "NewCell"
                  ])
df2 <- data.frame(tsne1 =reducedDims(sce.all)$TSNE[
                      colData(sce.all)$AnnotatedClusters == "NewCell",1],
                  tsne2 = reducedDims(sce.all)$TSNE[
                    colData(sce.all)$AnnotatedClusters == "NewCell",2],
                  cluster = colData(sce.all)$AnnotatedClusters[
                    colData(sce.all)$AnnotatedClusters == "NewCell"
                  ])
  
ggplot() +
  geom_point(data = df2, aes(tsne1, tsne2, colour = cluster)) +
  geom_point(data = df1, aes(tsne1, tsne2, colour = cluster)) +
  scale_color_manual(values = metadata(sce.all)$color_vector)
```

# Clustering of P15 cells

```{r}
sce.all <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
sce.P15 <- sce.all[,colData(sce.all)$Sample == "P15"]
sce.P15 <- normalize(sce.P15)

# HVG
HVgenes <- HVG(sce.P15)

g <- buildSNNGraph(logcounts(sce.P15)[HVgenes,], k = 10, pc.approx = TRUE)

clusters <- igraph::cluster_louvain(g)$membership

ggplot(data.frame(tsne1 = reducedDims(sce.P15)$TSNE[,1],
                  tsne2 = reducedDims(sce.P15)$TSNE[,2],
                  cluster = as.factor(clusters))) +
  geom_point(aes(tsne1, tsne2, colour = cluster)) + 
  scale_colour_manual(values = c(brewer.pal(n = 8, "Set1"), 
                                 brewer.pal(n = 8, "Set3")))

ggplot(data.frame(tsne1 = reducedDims(sce.P15)$TSNE[,1],
                  tsne2 = reducedDims(sce.P15)$TSNE[,2],
                  gene = logcounts(sce.P15)[rowData(sce.P15)$Symbol == "Cd14",])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + 
  scale_colour_viridis()

df1 <- data.frame(tsne1 =reducedDims(sce.P15)$TSNE[
                      colData(sce.P15)$AnnotatedClusters != "NewCell",1],
                  tsne2 = reducedDims(sce.P15)$TSNE[
                    colData(sce.P15)$AnnotatedClusters != "NewCell",2],
                  cluster = colData(sce.P15)$AnnotatedClusters[
                    colData(sce.P15)$AnnotatedClusters != "NewCell"
                  ])
df2 <- data.frame(tsne1 =reducedDims(sce.P15)$TSNE[
                      colData(sce.P15)$AnnotatedClusters == "NewCell",1],
                  tsne2 = reducedDims(sce.P15)$TSNE[
                    colData(sce.P15)$AnnotatedClusters == "NewCell",2],
                  cluster = colData(sce.P15)$AnnotatedClusters[
                    colData(sce.P15)$AnnotatedClusters == "NewCell"
                  ])

ggplot() +
  geom_point(data = df2, aes(tsne1, tsne2, colour = cluster)) +
  geom_point(data = df1, aes(tsne1, tsne2, colour = cluster)) +
  scale_color_manual(values = metadata(sce.all)$color_vector)


annotatedCluster <- clusters
annotatedCluster[annotatedCluster == 1] <- "Leydig_1"
annotatedCluster[annotatedCluster == 2] <- "Zygotene"
annotatedCluster[annotatedCluster == 3] <- "Differentiating_Spermatogonia_2"
annotatedCluster[annotatedCluster == 4] <- "Endothelial_cells"
annotatedCluster[annotatedCluster == 5] <- "PTM"
annotatedCluster[annotatedCluster == 6] <- "Pachytene_2"
annotatedCluster[annotatedCluster == 7] <- "Leptotene"
annotatedCluster[annotatedCluster == 8] <- "Immature_Leydig"
annotatedCluster[annotatedCluster == 9] <- "Differentiating_Spermatogonia_1"
annotatedCluster[annotatedCluster == 10] <- "Sertoli_1"
annotatedCluster[annotatedCluster == 11] <- "Sertoli_2"
annotatedCluster[annotatedCluster == 12] <- "Interstitial_tMg"
annotatedCluster[annotatedCluster == 13] <- "Outliers"
annotatedCluster[annotatedCluster == 14] <- "Pachytene_1"
annotatedCluster[annotatedCluster == 15] <- "Leydig_2"
annotatedCluster[annotatedCluster == 16] <- "Undifferentiated_Spermatogonia"


colData(sce.P15)$AnnotatedClusters <- factor(annotatedCluster,
      levels = c("Undifferentiated_Spermatogonia", "Differentiating_Spermatogonia_1", "Differentiating_Spermatogonia_2", 
                 "Leptotene", "Zygotene", "Pachytene_1", "Pachytene_2",                 "Immature_Leydig","Leydig_1", "Leydig_2", "Sertoli_1","Sertoli_2", 
                 "PTM", "Interstitial_tMg", "Endothelial_cells", "Outliers"))

colour_vector.P15 <- c("Undifferentiated_Spermatogonia" = "#00441b", 
                   "Differentiating_Spermatogonia_1" = "#238b45", 
                   "Differentiating_Spermatogonia_2" = "#66c2a4",
                   "Leptotene" = "#FBDCEA",
                   "Zygotene" = "#ceb3bf",
                   "Pachytene_1" = "#F2B2D1",
                   "Pachytene_2" = "#bc89a2",
                   "Sertoli_1" = "#643F18",
                   "Sertoli_2" = "sienna4",
                   "Leydig_1" = "#985D25",
                   "Leydig_2" = "#ad6827",
                   "Immature_Leydig" = "#A58D81",
                   "PTM" = "#C48C58", 
                   "Endothelial_cells" = "#FFDFB2",
                   "Interstitial_tMg" = "#F36E20",
                   "Outliers" = "#067277")

ggplot(data.frame(tsne1 = reducedDims(sce.P15)$TSNE[,1],
                  tsne2 = reducedDims(sce.P15)$TSNE[,2],
                  cluster = colData(sce.P15)$AnnotatedClusters)) +
  geom_point(aes(tsne1, tsne2, colour = cluster)) + 
  scale_colour_manual(values = colour_vector.P15)

P15.clusters <- rep(NA, ncol(sce.all))
P15.clusters[colData(sce.all)$Sample == "P15"] <- annotatedCluster

colData(sce.all)$P15Clusters <- factor(P15.clusters,
      levels = c("Undifferentiated_Spermatogonia", "Differentiating_Spermatogonia_1", "Differentiating_Spermatogonia_2", 
                 "Leptotene", "Zygotene", "Pachytene_1", "Pachytene_2",                 "Immature_Leydig","Leydig_1", "Leydig_2", "Sertoli_1","Sertoli_2", 
                 "PTM", "Interstitial_tMg", "Endothelial_cells", "Outliers", NA))
metadata(sce.all)$colour_vector.P15 <- colour_vector.P15

saveRDS(sce.all, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
```

# Write expression counts and metadata for ArrayExpress

```{r}
writeMM(counts(sce.all), "../../../../Desktop/raw_counts_emptyDropts.mtx")
write.table(colData(sce.all), "../../../../Desktop/cell_metadata_emptyDrops.txt")
write.table(rowData(sce.all)[,1:2], 
            "../../../../Desktop/genes_emptyDrops.txt", 
            sep = "\t", row.names = FALSE)
```

