---
title: "Filtering"
author: "eling01 Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/QualityControl/Parallel_filtering.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in data

```{r data, message=FALSE}
# Load libraries
library(Matrix)
library(scran)
library(plot3D)
library(Rtsne)
library(ggplot2)
library(DropletUtils)
library(scater)
library(viridis)
library(RColorBrewer)
library(irlba)
library(dynamicTreeCut)
library(RColorBrewer)
library(reshape2)
library(openxlsx)


# Read in data - Single cell experiment object per batch
# P5
sce.do26386 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do26386/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do26386)$Sample <- rep("P5", ncol(sce.do26386))
colData(sce.do26386)$Library <- rep("do26386", ncol(sce.do26386))

sce.do26387 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do26387/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do26387)$Sample <- rep("P5", ncol(sce.do26387))
colData(sce.do26387)$Library <- rep("do26387", ncol(sce.do26387))

# P10
sce.do17821 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17821/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17821)$Sample <- rep("P10", ncol(sce.do17821))
colData(sce.do17821)$Library <- rep("do17821", ncol(sce.do17821))

# P15
sce.do18195 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do18195/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do18195)$Sample <- rep("P15", ncol(sce.do18195))
colData(sce.do18195)$Library <- rep("do18195", ncol(sce.do18195))

# P20
sce.do17824 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17824/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17824)$Sample <- rep("P20", ncol(sce.do17824))
colData(sce.do17824)$Library <- rep("do17824", ncol(sce.do17824))

# P25
sce.do18196 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do18196/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do18196)$Sample <- rep("P25", ncol(sce.do18196))
colData(sce.do18196)$Library <- rep("do18196", ncol(sce.do18196))

# P30
sce.do17825 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17825/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17825)$Sample <- rep("P30", ncol(sce.do17825))
colData(sce.do17825)$Library <- rep("do17825", ncol(sce.do17825))

# P35
sce.do17827 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17827/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17827)$Sample <- rep("P35", ncol(sce.do17827))
colData(sce.do17827)$Library <- rep("do17827", ncol(sce.do17827))

# B6
sce.do17815 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17815/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17815)$Sample <- rep("B6", ncol(sce.do17815))
colData(sce.do17815)$Library <- rep("do17815", ncol(sce.do17815))

sce.do17816 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17816/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17816)$Sample <- rep("B6", ncol(sce.do17816))
colData(sce.do17816)$Library <- rep("do17816", ncol(sce.do17816))

# Tc0
sce.do15983 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do15983/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do15983)$Sample <- rep("Tc0", ncol(sce.do15983))
colData(sce.do15983)$Library <- rep("do15983", ncol(sce.do15983))

sce.do17622 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17622/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17622)$Sample <- rep("Tc0", ncol(sce.do17622))
colData(sce.do17622)$Library <- rep("do17622", ncol(sce.do17622))

sce.do18199 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do18199/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do18199)$Sample <- rep("Tc0", ncol(sce.do18199))
colData(sce.do18199)$Library <- rep("do18199", ncol(sce.do18199))

# Tc1
sce.do15984 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do15984/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do15984)$Sample <- rep("Tc1", ncol(sce.do15984))
colData(sce.do15984)$Library <- rep("do15984", ncol(sce.do15984))

sce.do17623 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17623/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17623)$Sample <- rep("Tc1", ncol(sce.do17623))
colData(sce.do17623)$Library <- rep("do17623", ncol(sce.do17623))

sce.do18197 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do18197/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do18197)$Sample <- rep("Tc1", ncol(sce.do18197))
colData(sce.do18197)$Library <- rep("do18197", ncol(sce.do18197))

sce.do18198 <- read10xCounts("../../../../Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do18198/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do18198)$Sample <- rep("Tc1", ncol(sce.do18198))
colData(sce.do18198)$Library <- rep("do18198", ncol(sce.do18198))
```

# Filtering 

```{r Other}
# Merge all datasets
sce.all <- cbind(sce.do17815, sce.do17816, sce.do15983, sce.do17622, sce.do18199,
                 sce.do15984, sce.do17623, sce.do18197, sce.do18198, sce.do26386,
                 sce.do26387, sce.do17821, sce.do18195, sce.do17824, sce.do18196,
                 sce.do17825, sce.do17827)
rm(list = setdiff(ls(), "sce.all"))

# Save statistics on these libraries
cur_stats <- melt(table(colData(sce.all)$Sample, colData(sce.all)$Library))
cur_stats <- cur_stats[cur_stats$value > 0,]
cur_stats <- cur_stats[order(cur_stats$Var1),]
stats.df <- data.frame(row.names = cur_stats$Var2,
                       Sample = cur_stats$Var1,
                       Library = cur_stats$Var2,
                       CellRanger_filter = cur_stats$value)

# Calculate QC metrics
sce.all <- calculateQCMetrics(sce.all)

# Remove cells with few features
plot(colData(sce.all)$total_counts, colData(sce.all)$total_features_by_counts, log = "x", pch = 16, 
     xlab = "Total counts", ylab = "Number of genes", 
     col = ifelse(colData(sce.all)$total_features_by_counts > 1000, "black", "red"))
sce.all <- sce.all[,colData(sce.all)$total_features_by_counts > 1000]

# Remove cells with high mitochondrial reads
mt <- read.table("../../Data/Mouse_genes.txt", sep = "\t", header = TRUE,
                 stringsAsFactors = FALSE)
mt <- mt[mt$Chromosome.scaffold.name == "MT",]
plot(Matrix::colSums(counts(sce.all)[mt$Gene.stable.ID,])/colData(sce.all)$total_counts, 
     pch = 16, ylab = "% mitochondrial reads")
sce.all <- sce.all[,Matrix::colSums(counts(sce.all)[mt$Gene.stable.ID,])/colData(sce.all)$total_counts < 0.1]

# Remove genes that are not expressed
sce.all <- sce.all[Matrix::rowSums(counts(sce.all)) > 0,]

# Add to stats data frame
cur_stats <- melt(table(colData(sce.all)$Sample, colData(sce.all)$Library))
cur_stats <- cur_stats[cur_stats$value > 0,]
cur_stats <- cur_stats[order(cur_stats$Var1),]
stats.df$AfterFiltering <- cur_stats$value
```

# Normalization

```{r normalization}
clusters <- quickCluster(sce.all, method = "igraph", irlba.args = c("work" = 100), 
                         max.size = 2000)

sce.all <- computeSumFactors(sce.all, clusters=clusters)

sce.all <- normalize(sce.all, return_log = TRUE)
```

# Quality visualization

```{r plotting}
source("../../Functions/auxiliary.R")
# Highly variable genes
HVgenes <- HVG(sce = sce.all[Matrix::rowMeans(logcounts(sce.all)) > 0.1,])

set.seed(12345)
pca <- prcomp_irlba(t(logcounts(sce.all[HVgenes,])), n = 50)
tsne <- Rtsne(pca$x, pca = FALSE)
reducedDims(sce.all)$TSNE <- tsne$Y

# Batches
col_vector <- c(brewer.pal(8, "Set1"), brewer.pal(8, "Set2"), brewer.pal(8, "Set3"))
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  batch = paste(colData(sce.all)$Sample, 
                                colData(sce.all)$Library))) +
  geom_point(aes(tsne1, tsne2, colour = batch)) +
  scale_color_manual(values = col_vector)

# Verification of chr21 expression: PDXK
# PDXK 
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)[rowData(sce.all)$ID[rowData(sce.all)$Symbol == "PDXK"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Library size
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = colData(sce.all)$log10_total_counts)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Number genes expressed
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = colData(sce.all)$total_features)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Mitochondrial genes
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  mito = Matrix::colSums(counts(sce.all)[rowData(sce.all)$ID %in%
                                                mt$Gene.stable.ID,])/
                    Matrix::colSums(counts(sce.all)))) + geom_point(aes(tsne1, tsne2, colour = mito)) + scale_colour_viridis()

# Stra8 
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)[rowData(sce.all)$ID[rowData(sce.all)$Symbol == "Stra8"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Defb19 
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)[rowData(sce.all)$ID[rowData(sce.all)$Symbol == "Defb19"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Prm1 
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)[rowData(sce.all)$ID[rowData(sce.all)$Symbol == "Prm1"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Dazl 
  ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)[rowData(sce.all)$ID[rowData(sce.all)$Symbol == "Dazl"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()


```

# Save at this stage

```{r}
saveRDS(sce.all, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")
sce.all <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")
```

# Batch correction

```{r correction}
source("../../Functions/auxiliary.R")
# Split datasets into batches
sce.single <- split.sce(sce = sce.all, groups = unique(colData(sce.all)$Library), 
                        colData.name = "Library")

# Batch correction
corrected <- batch.correction(sce.single)
rm(sce.single)

# Save batch corrected matrix in sce object
metadata(sce.all)$corrected <- corrected

saveRDS(sce.all, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")
```

# Visualize batch corrected data

```{r visualizataion}
set.seed(777)
pca <- prcomp_irlba(t(metadata(sce.all)$corrected), n = 50)
tsne <- Rtsne(pca$x, pca = FALSE, perplexity = 200)
reducedDims(sce.all)$TSNE <- tsne$Y

# Batches
col_vector <- c(brewer.pal(8, "Set1"), brewer.pal(8, "Set2"), brewer.pal(8, "Set3"))
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  batch = paste(colData(sce.all)$Sample,
                                colData(sce.all)$Library))) +
  geom_point(aes(tsne1, tsne2, colour = batch)) +
  scale_color_manual(values = col_vector)

# Verification of chr21 expression: PDXK
# PDXK 
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)[rowData(sce.all)$ID[rowData(sce.all)$Symbol == "PDXK"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

for(i in unique(sce.all$Library)){
  print(i)
  print(sum(counts(sce.all[grepl("ENSG", rowData(sce.all)$ID),sce.all$Library == i])))
}

# Library size
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = colData(sce.all)$log10_total_counts)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Number genes expressed
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = colData(sce.all)$total_features_by_counts)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Mitochondrial genes
mt <- read.table("../../Data/Mouse_genes.txt", sep = "\t", header = TRUE,
                 stringsAsFactors = FALSE)
mt <- mt[mt$Chromosome.scaffold.name == "MT",]
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  mito = Matrix::colSums(counts(sce.all)[rowData(sce.all)$ID %in%
                                                mt$Gene.stable.ID,])/
                    Matrix::colSums(counts(sce.all)))) + geom_point(aes(tsne1, tsne2, colour = mito)) + scale_colour_viridis()

cur_sce <- sce.all[,sce.all$Sample == "P10"]

# Plot gene expression
ggplot(data.frame(tsne1 = reducedDims(cur_sce)$TSNE[,1],
                  tsne2 = reducedDims(cur_sce)$TSNE[,2],
                  gene = logcounts(cur_sce)[rowData(sce.P15)$ID[rowData(cur_sce)$Symbol == "Tm4sf1"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()


# Cst12 - Sertoli, Col4a3 - PTM, Cd14 - tMg, Dlk1 - Immature Leydig, Tm4sf1 - Endothelial cells, Insl3 - Leydig
```

# Broad clustering

We first broadly cluster the cells into somatic and germ cells.

```{r clustering}
set.seed(1234)
g <- buildSNNGraph(metadata(sce.all)$corrected, k = 5, pc.approx=TRUE)
clusters <- igraph::cluster_louvain(g)$membership

clusters[clusters %in% as.numeric(names(table(clusters))[table(clusters) < 30])] <- "Outliers"

# Visualize clustering
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
#color_vector = c("#282B69", "#046735","#985D25","#C58C58","#FFDFB2","#A58D81","#643F18","#DCCE23","#F2B2D1","#9A1C59","#B3A9D3","#2A348B","#7E2678","#D8DAEC","#DCCE23","#E86FAA","#F36E21","#502269","#D31281","#6167AF","#22232B","#9595C9","#8470B1","#FBDCEA","#BCBCC2","#E5E5E3","#171447","#606B89","#056D6A","#9FA1A4","#444A5E","#DCCE23","#737884","#DCCE23","#DCCE23","#DCCE23","#DCCE23")
#names(color_vector) <- levels(as.factor(clusters))
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  cluster = as.factor(clusters))) +
  geom_point(aes(tsne1, tsne2, colour = cluster)) +
  scale_color_manual(values = col_vector)
```

# Annotation of broad clusters

```{r annotation}
# Annotate Germ cells and somatic cells first
new.clusters <- clusters
new.clusters[new.clusters == 1] <- "Germ"
new.clusters[new.clusters == 2] <- "Germ"
new.clusters[new.clusters == 3] <- "Germ"
new.clusters[new.clusters == 4] <- "Germ"
new.clusters[new.clusters == 5] <- "Germ"
new.clusters[new.clusters == 6] <- "Germ"
new.clusters[new.clusters == 7] <- "Germ"
new.clusters[new.clusters == 8] <- "Germ"
new.clusters[new.clusters == 9] <- "Outliers"
new.clusters[new.clusters == 10] <- "Somatic"
new.clusters[new.clusters == 11] <- "Somatic"
new.clusters[new.clusters == 12] <- "Somatic"
new.clusters[new.clusters == 13] <- "Somatic"
new.clusters[new.clusters == 14] <- "Somatic"
new.clusters[new.clusters == 15] <- "Germ"
new.clusters[new.clusters == 16] <- "Germ"
new.clusters[new.clusters == 17] <- "Germ"
new.clusters[new.clusters == 18] <- "Germ"
new.clusters[new.clusters == 19] <- "Germ"
new.clusters[new.clusters == 20] <- "Germ"
new.clusters[new.clusters == 21] <- "Outliers"
new.clusters[new.clusters == 22] <- "Somatic"
new.clusters[new.clusters == 23] <- "Germ"
new.clusters[new.clusters == 24] <- "Germ"
new.clusters[new.clusters == 25] <- "Germ"
new.clusters[new.clusters == 26] <- "Germ"
new.clusters[new.clusters == 27] <- "Germ"

ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  cluster = as.factor(new.clusters))) +
  geom_point(aes(tsne1, tsne2, colour = cluster)) +
  scale_color_manual(values = c("blue", "red", "black"))

colData(sce.all)$BroadClusters <- as.factor(new.clusters)
```

# Cluster somatic cells

```{r somatic}

```

```{r}
colData(sce.all)$AnnotatedClusters <- factor(new.clusters,
      levels = c("Spermatogonia_1", "Spermatogonia_2", "Spermatocytes_1",
                 "Spermatocytes_2","Spermatocytes_3","Spermatocytes_4",
                 "Spermatocytes_5","Spermatocytes_6","Spermatocytes_7", 
                 "S1", "S2", "S3", "S4", "S5", "S6", "S7", "S8", "S9",
                 "S10", "S11", "S12", "S13", "S14", "Sertoli", "Leydig", 
                 "Fetal_Leydig_1", "Fetal_Leydig_2", "Fetal_Leydig_3", "PTM", "Endothelial_cells", 
                 "Interstitial_tMg", "Outliers"))

colour_vector <- c("Spermatogonia_1" = "#046735", 
                   "Spermatogonia_2" = "#04331A",
                   "Spermatocytes_1" = "#FBDCEA", 
                   "Spermatocytes_2" = "#F2B2D1",
                   "Spermatocytes_3" = "#E86FAA", 
                   "Spermatocytes_4" = "#D31281",
                   "Spermatocytes_5" = "#9A1C59",
                   "Spermatocytes_6" = "#7E2678",
                   "Spermatocytes_7" = "#502269",
                   "S1" = "#B3A9D3",
                   "S2" = "#D8DAEC",
                   "S3" = "#9595C9",
                   "S4" = "#6167AF",
                   "S5" = "#2A348B",
                   "S6" = "#282B69",
                   "S7" = "#171447",
                   "S8" = "#22232B",
                   "S9" = "#444A5E",
                   "S10" = "#606B89",
                   "S11" = "#737884",
                   "S12" = "#9FA1A4",
                   "S13" = "#BCBCC2",
                   "S14" = "#E5E5E3",
                   "Sertoli" = "#643F18",
                   "Leydig" = "#985D25",
                   "Fetal_Leydig_1" = "#A58D81",
                   "Fetal_Leydig_2" = "#ceb0a1",
                   "Fetal_Leydig_3" = "#7a685f",
                   "PTM" = "#C48C58", 
                   "Endothelial_cells" = "#FFDFB2",
                   "Interstitial_tMg" = "#F36E20",
                   "Outliers" = "#067277")

ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  cluster = as.factor(colData(sce.all)$Sample))) +
  geom_point(aes(tsne1, tsne2, colour = colData(sce.all)$AnnotatedClusters)) +
  scale_color_manual(values = colour_vector)

metadata(sce.all)$color_vector <- colour_vector

# S4 contains contaminating cells

# Spermatocytes 3 contains tMg cells

# Add number of properly annotated cells to dataframe
cur_stats <- melt(table(colData(sce.all)$Sample[!(colData(sce.all)$AnnotatedClusters %in% 
                                  c(paste("NA", 1:5, sep = ""), "Outliers"))], 
                        colData(sce.all)$Library[!(colData(sce.all)$AnnotatedClusters %in% 
                                  c(paste("NA", 1:5, sep = ""), "Outliers"))]))
cur_stats <- cur_stats[cur_stats$value > 0,]
cur_stats <- cur_stats[order(cur_stats$Var1),]
stats.df <- read.table("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Tables/CellRangerFiltering_stats.txt", 
            sep = "\t")
stats.df$ProperlyAnnotated <- cur_stats$value

# Save dataframe
write.table(stats.df, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Tables/CellRangerFiltering_stats.txt", 
            sep = "\t", row.names = TRUE, col.names = TRUE, quote = FALSE)

```

# Save normalized and raw counts for ArrayExpress

```{r}
writeMM(counts(sce.all), "../../../../Desktop/raw_counts.mtx")
write.table(colData(sce.all), "../../../../Desktop/cell_metadata.txt")
```

```{r}
saveRDS(sce.all, "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")
sce.all <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")
```

# Generate overview table on filtering of libraries

```{r}
cellRanger <- read.table("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Tables/CellRangerFiltering_stats.txt")
emptyDrops <- read.table("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Tables/CellRangerFiltering_emptyDrops_stats.txt")
```