---
title: "Visualization of ChIP-Seq tracks"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/Chip-Seq/Promoter_visualization.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Read in data and libraries

```{r data}
library(Gviz)
library(rtracklayer)
library(EnsDb.Mmusculus.v79)
library(csaw)

names <- c("H3K4me1_ChIPSeq_SC.sorted.bam", "H3K4me1_ChIPSeq_ST.sorted.bam",
           "H3K4me3_ChIPSeq_SC.sorted.bam", "H3K4me3_ChIPSeq_ST.sorted.bam",
           "H3K27ac_ChIPSeq_SC.sorted.bam", "H3K27ac_ChIPSeq_ST.sorted.bam",
           "H27me3_ChIPSeq_SC.sorted.bam", "H27me3_ChIPSeq_ST.sorted.bam",
           "Input_ChIPSeq_SC.sorted.bam", "Input_ChIPSeq_ST.sorted.bam")

files <- paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Data/Hammoud_CSC/ChipSeq/bam/", names, sep = "")

# Blacklisted regions
black <- import("http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz")
seqlevels(black) <- sub("chr", "", seqlevels(black))
seqnames(black) <- sub("chr", "", seqnames(black))

# Parameter for reading in bam files
param <- readParam(dedup=TRUE, minq=10, discard=black, pe="none")

# Find large regions around promoters
cur_prom <- promoters(genes(EnsDb.Mmusculus.v79), 
                  upstream = 10000, downstream = 5000)

# Bin the mapped reads
bins <- windowCounts(files, bin = TRUE, width = 1000, param=param)

# Create objects for visualziation
gax <- GenomeAxisTrack(col="black", fontsize=15, size=2)
gr <- getGeneRegionTrackForGviz(EnsDb.Mmusculus.v79)
options(ucscChromosomeNames = FALSE)
```

## Visualization

```{r}
# Cypt1
cur_gene <- "ENSMUSG00000023257"
cur_name <- elementMetadata(cur_prom)$gene_name[
    elementMetadata(cur_prom)$gene_id == cur_gene]

cur_region <- cur_prom[elementMetadata(cur_prom)$gene_id == cur_gene]

collected <- list()
lib.sizes <- colSums(assays(bins)$counts)/1e6
for(i in 1:2){
  reads <- extractReads(bam.file=files[i], cur_region, param=param)
  cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
  collected[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,0.5),
       name=names[i],  col.axis="black", col.title="black",
     fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}

pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/Chip-Seq/Cypt1_H3K4me1_ChIP-Seq.pdf", 
            sep = ""),
      width = 12, height = 4)
  plotTracks(c(gax, collected[1:2],
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

for(i in 3:4){
  reads <- extractReads(bam.file=files[i], cur_region, param=param)
  cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
  collected[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,2),
       name=names[i],  col.axis="black", col.title="black",
     fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}

pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/Chip-Seq/Cypt1_H3K4me3_ChIP-Seq.pdf", 
            sep = ""),
      width = 12, height = 4)
  plotTracks(c(gax, collected[3:4],
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

for(i in 5:6){
  reads <- extractReads(bam.file=files[i], cur_region, param=param)
  cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
  collected[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,1),
       name=names[i],  col.axis="black", col.title="black",
     fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}

pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/Chip-Seq/Cypt1_H3K27ac_ChIP-Seq.pdf", 
            sep = ""),
      width = 12, height = 4)
  plotTracks(c(gax, collected[5:6],
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

for(i in 7:8){
  reads <- extractReads(bam.file=files[i], cur_region, param=param)
  cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
  collected[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,0.5),
       name=names[i],  col.axis="black", col.title="black",
     fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}

pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/Chip-Seq/Cypt1_H27me3_ChIP-Seq.pdf", 
            sep = ""),
      width = 12, height = 4)
  plotTracks(c(gax, collected[7:8],
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

for(i in 9:10){
  reads <- extractReads(bam.file=files[i], cur_region, param=param)
  cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
  collected[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,0.5),
       name=names[i],  col.axis="black", col.title="black",
     fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}

pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/Chip-Seq/Cypt1_Input_ChIP-Seq.pdf", 
            sep = ""),
      width = 12, height = 4)
  plotTracks(c(gax, collected[9:10],
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()
```