---
title: "Visualization of ChIP-Seq tracks"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/Chip-Seq/Promoter_visualization.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Read in data and libraries

```{r data}
library(Gviz)
library(rtracklayer)
library(EnsDb.Mmusculus.v79)
library(csaw)

names <- c("H3K4me1_ChIPSeq_SC.sorted.bam", "H3K4me1_ChIPSeq_ST.sorted.bam",
           "H3K4me3_ChIPSeq_SC.sorted.bam", "H3K4me3_ChIPSeq_ST.sorted.bam",
           "H3K27ac_ChIPSeq_SC.sorted.bam", "H3K27ac_ChIPSeq_ST.sorted.bam",
           "H27me3_ChIPSeq_SC.sorted.bam", "H27me3_ChIPSeq_ST.sorted.bam",
           "Input_ChIPSeq_SC.sorted.bam", "Input_ChIPSeq_ST.sorted.bam")

files <- paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Data/Hammoud_CSC/ChipSeq/bam/", names, sep = "")

# Blacklisted regions
black <- import("http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz")
seqlevels(black) <- sub("chr", "", seqlevels(black))
seqnames(black) <- sub("chr", "", seqnames(black))

# Parameter for reading in bam files
param <- readParam(dedup=TRUE, minq=10, discard=black, pe="none")

# Find large regions around promoters
cur_prom <- promoters(genes(EnsDb.Mmusculus.v79), 
                  upstream = 10000, downstream = 5000)

# Bin the mapped reads
bins <- windowCounts(files, bin = TRUE, width = 1000, param=param)
```

## Visualization

```{r}
# Akap4
cur_gene <- "ENSMUSG00000050089"
cur_name <- elementMetadata(cur_prom)$gene_name[
    elementMetadata(cur_prom)$gene_id == cur_gene]

cur_region <- cur_prom[elementMetadata(cur_prom)$gene_id == cur_gene]

collected.K4me1 <- list()
lib.sizes <- colSums(assays(bins.K9)$counts)/1e6
for (i in 1:length(bam.files.K9)) {
    reads <- extractReads(bam.file=bam.files.K9[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K9[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K9[i],  col.axis="black", col.title="black",
         fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K9 <- collected.K9[c(1,3,2,4)]

collected.K4 <- list()
lib.sizes <- colSums(assays(bins.K4)$counts)/1e6
for (i in 1:length(bam.files.K4)) {
    reads <- extractReads(bam.file=bam.files.K4[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K4[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K4[i],  col.axis="black", col.title="black",
         fill="dark blue", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K4 <- collected.K4[c(1,3,2,4)]


  pdf(paste("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/Akap4_forFigure_spermatocytes.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K9[1:2], collected.K4[1:2]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
  dev.off()
  
pdf(paste("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/Akap4_forFigure_spermatids.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K9[3:4], collected.K4[3:4]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()
```