---
title: "CnR Xchr in- and reactivation"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Variability/CnR_XChr.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r peaks, message=FALSE}
library(rtracklayer)
library(EnrichedHeatmap)
library(GenomicFeatures)
library(EnsDb.Mmusculus.v79)
library(scater)
library(plyr)
library(reshape2)
library(pheatmap)
library(viridis)

# Generate feature annotation
prom <- promoters(genes(EnsDb.Mmusculus.v79),
                  upstream = 1500, downstream = 500)
prom <- prom[seqnames(prom) %in% c(as.character(1:19), "X", "Y", "MT"),]
tss <- promoters(genes(EnsDb.Mmusculus.v79), 
                 upstream = 0, downstream = 1)
tss <- tss[seqnames(tss) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(tss) <- c(as.character(1:19), "X", "Y", "MT")
tss.X <- tss[seqnames(tss) == "X"]

# Read in blacklist file
blacklist <- import("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Cnr/Blacklist_mm10.bed")
seqlevels(blacklist) <- sub("^chr", '', seqlevels(blacklist))

# Read in sce data
sce <- readRDS("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all_clusters.rds")

# Select XChr genes
genenames <- read.table("../../Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
sce.XChr <- sce[rowData(sce)$ID %in% tss.X@elementMetadata@listData$gene_id,
                grepl("B6", colData(sce)$Sample)]
sce.XChr <- normalize(sce.XChr)
```

# Build mean expression matrix

```{r mean_expression}
df <- as.data.frame(t(as.matrix(logcounts(sce.XChr))))
df$groups <- colData(sce.XChr)$AnnotatedClusters
df.melt <- melt(df, id.vars = "groups")

# Collect mean expression for each gene in each group
mat <- ddply(df.melt, .(groups, variable), summarize, mean=mean(value))
mat.for.heatmap <- dcast(data = mat,formula = variable~groups,fun.aggregate = sum,value.var = "mean")
rownames(mat.for.heatmap) <- mat.for.heatmap$variable
mat.for.heatmap$variable <- NULL
mat.for.heatmap$Leydig <- NULL
mat.for.heatmap$Sertoli <- NULL
mat.for.heatmap$PTM_2 <- NULL
mat.for.heatmap$Outliers <- NULL

# Remove genes that are not expressed across all groups
mat.for.heatmap <- mat.for.heatmap[rowSums(mat.for.heatmap) > 0,]

# Order matrix based on mean expression in spermatids
mat.for.heatmap <- mat.for.heatmap[order(rowMeans(mat.for.heatmap[,9:21]), decreasing = TRUE),]
mat.for.heatmap <- mat.for.heatmap[rowMeans(mat.for.heatmap) > 0.1,]

# Split matrix in 3 categories:
# Inactivated and reactivated
mat.1 <- mat.for.heatmap[mat.for.heatmap[,1] > rowMeans(mat.for.heatmap[,2:8]) &
                           rowMeans(mat.for.heatmap[,2:8]) < rowMeans(mat.for.heatmap[,9:21]),]

# Inactivated, not reactivated
mat.2 <- mat.for.heatmap[mat.for.heatmap[,1] > rowMeans(mat.for.heatmap[,2:8]) &
                           rowMeans(mat.for.heatmap[,2:8]) >= rowMeans(mat.for.heatmap[,9:21]),]
# Not inactivated
mat.3 <- mat.for.heatmap[mat.for.heatmap[,1] <= rowMeans(mat.for.heatmap[,2:8]),]
```

## Visualize heatmaps

```{r}
pdf("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/Results/XChr/GeneExpression.pdf",
    width = 10, height = 15, onefile=FALSE)
pheatmap(rbind(mat.1, mat.2, mat.3), cluster_cols = FALSE, 
         cluster_rows = FALSE, show_rownames = FALSE,
         show_colnames = FALSE,
         color = viridis(100), 
         annotation_col = data.frame(row.names = colnames(mat.1),
                                     Cell_type = colnames(mat.1)),
         annotation_colors = list(Cell_type = metadata(sce.XChr)$color_vector),
         border_color = NA, gaps_row = c(nrow(mat.1), nrow(mat.1)+nrow(mat.2)))
dev.off()
```

# Plot corresponding Peak maps

## H3K4me3

```{r}
cur_files <- list.files("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bedgraphs/H3K4me3/", 
                        full.names = TRUE)
filenames <- sub(".bed.1-1000.bedgraph.gz$", ".pdf",
                 list.files("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bedgraphs/H3K4me3/"))

for(i in 1:length(cur_files)){
  cur_file <- import(cur_files[i])
  
  # Exclude blacklisted regions
  cur_removed <- overlapsAny(cur_file, blacklist)
  cur_file <- cur_file[!cur_removed]
  cur_file <- cur_file[seqnames(cur_file) %in%
                         c(as.character(1:19), "X", "Y", "MT"),]
  seqlevels(cur_file) <- c(as.character(1:19), "X", "Y", "MT")

  cur_bed <- cur_file[seqnames(cur_file) == "X"]

  mat.tss = normalizeToMatrix(cur_bed, tss.X, 
                        value_column = "score",
                        extend = 5000, mean_mode = "w0", 
                        w = 50, keep = c(0,0.99))
  
  # Flip the tss on minus strand
  # mat.tss[which(strand(tss.X) == "-"),] <- t(apply(mat.tss[which(strand(tss.X) == "-"),], 1, rev))
  pdf(paste("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/Results/XChr/H3K4me3/", 
            filenames[i], sep = ""),
    width = 10, height = 15, onefile=FALSE)
  pheatmap(mat.tss[c(rownames(mat.1), rownames(mat.2), rownames(mat.3)),],
           cluster_rows = FALSE, cluster_cols = FALSE, 
           show_rownames = FALSE, show_colnames = FALSE,
         color = colorRampPalette(c("gray90", "red", "black"))(100),
         gaps_row = c(nrow(mat.1), nrow(mat.1) + nrow(mat.2)))
  dev.off()
}
```

## H3K9me3

```{r}
cur_files <- list.files("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bedgraphs/H3K9me3/", 
                        full.names = TRUE)
filenames <- sub(".bed.1-1000.bedgraph.gz$", ".pdf",
                 list.files("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bedgraphs/H3K9me3/"))

for(i in 1:length(cur_files)){
  cur_file <- import(cur_files[i])
  
  # Exclude blacklisted regions
  cur_removed <- overlapsAny(cur_file, blacklist)
  cur_file <- cur_file[!cur_removed]
  cur_file <- cur_file[seqnames(cur_file) %in%
                         c(as.character(1:19), "X", "Y", "MT"),]
  seqlevels(cur_file) <- c(as.character(1:19), "X", "Y", "MT")

  cur_bed <- cur_file[seqnames(cur_file) == "X"]

  mat.tss = normalizeToMatrix(cur_bed, tss.X, 
                        value_column = "score",
                        extend = 5000, mean_mode = "w0", 
                        w = 50, keep = c(0,0.99))
  
  # Flip the tss on minus strand
  # mat.tss[which(strand(tss.X) == "-"),] <- t(apply(mat.tss[which(strand(tss.X) == "-"),], 1, rev))
  pdf(paste("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/Results/XChr/H3K9me3/", 
            filenames[i], sep = ""),
    width = 10, height = 15, onefile=FALSE)
  pheatmap(mat.tss[c(rownames(mat.1), rownames(mat.2), rownames(mat.3)),],
           cluster_rows = FALSE, cluster_cols = FALSE, 
           show_rownames = FALSE, show_colnames = FALSE,
         color = colorRampPalette(c("gray90", "blue", "black"))(100),
         gaps_row = c(nrow(mat.1), nrow(mat.1) + nrow(mat.2)))
  dev.off()
}
```

## H3K27ac

```{r}
cur_files <- list.files("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bedgraphs/H3K27ac/", 
                        full.names = TRUE)
filenames <- sub(".bed.1-1000.bedgraph.gz$", ".pdf",
                 list.files("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bedgraphs/H3K27ac/"))

for(i in 1:length(cur_files)){
  cur_file <- import(cur_files[i])
  
  # Exclude blacklisted regions
  cur_removed <- overlapsAny(cur_file, blacklist)
  cur_file <- cur_file[!cur_removed]
  cur_file <- cur_file[seqnames(cur_file) %in%
                         c(as.character(1:19), "X", "Y", "MT"),]
  seqlevels(cur_file) <- c(as.character(1:19), "X", "Y", "MT")

  cur_bed <- cur_file[seqnames(cur_file) == "X"]

  mat.tss = normalizeToMatrix(cur_bed, tss.X, 
                        value_column = "score",
                        extend = 5000, mean_mode = "w0", 
                        w = 50, keep = c(0,0.99))
  
  # Flip the tss on minus strand
  # mat.tss[which(strand(tss.X) == "-"),] <- t(apply(mat.tss[which(strand(tss.X) == "-"),], 1, rev))
  pdf(paste("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/Results/XChr/H3K27ac/", 
            filenames[i], sep = ""),
    width = 10, height = 15, onefile=FALSE)
  pheatmap(mat.tss[c(rownames(mat.1), rownames(mat.2), rownames(mat.3)),],
           cluster_rows = FALSE, cluster_cols = FALSE, 
           show_rownames = FALSE, show_colnames = FALSE,
         color = colorRampPalette(c("gray90", "green", "black"))(100),
         gaps_row = c(nrow(mat.1), nrow(mat.1) + nrow(mat.2)))
  dev.off()
}
```

## H3.3

```{r}
cur_files <- list.files("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bedgraphs/H3.3/", 
                        full.names = TRUE)
filenames <- sub(".bed.1-1000.bedgraph.gz$", ".pdf",
                 list.files("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bedgraphs/H3.3/"))

for(i in 1:length(cur_files)){
  cur_file <- import(cur_files[i])
  
  # Exclude blacklisted regions
  cur_removed <- overlapsAny(cur_file, blacklist)
  cur_file <- cur_file[!cur_removed]
  cur_file <- cur_file[seqnames(cur_file) %in%
                         c(as.character(1:19), "X", "Y", "MT"),]
  seqlevels(cur_file) <- c(as.character(1:19), "X", "Y", "MT")

  cur_bed <- cur_file[seqnames(cur_file) == "X"]

  mat.tss = normalizeToMatrix(cur_bed, tss.X, 
                        value_column = "score",
                        extend = 5000, mean_mode = "w0", 
                        w = 50, keep = c(0,0.99))
  
  # Flip the tss on minus strand
  # mat.tss[which(strand(tss.X) == "-"),] <- t(apply(mat.tss[which(strand(tss.X) == "-"),], 1, rev))
  pdf(paste("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/Results/XChr/H3.3/", 
            filenames[i], sep = ""),
    width = 10, height = 15, onefile=FALSE)
  pheatmap(mat.tss[c(rownames(mat.1), rownames(mat.2), rownames(mat.3)),],
           cluster_rows = FALSE, cluster_cols = FALSE, 
           show_rownames = FALSE, show_colnames = FALSE,
         color = colorRampPalette(c("gray90", "purple", "black"))(100),
         gaps_row = c(nrow(mat.1), nrow(mat.1) + nrow(mat.2)))
  dev.off()
}
```

# Order genes based on their position on the X chromosome

```{r positioning}
genes.X <- genes(EnsDb.Mmusculus.v79)
genes.X <- genes.X[seqnames(genes.X) == "X"]

mat.ordered <- mat.for.heatmap[genes.X@elementMetadata$gene_id[genes.X@elementMetadata$gene_id %in% rownames(mat.for.heatmap)],]

pdf("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/Results/XChr/GeneExpression_orderedByPosition.pdf",
    width = 10, height = 15, onefile=FALSE)
pheatmap(mat.ordered, cluster_cols = FALSE, 
         cluster_rows = FALSE, show_rownames = FALSE,
         show_colnames = FALSE,
         color = viridis(100), 
         annotation_col = data.frame(row.names = colnames(mat.ordered),
                                     Cell_type = colnames(mat.ordered)),
         annotation_colors = list(Cell_type = metadata(sce.XChr)$color_vector),
         border_color = NA)
dev.off()

# Visualize H3K4me3 peaks in the same order
cur_files <- list.files("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bedgraphs/H3K4me3/", 
                        full.names = TRUE)

filenames <- sub(".bed.1-1000.bedgraph.gz$", ".pdf",
                 list.files("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bedgraphs/H3K4me3/"))

# Spermatocytes
cur_file <- import(cur_files[11])
  
# Exclude blacklisted regions
cur_removed <- overlapsAny(cur_file, blacklist)
cur_file <- cur_file[!cur_removed]
cur_file <- cur_file[seqnames(cur_file) %in%
                       c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(cur_file) <- c(as.character(1:19), "X", "Y", "MT")

cur_bed <- cur_file[seqnames(cur_file) == "X"]
mat.tss = normalizeToMatrix(cur_bed, tss.X, 
                      value_column = "score",
                      extend = 5000, mean_mode = "w0", 
                      w = 50, keep = c(0,0.99))

# Flip the tss on minus strand
# mat.tss[which(strand(tss.X) == "-"),] <- t(apply(mat.tss[which(strand(tss.X) == "-"),], 1, rev))
pdf(paste("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/Results/XChr/OrderedByPosition_", 
          filenames[11], sep = ""),
  width = 10, height = 15, onefile=FALSE)
pheatmap(mat.tss[rownames(mat.ordered),],
         cluster_rows = FALSE, cluster_cols = FALSE, 
         show_rownames = FALSE, show_colnames = FALSE,
         color = colorRampPalette(c("gray90", "red", "black"))(100))
dev.off()

# Spermatids
cur_file <- import(cur_files[12])
  
# Exclude blacklisted regions
cur_removed <- overlapsAny(cur_file, blacklist)
cur_file <- cur_file[!cur_removed]
cur_file <- cur_file[seqnames(cur_file) %in%
                       c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(cur_file) <- c(as.character(1:19), "X", "Y", "MT")

cur_bed <- cur_file[seqnames(cur_file) == "X"]
mat.tss = normalizeToMatrix(cur_bed, tss.X, 
                      value_column = "score",
                      extend = 5000, mean_mode = "w0", 
                      w = 50, keep = c(0,0.99))
mat.tss <- mat.tss[rownames(mat.ordered),]

# Flip the tss on minus strand
# mat.tss[which(strand(tss.X) == "-"),] <- t(apply(mat.tss[which(strand(tss.X) == "-"),], 1, rev))
pdf(paste("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/Results/XChr/OrderedByPosition_", 
          filenames[12], sep = ""),
  width = 10, height = 15, onefile=FALSE)
pheatmap(mat.tss[rownames(mat.ordered),],
         cluster_rows = FALSE, cluster_cols = FALSE, 
         show_rownames = FALSE, show_colnames = FALSE,
         color = colorRampPalette(c("gray90", "red", "black"))(100))
EnrichedHeatmap(mat.tss, col = c("gray90", "red", "black"), 
                row_order = 1:nrow(mat.tss))
dev.off()

```
