---
title: "DB analysis of H3K9me3"
author: Aaron Lun
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    fig_caption: yes
---

```{r, echo=FALSE, message=FALSE, results="hide"}
library(BiocStyle)
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
```

# Defining the files

First we define the relevant files.

```{r}
library(cowplot)
library(reshape2)
library(csaw)
library(rtracklayer)
# Select the day for profiling
day <- "26" 

bam.files <- list.files("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K9me3/", full.names = TRUE,
                        pattern = paste(".+JP", day, ".+bam$", sep = ""))
```

We also define the conditions:

```{r}
batch <- gl(2, 2)
treatment <- rep(c("spermatocytes", "spermatids"), 2)
data.frame(BAM=bam.files, Batch=batch, Treatment=treatment)
```


# Counting reads in windows

Setting up a `readParam` object for read counting.
This is paired-end data, and we remove read pairs with insert sizes above 1 kbp.
We remove duplicated read pairs or reads with quality scores below 10.
We also use the ENCODE mm10 blacklist to ignore reads in problematic regions.

```{r, message=FALSE}
black <- import("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/Blacklist_mm10.bed")
param <- readParam(dedup=TRUE, minq=10, discard=black, pe="both", max.frag=1000)
param
```

We then count reads into 1000 and 5000 bp bins 
No read extension is necessary as the paired reads fully define each fragment.

```{r}
out <- windowCounts(bam.files, param=param, width=1000, bin = TRUE)
out
bins <- windowCounts(bam.files, param=param, width=5000, bin = TRUE)
```

# Filter out lowly abundant windows and removing the X chromosome

To further normalize the data, we remove lowly abundant windows. Also.
we exclude regions form the X chromosome due to unusually high enrichment.

```{r}
# Removing windows on X chromosome
out.noX <- out[!as.vector(seqnames(out) == "X"),]
bins.noX <- bins[!as.vector(seqnames(bins) == "X"),]

# Removing lowly abundant windows
fstat <- filterWindows(data=out.noX, background=bins.noX, type="global")
hist(fstat$filter, col="grey80", xlab="Log2-fold increase above background",
    ylab="Number of windows", freq=TRUE)
filtered <- out.noX[fstat$filter >= log2(2),]
```

# Normalization for efficiency biases

## Computing TMM factors

We normalize out efficiency biases, under the assumption that most histone marking does _not_ change between spermatids and spermatocytes.

```{r}
out <- normOffsets(filtered, se.out = out, type = "loess")

```

We can create some MA plots using the bins to examine the normalization factors.

```{r}
library(edgeR)
adjc <- calculateCPM(out, prior.count=0.5, log = TRUE, use.offsets = TRUE)
average <- scaledAverage(out)

par(mfrow=c(1, 2))
for (x in 2:ncol(adjc)) {
    smoothScatter(average, adjc[,x]-adjc[,1], xlab="Average", 
        ylab=sprintf("M (%i vs 1)", x))
}
```

Save normalized reads in 1Kb bins

```{r}
tmp <- asDGEList(out)
adjc <- cpm(tmp, log=TRUE, prior.count=0.5)
colnames(adjc) <- paste(treatment, batch, sep = "_")
saveRDS(file=paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/DB/H3K9me3/H3K9me3", day, "_1000kb_bins_normReads.rds", sep = ""), list(windows=rowRanges(out), 
    normReads=adjc)) 
```
