---
title: "Quality assessment of TC1/TC0 single spermatocyes and spermatogonia"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load libraries 

```{r}
library(RColorBrewer)
suppressMessages(library(scran))
suppressMessages(library(scater))
library(openxlsx)
```

## Collecting meta information on single cells

```{r}
# Raw counts
raw.counts <- read.table("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/raw_reads/All_fastqs_raw.txt", sep = "\t")

# Metadata 
meta <- read.xlsx("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/ArrayExpress/C1/sdrf.xlsx")

# Build sce object
sce <- SingleCellExperiment(assays = list(counts = as.matrix(raw.counts[!grepl("_", rownames(raw.counts)),])))
isSpike(sce, "ERCC") <- which(grepl("ERCC-", rownames(counts(sce)))) 

# Fill object with metddata
colData(sce) <- DataFrame(data.frame(Library = colnames(raw.counts),
                          Microscopy = meta$Description[match(colnames(raw.counts), meta$Source.Name)],
                          Genotype = meta$`Characteristics[genotype]`[match(colnames(raw.counts), meta$Source.Name)],
                          Cell_type = meta$`Characteristics[cell.type]`[match(colnames(raw.counts), meta$Source.Name)],
                          Unmapped = as.numeric(raw.counts["N_unmapped",]),
                          Multimapping = as.numeric(raw.counts["N_multimapping",]),
                          NoFeature = as.numeric(raw.counts["N_noFeature",]),
                          Ambiguous = as.numeric(raw.counts["N_ambiguous",])))

# Calculate QC metrics
sce <- calculateQCMetrics(sce)
```

# Visualization of features and filtering

We first plot the percentage of reads mapped to the exonic part of the genome

```{r}
# Colored by visual inspection
plot(Master.QC$Genomic/Master.QC$TotalReads, 
     pch=16, xlab="Cells", ylab="log10[Total mapped reads]", 
     col= ifelse(Master.QC$Microscopy == "-", "red", 
                 ifelse(Master.QC$Microscopy == "2 cells", "yellow", 
                        ifelse(Master.QC$Microscopy == "single", "black", "green"))), cex=0.7)
```

```{r}
# Colored by batches
plot(Master.QC$Genomic/Master.QC$TotalReads, 
     pch=16, xlab="Cells", ylab="log10[Total mapped reads]", 
     col= Master.QC$col, cex=0.7)
```

## Remove duplets, empty wells and debris

For further downstream analysis, we select well with only single cells.

```{r}
Singles <- Master.QC[Master.QC$Microscopy == "single",]
```

## Spike-in reads

Let's have a look at the percentage of reads mapped to ERCCs.

```{r}
plot(Singles$ERCC/Singles$TotalReads, col= Singles$col, ylab="%ERCC", pch=16)
plot(log10(Singles$ERCC + 1), col= Singles$col, ylab="Total number of ERCCs per well", pch=16)
```

It appears that two batches - SST4 Spermatogonia and SST6 Spermatogonia - have higher amounts of ERCCs. Therefore, I will not use ERCCs for QC and normalization.

## Exonic reads

Next we will filter on the percentage of reads mapped to exonic regions.

```{r}
plot(Singles$Genomic/Singles$TotalReads, col= Singles$col, ylab="%Genomic", pch=16)
Singles <- Singles[Singles$Genomic/Singles$TotalReads > 0.15,]
```

## Mapping stats

We remove cells with unusual mapping and counting stats.

```{r}
# Reads not mapped to exons
plot(Singles$NoFeature/Singles$TotalReads, 
     pch=16, xlab="Cells", ylab="log[No feature reads]", 
     col= Singles$col, cex=0.7)

# Multimapping reads
plot(Singles$Ambigous/Singles$TotalReads, 
     pch=16, xlab="Cells", ylab="log[Ambigous Reads]", 
     col= Singles$col, cex=0.7)

# Low quality reads
plot(Singles$LowQual/Singles$TotalReads, 
     pch=16, xlab="Cells", ylab="log[LowQual reads]", 
     col= Singles$col, cex=0.7)


plot(Singles$NotAligned/Singles$TotalReads, 
     pch=16, xlab="Cells", ylab="log[Not aligned reads]", 
     col= Singles$col, cex=0.7)

Singles <- Singles[Singles$NotAligned/Singles$TotalReads < 0.03,]
```

## Number of detected genes

Remove cells with unusually low number of genes detected

```{r}
plot(Singles$MouseGenesDetected, 
     pch=16, col=Singles$col, xlab="Cells", ylab="# Mouse Genes Detected")
Singles <- Singles[which(Singles$MouseGenesDetected > 3000),]
```

First round of filtering steps is done

```{r}
QC_pass <- input.SST3_7[,Singles$Libraries]
QC_pass <- QC_pass[grepl("E", rownames(input.SST3_7)),]
colnames(QC_pass) <- rownames(Singles)
```

## Pre-normalize data and remove spermatocytes that cluster with spermatogonia

We normlize data based on mouse genes expressed.

```{r}
Mouse <- QC_pass[grepl("ENSM", rownames(QC_pass)),]
Mouse <- Mouse[rowMeans(Mouse) > 1,]

# Generate SCE dataset
sce <- SingleCellExperiment(list(counts=as.matrix(Mouse)))

# Keep spermatocytes, spermatogonia and spermatids seperately during normalization
clusters <- factor(x = ifelse(grepl("Stem", colnames(Mouse)), 1, 
                              ifelse(grepl("Spermatocyte", colnames(Mouse)), 2, 3)))

# Calculate size factors
sce <- computeSumFactors(sce, clusters = clusters)
sf <- sizeFactors(sce)

# Normalize
norm.counts <- t(t(Mouse)/sf)
```

Calculate PCA of normalised data

```{r}
pca <- prcomp(log10(t(norm.counts) + 1))
plot(pca$x[,1], pca$x[,2], 
     col = ifelse(grepl("Stem", colnames(norm.counts)), "red", 
                  ifelse(grepl("Spermatids", colnames(norm.counts)), "green", "blue")), 
     pch=16)
plot(pca$x[,1], pca$x[,2], 
     col = Singles$col, pch=16)
```

Remove outlying cells

```{r}
cells <- colnames(norm.counts)[pca$x[,1] < 40 & pca$x[,1] > 0]
cells.sperm <- colnames(norm.counts)[pca$x[,1] > 40]
cells.sperm <- cells.sperm[grepl("Sperm", cells.sperm)]

QC_pass <- QC_pass[,-which(colnames(QC_pass) %in% c(cells, cells.sperm))]
```

## Remove Batch SST3

```{r}
QC_pass <- QC_pass[,!grepl("SST3", colnames(QC_pass))]
```

## Filtering on marker genes

We now remove cells that appear to be spermatid contaminations

```{r}
# Prm1 = ENSMUSG00000022501, Prm2 = ENSMUSG00000038015
plot(as.numeric(log10(QC_pass["ENSMUSG00000022501",] + 1)),
     as.numeric(log10(QC_pass["ENSMUSG00000038015",] + 1)),
     pch = 16, col = ifelse(grepl("Spermato", colnames(QC_pass)), "red", 
                            ifelse(grepl("Spermatid", colnames(QC_pass)), "green", "blue")))
QC_pass <- QC_pass[,-which(as.numeric(log10(QC_pass["ENSMUSG00000022501",] + 1)) > 1 &
                             as.numeric(log10(QC_pass["ENSMUSG00000038015",] + 1)) > 1 &
                             (grepl("Spermato", colnames(QC_pass)) | 
                                grepl("Stem", colnames(QC_pass))))]
```

## Filtering on genes

To filter on genes we split the dataset in three pertitions: Human genes and Mouse genes

```{r}
Mouse <- QC_pass[grepl("ENSM", rownames(QC_pass)),]
Human <- QC_pass[grepl("ENSG", rownames(QC_pass)),]
```

To filter mouse genes, we use an arbitrary threshold of an average count 1

```{r}
dim(Mouse)
Mouse <- Mouse[rowMeans(Mouse) > 1,]
dim(Mouse)
```
To filter human genes, we split cells into TC0 and TC1 partitions

```{r}
TC1 <- Human[,grepl("TC1", colnames(Human))]
TC0 <- Human[,grepl("TC0", colnames(Human))]
```

Remove genes that show cross mapping between human and mouse in TC0 mice

```{r}
plot(log10(rowSums(TC1) + 1), log10(rowSums(TC0) + 1), 
     pch = 16, xlab = "Chr21 expr TC1", ylab = "Chr21 expr TC0")
Human <- Human[-which(log10(rowSums(TC0) + 1) > 1 | rowSums(TC1) == 0),]
```

Merge Datasets again

```{r}
QC_pass <- rbind(Human, Mouse, QC_pass[grepl("ERCC", rownames(QC_pass)),])
```

Write file

```{r}
write.table(as.data.frame(QC_pass), "data/raw_reads/All_pass.txt", quote = FALSE, sep = "\t")

sessionInfo()
```
