---
title: "Figure S8"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S8.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data and libraries

```{r data, message=FALSE}
library(scater)
library(plyr)
library(reshape2)
library(pheatmap)
library(viridis)
library(plot3D)
library(ggsci)
library(cowplot)
library(openxlsx)
library(VennDiagram)
library(biomaRt)
library(seqinr)
source("../../Functions/auxiliary.R")

# Read in sce data
sce <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")
sce <- sce[,grepl("B6", colData(sce)$Sample) & 
             colData(sce)$AnnotatedClusters %in% levels(colData(sce)$AnnotatedClusters)[1:20]]
sce <- normalize(sce)

# Read in gene annotations
genenames <- read.table("../../Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
rownames(genenames) <- genenames$Gene.stable.ID
```

# Order cells over the differentiation time course

```{r ordering}
# Select Spermatids
sce.spermatids <- sce[, colData(sce)$AnnotatedClusters %in% 
                           paste("S", 1:11, sep = "")]
sce.spermatids <- normalize(sce.spermatids)

# Compute HVG
HVgenes <- HVG(sce.spermatids)

# Compute PCA
pca <- prcomp(t(logcounts(sce.spermatids)[HVgenes,]))

# Pseudo rank
prank <- PT(rd = pca$x[,1:3], 
            clusters = colData(sce.spermatids)$AnnotatedClusters,
            col_vector = metadata(sce.spermatids)$color_vector)
```

# Histone variants over spermiogenesis

Here, we visualize the histone variants taken from Kennani et al. Epigenetics & Chromatin, 2017.

```{r hist-var}
# Read in histone variants
Hist.all <- read.xlsx("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/Mouse_histones_annotated.xlsx")
rownames(Hist.all) <- Hist.all$Ensembl.Gene.ID

# Collect data for heatmap plotting
for.heatmap <- logcounts(sce.spermatids)[rownames(Hist.all),
                  order(prank[,"rank"], decreasing = TRUE)]
colnames(for.heatmap) <- paste(colData(sce.spermatids)$Barcode[order(prank[,"rank"],
                                                               decreasing = TRUE)],
                               colData(sce.spermatids)$Library[order(prank[,"rank"],
                                                               decreasing = TRUE)],
                               sep = "_")

# Plot these genes over pseudotime
pdf("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S8/HistoneVariants.pdf", onefile = FALSE)
pheatmap(for.heatmap, show_colnames = FALSE, show_rownames = FALSE,
         cluster_rows = FALSE, cluster_cols = FALSE, 
         color = colorRampPalette(c("#053061", "#4393c3", "#f7f7f7", "#d6604d", "#67001f"))(100),
         annotation_col = data.frame(row.names = colnames(for.heatmap),
                  cell_type = colData(sce.spermatids)$AnnotatedClusters[order(prank[,"rank"],
                      decreasing = TRUE)]),
         annotation_colors = list(cell_type = metadata(sce.spermatids)$color_vector,
                                  label = c("NA" = "white",
                                            "label" = "black"),
                                  Histone = c("H2A_variant" = "springgreen",
                                              "H2B_variant" = "springgreen3",
                                              "H3_variant" = "seagreen",
                                              "H1_variant" = "lightseagreen",
                                              "Canonical" = "indianred3",
                                              "Transition_protein" = "darkorange4",
                                              "Protamine" = "burlywood4"),
                                  Testis_specific = c("Testis-specific" = "sienna1",
                                                      "NA" = "white")), 
         scale = "row",
         annotation_row = data.frame(row.names = rownames(for.heatmap),
                                     Testis_specific = Hist.all$`Testis-specific`,
                                     Histone = Hist.all$Label,
                                     label = Hist.all$Gene_name))
dev.off()
```

# Visualize histone variants

```{r}
p.clusters <- ggplot(data.frame(tSNE1 = reducedDims(sce)$TSNE[,1],
                      tSNE2 = reducedDims(sce)$TSNE[,2],
                      cluster = colData(sce)$AnnotatedClusters)) +
      geom_point(aes(tSNE1, tSNE2, colour = cluster)) + 
      scale_color_manual(values = metadata(sce)$color_vector) + 
  guides(colour = FALSE)

# H3f3a
H3f3a.tsne <- ggplot(data.frame(tSNE1 = reducedDims(sce)$TSNE[,1],
                      tSNE2 = reducedDims(sce)$TSNE[,2],
                      Gene = logcounts(sce)[rowData(sce)$Symbol == "H3f3a",])) +
      geom_point(aes(tSNE1, tSNE2, colour = Gene)) + 
      scale_color_viridis() 

H3f3a.box <-  ggplot(data.frame(value = logcounts(sce)[rowData(sce)$Symbol == "H3f3a",],
                      cluster = colData(sce)$AnnotatedClusters)) +
      geom_boxplot(aes(x = cluster, y = value, fill = cluster)) + 
      scale_fill_manual(values = metadata(sce)$color_vector) + ylab("log2(Expr)") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1),
            panel.background = element_blank(), 
            axis.title.x = element_blank()) + guides(fill = FALSE) +
  ylim(c(0,9))

# H3f3b
H3f3b.tsne <- ggplot(data.frame(tSNE1 = reducedDims(sce)$TSNE[,1],
                      tSNE2 = reducedDims(sce)$TSNE[,2],
                      Gene = logcounts(sce)[rowData(sce)$Symbol == "H3f3b",])) +
      geom_point(aes(tSNE1, tSNE2, colour = Gene)) + 
      scale_color_viridis() 

H3f3b.box <-  ggplot(data.frame(value = logcounts(sce)[rowData(sce)$Symbol == "H3f3b",],
                      cluster = colData(sce)$AnnotatedClusters)) +
      geom_boxplot(aes(x = cluster, y = value, fill = cluster)) + 
      scale_fill_manual(values = metadata(sce)$color_vector) + ylab("log2(Expr)") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1),
            panel.background = element_blank(), 
            axis.title.x = element_blank()) + guides(fill = FALSE) +
  ylim(c(0,9))

# Hist1h4a 
Hist1h4a.tsne <- ggplot(data.frame(tSNE1 = reducedDims(sce)$TSNE[,1],
                      tSNE2 = reducedDims(sce)$TSNE[,2],
                      Gene = logcounts(sce)[rowData(sce)$Symbol == "Hist1h4a",])) +
      geom_point(aes(tSNE1, tSNE2, colour = Gene)) + 
      scale_color_viridis() 

Hist1h4a.box <-  ggplot(data.frame(value = logcounts(sce)[rowData(sce)$Symbol == "Hist1h4a",],
                      cluster = colData(sce)$AnnotatedClusters)) +
      geom_boxplot(aes(x = cluster, y = value, fill = cluster)) + 
      scale_fill_manual(values = metadata(sce)$color_vector) + ylab("log2(Expr)") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1),
            panel.background = element_blank(), 
            axis.title.x = element_blank()) + guides(fill = FALSE) +
  ylim(c(0,5))
```

# Save figure

```{r}
final <- plot_grid(p.clusters, H3f3a.tsne, H3f3b.tsne, 
          NULL, H3f3a.box, H3f3b.box,
          Hist1h4a.tsne, Hist1h4a.box, NULL, ncol = 3,
          nrow = 3)
ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S8/Fig_S8.pdf", final, width = 12, height = 12)
```

# RNA binding protein analysis

## Prepare files to use with RBPmap (http://rbpmap.technion.ac.il/)

Here, we select the fasta coding sequences and 3' and 5' UTR sequences of the genes in each category of genes that correlate or anticorrelate with the number of genes expressed over spermiogenesis (see Figure 6).

```{r}
# Read in the genes
genes <- read.xlsx("../../../../Dropbox (Personal)/Tc1_meiotic_silencing/Revisions/Figures/Supplemental Tables/Table_S7.xlsx")

# Initialize biomaRt
ensembl <- useMart("ensembl")
ensembl <- useDataset("mmusculus_gene_ensembl", mart=ensembl)

# Loop over genelists and write out files
for(i in unique(genes$group)){
  cur_genes <- genes$gene2[genes$group == i]
  res.5utr = biomaRt::getSequence(id = cur_genes, type = "ensembl_gene_id", mart=ensembl, 
                seqType = "5utr")
  res.3utr = biomaRt::getSequence(id = cur_genes, type = "ensembl_gene_id", mart=ensembl, 
                seqType = "3utr")
  res.coding = biomaRt::getSequence(id = cur_genes, type = "ensembl_gene_id", mart=ensembl, 
                seqType = "coding")
  
  # Write as fasta files 
  # 5UTR
  if(nrow(res.5utr) < 5000){
  write.fasta(sequences = as.list(res.5utr$`5utr`), names = res.5utr$ensembl_gene_id, 
              file.out = paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S8/RBPanalysis/Sequences/", sub(" ", "", i), "_5UTR.fa", sep = ""))
  }
  else{
    res.5utr_1 <- res.5utr[1:4000,]
    res.5utr_2 <- res.5utr[4001:nrow(res.5utr),]
    write.fasta(sequences = as.list(res.5utr_1$`5utr`), names = res.5utr_1$ensembl_gene_id, 
              file.out = paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S8/RBPanalysis/Sequences/", sub(" ", "", i), "_5UTR_1.fa", sep = ""))
    write.fasta(sequences = as.list(res.5utr_2$`5utr`), names = res.5utr_2$ensembl_gene_id, 
              file.out = paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S8/RBPanalysis/Sequences/", sub(" ", "", i), "_5UTR_2.fa", sep = ""))
  }
  
  # 3UTR  
  if(nrow(res.3utr) < 5000){
  write.fasta(sequences = as.list(res.3utr$`3utr`), names = res.3utr$ensembl_gene_id, 
              file.out = paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S8/RBPanalysis/Sequences/", sub(" ", "", i), "_3UTR.fa", sep = ""))
  }
  else{
    res.3utr_1 <- res.3utr[1:4000,]
    res.3utr_2 <- res.3utr[4001:nrow(res.3utr),]
    write.fasta(sequences = as.list(res.3utr_1$`3utr`), names = res.3utr_1$ensembl_gene_id, 
              file.out = paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S8/RBPanalysis/Sequences/", sub(" ", "", i), "_3UTR_1.fa", sep = ""))
    write.fasta(sequences = as.list(res.3utr_2$`3utr`), names = res.3utr_2$ensembl_gene_id, 
              file.out = paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S8/RBPanalysis/Sequences/", sub(" ", "", i), "_3UTR_2.fa", sep = ""))
  }
  
  # Coding
  if(nrow(res.coding) < 5000){
  write.fasta(sequences = as.list(res.coding$coding), names = res.coding$ensembl_gene_id, 
              file.out = paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S8/RBPanalysis/Sequences/", sub(" ", "", i), "_coding.fa", sep = ""))
  }
  else{
    res.coding_1 <- res.coding[1:4000,]
    res.coding_2 <- res.coding[4001:nrow(res.coding),]
    write.fasta(sequences = as.list(res.coding_1$coding), names = res.coding_1$ensembl_gene_id, 
              file.out = paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S8/RBPanalysis/Sequences/", sub(" ", "", i), "_coding_1.fa", sep = ""))
    write.fasta(sequences = as.list(res.coding_2$coding), names = res.coding_2$ensembl_gene_id, 
              file.out = paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S8/RBPanalysis/Sequences/", sub(" ", "", i), "_coding_2.fa", sep = ""))
  }
}


```