---
title: "Figure S11"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S11.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script recapitulates the replicated analysis of K4 and K9 marks at P24 and P28 and visualizes temporal peak patterns for Akap4 and Cypt1.
It also visualizes the enrichment of histone marks from Hammoud et. al. CSC.

# Load data and libraries

```{r data, message=FALSE}
library(rtracklayer)
library(EnrichedHeatmap)
library(GenomicFeatures)
library(EnsDb.Mmusculus.v79)
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
library(plyr)
library(reshape2)
library(pheatmap)
library(viridis)
library(plot3D)
library(ggsci)
library(cowplot)
library(openxlsx)
library(csaw)
library(edgeR)
library(Rsamtools)
library(Gviz)
source("../../Functions/auxiliary.R")

# Read in gene annotations
genenames <- read.table("../Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
rownames(genenames) <- genenames$Gene.stable.ID

# Generate feature annotation
prom <- promoters(genes(EnsDb.Mmusculus.v79))
gene.body <- genes(EnsDb.Mmusculus.v79)
prom <- prom[seqnames(prom) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(prom) <- c(as.character(1:19), "X", "Y", "MT")
gene.body <- gene.body[seqnames(gene.body) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(gene.body) <- c(as.character(1:19), "X", "Y", "MT")
prom.X <- prom[seqnames(prom) == "X"]
prom.Y <- prom[seqnames(prom) == "Y"]
prom.9 <- prom[seqnames(prom) == "9"]
gene.body.X <- gene.body[seqnames(gene.body) == "X"]

# K9 files P24
bam.files.K9.P24 <- list.files("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K9me3/", full.names = TRUE,
                        pattern = paste(".+JP24.+bam$", sep = ""))

# K9 files P28
bam.files.K9.P28 <- list.files("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K9me3/", full.names = TRUE,
                        pattern = paste(".+JP28.+bam$", sep = ""))

# K4 files P24
bam.files.K4.P24 <- list.files("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K4me3/", full.names = TRUE,
                        pattern = paste(".+JP24.+bam$", sep = ""))

# K4 files P24
bam.files.K4.P28 <- list.files("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K4me3/", full.names = TRUE,
                        pattern = paste(".+JP28.+bam$", sep = ""))

# Define conditions
batch <- gl(2, 2)
treatment <- rep(c("spermatocytes", "spermatids"), 2)

# Blacklisted regions
black <- import("http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz")
seqlevels(black) <- sub("chr", "", seqlevels(black))
seqnames(black) <- sub("chr", "", seqnames(black))

# Parameters for reading bam files
param <- readParam(dedup=TRUE, minq=10, discard=black, pe="both", max.frag=1000)

spermatid.specific <- read.xlsx("../../../../Dropbox (Personal)/Tc1_meiotic_silencing/Revisions/Figures/Supplemental Tables/Table_S8_bulkDetection.xlsx")
rownames(spermatid.specific) <- genenames[match(spermatid.specific$Genename, genenames[,2]),1]

# Create objects for visualziation
gax <- GenomeAxisTrack(col="black", fontsize=15, size=2)
gr <- getGeneRegionTrackForGviz(EnsDb.Mmusculus.v79)
options(ucscChromosomeNames = FALSE)

# Bin the genome in 1000Kb windows
bins.K9.P24 <- windowCounts(bam.files.K9.P24, bin = TRUE, width = 1000, param=param)
bins.K9.P28 <- windowCounts(bam.files.K9.P28, bin = TRUE, width = 1000, param=param)
bins.K4.P24 <- windowCounts(bam.files.K4.P24, bin = TRUE, width = 1000, param=param)
bins.K4.P28 <- windowCounts(bam.files.K4.P28, bin = TRUE, width = 1000, param=param)
```

# Count reads in promoter regions

```{r}
prom.X.K9.P24 <- regionCounts(bam.files.K9.P24, regions = prom.X, param=param)
prom.X.K9.P28 <- regionCounts(bam.files.K9.P28, regions = prom.X, param=param)
prom.X.K4.P24 <- regionCounts(bam.files.K4.P24, regions = prom.X, param=param)
prom.X.K4.P28 <- regionCounts(bam.files.K4.P28, regions = prom.X, param=param)

# Generate RPM per promoter - each promoter is 2200bp wide
cur_counts.K9 <- assays(prom.X.K9)$counts
cur_rpm.K9 <- t(t(cur_counts.K9)/(colSums(assays(bins.K9)$counts)/1000000))

cur_counts.K4 <- assays(prom.X.K4)$counts
cur_rpm.K4 <- t(t(cur_counts.K4)/(colSums(assays(bins.K4)$counts)/1000000))

# Compute average between spermatocytes and spermatids
df <- data.frame(Spermatocytes.K9 = rowMeans(cur_rpm.K9[,c(1,3)]),
                 Spermatids.K9 = rowMeans(cur_rpm.K9[,c(2,4)]),
                 avg.K9 = rowMeans(cur_rpm.K9),
                 Spermatocytes.K4 = rowMeans(cur_rpm.K4[,c(1,3)]),
                 Spermatids.K4 = rowMeans(cur_rpm.K4[,c(2,4)]),
                 avg.K4 = rowMeans(cur_rpm.K4))
df$Symbol <- genenames[rownames(df),2]

# Order based on K9 signal in spermatocytes
df <- df[order(df$Spermatocytes.K9, decreasing = TRUE),]

# Annotate genenames
df$spermatid_specific <- spermatid.specific[rownames(df),"Spermatid_specific"]
df$Rnf8 <- df$Symbol %in% K27writers.Rnf8$gene_id
df$Scml2 <- df$Symbol %in% K27writers.Scml2$gene_id

# Plot results
for.plot <- df[!is.na(df$spermatid_specific),]
spermatid.spec.K9.spermatocytes <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatocytes.K9 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama() + ylim(c(0,5))
ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_8/K9_spermatocytes_X_promoters.pdf", 
       spermatid.spec.K9.spermatocytes, width = 4, height = 4)

spermatid.spec.K9.spermatids <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatids.K9 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama()+ ylim(c(0,5))
ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_8/K9_spermatids_X_promoters.pdf", 
       spermatid.spec.K9.spermatids, width = 4, height = 4)

spermatid.spec.K4.spermatocytes <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatocytes.K4 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama() + ylim(c(0,5))
ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_8/K4_spermatocytes_X_promoters.pdf", 
       spermatid.spec.K4.spermatocytes, width = 4, height = 4)

spermatid.spec.K4.spermatids <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatids.K4 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama() + ylim(c(0,5))
ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_8/K4_spermatids_X_promoters.pdf", 
       spermatid.spec.K4.spermatids, width = 4, height = 4)

# Wilcoxon test
wilcox.test(for.plot$Spermatocytes.K9[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatocytes.K9[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatids.K9[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatids.K9[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatocytes.K4[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatocytes.K4[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatids.K4[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatids.K4[for.plot$spermatid_specific == FALSE])
```