---
title: "Figure S5"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S5/Fig_S5.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in the data

```{r data, message=FALSE}
library(scater)
library(ggplot2)
library(reshape2)
library(plyr)
library(openxlsx)
library(cowplot)
library(viridis)
library(pheatmap)
library(Rtsne)
library(irlba)
library(RColorBrewer)
library(monocle)
source("../../Functions/auxiliary.R")

# Single cell data
sce <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")
```

# Comparison of principal curve and monocle

```{r}
# Read in only spermatogonia from P10 and P15
sce.spermatogonia <- sce[,grepl("P10|P15", colData(sce)$Sample) &
                                       colData(sce)$AnnotatedClusters == "Spermatogonia"]
sce.spermatogonia <- normalize(sce.spermatogonia)

sce_RA <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_RA.rds")

# Batch correction
sce.spermatogonia <- sce.spermatogonia[match(intersect(rownames(sce_RA), 
                          rowData(sce.spermatogonia)$Symbol), rowData(sce.spermatogonia)$Symbol),]
sce_RA <- sce_RA[match(intersect(rownames(sce_RA), rowData(sce.spermatogonia)$Symbol), 
                       rownames(sce_RA)),]

rownames(sce_RA) <- rownames(sce.spermatogonia)
rowData(sce_RA)$Symbol <- rowData(sce.spermatogonia)$Symbol
rowData(sce_RA)$ID <- rowData(sce.spermatogonia)$ID

# Select specific cell types 
sce_early_RA <- sce_RA[,grepl("A1|ln|TypeBS|G1|TypeBG2M|ePL|mPL|lPL", colData(sce_RA)$cell_type)]
sce_early_RA <- normalize(sce_early_RA)

sce.single <- split.sce(sce = sce.spermatogonia, groups = unique(colData(sce.spermatogonia)$Library), 
                        colData.name = "Library")
sce.single$RA <- sce_early_RA

corrected <- batch.correction(sce.single)

# Visualize PCA 
pca <- prcomp(t(corrected))

# Plot sample distribution
sample <- ggplot() +
  geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                  PC2 = pca$x[1:ncol(sce.spermatogonia),2],
                  sample = sce.spermatogonia$Sample),
             aes(PC1, PC2, colour = sample)) +
            scale_colour_manual(values = c("dark blue", "coral"))

#Cluster spermatogonia
set.seed(1234)
g <- buildSNNGraph(corrected[,1:ncol(sce.spermatogonia)], k = 15, pc.approx = TRUE)
clusters <- igraph::cluster_louvain(g)$membership

clusters[clusters == 1] <- "Diff_In_B"
clusters[clusters == 2] <- "Undiff_2"
clusters[clusters == 3] <- "Diff_A"
clusters[clusters == 4] <- "Undiff_1"
clusters[clusters == 5] <- "Undiff_to_Diff"
clusters[clusters == 6] <- "pL"
clusters <- factor(clusters,
                   levels = c("Undiff_1", "Undiff_2", "Undiff_to_Diff",
                              "Diff_A", "Diff_In_B", "pL"))

col_vector <- c("Undiff_1" = "#0D3D21",
                "Undiff_2" = "#14713E",
                "Undiff_to_Diff" = "#41AE76",
                "Diff_A" = "#52E8A4",
                "Diff_In_B" = "#A3EFCE",
                "pL" = "#E5DFEA")

colour_vector_RA <- c("A1" = "#67001f",
                      "TypeBS" = "#49006a",
                      "ePL" = "#800026", 
                      "G1" = "#ae017e",        
                      "mPL" = "#fc4e2a",
                      "TypeBG2M" = "#e7298a",
                      "lPL" = "#fed976",
                      "ln" = "#fde0dd")

col_vector <- c(col_vector, colour_vector_RA)

# Calculate principal curve
pdf("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S5/Principal_curve.pdf")
prank <- PT(rd = pca$x[1:ncol(sce.spermatogonia),1:3], 
            clusters = clusters, col_vector = col_vector)
dev.off()

# Calculate monocle
pd <- new("AnnotatedDataFrame", data = data.frame(clusters = clusters))
fd <- new("AnnotatedDataFrame", data = data.frame(gene_short_name = rownames(corrected),
                                                  row.names = rownames(corrected)))
cds <- newCellDataSet(corrected[,1:ncol(sce.spermatogonia)], 
                      phenoData = pd, featureData = fd)
cds$Size_Factor <- rep(1, ncol(sce.spermatogonia))
cds <- reduceDimension(cds)
cds <- orderCells(cds)

# Rank cells based on PT
monocle.rank <- rank(cds$Pseudotime)

# Compare to principal curve
comp.monocle.PC <- ggplot(data.frame(PC.rank = prank[,"rank"],
                  monocle.rank = monocle.rank,
                  cluster = clusters)) +
  geom_point(aes(PC.rank, monocle.rank, colour = cluster)) +
  scale_color_manual(values = col_vector)
ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S5/Comp_Diffusionmap_PC.pdf", comp.monocle.PC)
```

# Visualize hand picked marker genes in heatmap of P10 and P15 samples

```{r}
cur_genes <- c("Id4", "Gfra1", "Lhx1", "Egr2", "Etv5", "Nanos2", "Ret", "Eomes",
               "Neurog3", "Rarg", "Nanos3", "Lin28a", "Upp1")

for_heatmap <- logcounts(sce.spermatogonia)[match(cur_genes, rowData(sce.spermatogonia)$Symbol),]
colnames(for_heatmap) <- paste(colData(sce.spermatogonia)$Library,
  colData(sce.spermatogonia)$Barcode, sep = "_")
rownames(for_heatmap) <- cur_genes

# Calculate the ratio of SSC marker and differetiation markers for each cell
dat <- expand.grid(blue=seq(0, 100, by=1), red=seq(0, 100, by=1))
dat <- within(dat, mix <- rgb(green=rev(blue), 
                              red=rev(red), blue=rev(blue), maxColorValue=100))
      
scaling <- function(x){(x-min(x))/(max(x)-min(x))}
      
SCC.vec <- apply(logcounts(sce.spermatogonia)[rowData(sce.spermatogonia)$Symbol %in% cur_genes[1:8],], 2, function(n){sum(n > 0)})
SCC.vec <- scaling(SCC.vec)
      
Diff.vec <- apply(logcounts(sce.spermatogonia)[rowData(sce.spermatogonia)$Symbol %in% cur_genes[9:13],], 2, function(n){sum(n > 0)})
Diff.vec <- scaling(Diff.vec)
        
col = vector(length = length(SCC.vec))
for(i in 1:length(SCC.vec)){
  col[i] <- dat$mix[dat$blue == round(SCC.vec[i]*100, digits = 0) & 
                      dat$red == round(Diff.vec[i]*100, digits = 0)]  
}

# Visualize the cell state on PCA
cur_pca <- pca$x[1:ncol(sce.spermatogonia),]
cur_plot <- ggplot() + 
  geom_point(data = data.frame(PC1 = cur_pca[,1], 
                               PC2 = cur_pca[,2]),
             aes(PC1, PC2), colour = "black", size = 2.5) +
  geom_point(data = data.frame(PC1 = cur_pca[SCC.vec - Diff.vec <= 0,1], 
                               PC2 = cur_pca[SCC.vec - Diff.vec <= 0,2]),
             aes(PC1, PC2), colour = col[SCC.vec - Diff.vec <= 0]) +
  geom_point(data = data.frame(PC1 = cur_pca[SCC.vec - Diff.vec > 0,1], 
                               PC2 = cur_pca[SCC.vec - Diff.vec > 0,2]),
             aes(PC1, PC2), colour = col[SCC.vec - Diff.vec > 0])

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S5/Stem_cells.pdf",cur_plot)

# Save the colour scale
colour_scale <- ggplot(dat, aes(x=red, y=blue)) + 
      geom_tile(aes(fill=mix), color="white") + 
      scale_fill_identity() + xlab("Progenitor markers") + ylab("Stem markers")
ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S5/Colour_scale.pdf", colour_scale)

# Visualize a clustering on these marker genes
pdf(file = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S5/Marker_heatmap.pdf", width = 12, height = 5)
pheatmap(for_heatmap, cluster_rows = FALSE,
         annotation_col = data.frame(row.names = colnames(for_heatmap),
                                     cluster = cur_cluster), 
         show_colnames = FALSE,
         color = viridis(100), annotation_colors = list(cluster = col_vector))
dev.off()
```

# Visualize spermatogonia from P5

```{r}
spermatogonia.P5 <- sce[,grepl("P5", colData(sce)$Sample) &
                                       colData(sce)$AnnotatedClusters == "Spermatogonia"]
spermatogonia.P5 <- spermatogonia.P5[rowMeans(counts(spermatogonia.P5)) > 0,]
spermatogonia.P5 <- normalize(spermatogonia.P5)

# Visualize non-batch corrected data
HVgenes <- HVG(sce = spermatogonia.P5)
pca.P5 <- prcomp(t(logcounts(spermatogonia.P5)[HVgenes,]))
ggplot(data.frame(PC1 = pca.P5$x[,1],
                  PC2 = pca.P5$x[,2],
                  libraries = colData(spermatogonia.P5)$Library)) +
  geom_point(aes(PC1, PC2, colour =  libraries)) + 
  scale_color_manual(values = c( "#FF9100", "#e31a1c"))

# We don't need to batch correct the P5 samples since there is no batch effect
set.seed(123)
g <- buildSNNGraph(logcounts(spermatogonia.P5)[HVgenes,], k = 15, 
                   pc.approx = TRUE)
clusters <- igraph::cluster_louvain(g)$membership

# Change cluster labels
clusters[clusters == 1] <- "Diff"
clusters[clusters == 2] <- "Undiff_2"
clusters[clusters == 3] <- "Prospermatogonia"
clusters[clusters == 4] <- "Undiff_1"

col_vector <- c("Prospermatogonia" = "#7A9606",
                "Undiff_1" = "#0D3D21",
                "Undiff_2" = "#076E38",
                "Diff" = "#5FA37D")

p5.cluster <- ggplot(data.frame(PC1 = pca.P5$x[,1],
                  PC2 = pca.P5$x[,2],
                  clusters = as.factor(clusters))) +
            geom_point(aes(PC1, PC2, colour = clusters)) + 
            scale_colour_manual(values = col_vector)

# Plot marker gene expression
gene <- "Gfra1"
cur_expr <- logcounts(spermatogonia.P5)[rowData(spermatogonia.P5)$Symbol == gene,] > 0
Gfra1.p5 <- ggplot() +
  geom_point(data = data.frame(PC1 = pca.P5$x[!cur_expr,1],
                                 PC2 = pca.P5$x[!cur_expr,2],
                                gene = logcounts(spermatogonia.P5)[
                                    rowData(spermatogonia.P5)$Symbol == gene,!cur_expr]), 
                 aes(PC1, PC2, colour = gene)) +
  geom_point(data = data.frame(PC1 = pca.P5$x[cur_expr,1],
                                 PC2 = pca.P5$x[cur_expr,2],
                                gene = logcounts(spermatogonia.P5)[
                                    rowData(spermatogonia.P5)$Symbol == gene,cur_expr]), 
                 aes(PC1, PC2, colour = gene)) + scale_colour_viridis()

gene <- "Stra8"
cur_expr <- logcounts(spermatogonia.P5)[rowData(spermatogonia.P5)$Symbol == gene,] > 0
Stra8.p5 <- ggplot() +
  geom_point(data = data.frame(PC1 = pca.P5$x[!cur_expr,1],
                                 PC2 = pca.P5$x[!cur_expr,2],
                                gene = logcounts(spermatogonia.P5)[
                                    rowData(spermatogonia.P5)$Symbol == gene,!cur_expr]), 
                 aes(PC1, PC2, colour = gene)) +
  geom_point(data = data.frame(PC1 = pca.P5$x[cur_expr,1],
                                 PC2 = pca.P5$x[cur_expr,2],
                                gene = logcounts(spermatogonia.P5)[
                                    rowData(spermatogonia.P5)$Symbol == gene,cur_expr]), 
                 aes(PC1, PC2, colour = gene)) + scale_colour_viridis()


gene <- "Eif2s2"
cur_expr <- logcounts(spermatogonia.P5)[rowData(spermatogonia.P5)$Symbol == gene,] > 0
Eif2s2.p5 <- ggplot() +
  geom_point(data = data.frame(PC1 = pca.P5$x[!cur_expr,1],
                                 PC2 = pca.P5$x[!cur_expr,2],
                                gene = logcounts(spermatogonia.P5)[
                                    rowData(spermatogonia.P5)$Symbol == gene,!cur_expr]), 
                 aes(PC1, PC2, colour = gene)) +
  geom_point(data = data.frame(PC1 = pca.P5$x[cur_expr,1],
                                 PC2 = pca.P5$x[cur_expr,2],
                                gene = logcounts(spermatogonia.P5)[
                                    rowData(spermatogonia.P5)$Symbol == gene,cur_expr]), 
                 aes(PC1, PC2, colour = gene)) + scale_colour_viridis()


spermatogonia.markers.P5 <- marker.detection(spermatogonia.P5, 
                                          clusters)
write.xlsx(spermatogonia.markers.P5, "../../../../Dropbox (Personal)/Tc1_meiotic_silencing/Revisions/Figures/Supplemental Tables/Table_S5.xlsx")


final_legend <- plot_grid(get_legend(p5.cluster), get_legend(Gfra1.p5) , get_legend(Stra8.p5), get_legend(Eif2s2.p5), ncol = 4 , nrow = 1)
final_plot <- plot_grid(p5.cluster + theme(legend.position= "NONE"), Gfra1.p5 + theme(legend.position= "NONE"), Stra8.p5 + theme(legend.position= "NONE"), Eif2s2.p5 + theme(legend.position= "NONE"), ncol = 4, nrow = 1)
final <- plot_grid(final_plot, final_legend, ncol = 1, nrow = 2, rel_heights = c(2,1))

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S5/Fig_S5.pdf",
       final, width = 15, height = 5)
```