---
title: "Figure S4"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S4.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script recapitulates the visualization of somatic cells.

# Load data and libraries

```{r data, message=FALSE}
# Libraries
library(scater)
library(ggplot2)
library(reshape2)
library(plyr)
library(openxlsx)
library(cowplot)
library(RColorBrewer)
library(pheatmap)
library(viridis)
library(Rtsne)
source("../../Functions/auxiliary.R")

# Single cell data
sce.all <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")

sce.all <- sce.all[,!(colData(sce.all)$Sample %in% c("Tc1", "Tc0")) & 
                     colData(sce.all)$AnnotatedClusters != "Outliers"]
```

# Visualize somatic cells from P5 and P10 sample before batch correction

```{r}
sce.somatic <- sce.all[,grepl("P5|P10", colData(sce.all)$Sample) &
                                       colData(sce.all)$AnnotatedClusters != "Spermatogonia" &
                                        colData(sce.all)$AnnotatedClusters != "Spermatocytes_1"]
sce.somatic <- normalize(sce.somatic)

# Compute PCA
pca.somatic <- runPCA(sce.somatic)
ggplot(data.frame(PC1 = reducedDims(pca.somatic)$PCA[,1],
                  PC2 = reducedDims(pca.somatic)$PCA[,2],
                  samples = colData(pca.somatic)$Sample)) +
  geom_point(aes(PC1, PC2, colour = samples)) + 
  scale_color_manual(values = c("#1f78b4", "#e31a1c"))

ggplot(data.frame(PC1 = reducedDims(pca.somatic)$PCA[,1],
                  PC2 = reducedDims(pca.somatic)$PCA[,2],
                  cell_type = colData(pca.somatic)$AnnotatedClusters)) +
  geom_point(aes(PC1, PC2, colour = cell_type)) + 
  scale_color_manual(values = metadata(sce.somatic)$color_vector)

# Compute tSNE
set.seed(123)
tsne.somatic <- runTSNE(sce.somatic)
# Visualize samples
tsne.sample.before <- ggplot(data.frame(tsne1 = reducedDims(tsne.somatic)$TSNE[,1],
                  tsne2 = reducedDims(tsne.somatic)$TSNE[,2],
                  samples = colData(tsne.somatic)$Sample)) +
  geom_point(aes(tsne1, tsne2, colour = samples), size = 0.5) + 
  scale_color_manual(values = c("P5"= "#e31a1c", "P10"="#1f78b4"))

# Visualize individual libraries
ggplot(data.frame(tsne1 = reducedDims(tsne.somatic)$TSNE[,1],
                  tsne2 = reducedDims(tsne.somatic)$TSNE[,2],
                  library = colData(tsne.somatic)$Library)) +
    geom_point(aes(tsne1, tsne2, colour = library), size = 0.5) + 
    scale_color_manual(values = c("#1f78b4", "#FF9100", "#e31a1c"))
# Visualize cell types
tsne.clusters.before <-ggplot(data.frame(tsne1 = reducedDims(tsne.somatic)$TSNE[,1],
                  tsne2 = reducedDims(tsne.somatic)$TSNE[,2],
                  cell_type = colData(sce.somatic)$AnnotatedClusters)) +
    geom_point(aes(tsne1, tsne2, colour = cell_type), size = 0.5) + 
    scale_color_manual(values = metadata(sce.somatic)$color_vector)
# Visualize gene expression
ggplot(data.frame(tsne1 = reducedDims(tsne.somatic)$TSNE[,1],
                  tsne2 = reducedDims(tsne.somatic)$TSNE[,2],
                  gene = logcounts(sce.somatic)[rowData(sce.somatic)$Symbol == "Igfbp4",]))               + geom_point(aes(tsne1, tsne2, colour = gene), size = 0.5) +
    scale_color_viridis()

# Differential expression between somatic cells types from P5 and P10
genenames <- read.table("../../Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
rownames(genenames) <- genenames$Gene.stable.ID

results.DE <- list()
cur_groups <- c("Sertoli", "Leydig_1", "Leydig_2", 
                "Fetal_Leydig_1", "Fetal_Leydig_2", 
                "PTM", "Endothelial_cells", 
                "Interstitial_tMg")
for(i in cur_groups){
  sce.test <- sce.somatic[,colData(sce.somatic)$AnnotatedClusters == i]
  sce.test <- sce.test[Matrix::rowMeans(logcounts(sce.test)) > 0,]
  sce.test <- normalize(sce.test)
  
  # Sum counts with each batch and group
  mat <- matrix(data = NA, ncol = length(unique(colData(sce.test)$Library)), 
                nrow = nrow(counts(sce.test)))
  rownames(mat) <- rownames(counts(sce.test))
  colnames(mat) <- unique(paste(colData(sce.test)$Sample, 
                                colData(sce.test)$Library, sep = "_"))
  
  for(j in colnames(mat)){
    cur_batch <- unlist(strsplit(j, "_"))[2]
    mat[,j] <- Matrix::rowSums(counts(sce.test)[,colData(sce.test)$Library == cur_batch]) 
  }
  
  # Perform differential testing
  y <- DGEList(counts=mat,
               group=sapply(colnames(mat), 
                            function(n){unlist(strsplit(n, "_"))[1]}))
  y <- calcNormFactors(y)
  design <- model.matrix(~0+sapply(colnames(mat), function(n){unlist(strsplit(n, "_"))[1]}))
  colnames(design) <- c("P10", "P5")
  y <- estimateDisp(y,design)
  
  fit <- glmQLFit(y,design, robust = TRUE)
  qlf <- glmTreat(fit,coef=2, lfc = 0.5, 
                  contrast = makeContrasts(P5 - P10, levels = design))
  cur_markers <- topTags(qlf, n = nrow(qlf$table))$table
  
  markers.P5 <- cur_markers[cur_markers$FDR < 0.1 & cur_markers$logFC > 0,]
  markers.P5$Genename <- rowData(sce.test)$Symbol[match(rownames(markers.P5),
                                                                 rowData(sce.test)$ID)]
  markers.P10 <- cur_markers[cur_markers$FDR < 0.1 & cur_markers$logFC < 0,]
  markers.P10$Genename <- rowData(sce.test)$Symbol[match(rownames(markers.P10),
                                                                 rowData(sce.test)$ID)]
  results.DE[[paste("Group_", i, "_P5", sep = "")]] <- markers.P5
  results.DE[[paste("Group_", i, "_P10", sep = "")]] <- markers.P10
}

write.xlsx(results.DE, "../../../../Dropbox/Tc1_meiotic_silencing/Revisions/Figures/Supplemental Tables/Table_S4.xlsx")
```

# Visualize batch corrected data

```{r}
# Batch correct the data
sce.single <- split.sce(sce = sce.somatic, 
                        groups = unique(colData(sce.somatic)$Library), 
                        colData.name = "Library")

# Use P10 as first sample
corrected <- batch.correction(sce.single[c("do17821", "do26386", "do26387")])

# Reassemble the sce object
sce.somatic <- do.call("cbind", sce.single[c("do17821", "do26386", "do26387")])
rm(sce.single)

# Recompute the tSNE of batch corrected data
set.seed(12345)
tSNE.somatic.after <- Rtsne(t(corrected))

# Plot sample distribution across somatic cells
sample_col_vector <- c("P5"= "#e31a1c", "P10"="#1f78b4")
tsne.sample.after <- ggplot(data = data.frame(tSNE1 = tSNE.somatic.after$Y[,1],
                             tSNE2 = tSNE.somatic.after$Y[,2],
                             sample = colData(sce.somatic)$Sample)) +     
                            geom_point(aes(tSNE1, tSNE2, colour = sample), size = 0.5) + 
                          scale_color_manual(values = sample_col_vector)

# Plot different somatic cell types
tsne.clusters.after <- ggplot(data = data.frame(tSNE1 = tSNE.somatic.after$Y[,1],
                             tSNE2 = tSNE.somatic.after$Y[,2],
                             cell_type = colData(sce.somatic)$AnnotatedClusters)) +     
  geom_point(aes(tSNE1, tSNE2, colour = cell_type), size = 0.5) + 
  scale_color_manual(values = metadata(sce.somatic)$color_vector)

# Visualize gene expression
ggplot(data.frame(tsne1 = tSNE.somatic.after$Y[,1],
                  tsne2 = tSNE.somatic.after$Y[,2],
                  gene = logcounts(sce.somatic)[rowData(sce.somatic)$Symbol == "Igfbp4",]))               + geom_point(aes(tsne1, tsne2, colour = gene), size = 0.5) +
    scale_color_viridis()

# Combine plots
final <- plot_grid(tsne.sample.before, tsne.clusters.before + theme(legend.position = c(0.5, 0.9)),
                   tsne.sample.after, tsne.clusters.after, ncol = 2)

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S4/Fig_S4.pdf",
       final, width = 10, height = 10)
```

# Collect marker genes for somatic cells in P10 and P5 

```{r}
# Find marker genes
somatic.markers <- marker.detection(sce.somatic,
                                    as.character(colData(sce.somatic)$AnnotatedClusters), 
                                    blocking = sce.somatic$Sample)

# Save this list as Table S1
write.xlsx(somatic.markers, "../../../../Dropbox/Tc1_meiotic_silencing/Revisions/Figures/Supplemental Tables/Table_S3.xlsx")
```

# Visualize marker genes

```{r}
# Calculate distance on logcounts
top.markers <- as.character(unlist(lapply(somatic.markers,
                                          function(n){n$GeneName[1:5]})))

# Genes to visualize
genes <- c(#Leydig_1
           "Insl3", top.markers[c(21:23, 25)],
           #Leydig_2
           "Igfbp4", top.markers[c(26, 28:30)],
          #Immature_Leydig1
          "Inhba", top.markers[6:9],
          #Immature_Leydig2
          "Stmn1", top.markers[c(11, 13:15)],
          #Sertoli
           "Gstm6", top.markers[c(36:39)],
          #Macrophages
           "Fcer1g", top.markers[c(16:18, 20)],
          #PTM
           "Acta2", top.markers[c(32:35)],
           #Endothelial cells
           "Col4a1", top.markers[c(1:3, 5)])

#genes <- c("Cst12", top.markers[26:29],  
#           "Sycp1", top.markers[32:35],
#           "Dmrt1", top.markers[c(36,38:40)],
#           "Dlk1", top.markers[6:9],
#           "Col4a3", top.markers[21:24],
#           "Tm4sf1", top.markers[1:4],
#           "Insl3", top.markers[c(16:17, 19,20)],
#           "Cd14", top.markers[11:14])

# Heatmap
# Visualize in heatmap
for.heatmap <- logcounts(sce.somatic)[match(genes, rowData(sce.somatic)$Symbol),]
rownames(for.heatmap) <- genes
colnames(for.heatmap) <- paste(colData(sce.somatic)$Barcode, 
                               colData(sce.somatic)$Library,
                               colData(sce.somatic)$Sample,
                               colData(sce.somatic)$AnnotatedClusters, sep = "__")

# Order cells
for.heatmap <- for.heatmap[,c(which(sce.somatic$AnnotatedClusters == "Leydig_1"),
                              which(sce.somatic$AnnotatedClusters == "Leydig_2"),
                              which(grepl("Fetal_Leydig_1", sce.somatic$AnnotatedClusters)),
                              which(grepl("Fetal_Leydig_2", sce.somatic$AnnotatedClusters)),
                              which(grepl("Sertoli", sce.somatic$AnnotatedClusters)),
                              which(grepl("Interstitial_tMg", sce.somatic$AnnotatedClusters)),
                              which(grepl("PTM", sce.somatic$AnnotatedClusters)),
                              which(grepl("Endothelial_cells", sce.somatic$AnnotatedClusters)))]

pdf("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S4/Somatic_markers_heatmap.pdf", 
    onefile = FALSE, width = 7, height = 10)
pheatmap(for.heatmap, cluster_rows = FALSE, cluster_cols = FALSE, 
         annotation_col = 
           data.frame(row.names = colnames(for.heatmap),
                      cell_type = sapply(colnames(for.heatmap), function(n){unlist(strsplit(n, "__"))[4]}),
                      sample = sapply(colnames(for.heatmap), function(n){unlist(strsplit(n, "__"))[3]})),
         annotation_row = 
           data.frame(row.names = rownames(for.heatmap),
                      cell_type = rep(c("Leydig_1", "Leydig_2", "Fetal_Leydig_1", "Fetal_Leydig_2",
                                        "Sertoli", "Interstitial_tMg", "PTM", "Endothelial_cells"), each = 5)),
         color = viridis(100),
         show_colnames = FALSE, 
         annotation_colors = list(cell_type = 
      metadata(sce.somatic)$color_vector[names(metadata(sce.somatic)$color_vector) %in%
                                       unique(colData(sce.somatic)$AnnotatedClusters)]),
      cellheight = 8, fontsize = 7, gaps_row = seq(5, 35, 5))
dev.off()
```

# Heatmap of genes differentially expressed between P5 and P10

```{r}
cur_genes <- c(rownames(results.DE$Group_Leydig_1_P5)[1:5], 
               rownames(results.DE$Group_Leydig_1_P10)[1:5],
               rownames(results.DE$Group_Leydig_2_P5)[1:5],
               rownames(results.DE$Group_Leydig_2_P10)[1:5],
               rownames(results.DE$Group_Fetal_Leydig_1_P5)[1:5],
               rownames(results.DE$Group_Fetal_Leydig_1_P10)[1:5],
               rownames(results.DE$Group_Sertoli_P5)[1:5], 
               rownames(results.DE$Group_Sertoli_P10)[1:5],
               rownames(results.DE$Group_Interstitial_tMg_P5)[1:5],
               rownames(results.DE$Group_Interstitial_tMg_P10)[1:5],
               rownames(results.DE$Group_PTM_P5)[1:5],
               rownames(results.DE$Group_PTM_P10)[1:5],
               rownames(results.DE$Group_Endothelial_cells_P5)[1:5],
               rownames(results.DE$Group_Endothelial_cells_P10)[1:5])

for.heatmap <- logcounts(sce.somatic)[cur_genes,]
rownames(for.heatmap) <- seq(1, nrow(for.heatmap))
colnames(for.heatmap) <- paste(colData(sce.somatic)$Barcode, 
                               colData(sce.somatic)$Library,
                               colData(sce.somatic)$Sample,
                               colData(sce.somatic)$AnnotatedClusters, sep = "__")

# Order cells
for.heatmap <- for.heatmap[,c(which(sce.somatic$AnnotatedClusters == "Leydig_1"),
                              which(sce.somatic$AnnotatedClusters == "Leydig_2"),
                              which(grepl("Fetal_Leydig_1", sce.somatic$AnnotatedClusters)),
                              which(grepl("Fetal_Leydig_2", sce.somatic$AnnotatedClusters)),
                              which(grepl("Sertoli", sce.somatic$AnnotatedClusters)),
                              which(grepl("Interstitial_tMg", sce.somatic$AnnotatedClusters)),
                              which(grepl("PTM", sce.somatic$AnnotatedClusters)),
                              which(grepl("Endothelial_cells", sce.somatic$AnnotatedClusters)))]

pdf("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S4/Somatic_DE_heatmap.pdf", 
    onefile = FALSE, width = 7, height = 10)
pheatmap(for.heatmap, cluster_rows = FALSE, cluster_cols = FALSE, 
         labels_row = rowData(sce.somatic)$Symbol[match(cur_genes, rowData(sce.somatic)$ID)],
         annotation_col = 
           data.frame(row.names = colnames(for.heatmap),
                      cell_type = sapply(colnames(for.heatmap), function(n){unlist(strsplit(n, "__"))[4]}),
                      sample = sapply(colnames(for.heatmap), function(n){unlist(strsplit(n, "__"))[3]})),
         annotation_row = 
           data.frame(row.names = rownames(for.heatmap),
                      cell_type = rep(c("Leydig_1", "Leydig_2", "Fetal_Leydig_1",
                                        "Sertoli", "Interstitial_tMg", "PTM", "Endothelial_cells"), each = 10),
                      sample = rep(rep(c("P5", "P10"), each = 5), 7)),
         color = viridis(100),
         show_colnames = FALSE, 
         annotation_colors = list(cell_type = 
      metadata(sce.somatic)$color_vector[names(metadata(sce.somatic)$color_vector) %in%
                                       unique(colData(sce.somatic)$AnnotatedClusters)]),
      cellheight = 8, fontsize = 7, gaps_row = seq(10, 70, 10))
dev.off()
```