---
title: "Figure S9"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S9.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load libraries

```{r data, message=FALSE}
library(scater)
library(plyr)
library(reshape2)
library(pheatmap)
library(viridis)
library(plot3D)
library(ggsci)
library(cowplot)
library(openxlsx)
library(VennDiagram)
source("../../Functions/auxiliary.R")

sce <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")
sce <- sce[,grepl("B6", colData(sce)$Sample) & 
             colData(sce)$AnnotatedClusters %in% levels(colData(sce)$AnnotatedClusters)[1:20]]
sce <- normalize(sce)

# Read in gene annotations
genenames <- read.table("../../Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
rownames(genenames) <- genenames$Gene.stable.ID
```

# X chromosomal expression in P15 sample of emptyDrops

```{r}
# Single cell data
sce.emptyDrops <- readRDS("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")

sce.emptyDrops <- sce.emptyDrops[,colData(sce.emptyDrops)$Sample == "P15"]
sce.emptyDrops <- sce.emptyDrops[,
                grepl("Pachytene|Differentiating|Zygotene|Leptotene|Undifferentiated", 
                          colData(sce.emptyDrops)$P15Clusters) &
                 colData(sce.emptyDrops)$AnnotatedClusters != "Outliers" &
             colData(sce.emptyDrops)$AnnotatedClusters != "Sertoli"]

# Select same genes as in Figure 7
genes <- rowData(sce)$ID[
  apply(logcounts(sce)[,colData(sce)$AnnotatedClusters == "Spermatogonia"], 1,
              function(n){length(which(n > 0))}) >
    0.3*sum(colData(sce)$AnnotatedClusters == "Spermatogonia") |
    apply(logcounts(sce)[,colData(sce)$AnnotatedClusters %in% 
          paste("S", 1:11, sep = "")], 1,
              function(n){length(which(n > 0))}) >
    0.3*sum(colData(sce)$AnnotatedClusters %in% 
              paste("S", 1:11, sep = ""))]

# Collect genes for some chromosomes
genes.X <- genes[genes %in% genenames[genenames[,3] == "X",1]]
genes.Y <- genes[genes %in% genenames[genenames[,3] == "Y",1]]
genes.9 <- genes[genes %in% genenames[genenames[,3] == "9",1]]
genes.A <- genes[genes %in% genenames[genenames[,3] != "X" & 
                                        genenames[,3] != "Y" &
                                        genenames[,3] != "MT",1]]

# Calculate X:A ratio and 9:X ratio
ratio.X.A <- colMeans(as.matrix(logcounts(sce.emptyDrops)[genes.X,]))/
  colMeans(as.matrix(logcounts(sce.emptyDrops)[genes.A,]))
ratio.X.A[is.nan(ratio.X.A)] <- 0

ratio.Y.A <- colMeans(as.matrix(logcounts(sce.emptyDrops)[genes.Y,]))/
  colMeans(as.matrix(logcounts(sce.emptyDrops)[genes.A,]))
ratio.Y.A[is.nan(ratio.Y.A)] <- 0

ratio.9.A <- colMeans(as.matrix(logcounts(sce.emptyDrops)[genes.9,]))/
  colMeans(as.matrix(logcounts(sce.emptyDrops)[genes.A,]))
ratio.9.A[is.nan(ratio.9.A)] <- 0

# Order cells in pseutotime
HVgenes <- HVG(sce.emptyDrops)
pca <- prcomp(t(logcounts(sce.emptyDrops)[HVgenes,]))
prank <- PT(rd = pca$x[,1:10], clusters = colData(sce.emptyDrops)$P15Clusters, 
            col_vector = metadata(sce.emptyDrops)$colour_vector.P15, )

# Build mean ratio matrix
df <- data.frame(ratios = c(ratio.X.A, ratio.Y.A, ratio.9.A),
                 groups = factor(c(as.character(colData(sce.emptyDrops)$P15Clusters), 
                            as.character(colData(sce.emptyDrops)$P15Clusters),
                            as.character(colData(sce.emptyDrops)$P15Clusters)), 
                            levels = levels(colData(sce.emptyDrops)$P15Clusters)),
                 chr = c(rep("X", length(ratio.X.A)), rep("Y", length(ratio.Y.A)),
                         rep("9", length(ratio.9.A))),
                 PT = c(prank[,"rank"], prank[,"rank"], prank[,"rank"]))

# Plot three different plots
X.chr <- ggplot(df[df$chr == "X",]) + geom_point(aes(PT, ratios, color = groups)) + 
  scale_colour_manual(values = metadata(sce.emptyDrops)$colour_vector.P15) +
  geom_smooth(aes(PT, ratios), colour = "blue") + scale_x_reverse() +
  ylim(c(0,1.5))

Y.chr <- ggplot(df[df$chr == "Y",]) + geom_point(aes(PT, ratios, color = groups)) + 
  scale_colour_manual(values = metadata(sce.emptyDrops)$colour_vector.P15) +
  geom_smooth(aes(PT, ratios), colour = "red") + scale_x_reverse()

chr.9 <- ggplot(df[df$chr == "9",]) + geom_point(aes(PT, ratios, color = groups)) + 
  scale_colour_manual(values = metadata(sce.emptyDrops)$colour_vector.P15) +
  geom_smooth(aes(PT, ratios), colour = "green") + scale_x_reverse() +
  ylim(c(0,1.5))

all <- plot_grid(X.chr + theme(legend.position = "none"),
                 Y.chr + theme(legend.position = "none"),
                 chr.9 + theme(legend.position = "none"), ncol = 3)
ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S9/Sex_chromosome_dynamics.pdf",all, width = 20, height = 5)
```

# Multicopy gene analysis

# Read in the BLAST hits for the different multicopy genes

```{r}
files <- as.list(list.files("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Multicopy/Mapping", full.names = TRUE))
names(files) <- sapply(list.files("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Multicopy/Mapping"), function(n){unlist(strsplit(n, "\\."))[1]})

# Filter genes that lay on chromosome X
mc_gene_table <- lapply(files, function(n){
  cur_file <- read.csv(n, stringsAsFactors = FALSE)
  cur_file <- cur_file[sapply(cur_file[,"Genomic.Location"], function(n){unlist(strsplit(n, "\\:"))[1]}) == "X",]
  cur_file
})

# Return gene names
mc_genes <- lapply(mc_gene_table, function(n){
  cur_genes <- n$Overlapping.Gene.s.
  cur_genes <- cur_genes[cur_genes != ""]
  unique(cur_genes)
})

df.multi <- data.frame(gene_group = rep(names(mc_genes), as.numeric(sapply(mc_genes, length))),
                 genes = as.character(unlist(sapply(mc_genes, function(n){n}))), 
                 stringsAsFactors = FALSE)
```

# Visualize metagenes of multicopy genes

```{r}
# Collect normalized expression counts of multicopy genes
m <- match(as.character(df.multi$genes), rowData(sce)$Symbol)
df.multi <- df.multi[!is.na(m),]
for.heatmap <- logcounts(sce)[m[!is.na(m)],]
rownames(for.heatmap) <- rowData(sce)$Symbol[match(rownames(for.heatmap),
                                                   rowData(sce)$ID)]

mat.meta <- matrix(data = NA, ncol = ncol(for.heatmap), nrow = length(unique(df.multi$gene_group)))
colnames(mat.meta) <- paste(colData(sce)$Library, colData(sce)$Barcode, sep = "")
rownames(mat.meta) <- unique(df.multi$gene_group)

for(i in unique(df.multi$gene_group)){
  if(sum(df.multi$gene_group == i) > 1){
    mat.meta[i,] <- Matrix::colSums(2^((for.heatmap)[df.multi$gene_group == i,]) - 1)
  }
  else{
    mat.meta[i,] <- 2^((for.heatmap)[df.multi$gene_group == i,]) - 1
  }
}

# Average this per cell type

# Build mean expression matrix
df <- as.data.frame(t(as.matrix(mat.meta)))
df$groups <- colData(sce)$AnnotatedClusters
df.melt <- melt(df, id.vars = "groups")

# Collect mean expression for each gene in each group
mat <- ddply(df.melt, .(groups, variable), summarize, mean=mean(value))
mat.for.heatmap <- dcast(data = mat,formula = variable~groups,fun.aggregate = sum,value.var = "mean")
rownames(mat.for.heatmap) <- mat.for.heatmap$variable
mat.for.heatmap <- mat.for.heatmap[,-1]

pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S7/MultiCopy_meta.pdf",   width = 7, height = 5)
pheatmap(mat.for.heatmap, show_colnames = FALSE, cluster_cols = FALSE, 
         cluster_rows = FALSE,
         color = viridis(100),
         annotation_col = data.frame(row.names = colnames(mat.for.heatmap),
                  cell_type = colnames(mat.for.heatmap)),
         annotation_colors = list(cell_type = metadata(sce)$color_vector),
         #scale = "row", 
         border_color = NA, cellheight = 8, fontsize = 7)
dev.off()

```



