---
title: "Figure S10"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S10.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data and libraries

```{r data, message=FALSE}
library(rtracklayer)
library(EnrichedHeatmap)
library(GenomicFeatures)
library(EnsDb.Mmusculus.v79)
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
library(plyr)
library(reshape2)
library(pheatmap)
library(viridis)
library(plot3D)
library(ggsci)
library(cowplot)
library(openxlsx)
library(csaw)
library(edgeR)
library(Rsamtools)
library(Gviz)
source("../../Functions/auxiliary.R")

# Generate feature annotation
prom <- promoters(genes(EnsDb.Mmusculus.v79))
gene.body <- genes(EnsDb.Mmusculus.v79)
prom <- prom[seqnames(prom) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(prom) <- c(as.character(1:19), "X", "Y", "MT")
gene.body <- gene.body[seqnames(gene.body) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(gene.body) <- c(as.character(1:19), "X", "Y", "MT")
prom.X <- prom[seqnames(prom) == "X"]
prom.Y <- prom[seqnames(prom) == "Y"]
prom.9 <- prom[seqnames(prom) == "9"]
gene.body.X <- gene.body[seqnames(gene.body) == "X"]

# K9 files
bam.files.K9 <- list.files("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K9me3/", full.names = TRUE,
                        pattern = paste(".+JP26.+bam$", sep = ""))

# Blacklisted regions
black <- import("http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz")
seqlevels(black) <- sub("chr", "", seqlevels(black))
seqnames(black) <- sub("chr", "", seqnames(black))

# Parameters for reading bam files
param <- readParam(dedup=TRUE, minq=10, discard=black, pe="both", max.frag=1000)

# Read in gene annotations
genenames <- read.table("../../Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
rownames(genenames) <- genenames$Gene.stable.ID

# Create objects for visualization
gax <- GenomeAxisTrack(col="black", fontsize=15, size=2)
gr <- getGeneRegionTrackForGviz(EnsDb.Mmusculus.v79)
options(ucscChromosomeNames = FALSE)

# Count reads in bins across the genome
bins.K9 <- windowCounts(bam.files.K9, bin = TRUE, width = 1000, param=param)
```

# Visualize regions with canonically high K9 signal

```{r}
# Regions of interest
# Cyp3a gene cluster
cur_region <- IRanges(start=145320874, end=146199915,
                      names="5")
cur_region <- GRanges(seqnames = Rle(values = "5", lengths = 1),
                      ranges = cur_region)

collected.K9 <- list()
lib.sizes <- colSums(assays(bins.K9)$counts)/1e6
for (i in 1:length(bam.files.K9)) {
    reads <- extractReads(bam.file=bam.files.K9[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K9[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K9[i],  col.axis="black", col.title="black",
         fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K9 <- collected.K9[c(1,3,2,4)]

pdf("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S10/Cyp3a_cluster_K9.pdf", width = 20, height = 7)
plotTracks(c(gax, c(collected.K9[1:4]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

# Olfactory receptors chr 7
cur_region <- IRanges(start=107696936, end=108953956,
                      names="7")
cur_region <- GRanges(seqnames = Rle(values = "7", lengths = 1),
                      ranges = cur_region)

collected.K9 <- list()
lib.sizes <- colSums(assays(bins.K9)$counts)/1e6
for (i in 1:length(bam.files.K9)) {
    reads <- extractReads(bam.file=bam.files.K9[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K9[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K9[i],  col.axis="black", col.title="black",
         fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K9 <- collected.K9[c(1,3,2,4)]

pdf("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S10/OR_cluster_chr7_K9.pdf", width = 20, height = 7)
plotTracks(c(gax, c(collected.K9[1:4]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

# Olfactory receptors chr 11
cur_region <- IRanges(start=73306044, end=74465623,
                      names="11")
cur_region <- GRanges(seqnames = Rle(values = "11", lengths = 1),
                      ranges = cur_region)

collected.K9 <- list()
lib.sizes <- colSums(assays(bins.K9)$counts)/1e6
for (i in 1:length(bam.files.K9)) {
    reads <- extractReads(bam.file=bam.files.K9[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K9[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K9[i],  col.axis="black", col.title="black",
         fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K9 <- collected.K9[c(1,3,2,4)]

pdf("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S10/OR_cluster_chr11_K9.pdf", width = 20, height = 7)
plotTracks(c(gax, c(collected.K9[1:4]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

# Olfactory receptors chr 16
cur_region <- IRanges(start=58691649, end=59486069,
                      names="16")
cur_region <- GRanges(seqnames = Rle(values = "16", lengths = 1),
                      ranges = cur_region)

collected.K9 <- list()
lib.sizes <- colSums(assays(bins.K9)$counts)/1e6
for (i in 1:length(bam.files.K9)) {
    reads <- extractReads(bam.file=bam.files.K9[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K9[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K9[i],  col.axis="black", col.title="black",
         fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K9 <- collected.K9[c(1,3,2,4)]

pdf("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S10/OR_cluster_chr16_K9.pdf", width = 20, height = 7)
plotTracks(c(gax, c(collected.K9[1:4]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()
```

# Whole chromosome view of K9 signal

```{r}
# Select chromosomal bins
bins.K9 <- windowCounts(bam.files.K9, bin = TRUE, width = 20000,
    param=param)
bins.K9 <- bins.K9[as.character(seqnames(bins.K9@rowRanges)) %in%
                             c(as.character(1:19), "X", "Y")]

cur_df <- data.frame(score = assays(bins.K9)$counts[,1]/(colSums(assays(bins.K9)$counts)[1]/1000000),
                     chr = factor(as.character(seqnames(bins.K9@rowRanges)),
                                  levels = c(as.character(1:19), "X", "Y")),
                     cat = rep(1, nrow(bins.K9)))

p.K9 <- ggplot(cur_df) + geom_bar(stat = "identity", aes(chr, cat, fill = log(score + 1))) +
  scale_fill_gradientn(colours = colorRampPalette(c(c("white", "dark blue", "black")))(100))
ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S10/K9_signal.pdf", p.K9)
```

# Analyse heterogeneity of K9 on X chromosome

```{r}
bins.K9.X <- bins.K9[seqnames(bins.K9) == "X",]

# Plot cumulative frequency
df.K9.X <- data.frame(spermatocytes_1 =
                      cumsum(assays(bins.K9.X)$counts[
                        order(assays(bins.K9.X)$counts[,1], 
                              decreasing = TRUE),1]),
                      spermatocytes_2 =
                      cumsum(assays(bins.K9.X)$counts[
                        order(assays(bins.K9.X)$counts[,3], 
                              decreasing = TRUE),3]),
                      spermatids_1 =
                      cumsum(assays(bins.K9.X)$counts[
                        order(assays(bins.K9.X)$counts[,2], 
                              decreasing = TRUE),2]),
                      spermatids_2 =
                      cumsum(assays(bins.K9.X)$counts[
                        order(assays(bins.K9.X)$counts[,4], 
                              decreasing = TRUE),4]))
df.K9.X <- t(t(df.K9.X)/as.numeric(tail(df.K9.X, 1)))
# Sample 10000 windows 
set.seed(1234)
sam <- sample(1:nrow(df.K9.X), 10000)
sam <- sam[order(sam)]

df.K9.X.melt <- melt(df.K9.X[c(1,sam),])

# Compare to Chr9
bins.K9.9 <- bins.K9[as.vector(seqnames(rowRanges(bins.K9)) == "9")]

# Plot cumulative frequency
df.K9.9 <- data.frame(spermatocytes_1 =
                      cumsum(assays(bins.K9.9)$counts[
                        order(assays(bins.K9.9)$counts[,1], 
                              decreasing = TRUE),1]),
                      spermatocytes_2 =
                      cumsum(assays(bins.K9.9)$counts[
                        order(assays(bins.K9.9)$counts[,3], 
                              decreasing = TRUE),3]),
                      spermatids_1 =
                      cumsum(assays(bins.K9.9)$counts[
                        order(assays(bins.K9.9)$counts[,2], 
                              decreasing = TRUE),2]),
                      spermatids_2 =
                      cumsum(assays(bins.K9.9)$counts[
                        order(assays(bins.K9.9)$counts[,4], 
                              decreasing = TRUE),4]))
df.K9.9 <- t(t(df.K9.9)/as.numeric(tail(df.K9.9, 1)))
set.seed(1234)
sam <- sample(1:nrow(df.K9.9), 10000)
sam <- sam[order(sam)]
df.K9.9.melt <- melt(df.K9.9[c(1,sam),])

cumsum.all <- ggplot() + 
  geom_line(data = df.K9.9.melt, aes(Var1, value, colour = Var2), lty = 3) +
  geom_line(data = df.K9.X.melt, aes(Var1, value, colour = Var2)) + 
  scale_color_manual(values = c("red", "dark red", "blue", "dark blue"))
ggsave(filename = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S10/Cumsum_all.pdf", 
       cumsum.all, width = 7, height = 4)
```


