---
title: "Figure S12"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S12.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script recapitulates the replicated analysis of K4 and K9 marks at P24 and P28 and visualizes temporal peak patterns for Akap4 and Cypt1.
We also analyse the K27 data here.
It also visualizes the enrichment of histone marks from Hammoud et. al. CSC.

# Load data and libraries

```{r data, message=FALSE}
library(rtracklayer)
library(EnrichedHeatmap)
library(GenomicFeatures)
library(EnsDb.Mmusculus.v79)
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
library(plyr)
library(reshape2)
library(pheatmap)
library(viridis)
library(plot3D)
library(ggsci)
library(cowplot)
library(openxlsx)
library(csaw)
library(edgeR)
library(Rsamtools)
library(Gviz)
source("../../Functions/auxiliary.R")

# Read in gene annotations
genenames <- read.table("../Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
rownames(genenames) <- genenames$Gene.stable.ID

# Generate feature annotation
prom <- promoters(genes(EnsDb.Mmusculus.v79))
gene.body <- genes(EnsDb.Mmusculus.v79)
prom <- prom[seqnames(prom) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(prom) <- c(as.character(1:19), "X", "Y", "MT")
gene.body <- gene.body[seqnames(gene.body) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(gene.body) <- c(as.character(1:19), "X", "Y", "MT")
prom.X <- prom[seqnames(prom) == "X"]
prom.Y <- prom[seqnames(prom) == "Y"]
prom.9 <- prom[seqnames(prom) == "9"]
gene.body.X <- gene.body[seqnames(gene.body) == "X"]

# K9 files P24
bam.files.K9.P24 <- list.files("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K9me3/", full.names = TRUE,
                        pattern = paste(".+JP24.+bam$", sep = ""))

# K9 files P28
bam.files.K9.P28 <- list.files("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K9me3/", full.names = TRUE,
                        pattern = paste(".+JP28.+bam$", sep = ""))

# K4 files P24
bam.files.K4.P24 <- list.files("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K4me3/", full.names = TRUE,
                        pattern = paste(".+JP24.+bam$", sep = ""))

# K4 files P28
bam.files.K4.P28 <- list.files("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K4me3/", full.names = TRUE,
                        pattern = paste(".+JP28.+bam$", sep = ""))

# K27 files P24
bam.files.K27.P26 <- list.files("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K27ac/", full.names = TRUE,
                        pattern = paste(".+JP26.+bam$", sep = ""))

# K27 files P28
bam.files.K27.P28 <- list.files("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K27ac/", full.names = TRUE,
                        pattern = paste(".+JP28.+bam$", sep = ""))

# Define conditions
batch <- gl(2, 2)
treatment <- rep(c("spermatocytes", "spermatids"), 2)

# Blacklisted regions
black <- import("http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz")
seqlevels(black) <- sub("chr", "", seqlevels(black))
seqnames(black) <- sub("chr", "", seqnames(black))

# Parameters for reading bam files
param <- readParam(dedup=TRUE, minq=10, discard=black, pe="both", max.frag=1000)

spermatid.specific <- read.xlsx("../../../../Dropbox (Personal)/Tc1_meiotic_silencing/Revisions/Figures/Supplemental Tables/Table_S8_bulkDetection.xlsx")
rownames(spermatid.specific) <- genenames[match(spermatid.specific$Genename, genenames[,2]),1]

# Create objects for visualziation
gax <- GenomeAxisTrack(col="black", fontsize=15, size=2)
gr <- getGeneRegionTrackForGviz(EnsDb.Mmusculus.v79)
options(ucscChromosomeNames = FALSE)

# Bin the genome in 1000Kb windows
bins.K9.P24 <- windowCounts(bam.files.K9.P24, bin = TRUE, width = 1000, param=param)
bins.K9.P28 <- windowCounts(bam.files.K9.P28, bin = TRUE, width = 1000, param=param)
bins.K4.P24 <- windowCounts(bam.files.K4.P24, bin = TRUE, width = 1000, param=param)
bins.K4.P28 <- windowCounts(bam.files.K4.P28, bin = TRUE, width = 1000, param=param)
bins.K27.P26 <- windowCounts(bam.files.K27.P26, bin = TRUE, width = 1000, param=param)
bins.K27.P28 <- windowCounts(bam.files.K27.P28, bin = TRUE, width = 1000, param=param)
```

# Count reads in promoter regions

```{r}
prom.X.K9.P24 <- regionCounts(bam.files.K9.P24, regions = prom.X, param=param)
prom.X.K9.P28 <- regionCounts(bam.files.K9.P28, regions = prom.X, param=param)
prom.X.K4.P24 <- regionCounts(bam.files.K4.P24, regions = prom.X, param=param)
prom.X.K4.P28 <- regionCounts(bam.files.K4.P28, regions = prom.X, param=param)
prom.X.K27.P26 <- regionCounts(bam.files.K27.P26, regions = prom.X, param=param)
prom.X.K27.P28 <- regionCounts(bam.files.K27.P28, regions = prom.X, param=param)


# Generate RPM per promoter - each promoter is 2200bp wide
cur_counts.K9.P24 <- assays(prom.X.K9.P24)$counts
cur_rpm.K9.P24 <- t(t(cur_counts.K9.P24)/(colSums(assays(bins.K9.P24)$counts)/1000000))

cur_counts.K9.P28 <- assays(prom.X.K9.P28)$counts
cur_rpm.K9.P28 <- t(t(cur_counts.K9.P28)/(colSums(assays(bins.K9.P28)$counts)/1000000))

cur_counts.K4.P24 <- assays(prom.X.K4.P24)$counts
cur_rpm.K4.P24 <- t(t(cur_counts.K4.P24)/(colSums(assays(bins.K4.P24)$counts)/1000000))

cur_counts.K4.P28 <- assays(prom.X.K4.P28)$counts
cur_rpm.K4.P28 <- t(t(cur_counts.K4.P28)/(colSums(assays(bins.K4.P28)$counts)/1000000))

cur_counts.K27.P26 <- assays(prom.X.K27.P26)$counts
cur_rpm.K27.P26 <- t(t(cur_counts.K27.P26)/(colSums(assays(bins.K27.P26)$counts)/1000000))

cur_counts.K27.P28 <- assays(prom.X.K27.P28)$counts
cur_rpm.K27.P28 <- t(t(cur_counts.K27.P28)/(colSums(assays(bins.K27.P28)$counts)/1000000))

# Compute average between spermatocytes and spermatids
df <- data.frame(Spermatocytes.K9.P24 = rowMeans(cur_rpm.K9.P24[,c(1,3)]),
                 Spermatids.K9.P24 = rowMeans(cur_rpm.K9.P24[,c(2,4)]),
                 Spermatocytes.K4.P24 = rowMeans(cur_rpm.K4.P24[,c(1,3)]),
                 Spermatids.K4.P24 = rowMeans(cur_rpm.K4.P24[,c(2,4)]),
                 Spermatocytes.K9.P28 = rowMeans(cur_rpm.K9.P28[,c(1,3)]),
                 Spermatids.K9.P28 = rowMeans(cur_rpm.K9.P28[,c(2,4)]),
                 Spermatocytes.K4.P28 = rowMeans(cur_rpm.K4.P28[,c(1,3)]),
                 Spermatids.K4.P28 = rowMeans(cur_rpm.K4.P28[,c(2,4)]),
                 Spermatocytes.K27.P26 = cur_rpm.K27.P26[,1],
                 Spermatids.K27.P26 = rowMeans(cur_rpm.K27.P26[,c(2,3)]),
                 Spermatocytes.K27.P28 = rowMeans(cur_rpm.K27.P28[,c(1,3)]),
                 Spermatids.K27.P28 = rowMeans(cur_rpm.K27.P28[,c(2,4)]))
df$Symbol <- genenames[rownames(df),2]

# Order based on K9 signal in spermatocytes at P24
df <- df[order(df$Spermatocytes.K9.P24, decreasing = TRUE),]

# Annotate genenames
df$spermatid_specific <- spermatid.specific[rownames(df),"Spermatid_specific"]

# Plot results
for.plot <- df[!is.na(df$spermatid_specific),]

# K9 spermatocytes P24
spermatid.spec.K9.spermatocytes.P24 <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatocytes.K9.P24 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama() + ylim(c(0,5))
ggsave(filename = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/K9_spermatocytes_P24_X_promoters.pdf", 
       spermatid.spec.K9.spermatocytes.P24, width = 4, height = 4)

# K9 spermatids P24
spermatid.spec.K9.spermatids.P24 <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatids.K9.P24 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama()+ ylim(c(0,5))
ggsave(filename = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/K9_spermatids_P24_X_promoters.pdf", 
       spermatid.spec.K9.spermatids.P24, width = 4, height = 4)

# K9 spermatocytes P28
spermatid.spec.K9.spermatocytes.P28 <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatocytes.K9.P28 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama() + ylim(c(0,5))
ggsave(filename = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/K9_spermatocytes_P28_X_promoters.pdf", 
       spermatid.spec.K9.spermatocytes.P28, width = 4, height = 4)

# K9 spermatids P28
spermatid.spec.K9.spermatids.P28 <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatids.K9.P28 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama()+ ylim(c(0,5))
ggsave(filename = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/K9_spermatids_P28_X_promoters.pdf", 
       spermatid.spec.K9.spermatids.P28, width = 4, height = 4)

# K4 spermatocytes P24
spermatid.spec.K4.spermatocytes.P24 <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatocytes.K4.P24 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama() + ylim(c(0,5))
ggsave(filename = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/K4_spermatocytes_P24_X_promoters.pdf", 
       spermatid.spec.K4.spermatocytes.P24, width = 4, height = 4)

# K4 spermatids P24
spermatid.spec.K4.spermatids.P24 <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatids.K4.P24 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama() + ylim(c(0,5))
ggsave(filename = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/K4_spermatids_P24_X_promoters.pdf", 
       spermatid.spec.K4.spermatids.P24, width = 4, height = 4)

# K4 spermatocytes P28
spermatid.spec.K4.spermatocytes.P28 <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatocytes.K4.P28 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama() + ylim(c(0,5))
ggsave(filename = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/K4_spermatocytes_P28_X_promoters.pdf", 
       spermatid.spec.K4.spermatocytes.P28, width = 4, height = 4)

# K4 spermatids P28
spermatid.spec.K4.spermatids.P28 <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatids.K4.P28 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama() + ylim(c(0,5))
ggsave(filename = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/K4_spermatids_P28_X_promoters.pdf", 
       spermatid.spec.K4.spermatids.P28, width = 4, height = 4)

# K27 spermatocytes P26
spermatid.spec.K27.spermatocytes.P26 <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatocytes.K27.P26 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama() + ylim(c(0,5))
ggsave(filename = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/K27_spermatocytes_P26_X_promoters.pdf", 
       spermatid.spec.K27.spermatocytes.P26, width = 4, height = 4)

# K27 spermatids P26
spermatid.spec.K27.spermatids.P26 <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatids.K27.P26 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama()+ ylim(c(0,5))
ggsave(filename = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/K27_spermatids_P26_X_promoters.pdf", 
       spermatid.spec.K27.spermatids.P26, width = 4, height = 4)

# K27 spermatocytes P28
spermatid.spec.K27.spermatocytes.P28 <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatocytes.K27.P28 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama() + ylim(c(0,5))
ggsave(filename = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/K27_spermatocytes_P28_X_promoters.pdf", 
       spermatid.spec.K27.spermatocytes.P28, width = 4, height = 4)

# K27 spermatids P28
spermatid.spec.K27.spermatids.P28 <- ggplot(for.plot) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatids.K27.P28 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama()+ ylim(c(0,5))
ggsave(filename = "../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/K27_spermatids_P28_X_promoters.pdf", 
       spermatid.spec.K27.spermatids.P28, width = 4, height = 4)


# Wilcoxon test
wilcox.test(for.plot$Spermatocytes.K9.P24[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatocytes.K9.P24[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatids.K9.P24[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatids.K9.P24[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatocytes.K9.P28[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatocytes.K9.P28[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatids.K9.P28[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatids.K9.P28[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatocytes.K4.P24[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatocytes.K4.P24[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatids.K4.P24[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatids.K4.P24[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatocytes.K4.P28[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatocytes.K4.P28[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatids.K4.P28[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatids.K4.P28[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatocytes.K27.P26[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatocytes.K27.P26[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatids.K27.P26[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatids.K27.P26[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatocytes.K27.P28[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatocytes.K27.P28[for.plot$spermatid_specific == FALSE])

wilcox.test(for.plot$Spermatids.K27.P28[for.plot$spermatid_specific == TRUE],
            for.plot$Spermatids.K27.P28[for.plot$spermatid_specific == FALSE])
```

# Akap4 and Cypt1 signal

Here, we visualize the signal of Akap4 and Cypt1 over the different timepoints.

# Gene Tracks specifically for Akap4 and Cypt1

```{r}
# Find promoter regions +- 1000bp
cur_prom <- promoters(genes(EnsDb.Mmusculus.v79), 
                  upstream = 10000, downstream = 5000)

# Akap4
cur_gene <- "ENSMUSG00000050089"
cur_name <- elementMetadata(cur_prom)$gene_name[
    elementMetadata(cur_prom)$gene_id == cur_gene]

cur_region <- cur_prom[elementMetadata(cur_prom)$gene_id == cur_gene]

# P24
collected.K9.P24 <- list()
lib.sizes <- colSums(assays(bins.K9.P24)$counts)/1e6
for (i in 1:length(bam.files.K9.P24)) {
    reads <- extractReads(bam.file=bam.files.K9.P24[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K9.P24[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K9.P24[i],  col.axis="black", col.title="black",
         fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K9.P24 <- collected.K9.P24[c(1,3,2,4)]

collected.K4.P24 <- list()
lib.sizes <- colSums(assays(bins.K4.P24)$counts)/1e6
for (i in 1:length(bam.files.K4.P24)) {
    reads <- extractReads(bam.file=bam.files.K4.P24[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K4.P24[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K4.P24[i],  col.axis="black", col.title="black",
         fill="dark blue", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K4.P24 <- collected.K4.P24[c(1,3,2,4)]


pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Akap4_P24_spermatocytes.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K9.P24[1:2], collected.K4.P24[1:2]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
  dev.off()
  
pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Akap4_P24_spermatids.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K9.P24[3:4], collected.K4.P24[3:4]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

# P28
collected.K9.P28 <- list()
lib.sizes <- colSums(assays(bins.K9.P28)$counts)/1e6
for (i in 1:length(bam.files.K9.P28)) {
    reads <- extractReads(bam.file=bam.files.K9.P28[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K9.P28[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K9.P28[i],  col.axis="black", col.title="black",
         fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K9.P28 <- collected.K9.P28[c(1,3,2,4)]

collected.K4.P28 <- list()
lib.sizes <- colSums(assays(bins.K4.P28)$counts)/1e6
for (i in 1:length(bam.files.K4.P28)) {
    reads <- extractReads(bam.file=bam.files.K4.P28[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K4.P28[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K4.P28[i],  col.axis="black", col.title="black",
         fill="dark blue", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K4.P28 <- collected.K4.P28[c(1,3,2,4)]


pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Akap4_P28_spermatocytes.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K9.P28[1:2], collected.K4.P28[1:2]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
  dev.off()
  
pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Akap4_P28_spermatids.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K9.P28[3:4], collected.K4.P28[3:4]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

# Cypt1
cur_gene <- "ENSMUSG00000023257"
cur_name <- elementMetadata(cur_prom)$gene_name[
    elementMetadata(cur_prom)$gene_id == cur_gene]

cur_region <- cur_prom[elementMetadata(cur_prom)$gene_id == cur_gene]

# P24
collected.K9.P24 <- list()
lib.sizes <- colSums(assays(bins.K9.P24)$counts)/1e6
for (i in 1:length(bam.files.K9.P24)) {
    reads <- extractReads(bam.file=bam.files.K9.P24[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K9.P24[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K9.P24[i],  col.axis="black", col.title="black",
         fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K9.P24 <- collected.K9.P24[c(1,3,2,4)]

collected.K4.P24 <- list()
lib.sizes <- colSums(assays(bins.K4.P24)$counts)/1e6
for (i in 1:length(bam.files.K4.P24)) {
    reads <- extractReads(bam.file=bam.files.K4.P24[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K4.P24[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K4.P24[i],  col.axis="black", col.title="black",
         fill="dark blue", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K4.P24 <- collected.K4.P24[c(1,3,2,4)]


pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Cypt1_P24_spermatocytes.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K9.P24[1:2], collected.K4.P24[1:2]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
  dev.off()
  
pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Cypt1_P24_spermatids.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K9.P24[3:4], collected.K4.P24[3:4]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

# P28
collected.K9.P28 <- list()
lib.sizes <- colSums(assays(bins.K9.P28)$counts)/1e6
for (i in 1:length(bam.files.K9.P28)) {
    reads <- extractReads(bam.file=bam.files.K9.P28[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K9.P28[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K9.P28[i],  col.axis="black", col.title="black",
         fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K9.P28 <- collected.K9.P28[c(1,3,2,4)]

collected.K4.P28 <- list()
lib.sizes <- colSums(assays(bins.K4.P28)$counts)/1e6
for (i in 1:length(bam.files.K4.P28)) {
    reads <- extractReads(bam.file=bam.files.K4.P28[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K4.P28[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K4.P28[i],  col.axis="black", col.title="black",
         fill="dark blue", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K4.P28 <- collected.K4.P28[c(1,3,2,4)]


pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Cypt1_P28_spermatocytes.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K9.P28[1:2], collected.K4.P28[1:2]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
  dev.off()
  
pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Cypt1_P28_spermatids.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K9.P28[3:4], collected.K4.P28[3:4]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

# K27 samples
# Akap4
cur_gene <- "ENSMUSG00000050089"
cur_name <- elementMetadata(cur_prom)$gene_name[
    elementMetadata(cur_prom)$gene_id == cur_gene]

cur_region <- cur_prom[elementMetadata(cur_prom)$gene_id == cur_gene]

# P26
collected.K27.P26 <- list()
lib.sizes <- colSums(assays(bins.K27.P26)$counts)/1e6
for (i in 1:length(bam.files.K27.P26)) {
    reads <- extractReads(bam.file=bam.files.K27.P26[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K27.P26[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K27.P26[i],  col.axis="black", col.title="black",
         fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K27.P26 <- collected.K27.P26[c(1,2,3)]

pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Akap4_K27_P26_spermatocytes.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K27.P26[1]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
  dev.off()
  
pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Akap4_K27_P26_spermatids.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K27.P26[2:3]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

# P28
collected.K27.P28 <- list()
lib.sizes <- colSums(assays(bins.K27.P28)$counts)/1e6
for (i in 1:length(bam.files.K27.P28)) {
    reads <- extractReads(bam.file=bam.files.K27.P28[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K27.P28[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K27.P28[i],  col.axis="black", col.title="black",
         fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K27.P28 <- collected.K27.P28[c(1,3,2,4)]

pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Akap4_K27_P28_spermatocytes.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K27.P28[1:2]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
  dev.off()
  
pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Akap4_K27_P28_spermatids.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K27.P28[3:4]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

# Cypt1
cur_gene <- "ENSMUSG00000023257"
cur_name <- elementMetadata(cur_prom)$gene_name[
    elementMetadata(cur_prom)$gene_id == cur_gene]

cur_region <- cur_prom[elementMetadata(cur_prom)$gene_id == cur_gene]

# P26
collected.K27.P26 <- list()
lib.sizes <- colSums(assays(bins.K27.P26)$counts)/1e6
for (i in 1:length(bam.files.K27.P26)) {
    reads <- extractReads(bam.file=bam.files.K27.P26[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K27.P26[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K27.P26[i],  col.axis="black", col.title="black",
         fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K27.P26 <- collected.K27.P26[c(1,2,3)]

pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Cypt1_K27_P26_spermatocytes.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K27.P26[1]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
  dev.off()
  
pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Cypt1_K27_P26_spermatids.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K27.P26[2:3]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

# P28
collected.K27.P28 <- list()
lib.sizes <- colSums(assays(bins.K27.P28)$counts)/1e6
for (i in 1:length(bam.files.K27.P28)) {
    reads <- extractReads(bam.file=bam.files.K27.P28[i], cur_region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected.K27.P28[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,5),
         name=bam.files.K27.P28[i],  col.axis="black", col.title="black",
         fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
}
collected.K27.P28 <- collected.K27.P28[c(1,3,2,4)]

pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Cypt1_K27_P28_spermatocytes.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K27.P28[1:2]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
  dev.off()
  
pdf(paste("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Cypt1_K27_P28_spermatids.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K27.P28[3:4]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
dev.off()

```

# Spermatid and non-spermatid specific promoter marks for the Hammoud data

## Read in the data

Reads in promoter regions were counted and normalized (CPM).
Here, we read in the counts in each promoter for each library generated by Hammoud et al CSC.

```{r}
library(openxlsx)
library(ggplot2)
library(reshape)

files <- list.files("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Data/Hammoud_CSC/ChipSeq/promoters/", full.names = TRUE)

all.data <- lapply(as.list(files), function(n){
  cur_dat <- read.table(n, sep = " ", header = TRUE, stringsAsFactors = FALSE)
})

all.data <- do.call("cbind", all.data)

genenames <- read.table("../../Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
rownames(genenames) <- genenames$Gene.stable.ID
```

# Visualize the data

Enrichment for epigenetic marks are visualized comparing spermatid-specific and non-specific genes.

```{r}
# Read in the spermatid-specific genes 
spermatid.spec <- read.xlsx("../../../../Dropbox (Personal)/Tc1_meiotic_silencing/Figures/Supplemental Tables/Table_S8.xlsx")
spermatid.spec$ID <- genenames[match(spermatid.spec$Symbol, genenames$Gene.name),1]
spermatid.spec <- spermatid.spec[!is.na(spermatid.spec$ID),]
```

## Visualize input control

```{r}
cur_df <- melt(all.data[,c("Input_ChIPSeq_SC", "Input_ChIPSeq_ST")])
cur_df$ID <- c(rownames(all.data),rownames(all.data))
cur_df$cell_type <- sapply(as.character(cur_df$variable), function(n){unlist(strsplit(n, "_"))[3]})
cur_df$spermatid_spec <- spermatid.spec[match(as.character(cur_df$ID), spermatid.spec$ID), 
                                        "spermatid_specific"]
cur_df <- cur_df[!is.na(cur_df$spermatid_spec),]
cur_spermatocytes <- ggplot(cur_df[cur_df$cell_type == "SC",]) + 
  geom_boxplot(aes(spermatid_spec, log2(value + 1), fill = spermatid_spec)) + 
  ylim(c(0,6)) + 
  scale_fill_jama()
cur_spermatids <- ggplot(cur_df[cur_df$cell_type == "ST",]) + 
  geom_boxplot(aes(spermatid_spec, log2(value + 1), fill = spermatid_spec)) + 
  ylim(c(0,6)) + 
  scale_fill_jama()

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Hammoud_Input_spermatocytes.pdf", cur_spermatocytes, width = 7, height = 5)

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Hammoud_Input_spermatids.pdf", cur_spermatids, width = 7, height = 5)

## Visualize H3K4me3

cur_df <- melt(all.data[,c("H3K4me3_ChIPSeq_SC", "H3K4me3_ChIPSeq_ST")])
cur_df$ID <- c(rownames(all.data),rownames(all.data))
cur_df$cell_type <- sapply(as.character(cur_df$variable), function(n){unlist(strsplit(n, "_"))[3]})
cur_df$spermatid_spec <- spermatid.spec[match(as.character(cur_df$ID), spermatid.spec$ID), 
                                        "spermatid_specific"]
cur_df <- cur_df[!is.na(cur_df$spermatid_spec),]
cur_spermatocytes <- ggplot(cur_df[cur_df$cell_type == "SC",]) + 
  geom_boxplot(aes(spermatid_spec, log2(value + 1), fill = spermatid_spec)) + 
  ylim(c(0,6)) + 
  scale_fill_jama()
cur_spermatids <- ggplot(cur_df[cur_df$cell_type == "ST",]) + 
  geom_boxplot(aes(spermatid_spec, log2(value + 1), fill = spermatid_spec)) + 
  ylim(c(0,6)) + 
  scale_fill_jama()

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Hammoud_H3K4me3_spermatocytes.pdf", cur_spermatocytes, width = 7, height = 5)

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Hammoud_H3K4me3_spermatids.pdf", cur_spermatids, width = 7, height = 5)


# Statistical testing
wilcox.test(log2(cur_df[cur_df$cell_type == "SC" & cur_df$spermatid_spec,"value"]) + 1,
            log2(cur_df[cur_df$cell_type == "SC" & !cur_df$spermatid_spec,"value"]) + 1)
wilcox.test(log2(cur_df[cur_df$cell_type == "ST" & cur_df$spermatid_spec,"value"]) + 1,
            log2(cur_df[cur_df$cell_type == "ST" & !cur_df$spermatid_spec,"value"]) + 1)

## Visualize H3K27me3

cur_df <- melt(all.data[,c("H27me3_ChIPSeq_SC", "H27me3_ChIPSeq_ST")])
cur_df$ID <- c(rownames(all.data),rownames(all.data))
cur_df$cell_type <- sapply(as.character(cur_df$variable), function(n){unlist(strsplit(n, "_"))[3]})
cur_df$spermatid_spec <- spermatid.spec[match(as.character(cur_df$ID), spermatid.spec$ID), 
                                        "spermatid_specific"]
cur_df <- cur_df[!is.na(cur_df$spermatid_spec),]
cur_spermatocytes <- ggplot(cur_df[cur_df$cell_type == "SC",]) + 
  geom_boxplot(aes(spermatid_spec, log2(value + 1), fill = spermatid_spec)) + 
  ylim(c(0,6)) + 
  scale_fill_jama()
cur_spermatids <- ggplot(cur_df[cur_df$cell_type == "ST",]) + 
  geom_boxplot(aes(spermatid_spec, log2(value + 1), fill = spermatid_spec)) + 
  ylim(c(0,6)) + 
  scale_fill_jama()

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Hammoud_H3K27me3_spermatocytes.pdf", cur_spermatocytes, width = 7, height = 5)

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Hammoud_H3K27me3_spermatids.pdf", cur_spermatids, width = 7, height = 5)


# Statistical testing
wilcox.test(log2(cur_df[cur_df$cell_type == "SC" & cur_df$spermatid_spec,"value"]) + 1,
            log2(cur_df[cur_df$cell_type == "SC" & !cur_df$spermatid_spec,"value"]) + 1)
wilcox.test(log2(cur_df[cur_df$cell_type == "ST" & cur_df$spermatid_spec,"value"]) + 1,
            log2(cur_df[cur_df$cell_type == "ST" & !cur_df$spermatid_spec,"value"]) + 1)

## Visualize H3K27ac

cur_df <- melt(all.data[,c("H3K27ac_ChIPSeq_SC", "H3K27ac_ChIPSeq_ST")])
cur_df$ID <- c(rownames(all.data),rownames(all.data))
cur_df$cell_type <- sapply(as.character(cur_df$variable), function(n){unlist(strsplit(n, "_"))[3]})
cur_df$spermatid_spec <- spermatid.spec[match(as.character(cur_df$ID), spermatid.spec$ID), 
                                        "spermatid_specific"]
cur_df <- cur_df[!is.na(cur_df$spermatid_spec),]
cur_spermatocytes <- ggplot(cur_df[cur_df$cell_type == "SC",]) + 
  geom_boxplot(aes(spermatid_spec, log2(value + 1), fill = spermatid_spec)) + 
  ylim(c(0,6)) + 
  scale_fill_jama()
cur_spermatids <- ggplot(cur_df[cur_df$cell_type == "ST",]) + 
  geom_boxplot(aes(spermatid_spec, log2(value + 1), fill = spermatid_spec)) + 
  ylim(c(0,6)) + 
  scale_fill_jama()

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Hammoud_H3K27ac_spermatocytes.pdf", cur_spermatocytes, width = 7, height = 5)

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Hammoud_H3K27ac_spermatids.pdf", cur_spermatids, width = 7, height = 5)

# Statistical testing
wilcox.test(log2(cur_df[cur_df$cell_type == "SC" & cur_df$spermatid_spec,"value"]) + 1,
            log2(cur_df[cur_df$cell_type == "SC" & !cur_df$spermatid_spec,"value"]) + 1)
wilcox.test(log2(cur_df[cur_df$cell_type == "ST" & cur_df$spermatid_spec,"value"]) + 1,
            log2(cur_df[cur_df$cell_type == "ST" & !cur_df$spermatid_spec,"value"]) + 1)

## Visualize H3K4me1

cur_df <- melt(all.data[,c("H3K4me1_ChIPSeq_SC", "H3K4me1_ChIPSeq_ST")])
cur_df$ID <- c(rownames(all.data),rownames(all.data))
cur_df$cell_type <- sapply(as.character(cur_df$variable), function(n){unlist(strsplit(n, "_"))[3]})
cur_df$spermatid_spec <- spermatid.spec[match(as.character(cur_df$ID), spermatid.spec$ID), 
                                        "spermatid_specific"]
cur_df <- cur_df[!is.na(cur_df$spermatid_spec),]
cur_spermatocytes <- ggplot(cur_df[cur_df$cell_type == "SC",]) + 
  geom_boxplot(aes(spermatid_spec, log2(value + 1), fill = spermatid_spec)) + 
  ylim(c(0,6)) + 
  scale_fill_jama()
cur_spermatids <- ggplot(cur_df[cur_df$cell_type == "ST",]) + 
  geom_boxplot(aes(spermatid_spec, log2(value + 1), fill = spermatid_spec)) + 
  ylim(c(0,6)) + 
  scale_fill_jama()

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Hammoud_H3K4me1_spermatocytes.pdf", cur_spermatocytes, width = 7, height = 5)

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Hammoud_H3K4me1_spermatids.pdf", cur_spermatids, width = 7, height = 5)

# Statistical testing
wilcox.test(log2(cur_df[cur_df$cell_type == "SC" & cur_df$spermatid_spec,"value"]) + 1,
            log2(cur_df[cur_df$cell_type == "SC" & !cur_df$spermatid_spec,"value"]) + 1)
wilcox.test(log2(cur_df[cur_df$cell_type == "ST" & cur_df$spermatid_spec,"value"]) + 1,
            log2(cur_df[cur_df$cell_type == "ST" & !cur_df$spermatid_spec,"value"]) + 1)

## Visualize H2AZ in spermatids

cur_df <- melt(all.data[,"H2AZ_ChIPSeq_ST_1"])
cur_df$ID <- rownames(all.data)
cur_df$cell_type <- "ST"
cur_df$spermatid_spec <- spermatid.spec[match(as.character(cur_df$ID), spermatid.spec$ID), 
                                        "spermatid_specific"]
cur_df <- cur_df[!is.na(cur_df$spermatid_spec),]
cur_spermatids <- ggplot(cur_df) + 
  geom_boxplot(aes(spermatid_spec, log2(value + 1), fill = spermatid_spec)) + 
  ylim(c(0,6)) + 
  scale_fill_jama()

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Hammoud_H2AZ_spermatids.pdf", cur_spermatids, width = 7, height = 5)

# Statistical testing
wilcox.test(log2(cur_df[cur_df$spermatid_spec,"value"]) + 1,
            log2(cur_df[!cur_df$spermatid_spec,"value"]) + 1)

## Visualize H3K9ac in spermatids

cur_df <- melt(all.data[,"H3K9ac_ChIPSeq_ST"])
cur_df$ID <- rownames(all.data)
cur_df$cell_type <- "ST"
cur_df$spermatid_spec <- spermatid.spec[match(as.character(cur_df$ID), spermatid.spec$ID), 
                                        "spermatid_specific"]
cur_df <- cur_df[!is.na(cur_df$spermatid_spec),]
cur_spermatids <- ggplot(cur_df) + 
  geom_boxplot(aes(spermatid_spec, log2(value + 1), fill = spermatid_spec)) + 
  ylim(c(0,6)) + 
  scale_fill_jama()

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Hammoud_H3K9ac_spermatids.pdf", cur_spermatids, width = 7, height = 5)

# Statistical testing
wilcox.test(log2(cur_df[cur_df$spermatid_spec,"value"]) + 1,
            log2(cur_df[!cur_df$spermatid_spec,"value"]) + 1)

## Visualize Pol2 in spermatids

cur_df <- melt(all.data[,"Pol2_ChIPSeq_ST_1"])
cur_df$ID <- rownames(all.data)
cur_df$cell_type <- "ST"
cur_df$spermatid_spec <- spermatid.spec[match(as.character(cur_df$ID), spermatid.spec$ID), 
                                        "spermatid_specific"]
cur_df <- cur_df[!is.na(cur_df$spermatid_spec),]
cur_spermatids <- ggplot(cur_df) + 
  geom_boxplot(aes(spermatid_spec, log2(value + 1), fill = spermatid_spec)) + 
  ylim(c(0,6)) + 
  scale_fill_jama()

ggsave("../../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S12/Hammoud_Pol2_spermatids.pdf", cur_spermatids, width = 7, height = 5)

# Statistical testing
wilcox.test(log2(cur_df[cur_df$spermatid_spec,"value"]) + 1,
            log2(cur_df[!cur_df$spermatid_spec,"value"]) + 1)
```