---
title: "P10 filtering"
author: "Nils Eling"
date: "08/03/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in data

```{r}
# Load libraries
library(Matrix)
library(scran)
library(plot3D)
library(Rtsne)
library(ggplot2)
library(DropletUtils)
library(scater)
library(viridis)

# Read in data - Single cell experiment object per batch
sce.do17821 <- read10xCounts("/Users/eling01/Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17821/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17821)$Sample <- rep("P10_do17821", ncol(sce.do17821))
```

# Find empty droplets

```{r}
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17821)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17821 <- sce.do17821[,is.cell]
```

# Further quality control

```{r}
# Apply more filtering steps on these cells
sce.do17821 <- calculateQCMetrics(sce.do17821)

# Number of genes detected
plot(colData(sce.do17821)$total_counts, colData(sce.do17821)$total_features, log = "x", pch = 16, 
     xlab = "Total counts", ylab = "Number of genes", 
     col = ifelse(colData(sce.do17821)$total_features > 200, "black", "red"))
sce.do17821 <- sce.do17821[,colData(sce.do17821)$total_features > 200]

# Mitochondrial genes
mt <- read.table("../../data/MT_genes.txt")
plot(Matrix::colSums(counts(sce.do17821)[mt$V1,])/colData(sce.do17821)$total_counts, 
     pch = 16, ylab = "% mitochondrial reads")
sce.do17821 <- sce.do17821[,Matrix::colSums(counts(sce.do17821)[mt$V1,])/colData(sce.do17821)$total_counts < 0.05]

# Remove genes that are not expressed
sce.do17821 <- sce.do17821[rowData(sce.do17821)$mean_counts > 0,]
```

# Normalization

```{r}
clusters <- quickCluster(sce.do17821, method = "igraph", irlba.args = c("work" = 100))

sce.do17821 <- computeSumFactors(sce.do17821, clusters=clusters)

sce.do17821 <- normalize(sce.do17821, return_log = TRUE)
```

```{r}
# Compute highly variable genes
HVG <- trendVar(sce.do17821, use.spikes = FALSE)
HVG.1 <- decomposeVar(sce.do17821, HVG)
rowData(sce.do17821)$HVG <- DataFrame(HVG.1)
HVG.1 <- HVG.1[order(HVG.1$bio, decreasing = TRUE),]
HVG.genes <- rownames(HVG.1)[1:1000]

# Compute tSNE
set.seed(12345)
sce.do17821 <- scater::runTSNE(sce.do17821)

# Library size
ggplot(data.frame(tsne1 = reducedDims(sce.do17821)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17821)$TSNE[,2],
                  gene = colData(sce.do17821)$log10_total_counts)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Number genes expressed
ggplot(data.frame(tsne1 = reducedDims(sce.do17821)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17821)$TSNE[,2],
                  gene = colData(sce.do17821)$total_features)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Stra8 
ggplot(data.frame(tsne1 = reducedDims(sce.do17821)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17821)$TSNE[,2],
                  gene = logcounts(sce.do17821)[rowData(sce.do17821)$ID[rowData(sce.do17821)$Symbol == "Stra8"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Defb19 
ggplot(data.frame(tsne1 = reducedDims(sce.do17821)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17821)$TSNE[,2],
                  gene = logcounts(sce.do17821)[rowData(sce.do17821)$ID[rowData(sce.do17821)$Symbol == "Defb19"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Col1a1 
ggplot(data.frame(tsne1 = reducedDims(sce.do17821)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17821)$TSNE[,2],
                  gene = logcounts(sce.do17821)[rowData(sce.do17821)$ID[rowData(sce.do17821)$Symbol == "Col1a1"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Nanos3
ggplot(data.frame(tsne1 = reducedDims(sce.do17821)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17821)$TSNE[,2],
                  gene = logcounts(sce.do17821)[rowData(sce.do17821)$ID[rowData(sce.do17821)$Symbol == "Nanos3"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Zbtb16
ggplot(data.frame(tsne1 = reducedDims(sce.do17821)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17821)$TSNE[,2],
                  gene = logcounts(sce.do17821)[rowData(sce.do17821)$ID[rowData(sce.do17821)$Symbol == "Zbtb16"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Prm1 
ggplot(data.frame(tsne1 = reducedDims(sce.do17821)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17821)$TSNE[,2],
                  gene = logcounts(sce.do17821)[rowData(sce.do17821)$ID[rowData(sce.do17821)$Symbol == "Prm1"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Dazl 
ggplot(data.frame(tsne1 = reducedDims(sce.do17821)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17821)$TSNE[,2],
                  gene = logcounts(sce.do17821)[rowData(sce.do17821)$ID[rowData(sce.do17821)$Symbol == "Dazl"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))
```

# Save sce object

```{r}
saveRDS(sce.do17821, "../../data/10X_data/SCE_P10.rds")
```