---
title: "P35 filtering"
author: "Nils Eling"
date: "11/04/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in data

```{r}
# Load libraries
library(Matrix)
library(scran)
library(plot3D)
library(Rtsne)
library(ggplot2)
library(DropletUtils)
library(scater)
library(viridis)

# Read in data - Single cell experiment object per batch
sce.do17827 <- read10xCounts("/Users/nils/Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17827/raw_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17827)$Sample <- rep("P35_do17827", ncol(sce.do17827))
```

# Find empty droplets

```{r}
set.seed(100)
empty.droplets <- emptyDrops(counts(sce.do17827)) 
is.cell <- empty.droplets$FDR <= 0.01
is.cell[is.na(is.cell)] <- FALSE
sce.do17827 <- sce.do17827[,is.cell]
```

# Further quality control

```{r}
# Apply more filtering steps on these cells
sce.do17827 <- calculateQCMetrics(sce.do17827)

# Number of genes detected
plot(colData(sce.do17827)$total_counts, colData(sce.do17827)$total_features, log = "x", pch = 16, 
     xlab = "Total counts", ylab = "Number of genes", 
     col = ifelse(colData(sce.do17827)$total_features > 500, "black", "red"))
sce.do17827 <- sce.do17827[,colData(sce.do17827)$total_features > 500]

plot(colData(sce.do17827)$total_counts, colData(sce.do17827)$total_features, log = "x", pch = 16, 
     xlab = "Total counts", ylab = "Number of genes", 
     col = ifelse(colData(sce.do17827)$total_counts > 3000, "black", "red"))
sce.do17827 <- sce.do17827[,colData(sce.do17827)$total_counts > 3000]

# Mitochondrial genes
genes <- read.table("../../data/Mouse_genes.txt", sep = "\t", header = TRUE)
mt <- genes[genes$Chromosome.scaffold.name == "MT",]
plot(Matrix::colSums(counts(sce.do17827)[mt$Gene.stable.ID,])/colData(sce.do17827)$total_counts, 
     pch = 16, ylab = "% mitochondrial reads")
sce.do17827 <- sce.do17827[,Matrix::colSums(counts(sce.do17827)[mt$Gene.stable.ID,])/colData(sce.do17827)$total_counts < 0.05]

# Remove genes that are not expressed
sce.do17827 <- sce.do17827[rowData(sce.do17827)$mean_counts > 0,]
```

# Normalization

```{r}
clusters <- quickCluster(sce.do17827, method = "igraph", irlba.args = c("work" = 100))

sce.do17827 <- computeSumFactors(sce.do17827, clusters=clusters, min.mean = 0.1)

sce.do17827 <- normalize(sce.do17827, return_log = TRUE)
```

```{r}
# Compute tSNE
set.seed(12345)
sce.do17827 <- scater::runTSNE(sce.do17827)

# Library size
ggplot(data.frame(tsne1 = reducedDims(sce.do17827)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17827)$TSNE[,2],
                  gene = colData(sce.do17827)$log10_total_counts)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# genes detected
ggplot(data.frame(tsne1 = reducedDims(sce.do17827)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17827)$TSNE[,2],
                  gene = colData(sce.do17827)$total_features)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Number genes expressed
ggplot(data.frame(tsne1 = reducedDims(sce.do17827)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17827)$TSNE[,2],
                  gene = colData(sce.do17827)$total_features)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Stra8 
ggplot(data.frame(tsne1 = reducedDims(sce.do17827)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17827)$TSNE[,2],
                  gene = logcounts(sce.do17827)[rowData(sce.do17827)$ID[rowData(sce.do17827)$Symbol == "Stra8"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Defb19 
ggplot(data.frame(tsne1 = reducedDims(sce.do17827)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17827)$TSNE[,2],
                  gene = logcounts(sce.do17827)[rowData(sce.do17827)$ID[rowData(sce.do17827)$Symbol == "Defb19"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Col1a1 
ggplot(data.frame(tsne1 = reducedDims(sce.do17827)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17827)$TSNE[,2],
                  gene = logcounts(sce.do17827)[rowData(sce.do17827)$ID[rowData(sce.do17827)$Symbol == "Col1a1"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Nanos3
ggplot(data.frame(tsne1 = reducedDims(sce.do17827)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17827)$TSNE[,2],
                  gene = logcounts(sce.do17827)[rowData(sce.do17827)$ID[rowData(sce.do17827)$Symbol == "Nanos3"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Zbtb16
ggplot(data.frame(tsne1 = reducedDims(sce.do17827)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17827)$TSNE[,2],
                  gene = logcounts(sce.do17827)[rowData(sce.do17827)$ID[rowData(sce.do17827)$Symbol == "Zbtb16"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Prm1 
ggplot(data.frame(tsne1 = reducedDims(sce.do17827)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17827)$TSNE[,2],
                  gene = logcounts(sce.do17827)[rowData(sce.do17827)$ID[rowData(sce.do17827)$Symbol == "Prm1"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Dazl 
ggplot(data.frame(tsne1 = reducedDims(sce.do17827)$TSNE[,1],
                  tsne2 = reducedDims(sce.do17827)$TSNE[,2],
                  gene = logcounts(sce.do17827)[rowData(sce.do17827)$ID[rowData(sce.do17827)$Symbol == "Dazl"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))
```

# Save sce object

```{r}
saveRDS(sce.do17827, "../../data/10X_data/SCE_P35.rds")
```