---
title: "Filtering"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Preprocessing/QualityControl/Parallel_filtering.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in data

```{r data, message=FALSE}
# Load libraries
library(Matrix)
library(scran)
library(plot3D)
library(Rtsne)
library(ggplot2)
library(DropletUtils)
library(scater)
library(viridis)
library(RColorBrewer)

# Read in data - Single cell experiment object per batch
# P10
sce.do17821 <- read10xCounts("/Users/nils/Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17821/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17821)$Sample <- rep("P10_do17821", ncol(sce.do17821))

# P15
sce.do17828 <- read10xCounts("/Users/nils/Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17828/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17828)$Sample <- rep("P15_do17828", ncol(sce.do17828))

# P20
sce.do17824 <- read10xCounts("/Users/nils/Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17824/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17824)$Sample <- rep("P20_do17824", ncol(sce.do17824))

# P30
sce.do17825 <- read10xCounts("/Users/nils/Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17825/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17825)$Sample <- rep("P30_do17825", ncol(sce.do17825))

# P35
sce.do17827 <- read10xCounts("/Users/nils/Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17827/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17827)$Sample <- rep("P35_do17827", ncol(sce.do17827))

# B6
sce.do17815 <- read10xCounts("/Users/nils/Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17815/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17815)$Sample <- rep("B6_do17815", ncol(sce.do17815))

sce.do17816 <- read10xCounts("/Users/nils/Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17816/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17816)$Sample <- rep("B6_do17816", ncol(sce.do17816))

# Tc0
sce.do15984 <- read10xCounts("/Users/nils/Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do15984/Mus_human_chr21/")
colData(sce.do15984)$Sample <- rep("Tc0_do15984", ncol(sce.do15984))

sce.do17622 <- read10xCounts("/Users/nils/Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17622/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17622)$Sample <- rep("Tc0_do17622", ncol(sce.do17622))

# Tc1
sce.do15983 <- read10xCounts("/Users/nils/Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do15983/Mus_human_chr21/")
colData(sce.do15983)$Sample <- rep("Tc1_do15983", ncol(sce.do15983))

sce.do17623 <- read10xCounts("/Users/nils/Google Drive File Stream/My Drive/Christina/Tc1_single_cell/data/raw_reads/10X/do17623/filtered_gene_bc_matrices/Mus_human_chr21/")
colData(sce.do17623)$Sample <- rep("Tc1_do17623", ncol(sce.do17623))
```

# Filtering 

```{r Other}
# Merge reamining datasets
sce.all <- cbind(sce.do17815, sce.do17816, sce.do15984, sce.do17622, 
                  sce.do15983, sce.do17623, sce.do17821, sce.do17828, sce.do17824, 
                 sce.do17825, sce.do17827)

# Calculate QC metrics
sce.all <- calculateQCMetrics(sce.all)

# Remove cells with few features
plot(colData(sce.all)$total_counts, colData(sce.all)$total_features, log = "x", pch = 16, 
     xlab = "Total counts", ylab = "Number of genes", 
     col = ifelse(colData(sce.all)$total_features > 2000, "black", "red"))
sce.all <- sce.all[,colData(sce.all)$total_features > 2000]

# Remove cells with high mitochondrial reads
mt <- read.table("../../Data/Mouse_genes.txt", sep = "\t", header = TRUE,
                 stringsAsFactors = FALSE)
mt <- mt[mt$Chromosome.scaffold.name == "MT",]
plot(Matrix::colSums(counts(sce.all)[mt$Gene.stable.ID,])/colData(sce.all)$total_counts, 
     pch = 16, ylab = "% mitochondrial reads")
sce.all <- sce.all[,Matrix::colSums(counts(sce.all)[mt$Gene.stable.ID,])/colData(sce.all)$total_counts < 0.1]

# Remove duplicated barcodes
#sce.all <- sce.all[,-which(colData(sce.all)$Barcode %in% colData(sce.all)$Barcode[which(duplicated(colData(sce.all)$Barcode))])]

# Remove genes that are not expressed
sce.all <- sce.all[Matrix::rowSums(counts(sce.all)) > 0,]
```

# Normalization

```{r normalization}
clusters <- quickCluster(sce.all, method = "igraph", irlba.args = c("work" = 100), 
                         max.size = 3000)

sce.all <- computeSumFactors(sce.all, clusters=clusters)

sce.all <- normalize(sce.all, return_log = TRUE)
```

# Quality visualization

```{r plotting}
set.seed(12345)
sce.all <- scater::runTSNE(sce.all, perplexity = 100)

# Batches
col_vector <- c(brewer.pal(8, "Set1"), brewer.pal(8, "Set2"), brewer.pal(8, "Set3"))
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  batch = colData(sce.all)$Sample)) +
  geom_point(aes(tsne1, tsne2, colour = batch)) +
  scale_color_manual(values = col_vector)

# Verification of chr21 expression: PDXK
# PDXK 
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)[rowData(sce.all)$ID[rowData(sce.all)$Symbol == "PDXK"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene))

# Library size
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = colData(sce.all)$log10_total_counts)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Number genes expressed
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = colData(sce.all)$total_features)) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Mitochondrial genes
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  mito = Matrix::colSums(counts(sce.all)[rowData(sce.all)$ID %in%
                                                mt$Gene.stable.ID,])/
                    Matrix::colSums(counts(sce.all)))) + geom_point(aes(tsne1, tsne2, colour = mito)) + scale_colour_viridis()

# Stra8 
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)[rowData(sce.all)$ID[rowData(sce.all)$Symbol == "Stra8"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Defb19 
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)[rowData(sce.all)$ID[rowData(sce.all)$Symbol == "Defb19"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Prm1 
ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)[rowData(sce.all)$ID[rowData(sce.all)$Symbol == "Prm1"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()

# Dazl 
  ggplot(data.frame(tsne1 = reducedDims(sce.all)$TSNE[,1],
                  tsne2 = reducedDims(sce.all)$TSNE[,2],
                  gene = logcounts(sce.all)[rowData(sce.all)$ID[rowData(sce.all)$Symbol == "Dazl"],])) +
  geom_point(aes(tsne1, tsne2, colour = gene)) + scale_colour_viridis()


```

```{r}
saveRDS(sce.all, "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all_QC.rds")
```