---
title: "Figure 5"
author: "nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5_v2.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data and libraries

```{r data, message=FALSE}
library(rtracklayer)
library(EnrichedHeatmap)
library(GenomicFeatures)
library(EnsDb.Mmusculus.v79)
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
library(scater)
library(plyr)
library(reshape2)
library(pheatmap)
library(viridis)
library(plot3D)
library(ggsci)
library(cowplot)
library(openxlsx)
library(csaw)

# Generate feature annotation
tss <- promoters(genes(EnsDb.Mmusculus.v79), 
                 upstream = 0, downstream = 1)
tss <- tss[seqnames(tss) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(tss) <- c(as.character(1:19), "X", "Y", "MT")
tss.X <- tss[seqnames(tss) == "X"]

prom <- promoters(genes(EnsDb.Mmusculus.v79))
prom <- prom[seqnames(prom) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(prom) <- c(as.character(1:19), "X", "Y", "MT")
prom.X <- prom[seqnames(prom) == "X"]

gen <- genes(EnsDb.Mmusculus.v79)
gen <- gen[seqnames(gen) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(gen) <- c(as.character(1:19), "X", "Y", "MT")
gen.X <- gen[seqnames(gen) == "X"]

# Make TXDB for mus musculus
#mmusculusEnsembl <- makeTxDbFromBiomart(dataset="mmusculus_gene_ensembl")

black <- import("http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz")

# Read in sce data
sce <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")
sce <- sce[,grepl("B6", colData(sce)$Sample) & 
             colData(sce)$AnnotatedClusters %in% levels(colData(sce)$AnnotatedClusters)[1:23]]
sce <- normalize(sce)

# Select XChr genes
genenames <- read.table("../Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
sce.XChr <- sce[rowData(sce)$ID %in% 
                  genenames[genenames$Chromosome.scaffold.name == "X",1],]
sce.XChr <- normalize(sce.XChr)

# Bulk data
bulk <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/bulk/norm/1st_wave_bulk_norm_reverse-stranded.rds")

# Meta info
meta <- as.data.frame(read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/ArrayExpress/Bulk/sdrf.xlsx"))
```

# Overlap DB for H3K4me3 and H3K27ac

```{r}
# Load data
H3K4me3 <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/DB/H3K4me3/H3K4me3_1000.rds")
H3K27ac <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/DB/H3K27ac/H3K27ac_1000.rds")

# Merge the windows
merged.K27 <- mergeWindows(H3K27ac$windows, tol=100L, max.width=5000L)
tabbest.K27 <- getBestTest(merged.K27$id, H3K27ac$DB)
tabcom.K27 <- combineTests(merged.K27$id, H3K27ac$DB)
colnames(tabcom.K27) <- paste(colnames(tabcom.K27), "_K27", sep = "")
colnames(tabbest.K27) <- paste(colnames(tabbest.K27), "_K27", sep = "")

merged.K4 <- mergeWindows(H3K4me3$windows, tol=100L, max.width=5000L)
tabbest.K4 <- getBestTest(merged.K4$id, H3K4me3$DB)
tabcom.K4 <- combineTests(merged.K4$id, H3K4me3$DB)
colnames(tabcom.K4) <- paste(colnames(tabcom.K4), "_K4", sep = "")
colnames(tabbest.K4) <- paste(colnames(tabbest.K4), "_K4", sep = "")

# Find overlaps to promoter - K4 
olap.prom <- findOverlaps(prom.X, merged.K4$region)
df.K4 <- data.frame(ID = prom.X$gene_id[olap.prom@from],
                    genename = prom.X$gene_name[olap.prom@from],
                    tabcom.K4[olap.prom@to,])

# Overlap for K27
olap.K27 <- findOverlaps(merged.K4$region[olap.prom@to], H3K27ac$windows)
tab.K27 <- combineOverlaps(olap.K27, H3K27ac$DB)
colnames(tab.K27) <- paste(colnames(tab.K27), "_K27", sep = "")
df.K27 <- data.frame(df.K4,tab.K27)
df.K27 <- df.K27[order(df.K27$direction_K4, df.K27$direction_K27, decreasing = FALSE),]

# Remove the two genes that are named twice
df.K27 <- df.K27[match(unique(df.K27$ID), df.K27$ID),]
rownames(df.K27) <- df.K27$ID
```

## Visualize bulk RNA-Seq data

```{r}
cur.genes <- as.character(df.K27$ID)
cur.genes <- cur.genes[!is.na(match(cur.genes, rownames(bulk)))]
for.heatmap <- bulk[cur.genes,
                   meta$Source.Name[order(meta$`Characteristics.[Age]`, decreasing = FALSE)]]
for.heatmap <- for.heatmap[rowMeans(for.heatmap) > 10,]

pheatmap(log2(for.heatmap + 1), 
         cluster_cols = FALSE, 
         cluster_rows = FALSE, show_rownames = FALSE,
         color = colorRampPalette(c("#053061", "#4393c3", "#f7f7f7", "#d6604d", "#67001f"))(100),
         border_color = NA,
         labels_col = paste("P", meta$`Characteristics.[Age]`[order(meta$`Characteristics.[Age]`, decreasing = FALSE)], sep = ""),
         annotation_row = data.frame(row.names = rownames(for.heatmap),
                                     H3K4me3 = df.K27[rownames(for.heatmap), "direction_K4"],
                                     H3K27ac = df.K27[rownames(for.heatmap), "direction_K27"]),
         scale = "row")

```

## Visualize scRNA-Seq data

```{r}
# Build mean expression matrix
df <- as.data.frame(t(as.matrix(logcounts(sce)[rownames(for.heatmap),])))
df$groups <- colData(sce.XChr)$AnnotatedClusters
df.melt <- melt(df, id.vars = "groups")

# Collect mean expression for each gene in each group
mat <- ddply(df.melt, .(groups, variable), summarize, mean=mean(value))
mat.for.heatmap <- dcast(data = mat,formula = variable~groups,fun.aggregate = sum,value.var = "mean")
rownames(mat.for.heatmap) <- mat.for.heatmap$variable
mat.for.heatmap <- mat.for.heatmap[,-1]


pheatmap(log2(mat.for.heatmap + 1), 
         cluster_cols = FALSE, 
         cluster_rows = FALSE, show_rownames = FALSE,
         color = colorRampPalette(c("#053061", "#4393c3", "#f7f7f7", "#d6604d", "#67001f"))(100),
         border_color = NA,
         annotation_row = data.frame(row.names = rownames(mat.for.heatmap),
                                     H3K4me3 = df.K27[rownames(for.heatmap), "direction_K4"],
                                     H3K27ac = df.K27[rownames(mat.for.heatmap), "direction_K27"]),
         scale = "row")

```

# Find DB regions for H3K4 and visualize H3K27ac enrichment

```{r}
# Load data
H3K4me3 <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/DB/H3K4me3/H3K4me3_1000.rds")
H3K27ac <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/DB/H3K27ac/H3K27ac_1000.rds")
H3K9me3 <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/DB/H3K9me3/H3K9me3_1000.rds")

merged.K4 <- mergeWindows(H3K4me3$windows, tol=100L, max.width=5000L)
tabbest.K4 <- getBestTest(merged.K4$id, H3K4me3$DB)
tabcom.K4 <- combineTests(merged.K4$id, H3K4me3$DB)
colnames(tabcom.K4) <- paste(colnames(tabcom.K4), "_K4", sep = "")
colnames(tabbest.K4) <- paste(colnames(tabbest.K4), "_K4", sep = "")

# Find overlaps to promoter - K4 
olap.prom <- findOverlaps(prom.X, merged.K4$region)
df.K4 <- data.frame(ID = prom.X$gene_id[olap.prom@from],
                    genename = prom.X$gene_name[olap.prom@from],
                    merged.id = olap.prom@to,
                    region = merged.K4$region[olap.prom@to],
                    tabcom.K4[olap.prom@to,])

# Mean K4 signal for these regions
olap.K4 <- findOverlaps(merged.K4$region[olap.prom@to], H3K4me3$windows, minoverlap = 1000L)
df.K4$K4_enrichment_spermatocytes <- ddply(data.frame(olap.K4, 
                enrichment.spermatocytes =
                  H3K4me3$enrichment[olap.K4@to,"spermatocytes"]
                ), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean
df.K4$K4_enrichment_spermatids <- ddply(data.frame(olap.K4, 
                enrichment.spermatids =
                  H3K4me3$enrichment[olap.K4@to,"spermatids"]
                ), .(queryHits), summarize, 
                mean = mean(enrichment.spermatids))$mean

# Overlap for K27
olap.K27 <- findOverlaps(merged.K4$region[olap.prom@to], H3K27ac$windows, minoverlap = 500L)
df.K4$K27_enrichment_spermatocytes <- ddply(data.frame(olap.K27, 
                                         enrichment.spermatocytes =
                                           H3K27ac$enrichment[olap.K27@to,"spermatocytes"]
                                         ), .(queryHits), summarize, mean = mean(enrichment.spermatocytes))$mean

df.K4$K27_enrichment_spermatids <- ddply(data.frame(olap.K27, 
                                         enrichment.spermatids =
                                           H3K27ac$enrichment[olap.K27@to,"spermatids"]
                                         ), .(queryHits), summarize, mean = mean(enrichment.spermatids))$mean

# Overlap for K9
olap.K9 <- findOverlaps(merged.K4$region[olap.prom@to], 
                         H3K9me3$windows, minoverlap = 500L)
df.K4$K9_enrichment_spermatocytes <- rep(NA, nrow(df.K4))
df.K4$K9_enrichment_spermatocytes[unique(olap.K9@from)] <- ddply(data.frame(olap.K9, 
      enrichment.spermatocytes = H3K9me3$enrichment[olap.K9@to,"spermatocytes"]
                          ), .(queryHits), summarize, mean =
      mean(enrichment.spermatocytes))$mean

df.K4$K9_enrichment_spermatids <- rep(NA, nrow(df.K4))
df.K4$K9_enrichment_spermatids[unique(olap.K9@from)] <- ddply(data.frame(olap.K9, 
      enrichment.spermatids = H3K9me3$enrichment[olap.K9@to,"spermatids"]
                          ), .(queryHits), summarize, mean =
      mean(enrichment.spermatids))$mean

df.K4 <- df.K4[order(df.K4$K4_enrichment_spermatocytes, df.K4$K9_enrichment_spermatocytes, decreasing = FALSE, na.last = FALSE),]

# Remove the two genes that are named twice
df.K4 <- df.K4[match(unique(df.K4$ID), df.K4$ID),]
rownames(df.K4) <- df.K4$ID

```

## Visualize bulk RNA-Seq data

```{r, fig.height=25, fig.width=5}
cur.genes <- df.K4
cur.genes <- cur.genes[!is.na(match(cur.genes$ID, rownames(bulk))),]
for.heatmap <- bulk[as.character(cur.genes$ID),
                   meta$Source.Name[order(meta$`Characteristics.[Age]`, decreasing = FALSE)]]
for.heatmap <- for.heatmap[rowMeans(for.heatmap) > 10,]

pheatmap(log2(for.heatmap + 1), 
         cluster_cols = FALSE, 
         cluster_rows = FALSE, labels_row = as.character(cur.genes$genename),
         color = colorRampPalette(c("#053061", "#4393c3", "#f7f7f7", "#d6604d", "#67001f"))(100),
         border_color = NA,
         labels_col = paste("P", meta$`Characteristics.[Age]`[order(meta$`Characteristics.[Age]`, decreasing = FALSE)], sep = ""),
         annotation_row = data.frame(row.names = rownames(for.heatmap),
        H3K4me3.spermatocytes = df.K4[rownames(for.heatmap),
                                      "K4_enrichment_spermatocytes"],
        H3K4me3.spermatids = df.K4[rownames(for.heatmap), 
                                      "K4_enrichment_spermatids"],
        H3K27ac.spermatocytes = df.K4[rownames(for.heatmap), 
                        "K27_enrichment_spermatocytes"],
        H3K27ac.spermatids = df.K4[rownames(for.heatmap), 
                        "K27_enrichment_spermatids"],
        H3K9me3.spermatocytes = df.K4[rownames(for.heatmap), 
                        "K9_enrichment_spermatocytes"],
        H3K9me3.spermatids = df.K4[rownames(for.heatmap), 
                        "K9_enrichment_spermatids"]),
         scale = "row", cellheight = 8, fontsize = 7)

```

```{r, fig.height=25, fig.width=5}
# Build mean expression matrix
df <- as.data.frame(t(as.matrix(logcounts(sce)[rownames(for.heatmap)[
  !is.na(match(rownames(for.heatmap), rownames(logcounts(sce))))],])))
df$groups <- colData(sce.XChr)$AnnotatedClusters
df.melt <- melt(df, id.vars = "groups")

# Collect mean expression for each gene in each group
mat <- ddply(df.melt, .(groups, variable), summarize, mean=mean(value))
mat.for.heatmap <- dcast(data = mat,formula = variable~groups,fun.aggregate = sum,value.var = "mean")
rownames(mat.for.heatmap) <- mat.for.heatmap$variable
mat.for.heatmap <- mat.for.heatmap[,-1]


pheatmap(log2(mat.for.heatmap + 1), 
         cluster_cols = FALSE, 
         cluster_rows = FALSE, labels_row = rowData(sce)$Symbol[
           match(rownames(mat.for.heatmap), rowData(sce)$ID)],
         color = colorRampPalette(c("#053061", "#4393c3", "#f7f7f7", "#d6604d", "#67001f"))(100),
         border_color = NA,
         annotation_row = data.frame(row.names = rownames(for.heatmap),
        H3K4me3.spermatocytes = df.K4[rownames(for.heatmap),
                                      "K4_enrichment_spermatocytes"],
        H3K4me3.spermatids = df.K4[rownames(for.heatmap), 
                                      "K4_enrichment_spermatids"],
        H3K27ac.spermatocytes = df.K4[rownames(for.heatmap), 
                        "K27_enrichment_spermatocytes"],
        H3K27ac.spermatids = df.K4[rownames(for.heatmap), 
                        "K27_enrichment_spermatids"],
        H3K9me3.spermatocytes = df.K4[rownames(for.heatmap), 
                        "K9_enrichment_spermatocytes"],
        H3K9me3.spermatids = df.K4[rownames(for.heatmap), 
                        "K9_enrichment_spermatids"]),
         scale = "row", cellheight = 8, fontsize = 7)

```

# Overall level of epigenetic enrichment of the different chromosomes

```{r}
# Generate matrix to store the average K9 enrichment
mat.K9 <- matrix(data = NA, ncol = length(seqlevels(H3K9me3$windows)), nrow = 2)
rownames(mat.K9) <- c("Spermatocytes", "Spermatids")
colnames(mat.K9) <- seqlevels(H3K9me3$windows)

# Loop trough the different chromosomes and find K9 enrichemnt for the whole chromosome
for(i in seqlevels(prom)){
  cur_windows <- seqnames(H3K9me3$windows) == i
  mat.K9[1,i] <- mean(H3K9me3$enrichment[as.vector(cur_windows),"spermatocytes"])
  mat.K9[2,i] <- mean(H3K9me3$enrichment[as.vector(cur_windows),"spermatids"])
}
mat.K9 <- mat.K9[,!is.na(mat.K9[1,])]
mat.K9 <- mat.K9[,!(colnames(mat.K9) == "MT")]
cur_df <- melt(mat.K9)
ggplot(cur_df) + geom_point(aes(Var2, value, colour = Var1)) + ggtitle("H3K9me3")

# Generate matrix to store the average K27 enrichment
mat.K27 <- matrix(data = NA, ncol = length(seqlevels(H3K27ac$windows)), nrow = 2)
rownames(mat.K27) <- c("Spermatocytes", "Spermatids")
colnames(mat.K27) <- seqlevels(H3K27ac$windows)

# Loop trough the different chromosomes and find K27 enrichemnt for the whole chromosome
for(i in seqlevels(prom)){
  cur_windows <- seqnames(H3K27ac$windows) == i
  mat.K27[1,i] <- mean(H3K27ac$enrichment[as.vector(cur_windows),"spermatocytes"])
  mat.K27[2,i] <- mean(H3K27ac$enrichment[as.vector(cur_windows),"spermatids"])
}
mat.K27 <- mat.K27[,!is.na(mat.K27[1,])]
mat.K27 <- mat.K27[,!(colnames(mat.K27) == "MT")]
cur_df <- melt(mat.K27)
ggplot(cur_df) + geom_point(aes(Var2, value, colour = Var1)) + ggtitle("H3K27ac")

# Generate matrix to store the average K27 enrichment
mat.K4 <- matrix(data = NA, ncol = length(seqlevels(H3K4me3$windows)), nrow = 2)
rownames(mat.K4) <- c("Spermatocytes", "Spermatids")
colnames(mat.K4) <- seqlevels(H3K4me3$windows)

# Loop trough the different chromosomes and find K4 enrichemnt for the whole chromosome
for(i in seqlevels(prom)){
  cur_windows <- seqnames(H3K4me3$windows) == i
  mat.K4[1,i] <- mean(H3K4me3$enrichment[as.vector(cur_windows),"spermatocytes"])
  mat.K4[2,i] <- mean(H3K4me3$enrichment[as.vector(cur_windows),"spermatids"])
}
mat.K4 <- mat.K4[,!is.na(mat.K4[1,])]
mat.K4 <- mat.K4[,!(colnames(mat.K4) == "MT")]
cur_df <- melt(mat.K4)
ggplot(cur_df) + geom_point(aes(Var2, value, colour = Var1)) + ggtitle("H3K4me3")
```

# Overall direction between spermatids and spermatocytes

```{r}
# Generate matrix to store the average K9 enrichment
mat.K9 <- matrix(data = NA, ncol = length(seqlevels(H3K9me3$windows)), nrow = 3)
rownames(mat.K9) <- c("down", "up", "mixed")
colnames(mat.K9) <- seqlevels(H3K9me3$windows)

# Loop trough the different chromosomes and find K9 enrichemnt for the whole chromosome
for(i in seqlevels(prom)){
  cur_windows <- seqnames(H3K9me3$windows) == i
  cur_merged <- mergeWindows(H3K9me3$windows[as.vector(cur_windows)], tol=100L)
  cur_tabcom <- combineTests(cur_merged$id, H3K9me3$DB[as.vector(cur_windows),])
  mat.K9[1,i] <- sum(cur_tabcom$direction == "down")/nrow(cur_tabcom)
  mat.K9[2,i] <- sum(cur_tabcom$direction == "up")/nrow(cur_tabcom)
  mat.K9[3,i] <- sum(cur_tabcom$direction == "mixed")/nrow(cur_tabcom)
}
mat.K9 <- mat.K9[,!is.na(mat.K9[1,])]
mat.K9 <- mat.K9[,!(colnames(mat.K9) == "MT")]
cur_df <- melt(mat.K9)
ggplot(cur_df) + geom_point(aes(Var2, value, colour = Var1)) + ggtitle("H3K9me3")

# Generate matrix to store the average K27 enrichment
mat.K27 <- matrix(data = NA, ncol = length(seqlevels(H3K27ac$windows)), nrow = 2)
rownames(mat.K27) <- c("Spermatocytes", "Spermatids")
colnames(mat.K27) <- seqlevels(H3K27ac$windows)

# Loop trough the different chromosomes and find K27 enrichemnt for the whole chromosome
for(i in seqlevels(prom)){
  cur_windows <- seqnames(H3K27ac$windows) == i
  mat.K27[1,i] <- mean(H3K27ac$enrichment[as.vector(cur_windows),"spermatocytes"])
  mat.K27[2,i] <- mean(H3K27ac$enrichment[as.vector(cur_windows),"spermatids"])
}
mat.K27 <- mat.K27[,!is.na(mat.K27[1,])]
mat.K27 <- mat.K27[,!(colnames(mat.K27) == "MT")]
cur_df <- melt(mat.K27)
ggplot(cur_df) + geom_point(aes(Var2, value, colour = Var1)) + ggtitle("H3K27ac")

# Generate matrix to store the average K27 enrichment
mat.K4 <- matrix(data = NA, ncol = length(seqlevels(H3K4me3$windows)), nrow = 2)
rownames(mat.K4) <- c("Spermatocytes", "Spermatids")
colnames(mat.K4) <- seqlevels(H3K4me3$windows)

# Loop trough the different chromosomes and find K4 enrichemnt for the whole chromosome
for(i in seqlevels(prom)){
  cur_windows <- seqnames(H3K4me3$windows) == i
  mat.K4[1,i] <- mean(H3K4me3$enrichment[as.vector(cur_windows),"spermatocytes"])
  mat.K4[2,i] <- mean(H3K4me3$enrichment[as.vector(cur_windows),"spermatids"])
}
mat.K4 <- mat.K4[,!is.na(mat.K4[1,])]
mat.K4 <- mat.K4[,!(colnames(mat.K4) == "MT")]
cur_df <- melt(mat.K4)
ggplot(cur_df) + geom_point(aes(Var2, value, colour = Var1)) + ggtitle("H3K4me3")
```
