---
title: "Figure 6"
author: "eling01 Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_6.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data and libraries

```{r data, message=FALSE}
# Libraries
library(scater)
library(ggplot2)
library(reshape2)
library(plyr)
library(openxlsx)
library(cowplot)
library(viridis)
library(pheatmap)
library(goseq)
library(GO.db)
library(org.Mm.eg.db)
library(ggsci)
source("../Functions/auxiliary.R")

# Single cell data
sce <- readRDS("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all_clusters.rds")

# Select Tc1, Tc0 and B6
sce <- sce[,grepl("B6|Tc0|Tc1", colData(sce)$Sample)]
sce <- normalise(sce)

# Read in length of genes
genelength <- read.table("../Data/Genelength.txt", header = TRUE, sep = "\t")

# Read in genenames
genenames <- read.table("../Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
```

# Ratios of cell types in each sample

```{r ratios}
# Calculate proprtions of cell types in each sample
df <- data.frame(sample = sub("_.*", "", colData(sce)$Sample),
                 group = colData(sce)$AnnotatedClusters,
                 value = rep(1, ncol(sce)))

proportions <- with(df, table(sample, group))/plyr::count(df = df, vars = "sample")$freq
```

## Visualize percentages for germ cells

```{r, visualization}
cur_df <- proportions[c("B6", "Tc0", "Tc1"),1:22]
cur_df.melt <- melt(cur_df)
levels(cur_df.melt$sample) <- rev(c("B6", "Tc0", "Tc1"))
levels(cur_df.melt$group) <- levels(colData(sce)$AnnotatedClusters)

proportions.p <- ggplot(cur_df.melt) + geom_point(aes(group, rev(sample), size = value, fill = group), shape = 22) +
  scale_fill_manual(values = metadata(sce)$color_vector) + 
  theme(panel.background = element_blank(), axis.title = element_blank(),
        axis.text.x = element_text(angle = 45, size = 12, hjust = 1),
        axis.text.y = element_text(size = 12))
```

# Display Genes that are differentially expressed between Tc0 and Tc1

```{r genes}
cur_gene <- "Aurkc"
cur_groups <- c("Early_Spermatocytes_1", "Early_Spermatocytes_2",
                "Mid_Spermatocytes_1", "Mid_Spermatocytes_2",
                "Late_Spermatocytes_1", "Late_Spermatocytes_2", "Meiosis")

df <- data.frame(sample = sub("_.*", "", colData(sce)$Sample)[colData(sce)$AnnotatedClusters %in% cur_groups],
                 group = colData(sce)$AnnotatedClusters[colData(sce)$AnnotatedClusters %in% cur_groups],
                 gene = logcounts(sce)[rowData(sce)$ID[rowData(sce)$Symbol == cur_gene],
                                       colData(sce)$AnnotatedClusters %in% cur_groups])

Aurkc <- ggplot(df) + geom_boxplot(aes(group, gene, fill = sample)) + 
  ylab(paste(cur_gene, "Expression")) +
  theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, size = 12, hjust = 1),
        axis.text.y = element_text(size = 12)) + scale_fill_manual(values = c("gray25", "snow4", "sienna3"))

# Calculate proportions of Aurkc expressing cells
Aurkc.prop <- ddply(df, .(sample, group), summarize, prop=sum(gene > 0))$prop/plyr::count(df = df, vars = c("sample", "group"))$freq
names(Aurkc.prop) <-  ddply(df, .(sample, group), summarize, prop=sum(gene > 0))$group
Aurkc.prop

# Boxplot of cells that express Aurkc
cur_sce <- sce[,logcounts(sce)[rowData(sce)$Symbol == "Aurkc",] > 0]
df <- data.frame(sample = sub("_.*", "", 
                              colData(cur_sce)$Sample)[colData(cur_sce)$AnnotatedClusters %in% cur_groups],
                 group = colData(cur_sce)$AnnotatedClusters[colData(cur_sce)$AnnotatedClusters %in% cur_groups],
                 gene = logcounts(cur_sce)[rowData(cur_sce)$ID[rowData(cur_sce)$Symbol == cur_gene],
                                       colData(cur_sce)$AnnotatedClusters %in% cur_groups])

Aurkc.expr <- ggplot(df) + geom_boxplot(aes(group, gene, fill = sample)) + 
  ylab(paste(cur_gene, "Expression")) +
  theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, size = 12, hjust = 1),
        axis.text.y = element_text(size = 12)) + scale_fill_manual(values = c("gray25", "snow4", "sienna3"))
Aurkc.expr

```

# Validation using C1 scRNAseq data

```{r C1}
# Read in C1 data
```

# Variability and mean expression of Chr21, X and 16 over the differentiation process

## Expression

```{r Chr21 inactivation and reactivation}
cur_sce <- sce[,grepl("Tc1", colData(sce)$Sample) & 
                 !(colData(sce)$AnnotatedClusters %in% 
                     c("Leydig", "Sertoli", "Outliers", "Endothelial_cells"))]
cur_sce <- normalize(cur_sce)

# Remove lowly expressed genes
genes <- rowData(cur_sce)$ID[
  apply(logcounts(cur_sce)[,colData(cur_sce)$AnnotatedClusters %in% 
          c("Spermatogonia", paste("S", 1:14, sep = ""))], 1,
              function(n){length(which(n > 0))}) >
    0.3*sum(colData(cur_sce)$AnnotatedClusters %in% 
              c("Spermatogonia", paste("S", 1:14, sep = "")))]

# Collect genes for some chromosomes
genes.X <- genes[genes %in% genenames[genenames[,3] == "X",1]]
genes.9 <- genes[genes %in% genenames[genenames[,3] == "9",1]]
genes.21 <- genes[grepl("ENSG", genes)]
genes.A <- genes[genes %in% genenames[genenames[,3] != "X" & 
                                        genenames[,3] != "Y" &
                                        genenames[,3] != "MT",1]]

# Calculate X:A ratio and 16:X ratio
ratio.X.A <- colMeans(as.matrix(logcounts(cur_sce)[genes.X,]))/
  colMeans(as.matrix(logcounts(cur_sce)[genes.A,]))
ratio.X.A[is.nan(ratio.X.A)] <- 0

ratio.21.A <- colMeans(as.matrix(logcounts(cur_sce)[genes.21,]))/
  colMeans(as.matrix(logcounts(cur_sce)[genes.A,]))
ratio.21.A[is.nan(ratio.21.A)] <- 0

ratio.9.A <- colMeans(as.matrix(logcounts(cur_sce)[genes.9,]))/
  colMeans(as.matrix(logcounts(cur_sce)[genes.A,]))
ratio.9.A[is.nan(ratio.9.A)] <- 0

# Build mean ratio matrix
df <- data.frame(ratios = c(ratio.X.A, ratio.9.A, ratio.21.A),
                 groups = factor(c(as.character(colData(cur_sce)$AnnotatedClusters), 
                            as.character(colData(cur_sce)$AnnotatedClusters),
                            as.character(colData(cur_sce)$AnnotatedClusters)), 
                            levels = levels(colData(cur_sce)$AnnotatedClusters)),
                 chr = c(rep("X", length(ratio.X.A)), 
                         rep("9", length(ratio.9.A)),
                         rep("21", length(ratio.21.A))))

# visualize in form of boxplots
x.21.inactivation <- ggplot(df) + geom_boxplot(aes(groups, ratios, fill = chr), outlier.colour = NA) + 
  scale_fill_jama() + ylab("Chr:Autosome ratio") + 
    theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, size = 12, hjust = 1),
        axis.text.y = element_text(size = 12))
ggsave(filename = "/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_6/21dynamics.pdf", 
       x.21.inactivation, width = 12, height = 4)
```

# Variability for X, 9 and 21

```{r}
# Read in variability measures
epsilons <- readRDS("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Variability/data/epsilons_Tc1.rds")

# Order variability measures
epsilons <- epsilons[,levels(colData(cur_sce)$AnnotatedClusters)[
  levels(colData(cur_sce)$AnnotatedClusters) %in% colnames(epsilons)
]]

# Plot variability measures for the 3 chromossomes
df.21 <- melt(epsilons[genes.21,])
df.X <- melt(epsilons[genes.X,])
df.9 <- melt(epsilons[genes.9,])

df <- rbind(df.21, df.X, df.9)
df$chr <- c(rep("21", nrow(df.21)), rep("X", nrow(df.X)), rep("9", nrow(df.9)))

x.21.variability <- ggplot(df) + geom_boxplot(aes(Var2, value, fill = chr), outlier.colour = NA) + 
  scale_fill_jama() + ylab("Variabilty") + 
    theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, size = 12, hjust = 1),
        axis.text.y = element_text(size = 12))
ggsave(filename = "/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_6/21variability.pdf", 
       x.21.variability, width = 12, height = 4)
```

# Visualize heatmap of Chr21

```{r}
# Build mean expression matrix
df <- as.data.frame(t(as.matrix(logcounts(cur_sce)[
  grepl("ENSG", rownames(cur_sce)),])))
df$groups <- colData(cur_sce)$AnnotatedClusters
df.melt <- melt(df, id.vars = "groups")

# Collect mean expression for each gene in each group
mat <- ddply(df.melt, .(groups, variable), summarize, mean=mean(value))
mat.for.heatmap <- dcast(data = mat,formula = variable~groups,fun.aggregate = sum,value.var = "mean")
rownames(mat.for.heatmap) <- mat.for.heatmap$variable
mat.for.heatmap <- mat.for.heatmap[,-1]
mat.for.heatmap <- mat.for.heatmap[genes.21,]

# Order genes based on mean expression
mat.for.heatmap <- mat.for.heatmap[order(rowMeans(mat.for.heatmap), 
                                         decreasing = TRUE),]

pdf("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_6/Chr21_GeneExpression.pdf", onefile=FALSE,
    width = 12)
pheatmap(mat.for.heatmap, cluster_cols = FALSE, 
         cluster_rows = FALSE,
         show_colnames = FALSE,
         color = viridis(100), 
         annotation_col = data.frame(row.names = colnames(mat.for.heatmap),
                                     Cell_type = colnames(mat.for.heatmap)),
         annotation_colors = list(Cell_type = metadata(cur_sce)$color_vector),
         border_color = NA, 
         labels_row = rowData(cur_sce)$Symbol[match(rownames(mat.for.heatmap), 
                                                    rowData(cur_sce)$ID)])
dev.off()
```


# Save final figure

```{r final}
final <- plot_grid(proportions.p, Aurkc, ncol = 1, nrow = 2)
ggsave(filename = "/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_6.pdf", final, width = 8, height = 8)
```
