---
title: "Figure 5"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data and libraries

```{r data, message=FALSE}
library(rtracklayer)
library(EnrichedHeatmap)
library(GenomicFeatures)
library(EnsDb.Mmusculus.v79)
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
library(scater)
library(plyr)
library(reshape2)
library(pheatmap)
library(viridis)
library(plot3D)
library(ggsci)
library(cowplot)
library(openxlsx)
library(csaw)

# Generate feature annotation
prom <- promoters(genes(EnsDb.Mmusculus.v79))
prom <- prom[seqnames(prom) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(prom) <- c(as.character(1:19), "X", "Y", "MT")
prom.X <- prom[seqnames(prom) == "X"]
prom.Y <- prom[seqnames(prom) == "Y"]
prom.9 <- prom[seqnames(prom) == "9"]

# Make TXDB for mus musculus
black <- import("http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz")

# Read in sce data
sce <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")
sce <- sce[,grepl("B6", colData(sce)$Sample) & 
             colData(sce)$AnnotatedClusters %in% levels(colData(sce)$AnnotatedClusters)[1:23]]
sce <- normalize(sce)

# Select XChr genes
genenames <- read.table("../Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
sce.XChr <- sce[rowData(sce)$ID %in% 
                  prom.X$gene_id,]
sce.XChr <- normalize(sce.XChr)

# Bulk data
bulk <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/bulk/norm/1st_wave_bulk_norm_reverse-stranded.rds")

# Meta info
meta <- as.data.frame(read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/ArrayExpress/Bulk/sdrf.xlsx"))

# Read in genelists
retro <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/X_retrogenes.xlsx")
retro <- retro[grepl("ENS", retro$Parental.Gene),]

K27writers.Rnf8 <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/Adams_et_al_down_in_RS_after_KO.xlsx", sheet = 1)
K27writers.Scml2 <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/Adams_et_al_down_in_RS_after_KO.xlsx", sheet = 2)
```

# Visualize the inactivation and reactivation of X and Y

```{r in-reactivation}
# Remove lowly expressed genes
genes <- rowData(sce)$ID[
  apply(logcounts(sce)[,colData(sce)$AnnotatedClusters %in% 
          c("Spermatogonia", paste("S", 1:14, sep = ""))], 1,
              function(n){length(which(n > 0))}) >
    0.3*sum(colData(sce)$AnnotatedClusters %in% 
              c("Spermatogonia", paste("S", 1:14, sep = "")))]

# Collect genes for some chromosomes
genes.X <- genes[genes %in% genenames[genenames[,3] == "X",1]]
genes.Y <- genes[genes %in% genenames[genenames[,3] == "Y",1]]
genes.9 <- genes[genes %in% genenames[genenames[,3] == "9",1]]
genes.A <- genes[genes %in% genenames[genenames[,3] != "X" & 
                                        genenames[,3] != "Y" &
                                        genenames[,3] != "MT",1]]

# Calculate X:A ratio and 16:X ratio
ratio.X.A <- colMeans(as.matrix(logcounts(sce)[genes.X,]))/
  colMeans(as.matrix(logcounts(sce)[genes.A,]))
ratio.X.A[is.nan(ratio.X.A)] <- 0

ratio.Y.A <- colMeans(as.matrix(logcounts(sce)[genes.Y,]))/
  colMeans(as.matrix(logcounts(sce)[genes.A,]))
ratio.Y.A[is.nan(ratio.Y.A)] <- 0

ratio.9.A <- colMeans(as.matrix(logcounts(sce)[genes.9,]))/
  colMeans(as.matrix(logcounts(sce)[genes.A,]))
ratio.9.A[is.nan(ratio.9.A)] <- 0

# Build mean ratio matrix
df <- data.frame(ratios = c(ratio.X.A, ratio.Y.A, ratio.9.A),
                 groups = factor(c(as.character(colData(sce)$AnnotatedClusters), 
                            as.character(colData(sce)$AnnotatedClusters),
                            as.character(colData(sce)$AnnotatedClusters)), 
                            levels = levels(colData(sce)$AnnotatedClusters)),
                 chr = c(rep("X", length(ratio.X.A)), rep("Y", length(ratio.Y.A)),
                         rep("9", length(ratio.9.A))))

# visualize in form of boxplots
x.inactivation <- ggplot(df) + geom_boxplot(aes(groups, ratios, fill = chr)) + 
  scale_fill_jama() + ylab("Chr:Autosome ratio") + 
    theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, size = 12, hjust = 1),
        axis.text.y = element_text(size = 12))
ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/Xdynamics.pdf", 
       x.inactivation, width = 12, height = 7)
```

# Visualize the K9 and K4 signal coming from the different chromosomes

```{r}
H3K4me3.normReads <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/DB/H3K4me3/H3K4me326_1000kb_bins_normReads.rds")
H3K9me3.normReads <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/DB/H3K9me3/H3K9me326_1000kb_bins_normReads.rds")

# For K9 signal
mat.K9 <- matrix(data = NA, ncol = length(seqlevels(H3K9me3.normReads$windows)), nrow = 2)
rownames(mat.K9) <- c("Spermatocytes", "Spermatids")
colnames(mat.K9) <- seqlevels(H3K9me3.normReads$windows)
mat.K9 <- mat.K9[,!grepl("GL|JH", colnames(mat.K9))]
mat.K9 <- mat.K9[,!(colnames(mat.K9) == "MT")]

# Loop trough the different chromosomes and find K4 enrichemnt for the whole chromosome
for(i in colnames(mat.K9)){
  cur_windows <- seqnames(H3K9me3.normReads$windows) == i
  mat.K9[1,i] <- mean(rowMeans(H3K9me3.normReads$normReads[as.vector(cur_windows),c(1,3)]))
  mat.K9[2,i] <- mean(rowMeans(H3K9me3.normReads$normReads[as.vector(cur_windows),c(2,3)]))
}

cur_df <- melt(mat.K9)
levels(cur_df$Var2) <- c(as.character(1:19), "X", "Y")

K9.signal <- ggplot(cur_df) + geom_point(aes(Var2, value, colour = Var1), size = 2) + 
  ggtitle("H3K9me3") + scale_color_aaas(name = "Cell type") + 
  ylab("Average log(CPM)") + xlab("")
ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/K9signal.pdf", 
       K9.signal, width = 7, height = 4)

# For K4 signal
mat.K4 <- matrix(data = NA, ncol = length(seqlevels(H3K4me3.normReads$windows)), nrow = 2)
rownames(mat.K4) <- c("Spermatocytes", "Spermatids")
colnames(mat.K4) <- seqlevels(H3K4me3.normReads$windows)
mat.K4 <- mat.K4[,!grepl("GL|JH", colnames(mat.K4))]
mat.K4 <- mat.K4[,!(colnames(mat.K4) == "MT")]

# Loop trough the different chromosomes and find K4 enrichemnt for the whole chromosome
for(i in colnames(mat.K4)){
  cur_windows <- seqnames(H3K4me3.normReads$windows) == i
  mat.K4[1,i] <- mean(rowMeans(H3K4me3.normReads$normReads[as.vector(cur_windows),c(1,3)]))
  mat.K4[2,i] <- mean(rowMeans(H3K4me3.normReads$normReads[as.vector(cur_windows),c(2,3)]))
}

cur_df <- melt(mat.K4)
levels(cur_df$Var2) <- c(as.character(1:19), "X", "Y")

K4.signal <- ggplot(cur_df) + geom_point(aes(Var2, value, colour = Var1), size = 2) + 
  ggtitle("H3K4me3") + scale_color_aaas(name = "Cell type") + 
  ylab("Average log(CPM)") + xlab("")
ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/K4signal.pdf", 
       K9.signal, width = 7, height = 4)
```

# X chromosome

# Visualize the normalized H3K9me3 and H3K4me3 normalized counts in promoter regions

```{r}
# Find spermatid specific genes first
# Select X chromosome genes
bulk.X <- bulk[intersect(prom.X$gene_id, rownames(bulk)),]
bulk.X <- bulk.X[rowMeans(bulk.X) > 10,]

# Collect day info on libraries
df <- data.frame(library = meta$Source.Name,
                 day = meta$`Characteristics.[Age]`)
bulk.X <- bulk.X[,df$library]

# Order genes based on their logFC between spermatocytes (< P20)
# and spermatids (> P20)
# Perform differential expression testing
y <- DGEList(counts=bulk.X,
             group=ifelse(df$day < 20, "Spermatocytes", "Spermatids"))
y <- calcNormFactors(y)
design <- model.matrix(~0+y$samples$group)
colnames(design) <- c("Spermatids", "Spermatocytes")
y <- estimateDisp(y,design)
  
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmTreat(fit,coef=2, lfc = 2, 
                contrast = makeContrasts(Spermatids - Spermatocytes, 
                                         levels = design))
cur_markers <- topTags(qlf, n = nrow(qlf$table))$table
cur_markers <- cur_markers[rownames(bulk.X),]

bulk.X <- bulk.X[order(cur_markers$logFC, decreasing = FALSE),]
cur_markers <- cur_markers[rownames(bulk.X),]

# Save info if gene is spermatid specifc or not
spermatid.spec <- ifelse(cur_markers$logFC > 2 & cur_markers$FDR < 0.1, TRUE, FALSE)
```

Collect H3K4 and H3K9 signal for X chromosome promoters

```{r}
cur_prom <- prom.X[match(rownames(bulk.X), prom.X$gene_id)]
cur_df <- data.frame(row.names = rownames(bulk.X),
                     genenames = prom.X$gene_name[match(rownames(bulk.X), 
                                                        prom.X$gene_id)],
                     spermatid_specific = spermatid.spec,
                     retro = ifelse(rownames(bulk.X) %in% retro$Parental.Gene, 1, 0),
                     Rnf8 = ifelse(prom.X$gene_name[match(rownames(bulk.X), 
                                                        prom.X$gene_id)] %in% 
                                     K27writers.Rnf8$gene_id, 1, 0),
                     Scml2 = ifelse(prom.X$gene_name[match(rownames(bulk.X), 
                                                        prom.X$gene_id)] %in% 
                                     K27writers.Scml2$gene_id, 1, 0))


# K9 signal
olap.K9 <- findOverlaps(cur_prom, H3K9me3.normReads$windows, minoverlap = 500L)
cur_df$K9.spermatocytes <- rep(NA, nrow(cur_df))
cur_df$K9.spermatocytes[unique(queryHits(olap.K9))] <- ddply(data.frame(olap.K9, 
                enrichment.spermatocytes =
                  rowMeans(H3K9me3.normReads$normReads[subjectHits(olap.K9),
                                    c("spermatocytes_1", "spermatocytes_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

cur_df$K9.spermatids <- rep(NA, nrow(cur_df))
cur_df$K9.spermatids[unique(queryHits(olap.K9))] <- ddply(data.frame(olap.K9, 
                enrichment.spermatocytes =
                  rowMeans(H3K9me3.normReads$normReads[subjectHits(olap.K9),
                                    c("spermatids_1", "spermatids_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

# K4 signal
olap.K4 <- findOverlaps(cur_prom, H3K4me3.normReads$windows, minoverlap = 500L)
cur_df$K4.spermatocytes <- rep(NA, nrow(cur_df))
cur_df$K4.spermatocytes[unique(queryHits(olap.K4))] <- ddply(data.frame(olap.K4, 
                enrichment.spermatocytes =
                  rowMeans(H3K4me3.normReads$normReads[subjectHits(olap.K4),
                                    c("spermatocytes_1", "spermatocytes_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

cur_df$K4.spermatids <- rep(NA, nrow(cur_df))
cur_df$K4.spermatids[unique(queryHits(olap.K4))] <- ddply(data.frame(olap.K4, 
                enrichment.spermatocytes =
                  rowMeans(H3K4me3.normReads$normReads[subjectHits(olap.K4),
                                    c("spermatids_1", "spermatids_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

```

Plot heatmaps

```{r, fig.height=30, fig.width=10}
cur_df <- cur_df[!is.na(cur_df$K9.spermatocytes) & !is.na(cur_df$K4.spermatocytes),]
cur_df <- cur_df[order(as.numeric(cur_df$spermatid_specific), cur_df$K9.spermatocytes),]
cur_mat <- bulk.X[rownames(cur_df),]

# Full heatmap
pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/Bulk_heatmap_X.pdf", 
    width = 7, height = 12)
pheatmap(log2(cur_mat + 1), 
         cluster_cols = FALSE, 
         cluster_rows = FALSE, show_rownames = FALSE,
         color = viridis(100),
         border_color = NA,
         labels_col = paste("P", meta$`Characteristics.[Age]`, sep = ""),
         annotation_row = data.frame(row.names = rownames(cur_mat),
        H3K4me3.spermatocytes = cur_df[rownames(cur_mat),"K4.spermatocytes"],
        H3K9me3.spermatocytes = cur_df[rownames(cur_mat),"K9.spermatocytes"],
        retro = cur_df[rownames(cur_mat),"retro"],
        Rnf8 = cur_df[rownames(cur_mat),"Rnf8"],
        Scml2 = cur_df[rownames(cur_mat),"Scml2"]),       
        cellheight = 0.5, cellwidth = 8, fontsize = 7, 
        annotation_colors = list(H3K9me3.spermatocytes = inferno(100)))
dev.off()

# Test for statistical enrichment of K4 and K9
# Collect mean K4 and K9 signal from spermtid specifc and non-specific genes  
ddply(cur_df, .(spermatid_specific), summarize, mean.K9.spermatocytes = mean(K9.spermatocytes, na.rm = TRUE),
      mean.K9.spermatids = mean(K9.spermatids, na.rm = TRUE),
      mean.K4.spermatocytes = mean(K4.spermatocytes, na.rm = TRUE),
      mean.K4.spermatids = mean(K4.spermatids, na.rm = TRUE))

# Fisher test
# Set up contigency table - K9 high > 0.5
K9.spermatocytes.table <- matrix(data = NA, ncol = 2, nrow = 2)
colnames(K9.spermatocytes.table) <- c("specific", "non-specific")
rownames(K9.spermatocytes.table) <- c("K9-high", "K9-low")
K9.spermatocytes.table[1,1] <- sum(cur_df$K9.spermatocytes > 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K9.spermatocytes.table[1,2] <- sum(cur_df$K9.spermatocytes > 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == FALSE) 
K9.spermatocytes.table[2,1] <- sum(cur_df$K9.spermatocytes < 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K9.spermatocytes.table[2,2] <- sum(cur_df$K9.spermatocytes < 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == FALSE)

fisher.test(K9.spermatocytes.table, alternative = "greater")

# Set up contigency table - K4 high > 0.5
K4.spermatocytes.table <- matrix(data = NA, ncol = 2, nrow = 2)
colnames(K4.spermatocytes.table) <- c("specific", "non-specific")
rownames(K4.spermatocytes.table) <- c("K4-high", "K4-low")
K4.spermatocytes.table[1,1] <- sum(cur_df$K4.spermatocytes > 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K4.spermatocytes.table[1,2] <- sum(cur_df$K4.spermatocytes > 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == FALSE) 
K4.spermatocytes.table[2,1] <- sum(cur_df$K4.spermatocytes < 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K4.spermatocytes.table[2,2] <- sum(cur_df$K4.spermatocytes < 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == FALSE)

fisher.test(K4.spermatocytes.table, alternative = "less")

```

# Y chromosome

# Visualize the normalized H3K9me3 and H3K4me3 normalized counts in promoter regions

```{r}
# Find spermatid specific genes first
# Select X chromosome genes
bulk.Y <- bulk[intersect(prom.Y$gene_id, rownames(bulk)),]
bulk.Y <- bulk.Y[rowMeans(bulk.Y) > 10,]

# Collect day info on libraries
df <- data.frame(library = meta$Source.Name,
                 day = meta$`Characteristics.[Age]`)
bulk.Y <- bulk.Y[,df$library]

# Order genes based on their logFC between spermatocytes (< P20)
# and spermatids (> P20)
# Perform differential expression testing
y <- DGEList(counts=bulk.Y,
             group=ifelse(df$day < 20, "Spermatocytes", "Spermatids"))
y <- calcNormFactors(y)
design <- model.matrix(~0+y$samples$group)
colnames(design) <- c("Spermatids", "Spermatocytes")
y <- estimateDisp(y,design)
  
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmTreat(fit,coef=2, lfc = 2, 
                contrast = makeContrasts(Spermatids - Spermatocytes, 
                                         levels = design))
cur_markers <- topTags(qlf, n = nrow(qlf$table))$table
cur_markers <- cur_markers[rownames(bulk.Y),]

bulk.Y <- bulk.Y[order(cur_markers$logFC, decreasing = FALSE),]
cur_markers <- cur_markers[rownames(bulk.Y),]

# Save info if gene is spermatid specifc or not
spermatid.spec <- ifelse(cur_markers$logFC > 2 & cur_markers$FDR < 0.1, TRUE, FALSE)
```

Collect H3K4 and H3K9 signal for these promoters

```{r}
cur_prom <- prom.Y[match(rownames(bulk.Y), prom.Y$gene_id)]
cur_df <- data.frame(row.names = rownames(bulk.Y),
                     genenames = prom.Y$gene_name[match(rownames(bulk.Y), 
                                                        prom.Y$gene_id)],
                     spermatid_specific = spermatid.spec)

# K9 signal
olap.K9 <- findOverlaps(cur_prom, H3K9me3.normReads$windows, minoverlap = 500L)
cur_df$K9.spermatocytes <- rep(NA, nrow(cur_df))
cur_df$K9.spermatocytes[unique(queryHits(olap.K9))] <- ddply(data.frame(olap.K9, 
                enrichment.spermatocytes =
                  rowMeans(H3K9me3.normReads$normReads[subjectHits(olap.K9),
                                    c("spermatocytes_1", "spermatocytes_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

cur_df$K9.spermatids <- rep(NA, nrow(cur_df))
cur_df$K9.spermatids[unique(queryHits(olap.K9))] <- ddply(data.frame(olap.K9, 
                enrichment.spermatocytes =
                  rowMeans(H3K9me3.normReads$normReads[subjectHits(olap.K9),
                                    c("spermatids_1", "spermatids_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

# K4 signal
olap.K4 <- findOverlaps(cur_prom, H3K4me3.normReads$windows, minoverlap = 500L)
cur_df$K4.spermatocytes <- rep(NA, nrow(cur_df))
cur_df$K4.spermatocytes[unique(queryHits(olap.K4))] <- ddply(data.frame(olap.K4, 
                enrichment.spermatocytes =
                  rowMeans(H3K4me3.normReads$normReads[subjectHits(olap.K4),
                                    c("spermatocytes_1", "spermatocytes_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

cur_df$K4.spermatids <- rep(NA, nrow(cur_df))
cur_df$K4.spermatids[unique(queryHits(olap.K4))] <- ddply(data.frame(olap.K4, 
                enrichment.spermatocytes =
                  rowMeans(H3K4me3.normReads$normReads[subjectHits(olap.K4),
                                    c("spermatids_1", "spermatids_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

```

Plot heatmaps

```{r, fig.height=30, fig.width=10}
cur_df <- cur_df[!is.na(cur_df$K9.spermatocytes) & !is.na(cur_df$K4.spermatocytes),]
cur_df <- cur_df[order(as.numeric(cur_df$spermatid_specific), cur_df$K9.spermatocytes),]
cur_mat <- bulk.Y[rownames(cur_df),]

# Full heatmap
pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/Bulk_heatmap_Y.pdf", 
    width = 7, height = 6)
pheatmap(log2(cur_mat + 1), 
         cluster_cols = FALSE, 
         cluster_rows = FALSE, labels_row = cur_df$genenames,
         color = viridis(100),
         border_color = NA,
         labels_col = paste("P", meta$`Characteristics.[Age]`, sep = ""),
         annotation_row = data.frame(row.names = rownames(cur_mat),
        H3K4me3.spermatocytes = cur_df[rownames(cur_mat),"K4.spermatocytes"],
        H3K9me3.spermatocytes = cur_df[rownames(cur_mat),"K9.spermatocytes"]),       
        cellheight = 8, cellwidth = 8, fontsize = 7, 
        annotation_colors = list(H3K9me3.spermatocytes = inferno(100)))
dev.off()

# Statistical testing
K9.spermatocytes.table <- matrix(data = NA, ncol = 2, nrow = 2)
colnames(K9.spermatocytes.table) <- c("specific", "non-specific")
rownames(K9.spermatocytes.table) <- c("K9-high", "K9-low")
K9.spermatocytes.table[1,1] <- sum(cur_df$K9.spermatocytes > 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K9.spermatocytes.table[1,2] <- sum(cur_df$K9.spermatocytes > 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == FALSE) 
K9.spermatocytes.table[2,1] <- sum(cur_df$K9.spermatocytes < 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K9.spermatocytes.table[2,2] <- sum(cur_df$K9.spermatocytes < 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == FALSE)

fisher.test(K9.spermatocytes.table, alternative = "greater")

# Set up contigency table - K4 high > 0.5
K4.spermatocytes.table <- matrix(data = NA, ncol = 2, nrow = 2)
colnames(K4.spermatocytes.table) <- c("specific", "non-specific")
rownames(K4.spermatocytes.table) <- c("K4-high", "K4-low")
K4.spermatocytes.table[1,1] <- sum(cur_df$K4.spermatocytes > 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K4.spermatocytes.table[1,2] <- sum(cur_df$K4.spermatocytes > 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == FALSE) 
K4.spermatocytes.table[2,1] <- sum(cur_df$K4.spermatocytes < 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K4.spermatocytes.table[2,2] <- sum(cur_df$K4.spermatocytes < 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == FALSE)

fisher.test(K4.spermatocytes.table, alternative = "less")
```

# Chromosome 9

# Visualize the H3K9me3 and H3K4me3 normalized counts in promoter regions

```{r}
# Find spermatid specific genes first
# Select X chromosome genes
bulk.9 <- bulk[intersect(prom.9$gene_id, rownames(bulk)),]
bulk.9 <- bulk.9[rowMeans(bulk.9) > 10,]

# Collect day info on libraries
df <- data.frame(library = meta$Source.Name,
                 day = meta$`Characteristics.[Age]`)
bulk.9 <- bulk.9[,df$library]

# Order genes based on their logFC between spermatocytes (< P20)
# and spermatids (> P20)
# Perform differential expression testing
y <- DGEList(counts=bulk.9,
             group=ifelse(df$day < 20, "Spermatocytes", "Spermatids"))
y <- calcNormFactors(y)
design <- model.matrix(~0+y$samples$group)
colnames(design) <- c("Spermatids", "Spermatocytes")
y <- estimateDisp(y,design)
  
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmTreat(fit,coef=2, lfc = 2, 
                contrast = makeContrasts(Spermatids - Spermatocytes, 
                                         levels = design))
cur_markers <- topTags(qlf, n = nrow(qlf$table))$table
cur_markers <- cur_markers[rownames(bulk.9),]

bulk.9 <- bulk.9[order(cur_markers$logFC, decreasing = FALSE),]
cur_markers <- cur_markers[rownames(bulk.9),]

# Save info if gene is spermatid specifc or not
spermatid.spec <- ifelse(cur_markers$logFC > 2 & cur_markers$FDR < 0.1, TRUE, FALSE)
```

Collect H3K4 and H3K9 signal for these promoters

```{r}
cur_prom <- prom.9[match(rownames(bulk.9), prom.9$gene_id)]
cur_df <- data.frame(row.names = rownames(bulk.9),
                     genenames = prom.9$gene_name[match(rownames(bulk.9), 
                                                        prom.9$gene_id)],
                     spermatid_specific = spermatid.spec)


# K9 signal
olap.K9 <- findOverlaps(cur_prom, H3K9me3.normReads$windows, minoverlap = 500L)
cur_df$K9.spermatocytes <- rep(NA, nrow(cur_df))
cur_df$K9.spermatocytes[unique(queryHits(olap.K9))] <- ddply(data.frame(olap.K9, 
                enrichment.spermatocytes =
                  rowMeans(H3K9me3.normReads$normReads[subjectHits(olap.K9),
                                    c("spermatocytes_1", "spermatocytes_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

cur_df$K9.spermatids <- rep(NA, nrow(cur_df))
cur_df$K9.spermatids[unique(queryHits(olap.K9))] <- ddply(data.frame(olap.K9, 
                enrichment.spermatocytes =
                  rowMeans(H3K9me3.normReads$normReads[subjectHits(olap.K9),
                                    c("spermatids_1", "spermatids_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

# K4 signal
olap.K4 <- findOverlaps(cur_prom, H3K4me3.normReads$windows, minoverlap = 500L)
cur_df$K4.spermatocytes <- rep(NA, nrow(cur_df))
cur_df$K4.spermatocytes[unique(queryHits(olap.K4))] <- ddply(data.frame(olap.K4, 
                enrichment.spermatocytes =
                  rowMeans(H3K4me3.normReads$normReads[subjectHits(olap.K4),
                                    c("spermatocytes_1", "spermatocytes_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

cur_df$K4.spermatids <- rep(NA, nrow(cur_df))
cur_df$K4.spermatids[unique(queryHits(olap.K4))] <- ddply(data.frame(olap.K4, 
                enrichment.spermatocytes =
                  rowMeans(H3K4me3.normReads$normReads[subjectHits(olap.K4),
                                    c("spermatids_1", "spermatids_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

# Statistical testing
K9.spermatocytes.table <- matrix(data = NA, ncol = 2, nrow = 2)
colnames(K9.spermatocytes.table) <- c("specific", "non-specific")
rownames(K9.spermatocytes.table) <- c("K9-high", "K9-low")
K9.spermatocytes.table[1,1] <- sum(cur_df$K9.spermatocytes > 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K9.spermatocytes.table[1,2] <- sum(cur_df$K9.spermatocytes > 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == FALSE) 
K9.spermatocytes.table[2,1] <- sum(cur_df$K9.spermatocytes < 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K9.spermatocytes.table[2,2] <- sum(cur_df$K9.spermatocytes < 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == FALSE)

fisher.test(K9.spermatocytes.table, alternative = "greater")

# Set up contigency table - K4 high > 0.5
K4.spermatocytes.table <- matrix(data = NA, ncol = 2, nrow = 2)
colnames(K4.spermatocytes.table) <- c("specific", "non-specific")
rownames(K4.spermatocytes.table) <- c("K4-high", "K4-low")
K4.spermatocytes.table[1,1] <- sum(cur_df$K4.spermatocytes > 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K4.spermatocytes.table[1,2] <- sum(cur_df$K4.spermatocytes > 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == FALSE) 
K4.spermatocytes.table[2,1] <- sum(cur_df$K4.spermatocytes < 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K4.spermatocytes.table[2,2] <- sum(cur_df$K4.spermatocytes < 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == FALSE)

fisher.test(K4.spermatocytes.table, alternative = "less")
```

