---
title: "Figure 7"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data and libraries

Aaron: Just use scaledAverage - don't need to care about fancy normalization schemes
regionCounts()

```{r data, message=FALSE}
library(rtracklayer)
library(EnrichedHeatmap)
library(GenomicFeatures)
library(EnsDb.Mmusculus.v79)
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
library(plyr)
library(reshape2)
library(pheatmap)
library(viridis)
library(plot3D)
library(ggsci)
library(cowplot)
library(openxlsx)
library(csaw)
library(edgeR)
library(Rsamtools)
library(Gviz)
source("../Functions/auxiliary.R")

# Generate feature annotation
prom <- promoters(genes(EnsDb.Mmusculus.v79))
prom <- prom[seqnames(prom) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(prom) <- c(as.character(1:19), "X", "Y", "MT")
prom.X <- prom[seqnames(prom) == "X"]
prom.Y <- prom[seqnames(prom) == "Y"]
prom.9 <- prom[seqnames(prom) == "9"]

# K9 files
bam.files.K9 <- list.files("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K9me3/", full.names = TRUE,
                        pattern = paste(".+JP26.+bam$", sep = ""))

# K4 files
bam.files.K4 <- list.files("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K4me3/", full.names = TRUE,
                        pattern = paste(".+JP26.+bam$", sep = ""))

# Define conditions
batch <- gl(2, 2)
treatment <- rep(c("spermatocytes", "spermatids"), 2)

# Blacklisted regions
black <- import("http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz")
seqlevels(black) <- sub("chr", "", seqlevels(black))
seqnames(black) <- sub("chr", "", seqnames(black))

# Parameters for reading bam files
param <- readParam(dedup=TRUE, minq=10, discard=black, pe="both", max.frag=1000)

# Read in genelists
retro <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/X_retrogenes.xlsx")
retro <- retro[grepl("ENS", retro$Parental.Gene),]

K27writers.Rnf8 <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/Adams_et_al_down_in_RS_after_KO.xlsx", sheet = 1)
K27writers.Scml2 <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/Adams_et_al_down_in_RS_after_KO.xlsx", sheet = 2)

spermatid.specific <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_6/SpermatidSpecificGenes.xlsx")

# Read in gene annotations
genenames <- read.table("../Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
rownames(genenames) <- genenames$Gene.stable.ID

# Create objects for visualziation
gax <- GenomeAxisTrack(col="black", fontsize=15, size=2)
gr <- getGeneRegionTrackForGviz(EnsDb.Mmusculus.v79)
options(ucscChromosomeNames = FALSE)
```

# Visualize the K9 and K4 signal coming from the different chromosomes

```{r}
# Count reads per chromosome
header <- scanBamHeader(bam.files.K9[1])
sizes <- header$`../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K9me3//do18097_H3K9me3_spermatocytes_ab8898_mm10C57BL6_JP26_18_5374_CRI0sorted.bam`$targets
sizes <- sizes[!grepl("JH|GL|MT", names(sizes))]

# Generate Chr ranges
chr.ranges <- GRanges(seqnames = names(sizes), 
                      ranges = IRanges(start = rep(1, length(sizes)),
                                       end = sizes))

# Count reads per chromosome
out.K9 <- regionCounts(bam.files.K9, regions = chr.ranges, param=param)

# Calculate RPKM
cur_data <- assays(out.K9)$counts
rownames(cur_data) <- names(sizes)
rpm <- t(t(cur_data)/(colSums(cur_data)/1000000))
rpkm <- rpm/(sizes/1000)
colnames(rpkm) <- paste(treatment, batch, sep = "_")

cur_data.melt <- melt(rpkm)
cur_data.melt$cell_type <- sapply(as.character(cur_data.melt$Var2), 
                                  function(n){unlist(strsplit(n, "_"))[1]})
cur_data.melt$replicate <- sapply(as.character(cur_data.melt$Var2), 
                                  function(n){unlist(strsplit(n, "_"))[2]})
levels(cur_data.melt$Var1) <- c(as.character(1:19), "X", "Y")

K9.signal <- ggplot(cur_data.melt) + geom_point(aes(Var1, value, 
                                                    colour = cell_type,
                                                    shape = replicate), size = 2) + 
  ggtitle("H3K9me3") + scale_color_aaas(name = "Cell type") + 
  ylab("RPKM") + xlab("")


ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/K9signal.pdf", 
       K9.signal, width = 7, height = 4)

# For K4 signal
# Count reads per chromosome
out.K4 <- regionCounts(bam.files.K4, regions = chr.ranges, param=param)

# Calculate RPKM
cur_data <- assays(out.K4)$counts
rownames(cur_data) <- names(sizes)
rpm <- t(t(cur_data)/(colSums(cur_data)/1000000))
rpkm <- rpm/(sizes/1000)
colnames(rpkm) <- paste(treatment, batch, sep = "_")

cur_data.melt <- melt(rpkm)
cur_data.melt$cell_type <- sapply(as.character(cur_data.melt$Var2), 
                                  function(n){unlist(strsplit(n, "_"))[1]})
cur_data.melt$replicate <- sapply(as.character(cur_data.melt$Var2), 
                                  function(n){unlist(strsplit(n, "_"))[2]})
levels(cur_data.melt$Var1) <- c(as.character(1:19), "X", "Y")

K4.signal <- ggplot(cur_data.melt) + geom_point(aes(Var1, value, 
                                                    colour = cell_type,
                                                    shape = replicate), size = 2) + 
  ggtitle("H3K9me3") + scale_color_aaas(name = "Cell type") + 
  ylab("RPKM") + xlab("")

ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/K4signal.pdf", 
       K9.signal, width = 7, height = 4)
```

# Analyse heterogeneity of K9 on X chromosome

```{r}
# Bin the genome in 1000Kb windows
bins.K9 <- windowCounts(bam.files.K9, bin = TRUE, width = 1000, param=param)
bins.K4 <- windowCounts(bam.files.K4, bin = TRUE, width = 1000, param=param)

# Filter X chromosome bins
bins.K9.X <- bins.K9[as.vector(seqnames(rowRanges(bins.K9)) == "X")]

# Plot cumulative frequency
df.K9.X <- data.frame(spermatocytes_1 =
                      cumsum(assays(bins.K9.X)$counts[
                        order(assays(bins.K9.X)$counts[,1], 
                              decreasing = TRUE),1]),
                      spermatocytes_2 =
                      cumsum(assays(bins.K9.X)$counts[
                        order(assays(bins.K9.X)$counts[,3], 
                              decreasing = TRUE),3]),
                      spermatids_1 =
                      cumsum(assays(bins.K9.X)$counts[
                        order(assays(bins.K9.X)$counts[,2], 
                              decreasing = TRUE),2]),
                      spermatids_2 =
                      cumsum(assays(bins.K9.X)$counts[
                        order(assays(bins.K9.X)$counts[,4], 
                              decreasing = TRUE),4]))
df.K9.X <- t(t(df.K9.X)/as.numeric(tail(df.K9.X, 1)))
df.K9.X.melt <- melt(df.K9.X)

cumsum.K9.X <- ggplot(df.K9.X.melt) + geom_line(aes(Var1, value, colour = Var2))
ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/Cumsum_K9_X.pdf", 
       cumsum.K9.X, width = 7, height = 4)
  
# Compare to Chr9
bins.K9.9 <- bins.K9[as.vector(seqnames(rowRanges(bins.K9)) == "9")]

# Plot cumulative frequency
df.K9.9 <- data.frame(spermatocytes_1 =
                      cumsum(assays(bins.K9.9)$counts[
                        order(assays(bins.K9.9)$counts[,1], 
                              decreasing = TRUE),1]),
                      spermatocytes_2 =
                      cumsum(assays(bins.K9.9)$counts[
                        order(assays(bins.K9.9)$counts[,3], 
                              decreasing = TRUE),3]),
                      spermatids_1 =
                      cumsum(assays(bins.K9.9)$counts[
                        order(assays(bins.K9.9)$counts[,2], 
                              decreasing = TRUE),2]),
                      spermatids_2 =
                      cumsum(assays(bins.K9.9)$counts[
                        order(assays(bins.K9.9)$counts[,4], 
                              decreasing = TRUE),4]))
df.K9.9 <- t(t(df.K9.9)/as.numeric(tail(df.K9.9, 1)))
df.K9.9.melt <- melt(df.K9.9)

cumsum.K9.9 <- ggplot(df.K9.9.melt) + geom_line(aes(Var1, value, colour = Var2))
ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/Cumsum_K9_9.pdf", 
       cumsum.K9.9, width = 7, height = 4)
```

# Find regions with highest K9

```{r}
# Spermatocytes
cur_high <- rowRanges(bins.K9.X)[order(assays(bins.K9.X)$counts[,1], decreasing = TRUE)]
cur_high <- cur_high[1:1000]
cur_high.merge <- mergeWindows(cur_high, tol = 500)

# Order by size
cur_high.merge <- cur_high.merge$region[order(cur_high.merge$region@ranges@width, 
                                              decreasing = TRUE)]
cur_high.merge

# The second hit is right next to Akap4. Number 8 is close to Tsga8 and on Dmd.

# Save as bedgraph
export(cur_high.merge, "../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/Spermatocytes_1_highK9regions.bed")

# Spermatids
cur_high <- rowRanges(bins.K9.X)[order(assays(bins.K9.X)$counts[,2], decreasing = TRUE)]
cur_high <- cur_high[1:1000]
cur_high.merge <- mergeWindows(cur_high, tol = 500)

# Order by size
cur_high.merge <- cur_high.merge$region[order(cur_high.merge$region@ranges@width, 
                                              decreasing = TRUE)]
cur_high.merge

# Save as bedgraph
export(cur_high.merge, "../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/Spermatids_1_highK9regions.bed")
```

# Count reads in promoter regions

```{r}
prom.X.K9 <- regionCounts(bam.files.K9, regions = prom.X, param=param)
prom.X.K4 <- regionCounts(bam.files.K4, regions = prom.X, param=param)

# Generate RPM per promoter - each promoter is 2200bp wide
cur_counts.K9 <- assays(prom.X.K9)$counts
cur_rpm.K9 <- t(t(cur_counts.K9)/(colSums(assays(bins.K9)$counts)/1000000))

cur_counts.K4 <- assays(prom.X.K4)$counts
cur_rpm.K4 <- t(t(cur_counts.K4)/(colSums(assays(bins.K4)$counts)/1000000))

# Compute average between spermatocytes and spermatids
df <- data.frame(Spermatocytes.K9 = rowMeans(cur_rpm.K9[,c(1,3)]),
                 Spermatids.K9 = rowMeans(cur_rpm.K9[,c(2,4)]),
                 avg.K9 = rowMeans(cur_rpm.K9),
                 Spermatocytes.K4 = rowMeans(cur_rpm.K4[,c(1,3)]),
                 Spermatids.K4 = rowMeans(cur_rpm.K4[,c(2,4)]),
                 avg.K4 = rowMeans(cur_rpm.K4))
df$Symbol <- genenames[rownames(df),2]

# Order based on K9 signal in spermatocytes
df <- df[order(df$Spermatocytes.K9, decreasing = TRUE),]

# Annotate genenames
df$spermatid_specific <- ifelse(rownames(df) %in% spermatid.specific$ID, TRUE, FALSE)
df$Rnf8 <- ifelse(df$Symbol %in% K27writers.Rnf8, TRUE, FALSE)
df$Scml2 <- ifelse(df$Symbol %in% K27writers.Scml2, TRUE, FALSE)

# Plot results
spermatid.spec.K9 <- ggplot(df) + geom_boxplot(aes(spermatid_specific,
                            log2(Spermatocytes.K9 + 1),
                            fill = spermatid_specific)) + 
  scale_fill_jama()
ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/K9_spermatocytes_X.pdf", 
       spermatid.spec.K9, width = 4, height = 4)

# Wilcoxon test
wilcox.test(df$Spermatocytes.K9[df$spermatid_specific == TRUE],
            df$Spermatocytes.K9[df$spermatid_specific == FALSE])

spermatid.spec.K4 <- ggplot(df) + geom_point(aes(log2(avg.K4),
                            log2(Spermatids.K4/Spermatocytes.K4),
                            colour = spermatid_specific, 
                            size = spermatid_specific)) + 
  scale_size_manual(values = c(1,3)) + scale_color_jama()
ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/K4_spermatocytesVSspermatids_X.pdf", 
       spermatid.spec.K4, width = 7, height = 4)

```

# Visualize genes with high K9

```{r}
# Select genes of interest
genes <- rownames(df)[df$spermatid_specific == TRUE & 
                        log2(df$Spermatocytes.K9 + 1) > 2.5]

# Find promoter regions +- 1000bp
cur_prom <- promoters(genes(EnsDb.Mmusculus.v79), 
                  upstream = 10000, downstream = 5000)

# Loop over the genelist and save annotation tracks
for(i in 1:length(genes)){
  cur_gene <- genes[i]
  cur_name <- elementMetadata(cur_prom)$gene_name[
    elementMetadata(cur_prom)$gene_id == cur_gene]

  cur_region <- cur_prom[elementMetadata(cur_prom)$gene_id == cur_gene]

  collected.K9 <- list()
  lib.sizes <- colSums(assays(bins.K9)$counts)/1e6
  for (i in 1:length(bam.files.K9)) {
      reads <- extractReads(bam.file=bam.files.K9[i], cur_region, param=param)
      cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
      collected.K9[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,10),
           name=bam.files.K9[i],  col.axis="black", col.title="black",
           fill="dark red", col.histogram=NA, options(ucscChromosomeNames=FALSE))
  }
  collected.K9 <- collected.K9[c(1,3,2,4)]

  collected.K4 <- list()
  lib.sizes <- colSums(assays(bins.K4)$counts)/1e6
  for (i in 1:length(bam.files.K4)) {
      reads <- extractReads(bam.file=bam.files.K4[i], cur_region, param=param)
      cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
      collected.K4[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,10),
           name=bam.files.K4[i],  col.axis="black", col.title="black",
           fill="dark blue", col.histogram=NA, options(ucscChromosomeNames=FALSE))
  }
  collected.K4 <- collected.K4[c(1,3,2,4)]

  pdf(paste("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/", cur_name, "_spermatocytes.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K9[1:2], collected.K4[1:2]),
                                    GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
  dev.off()
  
  pdf(paste("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/", cur_name, "_spermatids.pdf", 
            sep = ""),
      width = 12, height = 7)
  plotTracks(c(gax, c(collected.K9[3:4], collected.K4[3:4]),
                                 GeneRegionTrack(gr)),
             chromosome=as.character(seqnames(cur_region)),
     from=start(cur_region), to=end(cur_region), showId=TRUE)
  dev.off()

}

```
