---
title: "Figure 4"
author: "nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_4/Fig_4.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data and libraries

```{r data, message=FALSE}
# Libraries
library(scater)
library(ggplot2)
library(reshape2)
library(plyr)
library(openxlsx)
library(cowplot)
library(viridis)
library(pheatmap)
library(goseq)
library(GO.db)
library(org.Mm.eg.db)
library(RColorBrewer)
library(edgeR)
source("../Functions/auxiliary.R")

# Single cell data
sce <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")
marker.genes <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/10X/Results/Marker_genes/Marker_genes_B6.rds")

# Select spermatogonia and spermatocytes
sce <- sce[,colData(sce)$Sample == "B6"]
sce <- normalize(sce)
```

# Spermatocytes

# Order cells in pseudotime

```{r PT}
sce.spermatocytes <- sce[, colData(sce)$AnnotatedClusters %in% levels(colData(sce)$AnnotatedClusters)[2:9]]
sce.spermatocytes <- normalize(sce.spermatocytes)

# Compute HVG
HVgenes <- HVG(sce.spermatocytes)

# Compute PCA
pca <- prcomp(t(logcounts(sce.spermatocytes)[HVgenes,]))

# Pseudo rank
prank <- PT(rd = pca$x[,1:3], clusters = colData(sce.spermatocytes)$AnnotatedClusters,
            col_vector = metadata(sce.spermatocytes)$color_vector)

```


```{r}
PT <-  prank[,"rank"]

set.seed(123)
y = rnorm(length(PT), mean = 0, sd = 0.01)

p.PT <- ggplot(data.frame(x = PT,
                  y = y,
                  group = colData(sce.spermatocytes)$AnnotatedClusters)) +
  geom_point(aes(x, y , fill = group), shape = 21, size = 3) + 
  scale_fill_manual(values = metadata(sce.spermatocytes)$color_vector) + ylim(c(-0.033,0.033)) +
  theme(legend.position = "none", panel.background = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.title.y = element_blank(), 
        axis.title.x = element_blank(), axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), axis.line.y = element_blank(),
        axis.text.y = element_blank(), axis.ticks.y = element_blank())

# Visualize number of genes expressed
number.genes <- apply(logcounts(sce.spermatocytes), 2, function(n){length(which(n>0))})

p.number.genes <- ggplot(
  data.frame(x = PT,
             y = number.genes,
             group = colData(sce.spermatocytes)$AnnotatedClusters)) + 
  geom_point(aes(x, y , fill = group), shape = 21, size = 3) +
  geom_smooth(aes(x = x, y = y), colour = "black") + ylab("# genes expressed") +
  scale_fill_manual(values = metadata(sce.spermatocytes)$color_vector) + 
  theme(legend.position = "none", panel.background = element_blank(), 
        panel.grid.major = element_line(colour="grey",size = rel(0.5)), 
        panel.grid.minor = element_blank(), axis.title.y = element_text(size = 15), 
        axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank())

# Plot marker genes
genes <- c("Hormad1", "Sycp3", "Pou5f2", "Tcte2")
all.max <- max(logcounts(sce.spermatocytes)[rowData(sce.spermatocytes)$ID[match(genes, rowData(sce.spermatocytes)$Symbol)],])

# Hormad1
Hormad1.p <- ggplot(data.frame(x = PT,
                  y = y,
                  gene = logcounts(sce.spermatocytes)[rowData(sce.spermatocytes)$ID[match(genes[1],
                                           rowData(sce.spermatocytes)$Symbol)],])) +
  geom_point(aes(x, y , fill = gene), shape = 21, size = 3) + 
  scale_fill_viridis(name = NULL) + ylim(c(-0.033,0.033)) + 
  ylab(genes[1]) +
  theme(panel.background = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.text.y = element_blank(), 
        axis.title.x = element_blank(), axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), axis.line.y = element_blank(),
        axis.title.y = element_text(size = 15, face = "bold"), axis.ticks.y = element_blank(),
        axis.line.x = element_blank())

# Sycp3
Sycp3.p <- ggplot(data.frame(x = PT,
                  y = y,
                  gene = logcounts(sce.spermatocytes)[rowData(sce.spermatocytes)$ID[match(genes[2],
                                           rowData(sce.spermatocytes)$Symbol)],])) +
  geom_point(aes(x, y , fill = gene), shape = 21, size = 3) + 
  scale_fill_viridis(name = NULL) + ylim(c(-0.033,0.033)) + 
  ylab(genes[2]) +
  theme(panel.background = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.text.y = element_blank(), 
        axis.title.x = element_blank(), axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), axis.line.y = element_blank(),
        axis.title.y = element_text(size = 15, face = "bold"), axis.ticks.y = element_blank(),
        axis.line.x = element_blank())

# Pou5f2
Pou5f2.p <- ggplot(data.frame(x = PT,
                  y = y,
                  gene = logcounts(sce.spermatocytes)[rowData(sce.spermatocytes)$ID[match(genes[3],
                                           rowData(sce.spermatocytes)$Symbol)],])) +
  geom_point(aes(x, y , fill = gene), shape = 21, size = 3) + 
  scale_fill_viridis(name = NULL) + ylim(c(-0.033,0.033)) + 
  ylab(genes[3]) +
  theme(panel.background = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.text.y = element_blank(), 
        axis.title.x = element_blank(), axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), axis.line.y = element_blank(),
        axis.title.y = element_text(size = 15, face = "bold"), axis.ticks.y = element_blank(),
        axis.line.x = element_blank())

# Tcte2
Tcte2.p <- ggplot(data.frame(x = PT,
                  y = y,
                  gene = logcounts(sce.spermatocytes)[rowData(sce.spermatocytes)$ID[match(genes[4],
                                           rowData(sce.spermatocytes)$Symbol)],])) +
  geom_point(aes(x, y , fill = gene), shape = 21, size = 3) + 
  scale_fill_viridis(name = NULL) + ylim(c(-0.033,0.033)) + 
  ylab(genes[4]) +
  theme(panel.background = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.text.y = element_blank(), 
        axis.title.x = element_blank(), axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), axis.line.y = element_blank(),
        axis.title.y = element_text(size = 15, face = "bold"), axis.ticks.y = element_blank(),
        axis.line.x = element_blank())

spermatocytes.p <- plot_grid(p.number.genes, Hormad1.p, Sycp3.p, Pou5f2.p, Tcte2.p, ncol = 1, nrow = 5)
ggsave("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_4/Fig_4.pdf", 
       spermatocytes.p, width = 6, height = 8)
```

# Map RA blocked cells to spermatocytes

```{r}
# Data from Chen et al, Cell Research
files <- list.files("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Data/Chen_CellResearch/GSE107644_RAW/", full.names = TRUE)

all_cells <- lapply(as.list(files), function(n){
  cur_file <- read.table(n, sep = "\t", header = TRUE)
  rownames(cur_file) <- cur_file[,1]
  cur_file <- cur_file[,-1]
})

all_cells <- do.call("cbind", all_cells)

# Form a sce 
sce_RA <- SingleCellExperiment(assays=list(counts=Matrix::Matrix(as.matrix(all_cells), sparse = TRUE)))
colData(sce_RA)$time_point <- factor(as.character(sapply(colnames(sce_RA), 
                                                         function(n){unlist(strsplit(n, "_"))[1]})),
                                        levels = paste("SpSC", c(0, 2:37), sep = "")) 
colData(sce_RA)$cell_type <- as.character(sapply(colnames(sce_RA), function(n){unlist(strsplit(n, "_"))[2]})) 

# Quality check 
sce_RA <- calculateQCMetrics(sce_RA)

# Remove 2 cells with few features detected
sce_RA <- sce_RA[,colData(sce_RA)$total_features_by_counts > 1250]

clusters <- quickCluster(sce_RA, method = "igraph", irlba.args = c("work" = 100), 
                         max.size = 2000)

sce_RA <- computeSumFactors(sce_RA, clusters=clusters)

sce_RA <- normalize(sce_RA, return_log = TRUE)

# Batch correction
sce.spermatocytes.1 <- sce.spermatocytes[match(intersect(rownames(sce_RA), rowData(sce.spermatocytes)$Symbol),
                                               rowData(sce.spermatocytes)$Symbol),]
sce_RA <- sce_RA[match(intersect(rownames(sce_RA), rowData(sce.spermatocytes.1)$Symbol), 
                       rownames(sce_RA)),]
rownames(sce_RA) <- rownames(sce.spermatocytes.1)
rowData(sce_RA)$Symbol <- rowData(sce.spermatocytes.1)$Symbol
rowData(sce_RA)$ID <- rowData(sce.spermatocytes.1)$ID

# Select specific cell types 
sce_mid_RA <- sce_RA[,grepl("D|MI|MII|RS1o2", colData(sce_RA)$cell_type) | colData(sce_RA)$cell_type == "lP" | colData(sce_RA)$cell_type == "mP"]
sce_mid_RA <- normalize(sce_mid_RA)

sce.single <- split.sce(sce = sce.spermatocytes.1, groups = unique(colData(sce.spermatocytes.1)$Library), 
                        colData.name = "Library")
sce.single$RA <- sce_mid_RA

corrected <- batch.correction(sce.single)

# Compute PCA
pca <- prcomp(t(corrected))

new_vector <- c(brewer.pal(n = 8, name = "Set1"), brewer.pal(n = 8, name = "Set2"))
names(new_vector) <- unique(colData(sce_mid_RA)$cell_type)
col_vector <- c(metadata(sce)$color_vector, new_vector)

ggplot() +
  geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatocytes.1),1],
                  PC2 = pca$x[1:ncol(sce.spermatocytes.1),2],
                  cluster = colData(sce.spermatocytes.1)$AnnotatedClusters),
             aes(PC1, PC2, colour = cluster)) + 
    geom_point(data = data.frame(PC1 = pca$x[(ncol(sce.spermatocytes.1)+1):(ncol(sce.spermatocytes.1)+ncol(sce_mid_RA)),1],
                  PC2 = pca$x[(ncol(sce.spermatocytes.1)+1):(ncol(sce.spermatocytes.1)+ncol(sce_mid_RA)),2],
                  cell_type = colData(sce_mid_RA)$cell_type),
             aes(PC1, PC2, colour = cell_type, size = 3)) +
  scale_color_manual(values = col_vector)

# Pseudo rank
prank <- PT(rd = pca$x[,1:3], clusters = c(as.character(colData(sce.spermatocytes.1)$AnnotatedClusters),
                                           colData(sce_mid_RA)$cell_type),
            col_vector = col_vector)

# Visualize number of genes expressed
cur_sce <- do.call("cbind", sce.single)
number.genes <- apply(logcounts(cur_sce), 2, function(n){length(which(n>0))})



```

# Correlation analysis to number of genes expressed

```{r}
# Remove lowly expressed genes
sce.spermatocytes <- sce.spermatocytes[Matrix::rowMeans(logcounts(sce.spermatocytes)) > 0.1,]

cur_mat <- rbind(logcounts(sce.spermatocytes), number.genes)

null.dist <- correlateNull(ncells = ncol(sce.spermatocytes), iters = 100000)
cors <- correlatePairs(cur_mat, null.dist=null.dist, pairings = list(c("number.genes"), rownames(cur_mat)))
cors$genename <- rowData(sce.spermatocytes)$Symbol[match(cors$gene2, rowData(sce.spermatocytes)$ID)]


# Perform DE analysis between spermatocytes and spermatids
sce.for.testing <- sce[match(rowData(sce.spermatocytes)$ID, 
                             rowData(sce)$ID),
                       !(colData(sce)$AnnotatedClusters %in% 
                            c("Endothelial_cells", "Outliers",
                              "Leydig", "Sertoli", "Spermatogonia"))]
sce.for.testing <- normalize(sce.for.testing)
DE <- findMarkers(sce.for.testing, 
    clusters = ifelse(colData(sce.for.testing)$AnnotatedClusters %in% 
                            levels(colData(sce)$AnnotatedClusters)[2:8], "Spermatocytes", "Spermatids"))

cors$FDR.DE <- DE$Spermatocytes$FDR[match(cors$gene2, 
                                          rownames(DE$Spermatocytes))]
cors$logFC.Spermatids <- DE$Spermatocytes$logFC.Spermatids[match(cors$gene2, 
                                          rownames(DE$Spermatocytes))]

cors.out <- cors[order(cors$rho, decreasing = TRUE),]
cors.out <- cors.out[cors$FDR < 0.1 & (cors$rho > 0.3 | cors$rho < -0.3), ]

write.xlsx(cors, "../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_tables/Table_S6.xlsx")

```

# Enrichment analysis for fertility genes

```{r enichment}
# Read in fertility related genes
# These can be downloaded at: https://media.nature.com/original/nature-assets/nm/journal/v14/n11/extref/nm.f.1895-S1.pdf
library(rJava)
library(tabulizer)
# There are pages which are diffucult to read in: 1, 16, 18, 21, 
# We extracted the information on these pages by hand
fertility.genes <- tabulizer::extract_tables("https://media.nature.com/original/nature-assets/nm/journal/v14/n11/extref/nm.f.1895-S1.pdf", pages = c(2:15, 17, 19, 20, 22))
fertility.genes.1 <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_4/NatMed_fertility_remaining.xlsx")

# Parse the output and save in new data.frame
fertility.genes.clean <- lapply(fertility.genes, function(n){
  cur_n <- n[,c(1,2,4)]
  cur_n[,1] <- sub("[\\)\\;].*", "", sub(".*\\(", "", cur_n[,1]))
  cur_n
})
df.fertility <- do.call("rbind", fertility.genes.clean)
colnames(df.fertility) <- colnames(fertility.genes.1)
df.fertility <- rbind(df.fertility, fertility.genes.1)
df.fertility <- df.fertility[match(unique(df.fertility$Genename), df.fertility$Genename),]
rownames(df.fertility) <- df.fertility$Genename
df.fertility[grepl("Male", df.fertility[,"Sex"]),"Sex"] <- "Male"
df.fertility[grepl("Female", df.fertility[,"Sex"]),"Sex"] <- "Female"
df.fertility[grepl("Both", df.fertility[,"Sex"]),"Sex"] <- "Both"

# Select the genes that show male or both fertility effect
df.fertility.male <- df.fertility[grepl("Both|Male", df.fertility[,2]),]

# Calculate the enrichment of the expressed fertility genes within the set of 
# group specific genes
df.stats <- matrix(NA, ncol = 3, nrow = length(marker.genes))
rownames(df.stats) <- names(marker.genes)
colnames(df.stats) <- c("Percent fertility in marker genes", "pvalue", "BHvalue")

for(i in 1:nrow(df.stats)){
  cur_sce <- sce[,sce$AnnotatedClusters == rownames(df.stats)[i]]
  cur_sce <- cur_sce[Matrix::rowMeans(logcounts(cur_sce)) > log2(2),]
  
  # Enrichment testing
  # Build contigency table
  cur_mat <- matrix(data = NA, 2, 2)
  rownames(cur_mat) <- c("Fertility-genes", "Non-fertility-genes")
  colnames(cur_mat) <- c("Group-spec", "Non-group-spec")
  cur_markers <- marker.genes[[rownames(df.stats)[i]]]$GeneName
  cur_markers <- cur_markers[cur_markers %in% rowData(cur_sce)$Symbol]
  cur_symbol <- rowData(cur_sce)$Symbol
  cur_mat[1,1] <- sum(cur_symbol %in% as.character(df.fertility$Genename) & 
                        cur_symbol %in% cur_markers)
  cur_mat[1,2] <- sum(cur_symbol %in% as.character(df.fertility$Genename) & 
                        !(cur_symbol %in% cur_markers))
  cur_mat[2,1] <- sum(!(cur_symbol %in% as.character(df.fertility$Genename)) & 
                        cur_symbol %in% cur_markers)
  cur_mat[2,2] <- sum(!(rowData(cur_sce)$Symbol %in% as.character(df.fertility$Genename)) & 
                        !(cur_symbol %in% cur_markers))
  # Odds ratio
  (cur_mat[1,1]/cur_mat[1,2])/(cur_mat[2,1]/cur_mat[2,2])
  test <- fisher.test(cur_mat, alternative = "greater")
  df.stats[i,2] <- test$p.value
  df.stats[i,1] <- cur_mat[1,1]/(cur_mat[1,1] + cur_mat[2,1])
}

df.stats[,3] <- p.adjust(df.stats[,2], method = "BH")
df.stats <- as.data.frame(df.stats)
df.stats$Celltype <- rownames(df.stats)

df.stats.red <- df.stats[grepl("Spermatocytes", rownames(df.stats)) |
                           rownames(df.stats) == "S1",]

# Statistic visualization
p.stats <- ggplot(df.stats.red) + geom_col(aes(factor(Celltype,
                        levels = c(paste("Spermatocytes_", 1:7, sep = ""), "S1")), -log10(BHvalue), 
                                           fill = Celltype)) +
  scale_fill_manual(values = metadata(sce)$color_vector) + 
  guides(fill = FALSE) + geom_hline(yintercept = 1, color = "dark red") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  ylab("-log10(BH-corrected p-value)")

ggsave("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_4/Fig_4_enrichment.pdf", 
       p.stats, width = 5, height = 4)

# See if genes are enriched in all marker genes
cur_sce <- sce[Matrix::rowMeans(logcounts(sce)) > 0,]
cur_mat <- matrix(data = NA, 2, 2)  
rownames(cur_mat) <- c("Fertility-genes", "Non-fertility-genes")
colnames(cur_mat) <- c("Group-spec", "Non-group-spec")
cur_markers <- lapply(marker.genes, function(n){n$GeneName})
cur_markers <- as.character(do.call("c", cur_markers))
cur_markers <- cur_markers[cur_markers %in% rowData(cur_sce)$Symbol]
cur_symbol <- rowData(cur_sce)$Symbol
cur_mat[1,1] <- sum(cur_symbol %in% as.character(df.fertility$Genename) & 
                      cur_symbol %in% cur_markers)
cur_mat[1,2] <- sum(cur_symbol %in% as.character(df.fertility$Genename) & 
                      !(cur_symbol %in% cur_markers))
cur_mat[2,1] <- sum(!(cur_symbol %in% as.character(df.fertility$Genename)) & 
                      cur_symbol %in% cur_markers)
cur_mat[2,2] <- sum(!(rowData(cur_sce)$Symbol %in% as.character(df.fertility$Genename)) & 
                      !(cur_symbol %in% cur_markers))
# Odds ratio
(cur_mat[1,1]/cur_mat[1,2])/(cur_mat[2,1]/cur_mat[2,2])
fisher.test(cur_mat, alternative = "greater")
```

# Visualize cluster specifc marker genes

```{r markers}
# Visualise the top15 markers per cell type and label based on their fertility phenotype
markers <- lapply(marker.genes[rownames(df.stats.red)], function(n){n$GeneName[1:15]})
markers <- as.character(do.call("c", markers))

for.heatmap <- logcounts(sce.spermatocytes)[rowData(sce.spermatocytes)$ID[match(markers,
                                                                    rowData(sce.spermatocytes)$Symbol)],
                                            order(prank[,"rank"])]
rownames(for.heatmap) <- markers
colnames(for.heatmap) <- colData(sce.spermatocytes)$Barcode[order(prank[,"rank"])]

# Marker genes to label
to.label <- c("Hmgb2", "Tex101", "Cks2", "Sycp1", "Sycp2", "Piwil1", "Ybx2", "Pou5f2", "Cdk1", "Aurka", "H3f3b")

pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_4/Fig_4_MarkerGenes.pdf", onefile = FALSE, height = 5, width = 10)
pheatmap(for.heatmap, show_colnames = FALSE, show_rownames = FALSE,
         cluster_rows = FALSE, cluster_cols = FALSE,
         color = colorRampPalette(c("#053061", "#4393c3", "#f7f7f7", "#d6604d", "#67001f"))(100), 
         annotation_col = data.frame(row.names = colnames(for.heatmap),
                  cell_type = colData(sce.spermatocytes)$AnnotatedClusters[order(prank[,"rank"])]),
         annotation_colors = list(cell_type = metadata(sce.spermatocytes)$color_vector,
                                  infertility = c("Both" = "green",
                                                  "Female" = "red",
                                                  "Male" = "blue"),
                                  enrichment = viridis(100)),
         annotation_row = data.frame(row.names = rownames(for.heatmap),
                                     cell_type = factor(rep(rownames(df.stats.red), each = 15),
                                                        labels = rownames(df.stats.red)),
                                     infertility = df.fertility[rownames(for.heatmap),"Sex"],
                                     enrichment = rep(-log10(df.stats.red$BHvalue), each = 15),
                                     label = ifelse(rownames(for.heatmap) %in% to.label, "Yes", "No")), 
         scale = "row", annotation_legend = TRUE,
         gaps_row = seq(15,105, 15), cellheight = 2)
dev.off()
rownames(for.heatmap)[rownames(for.heatmap) %in% to.label]
```

# Tc1 and Tc0 analysis

## Read in data

```{r Tc1}
# Read in Tc1 data
# Single cell data
sce <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")

# Select spermatogonia and spermatocytes
sce <- sce[,grepl("Tc1|Tc0", colData(sce)$Sample)]
sce <- normalize(sce)
```

## Calculate proportions

```{r proportions}
df <- data.frame(sample = paste(colData(sce)$Sample, colData(sce)$Library, sep = "_"),
                 group = colData(sce)$AnnotatedClusters,
                 value = rep(1, ncol(sce)))

proportions <- with(df, table(sample, group))/plyr::count(df = df, vars = "sample")$freq

cur_df <- proportions[,1:22]
cur_df.melt <- melt(cur_df)
cur_df.melt$Genotype <- sapply(as.character(cur_df.melt$sample), function(n){
  unlist(strsplit(n, "_"))[1]
})
levels(cur_df.melt$Genotype) <- c("Tc0", "Tc1")
levels(cur_df.melt$sample) <- rev(rownames(cur_df))
levels(cur_df.melt$group) <- levels(colData(sce)$AnnotatedClusters)

proportions.p <- ggplot(cur_df.melt) + geom_point(aes(group, rev(sample), size = value, fill = group), shape = 22) +
  scale_fill_manual(values = metadata(sce)$color_vector) + 
  theme(panel.background = element_blank(), axis.title = element_blank(),
        axis.text.x = element_text(angle = 45, size = 12, hjust = 1),
        axis.text.y = element_text(size = 12))

# Save plot
ggsave("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_4/CellTypeProportion_Tc1.pdf",
       proportions.p, width = 10, height = 5)

# Statistical testing between conditions
sample.counts <- with(df, table(sample, group))
sample.counts <- sample.counts[,1:22]
overall.counts <- plyr::count(df = df, vars = "sample")$freq

y <- DGEList(counts = t(sample.counts), lib.size = overall.counts, 
               group = c(rep("Tc0", 3), rep("Tc1", 4)))
design <- model.matrix(~0+y$samples$group)
colnames(design) <- c("Tc0", "Tc1")
y <- estimateDisp(y,design)
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmQLFTest(fit,coef=2, 
                  contrast = makeContrasts(Tc1 - Tc0, levels = design))
res <- topTags(qlf, n = nrow(qlf$table))$table

# Tc1 
res[res$logFC > 0 & res$FDR < 0.1,]
res[res$logFC < 0 & res$FDR < 0.1,]
```
