---
title: "Figure 4"
author: "nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_4/Fig_4.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data and libraries

```{r data, message=FALSE}
# Libraries
library(scater)
library(ggplot2)
library(reshape2)
library(plyr)
library(openxlsx)
library(cowplot)
library(viridis)
library(pheatmap)
library(goseq)
library(GO.db)
library(org.Mm.eg.db)
library(RColorBrewer)
library(edgeR)
source("../Functions/auxiliary.R")

# Single cell data
sce <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")

# Select spermatogonia and spermatocytes
sce <- sce[,colData(sce)$Sample == "B6"]
sce <- normalize(sce)

# Read in length of genes
genelength <- read.table("../Data/Genelength.txt", header = TRUE, sep = "\t")
```

# Spermatocytes

# Order cells in pseudotime

```{r PT}
sce.spermatocytes <- sce[, colData(sce)$AnnotatedClusters %in% levels(colData(sce)$AnnotatedClusters)[2:9]]
sce.spermatocytes <- normalize(sce.spermatocytes)

# Compute HVG
HVgenes <- HVG(sce.spermatocytes)

# Compute PCA
pca <- prcomp(t(logcounts(sce.spermatocytes)[HVgenes,]))

# Pseudo rank
prank <- PT(rd = pca$x[,1:3], clusters = colData(sce.spermatocytes)$AnnotatedClusters,
            col_vector = metadata(sce.spermatocytes)$color_vector)

```


```{r}
PT <-  prank[,"rank"]

set.seed(123)
y = rnorm(length(PT), mean = 0, sd = 0.01)

p.PT <- ggplot(data.frame(x = PT,
                  y = y,
                  group = colData(sce.spermatocytes)$AnnotatedClusters)) +
  geom_point(aes(x, y , fill = group), shape = 21, size = 3) + 
  scale_fill_manual(values = metadata(sce.spermatocytes)$color_vector) + ylim(c(-0.033,0.033)) +
  theme(legend.position = "none", panel.background = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.title.y = element_blank(), 
        axis.title.x = element_blank(), axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), axis.line.y = element_blank(),
        axis.text.y = element_blank(), axis.ticks.y = element_blank())

# Visualize number of genes expressed
number.genes <- apply(logcounts(sce.spermatocytes), 2, function(n){length(which(n>0))})

p.number.genes <- ggplot(
  data.frame(x = PT,
             y = number.genes,
             group = colData(sce.spermatocytes)$AnnotatedClusters)) + 
  geom_point(aes(x, y , fill = group), shape = 21, size = 3) +
  geom_smooth(aes(x = x, y = y), colour = "black") + ylab("# genes expressed") +
  scale_fill_manual(values = metadata(sce.spermatocytes)$color_vector) + 
  theme(legend.position = "none", panel.background = element_blank(), 
        panel.grid.major = element_line(colour="grey",size = rel(0.5)), 
        panel.grid.minor = element_blank(), axis.title.y = element_text(size = 15), 
        axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank())

# Plot marker genes
genes <- c("Hormad1", "Sycp3", "Pou5f2", "Tcte2")
all.max <- max(logcounts(sce.spermatocytes)[rowData(sce.spermatocytes)$ID[match(genes, rowData(sce.spermatocytes)$Symbol)],])

# Hormad1
Hormad1.p <- ggplot(data.frame(x = PT,
                  y = y,
                  gene = logcounts(sce.spermatocytes)[rowData(sce.spermatocytes)$ID[match(genes[1],
                                           rowData(sce.spermatocytes)$Symbol)],])) +
  geom_point(aes(x, y , fill = gene), shape = 21, size = 3) + 
  scale_fill_viridis(name = NULL) + ylim(c(-0.033,0.033)) + 
  ylab(genes[1]) +
  theme(panel.background = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.text.y = element_blank(), 
        axis.title.x = element_blank(), axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), axis.line.y = element_blank(),
        axis.title.y = element_text(size = 15, face = "bold"), axis.ticks.y = element_blank(),
        axis.line.x = element_blank())

# Sycp3
Sycp3.p <- ggplot(data.frame(x = PT,
                  y = y,
                  gene = logcounts(sce.spermatocytes)[rowData(sce.spermatocytes)$ID[match(genes[2],
                                           rowData(sce.spermatocytes)$Symbol)],])) +
  geom_point(aes(x, y , fill = gene), shape = 21, size = 3) + 
  scale_fill_viridis(name = NULL) + ylim(c(-0.033,0.033)) + 
  ylab(genes[2]) +
  theme(panel.background = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.text.y = element_blank(), 
        axis.title.x = element_blank(), axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), axis.line.y = element_blank(),
        axis.title.y = element_text(size = 15, face = "bold"), axis.ticks.y = element_blank(),
        axis.line.x = element_blank())

# Pou5f2
Pou5f2.p <- ggplot(data.frame(x = PT,
                  y = y,
                  gene = logcounts(sce.spermatocytes)[rowData(sce.spermatocytes)$ID[match(genes[3],
                                           rowData(sce.spermatocytes)$Symbol)],])) +
  geom_point(aes(x, y , fill = gene), shape = 21, size = 3) + 
  scale_fill_viridis(name = NULL) + ylim(c(-0.033,0.033)) + 
  ylab(genes[3]) +
  theme(panel.background = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.text.y = element_blank(), 
        axis.title.x = element_blank(), axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), axis.line.y = element_blank(),
        axis.title.y = element_text(size = 15, face = "bold"), axis.ticks.y = element_blank(),
        axis.line.x = element_blank())

# Tcte2
Tcte2.p <- ggplot(data.frame(x = PT,
                  y = y,
                  gene = logcounts(sce.spermatocytes)[rowData(sce.spermatocytes)$ID[match(genes[4],
                                           rowData(sce.spermatocytes)$Symbol)],])) +
  geom_point(aes(x, y , fill = gene), shape = 21, size = 3) + 
  scale_fill_viridis(name = NULL) + ylim(c(-0.033,0.033)) + 
  ylab(genes[4]) +
  theme(panel.background = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.text.y = element_blank(), 
        axis.title.x = element_blank(), axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), axis.line.y = element_blank(),
        axis.title.y = element_text(size = 15, face = "bold"), axis.ticks.y = element_blank(),
        axis.line.x = element_blank())

spermatocytes.p <- plot_grid(p.number.genes, Hormad1.p, Sycp3.p, Pou5f2.p, Tcte2.p, ncol = 1, nrow = 5)
ggsave("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_4/Fig_4.pdf", 
       spermatocytes.p, width = 6, height = 8)
```

# Correlation analysis to number of genes expressed

```{r}
# Remove lowly expressed genes
sce.spermatocytes <- sce.spermatocytes[Matrix::rowMeans(logcounts(sce.spermatocytes)) > 0.1,]

cur_mat <- rbind(logcounts(sce.spermatocytes), number.genes)

null.dist <- correlateNull(ncells = ncol(sce.spermatocytes), iters = 100000)
cors <- correlatePairs(cur_mat, null.dist=null.dist, pairings = list(c("number.genes"), rownames(cur_mat)))
cors$genename <- rowData(sce.spermatocytes)$Symbol[match(cors$gene2, rowData(sce.spermatocytes)$ID)]


# Perform DE analysis between spermatocytes and spermatids
sce.for.testing <- sce[match(rowData(sce.spermatocytes)$ID, 
                             rowData(sce)$ID),
                       !(colData(sce)$AnnotatedClusters %in% 
                            c("Endothelial_cells", "Outliers",
                              "Leydig", "Sertoli", paste("NA", 1:5, sep = "")))]
sce.for.testing <- normalize(sce.for.testing)
DE <- findMarkers(sce.for.testing, 
    clusters = ifelse(colData(sce.for.testing)$AnnotatedClusters %in% 
                            levels(colData(sce)$AnnotatedClusters)[2:9], "Spermatocytes", "Spermatids"))

cors$FDR.DE <- DE$Spermatocytes$FDR[match(cors$gene2, 
                                          rownames(DE$Spermatocytes))]
cors$logFC.Spermatids <- DE$Spermatocytes$logFC.Spermatids[match(cors$gene2, 
                                          rownames(DE$Spermatocytes))]

cors.out <- cors[order(cors$rho, decreasing = TRUE),]
cors.out <- cors.out[cors$FDR < 0.1 & (cors$rho > 0.3 | cors$rho < -0.3), ]

write.xlsx(cors, "../../../Dropbox (Personal)/Tc1_meiotic_silencing/Figures/Supplemental Tables/Table_S6.xlsx")

# Split gene list into correlated and not correlated with number of genes expressed
# Focus on the genes that are spermatocytes specific
pos.cor <- cors[cors$FDR < 0.1 & cors$rho > 0.3 & cors$logFC.Spermatids > 1,]
neg.cor <- cors[cors$FDR < 0.1 & cors$rho < -0.3 &cors$logFC.Spermatids > 1,]

# Visualize top 20 genes in heatmap
for.heatmap <- logcounts(sce.spermatocytes)[pos.cor$gene2[1:20],order(prank[,"rank"])]
colnames(for.heatmap) <- colData(sce.spermatocytes)$Barcode[order(prank[,"rank"])]
pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_4/Fig_4_PosCorrelationHeatmap.pdf", onefile = FALSE)
pheatmap(for.heatmap, show_colnames = FALSE,
         cluster_rows = FALSE, cluster_cols = FALSE, color = viridis(100), 
         labels_row = pos.cor$genename[1:20], 
         annotation_col = data.frame(row.names = colnames(for.heatmap),
                  cell_type = colData(sce.spermatocytes)$AnnotatedClusters[order(prank[,"rank"])]),
         annotation_colors = list(cell_type = metadata(sce.spermatocytes)$color_vector),
         annotation_row = data.frame(row.names = rownames(for.heatmap),
                                     Rho = pos.cor$rho[1:20]))
dev.off()

for.heatmap <- logcounts(sce.spermatocytes)[neg.cor$gene2[1:20],order(prank[,"rank"])]
colnames(for.heatmap) <- colData(sce.spermatocytes)$Barcode[order(prank[,"rank"])]
pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_4/Fig_4_NegCorrelationHeatmap.pdf", onefile = FALSE)
pheatmap(for.heatmap, show_colnames = FALSE,
         cluster_rows = FALSE, cluster_cols = FALSE, color = viridis(100), 
         labels_row = neg.cor$genename[1:20], 
         annotation_col = data.frame(row.names = colnames(for.heatmap),
                  cell_type = colData(sce.spermatocytes)$AnnotatedClusters[order(prank[,"rank"])]),
         annotation_colors = list(cell_type = metadata(sce.spermatocytes)$color_vector),
         annotation_row = data.frame(row.names = rownames(for.heatmap),
                                     Rho = neg.cor$rho[1:20]))
dev.off()

```

# Visualize cluster specifc marker genes

```{r markers}
# This list of marker genes consists of the top 15 markers per cell type
# Based on a litereature search, they were annotated reagrding their fertility phenotype
markers <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_4/Spermatocytes_Top15.xlsx")

for.heatmap <- logcounts(sce.spermatocytes)[rowData(sce.spermatocytes)$ID[match(markers$GeneName,
                                                                    rowData(sce.spermatocytes)$Symbol)],
                                            order(prank[,"rank"])]
colnames(for.heatmap) <- colData(sce.spermatocytes)$Barcode[order(prank[,"rank"])]

pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_4/Fig_4_MarkerGenes.pdf", onefile = FALSE, height = 5, width = 5)
pheatmap(for.heatmap, show_colnames = FALSE, show_rownames = FALSE,
         cluster_rows = FALSE, cluster_cols = FALSE,
         color = colorRampPalette(c("#053061", "#4393c3", "#f7f7f7", "#d6604d", "#67001f"))(100), 
         annotation_col = data.frame(row.names = colnames(for.heatmap),
                  cell_type = colData(sce.spermatocytes)$AnnotatedClusters[order(prank[,"rank"])]),
         annotation_colors = list(cell_type = metadata(sce.spermatocytes)$color_vector),
         annotation_row = data.frame(row.names = rownames(for.heatmap),
                                     cell_type = markers$Cluster,
                                     infertility = markers$Sterility_Phenotype,
                                     label = markers$Label), 
         scale = "row",
         gaps_row = seq(15,120, 15), cellheight = 2)
dev.off()
```

# Tc1 and Tc0 analysis

## Read in data

```{r Tc1}
# Read in Tc1 data
# Single cell data
sce <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")

# Select spermatogonia and spermatocytes
sce <- sce[,grepl("Tc1|Tc0", colData(sce)$Sample)]
sce <- normalize(sce)
```

## Calculate proportions

```{r proportions}
df <- data.frame(sample = paste(colData(sce)$Sample, colData(sce)$Library, sep = "_"),
                 group = colData(sce)$AnnotatedClusters,
                 value = rep(1, ncol(sce)))

proportions <- with(df, table(sample, group))/plyr::count(df = df, vars = "sample")$freq

cur_df <- proportions[,1:23]
cur_df.melt <- melt(cur_df)
cur_df.melt$Genotype <- sapply(as.character(cur_df.melt$sample), function(n){
  unlist(strsplit(n, "_"))[1]
})
levels(cur_df.melt$Genotype) <- c("Tc0", "Tc1")
levels(cur_df.melt$sample) <- rev(rownames(cur_df))
levels(cur_df.melt$group) <- levels(colData(sce)$AnnotatedClusters)

proportions.p <- ggplot(cur_df.melt) + geom_point(aes(group, rev(sample), size = value, fill = group), shape = 22) +
  scale_fill_manual(values = metadata(sce)$color_vector) + 
  theme(panel.background = element_blank(), axis.title = element_blank(),
        axis.text.x = element_text(angle = 45, size = 12, hjust = 1),
        axis.text.y = element_text(size = 12))

# Save plot
ggsave("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_4/CellTypeProportion_Tc1.pdf",
       proportions.p, width = 10, height = 5)

# Sattistical testing between conditions
sample.counts <- with(df, table(sample, group))
sample.counts <- sample.counts[,1:23]
overall.counts <- plyr::count(df = df, vars = "sample")$freq

y <- DGEList(counts = t(sample.counts), lib.size = overall.counts, 
               group = c(rep("Tc0", 3), rep("Tc1", 4)))
design <- model.matrix(~0+y$samples$group)
colnames(design) <- c("Tc0", "Tc1")
y <- estimateDisp(y,design)
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmQLFTest(fit,coef=2, 
                  contrast = makeContrasts(Tc1 - Tc0, levels = design))
res <- topTags(qlf, n = nrow(qlf$table))$table

# Tc1 
res[res$logFC > 0 & res$FDR < 0.1,]
res[res$logFC < 0 & res$FDR < 0.1,]
```
