---
title: "Figure 3"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S3/Fig_3.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in the data

```{r data, message=FALSE}
library(scater)
library(ggplot2)
library(reshape2)
library(plyr)
library(openxlsx)
library(cowplot)
library(viridis)
library(pheatmap)
library(Rtsne)
library(irlba)
library(RColorBrewer)
source("../Functions/auxiliary.R")

# Single cell data
sce <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")

# Read in markers for spermatogonia
sperm.markers <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_3/Spermatogonia_Top15.xlsx")
```

# Visualize spermatogonia from P10 and P15 and RA syncrhonised cells 

```{r}
# Read in only spermatogonia from P10 and P15
sce.spermatogonia <- sce[,grepl("P10|P15", colData(sce)$Sample) &
                                       colData(sce)$AnnotatedClusters == "Spermatogonia"]
sce.spermatogonia <- normalize(sce.spermatogonia)

sce_RA <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_RA.rds")

# Batch correction
sce.spermatogonia <- sce.spermatogonia[match(intersect(rownames(sce_RA), 
                          rowData(sce.spermatogonia)$Symbol), rowData(sce.spermatogonia)$Symbol),]
sce_RA <- sce_RA[match(intersect(rownames(sce_RA), rowData(sce.spermatogonia)$Symbol), 
                       rownames(sce_RA)),]

rownames(sce_RA) <- rownames(sce.spermatogonia)
rowData(sce_RA)$Symbol <- rowData(sce.spermatogonia)$Symbol
rowData(sce_RA)$ID <- rowData(sce.spermatogonia)$ID

# Select specific cell types 
sce_early_RA <- sce_RA[,grepl("A1|ln|TypeBS|G1|TypeBG2M|ePL|mPL|lPL", colData(sce_RA)$cell_type)]
sce_early_RA <- normalize(sce_early_RA)

sce.single <- split.sce(sce = sce.spermatogonia, groups = unique(colData(sce.spermatogonia)$Library), 
                        colData.name = "Library")
sce.single$RA <- sce_early_RA

corrected <- batch.correction(sce.single)

# Visualize PCA 
pca <- prcomp(t(corrected))

# Plot sample distribution
sample <- ggplot() +
  geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                  PC2 = pca$x[1:ncol(sce.spermatogonia),2],
                  sample = sce.spermatogonia$Sample),
             aes(PC1, PC2, colour = sample)) + 
            scale_color_manual(values = c("#1f78b4", "#33a02c"))

# Plot number of genes expressed
no.genes <- nexprs(sce.spermatogonia, detection_limit = 0)
number_genes <- ggplot() +
  geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                  PC2 = pca$x[1:ncol(sce.spermatogonia),2]),
             aes(PC1, PC2, colour = no.genes)) +
  scale_color_gradientn(colours = inferno(100))


#Cluster spermatogonia
set.seed(1234)
g <- buildSNNGraph(corrected[,1:ncol(sce.spermatogonia)], k = 15, pc.approx = TRUE)
clusters <- igraph::cluster_louvain(g)$membership

clusters[clusters == 1] <- "Diff_In_B"
clusters[clusters == 2] <- "Diff_A"
clusters[clusters == 3] <- "pL"
clusters[clusters == 4] <- "Undiff_2"
clusters[clusters == 5] <- "Undiff_3"
clusters[clusters == 6] <- "Undiff_1"

col_vector <- c("Undiff_1" = "#0D3D21",
                "Undiff_2" = "#14713E",
                "Undiff_3" = "#41AE76",
                "Diff_A" = "#52E8A4",
                "Diff_In_B" = "#A3EFCE",
                "pL" = "#E5DFEA")

# Plot cluster
cluster <- ggplot() +
  geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                  PC2 = pca$x[1:ncol(sce.spermatogonia),2],
                  cluster = clusters),
             aes(PC1, PC2, colour = cluster)) + 
            scale_color_manual(values = c(col_vector))

# Include RA-cells

cluster_RA <- ggplot() +
  geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                                 PC2 = pca$x[1:ncol(sce.spermatogonia),2],
                                 cluster = clusters),
               aes(PC1, PC2, colour = cluster)) + 
   geom_point(data = data.frame(PC1 = pca$x[(ncol(sce.spermatogonia)+1):(ncol(sce.spermatogonia)+ncol(sce_early_RA)),1],
                                 PC2 = pca$x[(ncol(sce.spermatogonia)+1):(ncol(sce.spermatogonia)+ncol(sce_early_RA)),2],
                                 cell_type = colData(sce_early_RA)$cell_type),
               aes(PC1, PC2, colour = cell_type), shape = 8) +
    scale_color_manual(values = c(col_vector, metadata(sce_early_RA)$colour_vector_RA))

# Plot marker gene expression

# Gfra1 and Id4 for undifferentiated spermatogonia with stem-like features
Gfra1 <- ggplot() +
      geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                                 PC2 = pca$x[1:ncol(sce.spermatogonia),2],
                                Gfra1 = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == "Gfra1",]), aes(PC1, PC2, colour = Gfra1)) + 
                                    scale_colour_viridis()
Id4 <- ggplot() +
      geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                                 PC2 = pca$x[1:ncol(sce.spermatogonia),2],
                                Id4 = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == "Id4",]), aes(PC1, PC2, colour = Id4)) + 
                                    scale_colour_viridis()

# Nanos 3 and ZBTB16 for undifferentiated spermatogonia
Nanos3 <- ggplot() +
      geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                                 PC2 = pca$x[1:ncol(sce.spermatogonia),2],
                                Nanos3 = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == "Nanos3",]), aes(PC1, PC2, colour = Nanos3)) + 
                                    scale_colour_viridis()

Zbtb16 <- ggplot() +
      geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                                 PC2 = pca$x[1:ncol(sce.spermatogonia),2],
                                Zbtb16 = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == "Zbtb16",]), aes(PC1, PC2, colour = Zbtb16)) + 
                                    scale_colour_viridis()

# Stra8 as a marker for differentiating spemratogonia

Stra8 <- ggplot() +
      geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                                 PC2 = pca$x[1:ncol(sce.spermatogonia),2],
                                Stra8 = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == "Stra8",]), aes(PC1, PC2, colour = Stra8)) + 
                                    scale_colour_viridis()

# Dmrtb1 medites mitosis-to-meiosis transition
Dmrtb1 <- ggplot() +
      geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                                 PC2 = pca$x[1:ncol(sce.spermatogonia),2],
                                Dmrtb1 = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == "Dmrtb1",]), aes(PC1, PC2, colour = Dmrtb1)) + 
                                    scale_colour_viridis()

legend_RA <- plot_grid(get_legend(cluster_RA + theme(legend.position = "right")))

A <- plot_grid(sample + theme(legend.position=c(.8,.8)), cluster + theme(legend.position=c(.8,.8)), cluster_RA + theme(legend.position= "NULL"), legend_RA,  Gfra1 + theme(legend.position=c(.8,.8)), Zbtb16 + theme(legend.position=c(.8,.8)), Stra8 + theme(legend.position=c(.8,.8)), Dmrtb1 + theme(legend.position=c(.8,.8)), ncol = 4, nrow = 2)


```

# Calculate Entropy

```{r}
sce.P15 <- sce.spermatogonia[,grepl("P15", colData(sce.spermatogonia)$Sample)]
clusters.P15 <- clusters[grepl("P15", colData(sce.spermatogonia)$Sample)]
cur_counts <- 2^(logcounts(sce.P15)) - 1 + 0.1
probs   <- t(t(cur_counts)/apply(cur_counts,2,sum))
entropy <- -apply(probs*log(probs)/log(nrow(sce)),2,sum)

sce.P15 <- runPCA(sce.P15)

ggplot() +
  geom_point(data = data.frame(PC1 = reducedDims(sce.P15)$PCA[,1],
                  PC2 = reducedDims(sce.P15)$PCA[,2],
                  entropy = entropy),
             aes(PC1, PC2, colour = entropy)) + 
            scale_color_gradientn(colours = magma(100))

ggplot() +
  geom_point(data = data.frame(PC1 = reducedDims(sce.P15)$PCA[,1],
                  PC2 = reducedDims(sce.P15)$PCA[,2],
                  clusters = clusters.P15),
             aes(PC1, PC2, colour = clusters)) + 
            scale_color_manual(values = col_vector)
```

# Find marker genes for these groups

```{r}
spermatogonia.markers <- marker.detection(sce.spermatogonia, 
                                          clusters)

write.xlsx(spermatogonia.markers, "../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_tables/Table_S4.xlsx")

```

# Heatmap of genes during spermatogonial differentiation

```{r spermatogonia}

# First we order the cells based on their differerentiation trajectory
prank <- PT(rd = pca$x[1:ncol(sce.spermatogonia),1:3], clusters = clusters, col_vector = col_vector)

# Find marker genes to visualize
spermatogonia.markers <- spermatogonia.markers[c("Undiff_1",
                                                 "Undiff_2",
                                                 "Undiff_3",
                                                 "Diff_A",
                                                 "Diff_In_B",
                                                 "pL")]


markers <- data.frame(symbol = unlist(lapply(spermatogonia.markers, 
                                             function(n){n[1:10,"GeneName"]})),
                      ID = unlist(lapply(spermatogonia.markers, 
                                             function(n){rownames(n)[1:10]})),
                      group = rep(names(spermatogonia.markers), each = 10))

# Collect count for heatmap
for.heatmap <- logcounts(sce.spermatogonia)[match(markers$symbol, 
                                                         rowData(sce.spermatogonia)$Symbol),
                                      order(prank[,"rank"], decreasing = TRUE)]
colnames(for.heatmap) <- paste(colData(sce.spermatogonia)$Library[
  order(prank[,"rank"], decreasing = TRUE)],
  colData(sce.spermatogonia)$Barcode[
  order(prank[,"rank"], decreasing = TRUE)], sep = "_")
rownames(for.heatmap) <- markers$symbol

# Visualize top genes in form of heatmap
pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S4/Spermatogonia_heatmap_P10_P15.pdf", onefile = FALSE, width = 10, height = 15)
pheatmap(for.heatmap, 
         cluster_rows = FALSE, cluster_cols = FALSE,
         scale = "row", 
         show_colnames = FALSE, gaps_row = seq(10,60, 10),
         cellheight = 8, fontsize = 7,
         color = colorRampPalette(c("#053061", "#4393c3", 
                                    "#f7f7f7", "#d6604d", "#67001f"))(100),
         annotation_col = data.frame(row.names = colnames(for.heatmap),
                                     cell_type = clusters[order(prank[,"rank"], 
                                                                decreasing = TRUE)]),
         annotation_row = data.frame(row.names = rownames(for.heatmap),
                                     gene_list = markers$group),
         annotation_colors = list(cell_type = col_vector,
                                  gene_list = col_vector))
dev.off()
```

# Visualize spermatogonia from P5

```{r}
spermatogonia.P5 <- sce[,grepl("P5", colData(sce)$Sample) &
                                       colData(sce)$AnnotatedClusters == "Spermatogonia"]
spermatogonia.P5 <- normalize(spermatogonia.P5)

# Visualize non-batch corrected data
spermatogonia.P5 <- runPCA(spermatogonia.P5)
ggplot(data.frame(PC1 = reducedDims(spermatogonia.P5)$PCA[,1],
                  PC2 = reducedDims(spermatogonia.P5)$PCA[,2],
                  libraries = colData(spermatogonia.P5)$Library)) +
  geom_point(aes(PC1, PC2, colour =  libraries)) + 
  scale_color_manual(values = c( "#FF9100", "#e31a1c"))

# Batch correction
sce.single <- split.sce(sce = spermatogonia.P5, 
                        groups = unique(colData(spermatogonia.P5)$Library), 
                        colData.name = "Library")

corrected <- batch.correction(sce.single[c("do26386", "do26387")])

spermatogonia.P5 <- do.call(cbind, sce.single[c("do26386", "do26387")])

# Compute pca 
pca <- prcomp(t(corrected))

# Visualize PCA with marker genes
libraries <- ggplot(data.frame(PC1 = pca$x[,1],
                  PC2 = pca$x[,2],
                  libraries = colData(spermatogonia.P5)$Library)) +
  geom_point(aes(PC1, PC2, colour = libraries)) + 
  scale_color_manual(values = c("#FF9100", "#e31a1c")) 

set.seed(123)
g <- buildSNNGraph(corrected, k = 15, pc.approx = TRUE)
clusters <- igraph::cluster_louvain(g)$membership


# Change cluster labels
clusters[clusters == 1] <- "Undiff_1"
clusters[clusters == 2] <- "Prospermatogonia"
clusters[clusters == 3] <- "Undiff_2"
clusters[clusters == 4] <- "Diff"


col_vector <- c("Prospermatogonia" = "#7A9606",
                "Undiff_1" = "#0D3D21",
                "Undiff_2" = "#076E38",
                "Diff" = "#5FA37D")

p5.cluster <- ggplot(data.frame(PC1 = pca$x[,1],
                  PC2 = pca$x[,2],
                  clusters = as.factor(clusters))) +
            geom_point(aes(PC1, PC2, colour = clusters)) + 
            scale_colour_manual(values = col_vector)

# Plot marker gene expression

Gfra1.p5 <- ggplot(data.frame(PC1 = pca$x[,1],
                  PC2 = pca$x[,2],
                  Gfra1 = logcounts(spermatogonia.P5)[
                    rowData(spermatogonia.P5)$Symbol == "Gfra1",])) +
  geom_point(aes(PC1, PC2, colour = Gfra1)) + scale_colour_viridis()

Stra8.p5 <- ggplot(data.frame(PC1 = pca$x[,1],
                  PC2 = pca$x[,2],
                  Stra8 = logcounts(spermatogonia.P5)[
                    rowData(spermatogonia.P5)$Symbol == "Stra8",])) +
  geom_point(aes(PC1, PC2, colour = Stra8)) + scale_colour_viridis()

Eif2s2.p5 <- ggplot(data.frame(PC1 = pca$x[,1],
                  PC2 = pca$x[,2],
                  Eif2s2 = logcounts(spermatogonia.P5)[
                    rowData(spermatogonia.P5)$Symbol == "Eif2s2",])) +
  geom_point(aes(PC1, PC2, colour = Eif2s2)) + scale_colour_viridis()


spermatogonia.markers.P5 <- marker.detection(spermatogonia.P5, 
                                          clusters)
write.xlsx(spermatogonia.markers.P5, "../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_tables/Table_S4_P5.xlsx")


legend_B <- plot_grid(get_legend(p5.cluster), get_legend(Gfra1.p5) , get_legend(Stra8.p5), get_legend(Eif2s2.p5), ncol = 4 , nrow = 1)
B_plot <- plot_grid(p5.cluster + theme(legend.position= "NONE"), Gfra1.p5 + theme(legend.position= "NONE"), Stra8.p5 + theme(legend.position= "NONE"), Eif2s2.p5 + theme(legend.position= "NONE"), ncol = 4, nrow = 1)
B <- plot_grid(B_plot, legend_B, ncol = 1, nrow = 2)
final <- plot_grid(A, B, ncol = 1, nrow = 2, rel_heights = c(3,1))

ggsave("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_3/Fig_3.pdf",
       final, width = 15, height = 12.5)
```



# Visualize RA blocked cells on empty drops

```{r}
# P15 data
sce <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")
sce <- sce[,grepl("P15", colData(sce)$Sample)]

# Select only the germ cells
sce.P15 <- sce[,grepl("Pachytene|Differentiating|Zygotene|Leptotene|Undifferentiated", colData(sce)$P15Clusters) &
                 colData(sce)$AnnotatedClusters != "Outliers"]

# Select Early germ cells from RA blocked cells
sce_early_RA <- sce_RA[,grepl("TypeB|A1|ePL|G1|mPL|eP|mP|lPL|Z|L|ln|Spo11KOL1", colData(sce_RA)$cell_type)]
sce_early_RA <- sce_early_RA[intersect(rownames(sce_early_RA), rownames(sce.P15)),]
sce.P15 <- sce.P15[rownames(sce_early_RA),]

sce.P15 <- normalize(sce.P15)
sce_early_RA <- normalize(sce_early_RA)

# Batch correction
# Calculate highly variable genes
HVG.genes <- lapply(list(sce.P15, sce_early_RA), function(n){
    HVG <- trendVar(n, use.spikes = FALSE)
    decomposeVar(n, HVG)
  })
  
HVG.df <- do.call("combineVar", HVG.genes)
HVG.df <- HVG.df[order(HVG.df$bio, decreasing = TRUE),]
genes <- rownames(HVG.df)[1:1000]
  
# Batch correction
corrected <- mnnCorrect(as.matrix(logcounts(sce.P15)[genes,]),
                        as.matrix(logcounts(sce_early_RA)[genes,]), 
                        cos.norm.in=TRUE, cos.norm.out=TRUE, sigma=0.1)

# Assign mutual nearest neighbours in P15 sample with cell identity from RA blocked cells
new.cellID <- vector(length = ncol(sce.P15))
new.cellID[as.integer(corrected$pairs[[2]]$other.cell)] <- sce_early_RA$cell_type[as.integer(corrected$pairs[[2]]$current.cell)]
new.cellID[new.cellID == "FALSE"] <- "NewCell"
colData(sce.P15)$RA_blocked_identity <- new.cellID

# Visualization
set.seed(111)
pca <- prcomp_irlba(t(do.call("cbind", corrected$corrected)), n = 50)
tsne <- Rtsne(pca$x, pca = FALSE, perplexity = 50)

# CellRanger threshold
empty_drops <- ggplot() +
  geom_point(data = data.frame(tSNE1 = tsne$Y[1:ncol(sce.P15),1],
                  tSNE2 = tsne$Y[1:ncol(sce.P15),2],
                  cluster = colData(sce.P15)$AnnotatedClusters),
             aes(tSNE1, tSNE2, colour = cluster)) + 
  scale_color_manual(values = c(metadata(sce.P15)$color_vector,
                                metadata(sce_early_RA)$colour_vector_RA, "NewCell" = "black"))

# CellRanger threshold with RA mapped cells
empty_drops_RA <- ggplot() +
  geom_point(data = data.frame(tSNE1 = tsne$Y[1:ncol(sce.P15),1],
                  tSNE2 = tsne$Y[1:ncol(sce.P15),2],
                  cluster = colData(sce.P15)$AnnotatedClusters),
             aes(tSNE1, tSNE2, colour = cluster)) + 
    geom_point(data = data.frame(tSNE1 = tsne$Y[(ncol(sce.P15)+1):(ncol(sce.P15)+ncol(sce_early_RA)),1],
                  tSNE2 = tsne$Y[(ncol(sce.P15)+1):(ncol(sce.P15)+ncol(sce_early_RA)),2],
                  cell_type = colData(sce_early_RA)$cell_type),
             aes(tSNE1, tSNE2, colour = cell_type, size = 3)) +
  scale_color_manual(values = c(metadata(sce.P15)$color_vector,
                                metadata(sce_early_RA)$colour_vector_RA, "NewCell" = "black"))

# EmptyDrops threshold
empty_drops_clustered <- ggplot() +
  geom_point(data = data.frame(tSNE1 = tsne$Y[1:ncol(sce.P15),1],
                  tSNE2 = tsne$Y[1:ncol(sce.P15),2],
                  cluster = colData(sce.P15)$P15Clusters),
             aes(tSNE1, tSNE2, colour = cluster)) + 
    geom_point(data = data.frame(tSNE1 = tsne$Y[(ncol(sce.P15)+1):(ncol(sce.P15)+ncol(sce_early_RA)),1],
                  tSNE2 = tsne$Y[(ncol(sce.P15)+1):(ncol(sce.P15)+ncol(sce_early_RA)),2],
                  cell_type = colData(sce_early_RA)$cell_type),
             aes(tSNE1, tSNE2, colour = cell_type, size = 3)) +
  scale_color_manual(values = c(metadata(sce.P15)$colour_vector.P15,
                                metadata(sce_early_RA)$colour_vector_RA, "NewCell" = "black"))
```

# Save final figure

```{r save}
A <- plot_grid(Gfra1 + theme(legend.position=c(.05,.8)), Nanos3+ theme(legend.position=c(.05,.8)), Stra8+ theme(legend.position=c(.05,.8)), Dmrtb1+ theme(legend.position=c(.05,.8)), samples+ theme(legend.position=c(.05,.8)), p.clusters + theme(legend.position=c(.05,.8)), ncol = 6)
B_legend1 <- get_legend(empty_drops + theme(legend.position="bottom"))
B_legend2 <- get_legend(empty_drops_RA + theme(legend.position="bottom"))
B_legend3 <- get_legend(empty_drops_clustered + theme(legend.position = "bottom"))
B <- plot_grid(empty_drops+ theme(legend.position="none"), empty_drops_RA+ theme(legend.position="none"), empty_drops_clustered+ theme(legend.position="none"), B_legend1, B_legend2, B_legend3, ncol = 3, nrow = 2, rel_heights = c(4,1))
final <- plot_grid(A, B, nrow=2, rel_heights = c(1,2))

ggsave("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_3/Fig_3.pdf",
       final, width = 25, height = 20)
```