---
title: "Figure 5"
author: "nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/Fig_5.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in data

```{r data, message=FALSE}
# Libraries
library(scater)
library(ggplot2)
library(reshape2)
library(plyr)
library(openxlsx)
library(cowplot)
library(viridis)
library(pheatmap)
library(goseq)
library(GO.db)
library(org.Mm.eg.db)
library(RColorBrewer)
source("../Functions/auxiliary.R")

# Single cell data
sce <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")

# Select spermatogonia and spermatocytes
sce <- sce[,grepl("B6", colData(sce)$Sample)]
sce <- normalize(sce)
```

# Order cells over the differentiation time course

```{r ordering}
# Select Spermatids
sce.spermatids <- sce[, colData(sce)$AnnotatedClusters %in% 
                           paste("S", 1:11, sep = "")]
sce.spermatids <- normalize(sce.spermatids)

# Compute HVG
HVgenes <- HVG(sce.spermatids)

# Compute PCA
pca <- prcomp(t(logcounts(sce.spermatids)[HVgenes,]))

# Pseudo rank
prank <- PT(rd = pca$x[,1:3], 
            clusters = colData(sce.spermatids)$AnnotatedClusters,
            col_vector = metadata(sce.spermatids)$color_vector)
```

# Histone variants over spermiogenesis

Here, we visualize the histone variants taken from Kennani et al. Epigenetics & Chromatin, 2017.

```{r hist-var}
# Read in histone variants
Hist.all <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/Mouse_histones_annotated.xlsx")
rownames(Hist.all) <- Hist.all$Ensembl.Gene.ID

# Collect data for heatmap plotting
for.heatmap <- logcounts(sce.spermatids)[rownames(Hist.all),
                  order(prank[,"rank"], decreasing = TRUE)]
colnames(for.heatmap) <- paste(colData(sce.spermatids)$Barcode[order(prank[,"rank"],
                                                               decreasing = TRUE)],
                               colData(sce.spermatids)$Library[order(prank[,"rank"],
                                                               decreasing = TRUE)],
                               sep = "_")

# Plot these genes over pseudotime
pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_figures/Fig_S8/HistoneVariants.pdf", onefile = FALSE)
pheatmap(for.heatmap, show_colnames = FALSE, show_rownames = FALSE,
         cluster_rows = FALSE, cluster_cols = FALSE, 
         color = colorRampPalette(c("#053061", "#4393c3", "#f7f7f7", "#d6604d", "#67001f"))(100),
         annotation_col = data.frame(row.names = colnames(for.heatmap),
                  cell_type = colData(sce.spermatids)$AnnotatedClusters[order(prank[,"rank"],
                      decreasing = TRUE)]),
         annotation_colors = list(cell_type = metadata(sce.spermatids)$color_vector,
                                  label = c("NA" = "white",
                                            "label" = "black"),
                                  Histone = c("H2A_variant" = "springgreen",
                                              "H2B_variant" = "springgreen3",
                                              "H3_variant" = "seagreen",
                                              "H1_variant" = "lightseagreen",
                                              "Canonical" = "indianred3",
                                              "Transition_protein" = "darkorange4",
                                              "Protamine" = "burlywood4"),
                                  Testis_specific = c("Testis-specific" = "sienna1",
                                                      "NA" = "white")), 
         scale = "row",
         annotation_row = data.frame(row.names = rownames(for.heatmap),
                                     Testis_specific = Hist.all$`Testis-specific`,
                                     Histone = Hist.all$Label,
                                     label = Hist.all$Gene_name))
dev.off()
```

# Find transcriptional shutdown

```{r}
PT <-  prank[,"rank"]

number.genes <- apply(logcounts(sce.spermatids), 2, function(n){length(which(n>0))})

p.number.genes <- ggplot(
  data.frame(x = PT,
             y = number.genes,
             group = colData(sce.spermatids)$AnnotatedClusters)) + 
  geom_point(aes(x, y , fill = group), shape = 21, size = 3) +
  scale_x_reverse() + 
  geom_smooth(aes(x = x, y = y), colour = "black") + ylab("# genes expressed") +
  scale_fill_manual(values = metadata(sce.spermatids)$color_vector) + 
  theme(legend.position = "none", panel.background = element_blank(), 
        panel.grid.major = element_line(colour="grey",size = rel(0.5)), 
        panel.grid.minor = element_blank(), axis.title.y = element_text(size = 15), 
        axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank())

ggsave("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_6/NoGenes.pdf", 
       p.number.genes, width = 12, height = 4)
```

# Correlate genes to number of genes expressed

```{r}
cur_mat <- rbind(logcounts(sce.spermatids), number.genes)
cur_mat <- cur_mat[Matrix::rowMeans(cur_mat) > 0.1,]

null.dist <- correlateNull(ncells = ncol(sce.spermatids), iters = 100000)
cors <- correlatePairs(cur_mat, null.dist=null.dist, 
                       pairings = list(c("number.genes"), rownames(cur_mat)))
cors$genename <- rowData(sce.spermatids)$Symbol[match(cors$gene2, rowData(sce.spermatids)$ID)]

# Visualize results
# Calculate the Z score
z.score <- t(apply(as.matrix(cur_mat)[rownames(cur_mat) != "number.genes",],
                   1, function(n){(n-mean(n))/sd(n)}))
cors <- cors[match(rownames(z.score), cors$gene2),]

# Generate meta profiles 
splits <- 10
ran <- rev(seq(min(cors$rho) - 0.001, max(cors$rho) + 0.001, length.out = splits))
mat <- matrix(data = NA, nrow = splits - 1, ncol = ncol(z.score))

for(i in 1:(splits - 1)){
  mat[i,] <- colMeans(z.score[cors$rho <= ran[i] & cors$rho > ran[i + 1],])
}

colnames(mat) <- paste(colData(sce.spermatids)$Barcode,
                       colData(sce.spermatids)$Library)
rownames(mat) <- paste("GeneSet", 1:nrow(mat))

pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_6/TranscriptionalShotdown.pdf", onefile = FALSE)
pheatmap(mat[,order(PT, decreasing = TRUE)], breaks = seq(-2, 2, length.out = 100),
         cluster_rows = FALSE, cluster_cols = FALSE, 
         color = colorRampPalette(c("#053061", "#4393c3", "#f7f7f7", "#d6604d", "#67001f"))(100),
         annotation_col = data.frame(row.names = colnames(mat)[order(PT, decreasing = TRUE)],
                  cell_type = colData(sce.spermatids)$AnnotatedClusters[order(PT, decreasing = TRUE)]),
         annotation_colors = list(cell_type = metadata(sce.spermatids)$color_vector), show_colnames = FALSE)
dev.off()

groups <- vector(length = nrow(z.score))

for(i in 1:(splits - 1)){
  groups[cors$rho <= ran[i] & cors$rho > ran[i + 1]] <- paste("GeneSet", i)
}

cors$group <- groups
cors <- cors[order(cors$rho, decreasing = TRUE),]
write.xlsx(cors, "../../../Dropbox (Personal)/Tc1_meiotic_silencing/Revisions/Figures/Supplemental Tables/Table_S10.xlsx")
```