---
title: "Figure 4"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_4/Fig_4.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in the data

```{r data, message=FALSE}
library(scater)
library(ggplot2)
library(reshape2)
library(plyr)
library(openxlsx)
library(cowplot)
library(viridis)
library(pheatmap)
library(irlba)
library(Rtsne)
library(randomForest)
source("../Functions/auxiliary.R")

# Single cell data
sce <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_emptyDrops.rds")

# Read in RA blocked cells
sce_RA <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_RA.rds")

# Select germ cells from P15
sce <- sce[,colData(sce)$Sample == "P15"]
sce <- sce[,grepl("Pachytene|Differentiating|Zygotene|Leptotene|Undifferentiated", 
                          colData(sce)$P15Clusters) &
                 colData(sce)$AnnotatedClusters != "Outliers" &
             colData(sce)$AnnotatedClusters != "Sertoli"]

# Match the genesnames
sce.P15 <- sce[match(intersect(rownames(sce_RA), 
                          rowData(sce)$Symbol), rowData(sce)$Symbol),]
sce_RA <- sce_RA[match(intersect(rownames(sce_RA), rowData(sce.P15)$Symbol), 
                       rownames(sce_RA)),]

rownames(sce_RA) <- rownames(sce.P15)
rowData(sce_RA)$Symbol <- rowData(sce.P15)$Symbol
rowData(sce_RA)$ID <- rowData(sce.P15)$ID

# Bulk data
bulk <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/bulk/norm/1st_wave_bulk_norm_reverse-stranded.rds")

# Meta info for bulk data
meta <- read.table("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Tables/BulkSamples_stats.txt", sep = "\t")
```

# Map RA cells to germ cells of P15 sample

```{r}
# Select Early germ cells from RA blocked cells
sce_early_RA <- sce_RA[,grepl("TypeB|A1|ePL|G1|mPL|eP|mP|lPL|Z|L|ln|lP", colData(sce_RA)$cell_type) &
                         colData(sce_RA)$cell_type != "Spo11KOL1"]

sce.P15 <- normalize(sce.P15)
sce_early_RA <- normalize(sce_early_RA)

# Batch correction
# Calculate highly variable genes
HVG.genes <- lapply(list(sce.P15, sce_early_RA), function(n){
    HVG <- trendVar(n, use.spikes = FALSE)
    decomposeVar(n, HVG)
  })
  
HVG.df <- do.call("combineVar", HVG.genes)
HVG.df <- HVG.df[order(HVG.df$bio, decreasing = TRUE),]
genes <- rownames(HVG.df)[1:1000]
  
# Batch correction
corrected <- mnnCorrect(as.matrix(logcounts(sce.P15)[genes,]),
                        as.matrix(logcounts(sce_early_RA)[genes,]), 
                        cos.norm.in=TRUE, cos.norm.out=TRUE, sigma=0.1)
# Visualization
set.seed(3456)
pca <- prcomp_irlba(t(do.call("cbind", corrected$corrected)), n = 50)
tsne <- Rtsne(pca$x, pca = FALSE, perplexity = 50)

# Plot cellRanger and emptyDrops distribution
cur_tsne <- tsne$Y[1:ncol(sce.P15),]
emptyDrops <- ggplot() +
  geom_point(data = data.frame(tSNE1 = cur_tsne[sce.P15$AnnotatedClusters == "NewCell",1],
                  tSNE2 = cur_tsne[sce.P15$AnnotatedClusters == "NewCell",2],
                  clusters = sce.P15$AnnotatedClusters[sce.P15$AnnotatedClusters == "NewCell"]),
             aes(tSNE1, tSNE2, colour = clusters)) +
    geom_point(data = data.frame(tSNE1 = cur_tsne[sce$AnnotatedClusters != "NewCell",1],
                  tSNE2 = cur_tsne[sce.P15$AnnotatedClusters != "NewCell",2],
                  clusters = sce.P15$AnnotatedClusters[sce.P15$AnnotatedClusters != "NewCell"]),
             aes(tSNE1, tSNE2, colour = clusters)) +
            scale_colour_manual(values = metadata(sce.P15)$color_vector) +
  guides(colour = FALSE)

# Number of genes expressed
no.genes <- nexprs(sce.P15, detection_limit = 0)
number_genes <- ggplot() +
  geom_point(data = data.frame(tSNE1 = cur_tsne[,1],
                  tSNE2 = cur_tsne[,2],
                  genes = no.genes),
             aes(tSNE1, tSNE2, colour = genes)) +
            scale_colour_gradientn(colours = inferno(100))

# Plot emptyDrops clustering with RA cells
empty_drops_clustered <- ggplot() +
  geom_point(data = data.frame(tSNE1 = tsne$Y[1:ncol(sce.P15),1],
                  tSNE2 = tsne$Y[1:ncol(sce.P15),2],
                  cluster = colData(sce.P15)$P15Clusters),
             aes(tSNE1, tSNE2, colour = cluster)) + 
      geom_point(data = data.frame(tSNE1 = tsne$Y[(ncol(sce.P15)+1):(ncol(sce.P15)+ncol(sce_early_RA)),1],
                  tSNE2 = tsne$Y[(ncol(sce.P15)+1):(ncol(sce.P15)+ncol(sce_early_RA)),2],
                  cell_type = colData(sce_early_RA)$cell_type),
             aes(tSNE1, tSNE2, size = 2.5)) +
    geom_point(data = data.frame(tSNE1 = tsne$Y[(ncol(sce.P15)+1):(ncol(sce.P15)+ncol(sce_early_RA)),1],
                  tSNE2 = tsne$Y[(ncol(sce.P15)+1):(ncol(sce.P15)+ncol(sce_early_RA)),2],
                  cell_type = colData(sce_early_RA)$cell_type),
             aes(tSNE1, tSNE2, colour = cell_type), size = 2.3) +
  scale_color_manual(values = c(metadata(sce.P15)$colour_vector.P15,
                                metadata(sce_early_RA)$colour_vector_RA, "NewCell" = "black"))
```

# Find marker genes

```{r}
sce <- sce[rowMeans(counts(sce)) > 0,]
Lep.Zygot.markers <- marker.detection(sce, 
                                  as.character(colData(sce)$P15Clusters))

write.xlsx(Lep.Zygot.markers, "../../../Dropbox (Personal)/Tc1_meiotic_silencing/Revisions/Figures/Supplemental Tables/Table_S5.xlsx")
```

# Random forrest mapping of bulk RNA-Seq data

```{r}
# Select marker genes
cur_markers <- as.character(unlist(lapply(Lep.Zygot.markers, function(n){rownames(n)[1:50]})))
cur_markers <- cur_markers[cur_markers %in% rownames(bulk)]

# Train the classifier on P15 data 
labels <- factor(colData(sce)$P15Clusters,
                    levels = levels(colData(sce)$P15Clusters)[
                      levels(colData(sce)$P15Clusters) %in%
                        unique(colData(sce)$P15Clusters)])

# Create training and testing datasets
set.seed(1234)
sam <- sample(1:ncol(sce), 4000)

train.data <- as.matrix(logcounts(sce)[cur_markers,sam])
train.labels <- labels[sam] 
test.data <- as.matrix(logcounts(sce)[cur_markers,-sam])
test.labels <- labels[-sam] 

# Scale the counts prior to training
# Z score
train.data.Z <- t(apply(train.data, 1, function(n){(n - mean(n))/sd(n)}))
test.data.Z <- t(apply(test.data, 1, function(n){(n - mean(n))/sd(n)}))

# Random forrest
Tree <- randomForest(x = t(as.matrix(train.data.Z)), y = train.labels, 
                     xtest = t(as.matrix(test.data.Z)), 
                     ytest = test.labels, keep.forest = TRUE)

# Prediction 
# Select only bulk data from P6 - P20
cur_bulk <- bulk[,rownames(meta)[1:15]]
cur_meta <- meta[rownames(meta)[1:15],]
new.data <- log2(cur_bulk[cur_markers,] + 1)
new.data.Z <- t(apply(new.data, 1, function(n){(n - mean(n))/sd(n)}))

pred.bulk <- predict(object = Tree, newdata = t(new.data.Z), type="prob")

# Collect day info on libraries
df <- data.frame(library = rownames(pred.bulk),
                 day = as.numeric(sub(" days", "", cur_meta[rownames(pred.bulk),"Age"])))

pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_4/Fig_4_bulkmapping_heatmap.pdf", 
    onefile = FALSE)
pheatmap(pred.bulk[order(df$day),], cluster_cols = FALSE, cluster_rows = FALSE,
         col = inferno(100),
         annotation_row = data.frame(row.names = rownames(pred.bulk),
                    sample = factor(df$day, levels = unique(df$day)[order(unique(df$day))])),
         labels_row = df$day[order(df$day)], border_color = NA)
dev.off()
```

# Heatmap of marker genes

```{r}
# First we order the cells based on their differerentiation trajectory
prank <- PT(rd = pca$x[1:ncol(sce),1:10], clusters = colData(sce)$P15Clusters, 
            col_vector = metadata(sce)$colour_vector.P15)

# Find marker genes to visualize
Lep.Zygot.markers <- Lep.Zygot.markers[c("Undifferentiated_Spermatogonia",
                                                 "Differentiating_Spermatogonia_1",
                                                 "Differentiating_Spermatogonia_2",
                                                 "Leptotene",
                                                 "Zygotene",
                                                 "Pachytene_1",
                                         "Pachytene_2")]

# Hand picked markers from gene lists
cur_markers <- c("Egr4", "Sox4", "Sohlh1", 
                 "Uchl1", "Stra8", "Dmrt1",
                 "Ccnd1", "Plk1", "Esx1",
                 "Prss50", "Smc3", "Tex15", "Smc1b", "Syce2", "Dmc1", "Prdm9", "Fbxw2", "Brca2", "Rec8",
                 "Ly6k", "Sycp3", "Sycp1", "Tex101", "Emc7", "Rad51ap2", "Syce3", "H2afx", "Sycp2", "Meiob",
                 "Clgn", "Hormad1", "Cntd1", "Tdrd9", "Piwil2", "Atr", "Scmh1", "Catsperb", "Kdm3a" , "Msh4",
                 "Spata16", "Spink2", "Pgam2", "Nme5", "Tex40")

markers <- data.frame(symbol = cur_markers,
                      ID = rowData(sce)$ID[match(cur_markers, rowData(sce)$Symbol)],
                      group = rep(names(Lep.Zygot.markers), c(3,3,3,10,10,10,5)))

# Collect count for heatmap
for.heatmap <- logcounts(sce)[match(markers$symbol, 
                                                         rowData(sce)$Symbol),
                                      order(prank[,"rank"], decreasing = TRUE)]
colnames(for.heatmap) <- paste(colData(sce)$Library[
  order(prank[,"rank"], decreasing = TRUE)],
  colData(sce)$Barcode[
  order(prank[,"rank"], decreasing = TRUE)], sep = "_")
rownames(for.heatmap) <- markers$symbol

# Visualize top genes in form of heatmap
pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_4/EmptyDrops_heatmap_P15.pdf", onefile = FALSE, width = 10, height = 15)
pheatmap(for.heatmap, 
         cluster_rows = FALSE, cluster_cols = FALSE,
         scale = "row", 
         show_colnames = FALSE, gaps_row = c(3,6,9,19,29,39),
         cellheight = 8, fontsize = 7,
         color = colorRampPalette(c("#053061", "#4393c3", 
                                    "#f7f7f7", "#d6604d", "#67001f"))(100),
         annotation_col = data.frame(row.names = colnames(for.heatmap),
                                     cell_type = colData(sce)$P15Clusters[order(prank[,"rank"], 
                                                                decreasing = TRUE)]),
         annotation_row = data.frame(row.names = rownames(for.heatmap),
                                     gene_list = markers$group),
         annotation_colors = list(cell_type = metadata(sce)$colour_vector.P15,
                                  gene_list = metadata(sce)$colour_vector.P15))
dev.off()
```

# Save final figure

```{r save}
B_legend1 <- get_legend(empty_drops_clustered + theme(legend.position="right"))
final <- plot_grid(emptyDrops,
               number_genes + theme(legend.position = "right"),
               empty_drops_clustered + theme(legend.position = "none"),
               B_legend1,  ncol = 2, nrow = 2)

ggsave("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_4/Fig_4.pdf",
       final, width = 15, height = 15)
```