---
title: "Figure 5"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data and libraries

```{r data, message=FALSE}
library(rtracklayer)
library(EnrichedHeatmap)
library(GenomicFeatures)
library(EnsDb.Mmusculus.v79)
library(scater)
library(plyr)
library(reshape2)
library(pheatmap)
library(viridis)
library(ggsci)
library(cowplot)
library(openxlsx)

# Generate feature annotation
tss <- promoters(genes(EnsDb.Mmusculus.v79), 
                 upstream = 0, downstream = 1)
tss <- tss[seqnames(tss) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(tss) <- c(as.character(1:19), "X", "Y", "MT")
tss.X <- tss[seqnames(tss) == "X"]

blacklist <- import("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Cnr/Blacklist_mm10.bed")
seqlevels(blacklist) <- sub("^chr", '', seqlevels(blacklist))

# Read in sce data
sce <- readRDS("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all_clusters.rds")
sce <- sce[,grepl("B6", colData(sce)$Sample) & 
             !(colData(sce)$AnnotatedClusters %in% c("variable", "Leydig", "Sertoli",
                                          "Endothelial_cells", "Outliers"))]
sce <- normalize(sce)

# Select XChr genes
genenames <- read.table("../Data/Mouse_genes.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
sce.XChr <- sce[rowData(sce)$ID %in% 
                  genenames[genenames$Chromosome.scaffold.name == "X",1],]
sce.XChr <- normalize(sce.XChr)

# Bulk data
bulk <- readRDS("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/bulk/norm/1st_wave_bulk_norm_reverse-stranded.rds")
# Use genes that are also detected in scRNAseq data
bulk <- bulk[rownames(bulk) %in% rowData(sce)$ID,]

# Meta info
meta <- as.data.frame(read.xlsx("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Metadata.xlsx", sheet = 2))
meta <- meta[!is.na(meta$DO.number),]
rownames(meta) <- meta$DO.number
```

# Visualize X inactivation and reactivation

## Visualize X inactivation and reactivation

This is done with single cells

```{r in-reactivation}
# Remove lowly expressed genes
genes <- rowData(sce)$ID[
  apply(logcounts(sce)[,colData(sce)$AnnotatedClusters %in% 
          c("Spermatogonia", paste("S", 1:14, sep = ""))], 1,
              function(n){length(which(n > 0))}) >
    0.3*sum(colData(sce)$AnnotatedClusters %in% 
              c("Spermatogonia", paste("S", 1:14, sep = "")))]

# Collect genes for some chromosomes
genes.X <- genes[genes %in% genenames[genenames[,3] == "X",1]]
genes.9 <- genes[genes %in% genenames[genenames[,3] == "9",1]]
genes.A <- genes[genes %in% genenames[genenames[,3] != "X" & 
                                        genenames[,3] != "Y" &
                                        genenames[,3] != "MT",1]]

# Calculate X:A ratio and 16:X ratio
ratio.X.A <- colMeans(as.matrix(logcounts(sce)[genes.X,]))/
  colMeans(as.matrix(logcounts(sce)[genes.A,]))
ratio.X.A[is.nan(ratio.X.A)] <- 0

ratio.9.A <- colMeans(as.matrix(logcounts(sce)[genes.9,]))/
  colMeans(as.matrix(logcounts(sce)[genes.A,]))
ratio.9.A[is.nan(ratio.9.A)] <- 0

# Build mean ratio matrix
df <- data.frame(ratios = c(ratio.X.A, ratio.9.A),
                 groups = factor(c(as.character(colData(sce)$AnnotatedClusters), 
                            as.character(colData(sce)$AnnotatedClusters)), 
                            levels = levels(colData(sce)$AnnotatedClusters)),
                 chr = c(rep("X", length(ratio.X.A)), rep("9", length(ratio.9.A))))

# visualize in form of boxplots
x.inactivation <- ggplot(df) + geom_boxplot(aes(groups, ratios, fill = chr)) + 
  scale_fill_jama() + ylab("Chr:Autosome ratio") + 
    theme(panel.background = element_blank(), axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, size = 12, hjust = 1),
        axis.text.y = element_text(size = 12))
ggsave(filename = "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/Xdynamics.pdf", 
       x.inactivation, width = 12, height = 7)
```

## Visualize in bulk data

```{r bulk}
# Select X chromosome genes
bulk.X <- bulk[intersect(rowData(sce.XChr)$ID, rownames(bulk)),]
bulk.X <- bulk.X[rowMeans(bulk.X) > 10,]

# Collect day info on libraries
df <- data.frame(library = colnames(bulk.X),
                 day = meta$`Age.(days)`[order(rownames(meta))])

# Order genes based on their logFC between spermatocytes (< P20)
# and spermatids (> P20)
bulk.X <- bulk.X[order(log(rowMeans(bulk.X[,df$day < 20])/
                   rowMeans(bulk.X[,df$day > 20])), decreasing = TRUE),]

# Order based on where gene expression reaches 10 counts for the first time
bulk.X <- bulk.X[order(apply(bulk.X, 1, function(n){
  which(n > 10)[1]
}), decreasing = FALSE),]

pdf("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/BulkExpression.pdf",
    width = 10, height = 15, onefile=FALSE)
pheatmap(log10(bulk.X[,order(df$day)] + 1), cluster_cols = FALSE, 
         cluster_rows = FALSE, show_rownames = FALSE,
         color = viridis(100),
         border_color = NA,
         labels_col = paste("P", df$day[order(df$day)], sep = ""))
dev.off()
```

## Visualize different groups of genes

```{r genesets}
# Build mean expression matrix
df <- as.data.frame(t(as.matrix(logcounts(sce.XChr))))
df$groups <- colData(sce.XChr)$AnnotatedClusters
df.melt <- melt(df, id.vars = "groups")

# Collect mean expression for each gene in each group
mat <- ddply(df.melt, .(groups, variable), summarize, mean=mean(value))
mat.for.heatmap <- dcast(data = mat,formula = variable~groups,fun.aggregate = sum,value.var = "mean")
rownames(mat.for.heatmap) <- mat.for.heatmap$variable
mat.for.heatmap <- mat.for.heatmap[,-1]

# Match genes with bulk libraries
mat.for.heatmap <- mat.for.heatmap[rownames(bulk.X),]
```

## Save different heatmaps

```{r}
pdf("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/GeneExpression.pdf",
    width = 10, height = 15, onefile=FALSE)
pheatmap(mat.for.heatmap, cluster_cols = FALSE, 
         cluster_rows = FALSE, show_rownames = FALSE,
         show_colnames = FALSE,
         color = viridis(100), 
         annotation_col = data.frame(row.names = colnames(mat.for.heatmap),
                                     Cell_type = colnames(mat.for.heatmap)),
         annotation_colors = list(Cell_type = metadata(sce.XChr)$color_vector),
         border_color = NA)
dev.off()
```

# Visualize peaks for these genes

## H3K4me3

```{r peaks}
cur_files <- list.files("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bedgraphs/H3K4me3/", 
                        full.names = TRUE)
filenames <- list.files("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bedgraphs/H3K4me3/", 
                        full.names = FALSE)

cur_files <- cur_files[sapply(filenames, 
                function(n){all(sapply(c("spermatids|spermatocytes", "JP18|JP20|JP22|JP24|JP26|JP28|JP30"), 
                                   grepl, n))})]


filenames <- as.character(sub(".bed.1-1000.bedgraph.gz$", ".pdf",
                 sapply(cur_files, function(n){unlist(strsplit(n, "/"))[11]})))

cur_genes <- rownames(mat.for.heatmap)
cur_tss <- tss[match(cur_genes, tss@elementMetadata$gene_id)[
  !is.na(match(cur_genes, tss@elementMetadata$gene_id))
]]

for(i in 1:length(cur_files)){
  cur_file <- import(cur_files[i])
  
  # Exclude blacklisted regions
  cur_removed <- overlapsAny(cur_file, blacklist)
  cur_file <- cur_file[!cur_removed]
  cur_file <- cur_file[seqnames(cur_file) %in%
                         c(as.character(1:19), "X", "Y", "MT"),]
  seqlevels(cur_file) <- c(as.character(1:19), "X", "Y", "MT")

  cur_bed <- cur_file[seqnames(cur_file) == "X"]

  cur_mat = normalizeToMatrix(cur_bed, cur_tss, 
                        value_column = "score",
                        extend = 5000, mean_mode = "w0", 
                        w = 50, keep = c(0,0.99))
  
  pdf(paste("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/", 
          filenames[i], sep = ""), width = 10, height = 15, onefile=FALSE)
  print(EnrichedHeatmap(cur_mat, col = c("white", "red", "black"),
                  top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4))),
                  row_order = 1:nrow(cur_mat)))
  dev.off()
}
```

## H3K27ac

```{r peaks}
cur_files <- list.files("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bedgraphs/H3K27ac/", 
                        full.names = TRUE)

filenames <- as.character(sub(".bed.1-1000.bedgraph.gz$", ".pdf",
                 sapply(cur_files, function(n){unlist(strsplit(n, "/"))[11]})))

cur_genes <- rownames(rbind(mat.1, mat.2, mat.3))
cur_tss <- tss[match(cur_genes, tss@elementMetadata$gene_id)]

for(i in 1:length(cur_files)){
  cur_file <- import(cur_files[i])
  
  # Exclude blacklisted regions
  cur_removed <- overlapsAny(cur_file, blacklist)
  cur_file <- cur_file[!cur_removed]
  cur_file <- cur_file[seqnames(cur_file) %in%
                         c(as.character(1:19), "X", "Y", "MT"),]
  seqlevels(cur_file) <- c(as.character(1:19), "X", "Y", "MT")

  cur_bed <- cur_file[seqnames(cur_file) == "X"]

  cur_mat = normalizeToMatrix(cur_bed, cur_tss, 
                        value_column = "score",
                        extend = 5000, mean_mode = "w0", 
                        w = 50, keep = c(0,0.99))
  
  pdf(paste("/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5/", 
          filenames[i], sep = ""), width = 10, height = 15, onefile=FALSE)
  print(EnrichedHeatmap(cur_mat[cur_genes,], col = c("white", "red", "black"), 
                  split = c(rep(1, nrow(mat.1)), rep(2, nrow(mat.2)), rep(3, nrow(mat.3))),
                  top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4))),
                  row_order = 1:length(cur_genes)))
  dev.off()
}
```

# Save final figure

```{r final}
final <- plot_grid(proportions.p, Aurkc, ncol = 1, nrow = 2)
ggsave(filename = "/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_5.pdf", final, width = 8, height = 8)
```
