---
title: "Figure 7"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data and libraries

Aaron: Just use scaledAverage - don't need to care about fancy normalization schemes
regionCounts()

```{r data, message=FALSE}
library(rtracklayer)
library(EnrichedHeatmap)
library(GenomicFeatures)
library(EnsDb.Mmusculus.v79)
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
library(plyr)
library(reshape2)
library(pheatmap)
library(viridis)
library(plot3D)
library(ggsci)
library(cowplot)
library(openxlsx)
library(csaw)
library(edgeR)
library(Rsamtools)
source("../Functions/auxiliary.R")

# Generate feature annotation
prom <- promoters(genes(EnsDb.Mmusculus.v79))
prom <- prom[seqnames(prom) %in% c(as.character(1:19), "X", "Y", "MT"),]
seqlevels(prom) <- c(as.character(1:19), "X", "Y", "MT")
prom.X <- prom[seqnames(prom) == "X"]
prom.Y <- prom[seqnames(prom) == "Y"]
prom.9 <- prom[seqnames(prom) == "9"]

# K9 files
bam.files.K9 <- list.files("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K9me3/", full.names = TRUE,
                        pattern = paste(".+JP26.+bam$", sep = ""))

# K4 files
bam.files.K4 <- list.files("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K4me3/", full.names = TRUE,
                        pattern = paste(".+JP26.+bam$", sep = ""))

# Define conditions
batch <- gl(2, 2)
treatment <- rep(c("spermatocytes", "spermatids"), 2)

# Blacklisted regions
black <- import("http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz")
seqlevels(black) <- sub("chr", "", seqlevels(black))
seqnames(black) <- sub("chr", "", seqnames(black))

# Parameters for reading bam files
param <- readParam(dedup=TRUE, minq=10, discard=black, pe="both", max.frag=1000)

# Read in genelists
retro <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/X_retrogenes.xlsx")
retro <- retro[grepl("ENS", retro$Parental.Gene),]

K27writers.Rnf8 <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/Adams_et_al_down_in_RS_after_KO.xlsx", sheet = 1)
K27writers.Scml2 <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/Adams_et_al_down_in_RS_after_KO.xlsx", sheet = 2)

spermatid.specific <- read.xlsx("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_6/SpermatidSpecificGenes.xlsx")
```

# Visualize the K9 and K4 signal coming from the different chromosomes

```{r}
# Count reads per chromosome
header <- scanBamHeader(bam.files.K9[1])
sizes <- header$`../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/CnR/bam/H3K9me3//do18097_H3K9me3_spermatocytes_ab8898_mm10C57BL6_JP26_18_5374_CRI0sorted.bam`$targets
sizes <- sizes[!grepl("JH|GL|MT", names(sizes))]

# Generate Chr ranges
chr.ranges <- GRanges(seqnames = names(sizes), 
                      ranges = IRanges(start = rep(1, length(sizes)),
                                       end = sizes))

# Count reads per chromosome
out.K9 <- regionCounts(bam.files.K9, regions = chr.ranges, param=param)

# Calculate RPKM
cur_data <- assays(out.K9)$counts
rownames(cur_data) <- names(sizes)
rpm <- t(t(cur_data)/(colSums(cur_data)/1000000))
rpkm <- rpm/(sizes/1000)
colnames(rpkm) <- paste(treatment, batch, sep = "_")

cur_data.melt <- melt(rpkm)
cur_data.melt$cell_type <- sapply(as.character(cur_data.melt$Var2), 
                                  function(n){unlist(strsplit(n, "_"))[1]})
cur_data.melt$replicate <- sapply(as.character(cur_data.melt$Var2), 
                                  function(n){unlist(strsplit(n, "_"))[2]})
levels(cur_data.melt$Var1) <- c(as.character(1:19), "X", "Y")

K9.signal <- ggplot(cur_data.melt) + geom_point(aes(Var1, value, 
                                                    colour = cell_type,
                                                    shape = replicate), size = 2) + 
  ggtitle("H3K9me3") + scale_color_aaas(name = "Cell type") + 
  ylab("RPKM") + xlab("")


ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/K9signal.pdf", 
       K9.signal, width = 7, height = 4)

# For K4 signal
# Count reads per chromosome
out.K4 <- regionCounts(bam.files.K4, regions = chr.ranges, param=param)

# Calculate RPKM
cur_data <- assays(out.K4)$counts
rownames(cur_data) <- names(sizes)
rpm <- t(t(cur_data)/(colSums(cur_data)/1000000))
rpkm <- rpm/(sizes/1000)
colnames(rpkm) <- paste(treatment, batch, sep = "_")

cur_data.melt <- melt(rpkm)
cur_data.melt$cell_type <- sapply(as.character(cur_data.melt$Var2), 
                                  function(n){unlist(strsplit(n, "_"))[1]})
cur_data.melt$replicate <- sapply(as.character(cur_data.melt$Var2), 
                                  function(n){unlist(strsplit(n, "_"))[2]})
levels(cur_data.melt$Var1) <- c(as.character(1:19), "X", "Y")

K4.signal <- ggplot(cur_data.melt) + geom_point(aes(Var1, value, 
                                                    colour = cell_type,
                                                    shape = replicate), size = 2) + 
  ggtitle("H3K9me3") + scale_color_aaas(name = "Cell type") + 
  ylab("RPKM") + xlab("")

ggsave(filename = "../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/K4signal.pdf", 
       K9.signal, width = 7, height = 4)
```



# X chromosome

# Visualize the normalized H3K9me3 and H3K4me3 normalized counts in promoter regions

```{r}
# Find spermatid specific genes first
# Select X chromosome genes
bulk.X <- bulk[intersect(prom.X$gene_id, rownames(bulk)),]
bulk.X <- bulk.X[rowMeans(bulk.X) > 10,]

# Collect day info on libraries
df <- data.frame(library = meta$Source.Name,
                 day = meta$`Characteristics.[Age]`)
bulk.X <- bulk.X[,df$library]

# Order genes based on their logFC between spermatocytes (< P20)
# and spermatids (> P20)
# Perform differential expression testing
y <- DGEList(counts=bulk.X,
             group=ifelse(df$day < 20, "Spermatocytes", "Spermatids"))
y <- calcNormFactors(y)
design <- model.matrix(~0+y$samples$group)
colnames(design) <- c("Spermatids", "Spermatocytes")
y <- estimateDisp(y,design)
  
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmTreat(fit,coef=2, lfc = 2, 
                contrast = makeContrasts(Spermatids - Spermatocytes, 
                                         levels = design))
cur_markers.X <- topTags(qlf, n = nrow(qlf$table))$table
cur_markers.X <- cur_markers.X[rownames(bulk.X),]

bulk.X <- bulk.X[order(cur_markers.X$logFC, decreasing = FALSE),]
cur_markers.X <- cur_markers.X[rownames(bulk.X),]

# Save info if gene is spermatid specifc or not
spermatid.spec <- ifelse(cur_markers.X$logFC > 2 & cur_markers.X$FDR < 0.1, TRUE, FALSE)
```

Collect H3K4 and H3K9 signal for X chromosome promoters

```{r}
cur_prom <- prom.X[match(rownames(bulk.X), prom.X$gene_id)]
cur_df <- data.frame(row.names = rownames(bulk.X),
                     genenames = prom.X$gene_name[match(rownames(bulk.X), 
                                                        prom.X$gene_id)],
                     spermatid_specific = spermatid.spec,
                     retro = ifelse(rownames(bulk.X) %in% retro$Parental.Gene, 1, 0),
                     Rnf8 = ifelse(prom.X$gene_name[match(rownames(bulk.X), 
                                                        prom.X$gene_id)] %in% 
                                     K27writers.Rnf8$gene_id, 1, 0),
                     Scml2 = ifelse(prom.X$gene_name[match(rownames(bulk.X), 
                                                        prom.X$gene_id)] %in% 
                                     K27writers.Scml2$gene_id, 1, 0))


# K9 signal
olap.K9 <- findOverlaps(cur_prom, H3K9me3.normReads$windows, minoverlap = 500L)
cur_df$K9.spermatocytes <- rep(NA, nrow(cur_df))
cur_df$K9.spermatocytes[unique(queryHits(olap.K9))] <- ddply(data.frame(olap.K9, 
                enrichment.spermatocytes =
                  rowMeans(H3K9me3.normReads$normReads[subjectHits(olap.K9),
                                    c("spermatocytes_1", "spermatocytes_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

cur_df$K9.spermatids <- rep(NA, nrow(cur_df))
cur_df$K9.spermatids[unique(queryHits(olap.K9))] <- ddply(data.frame(olap.K9, 
                enrichment.spermatocytes =
                  rowMeans(H3K9me3.normReads$normReads[subjectHits(olap.K9),
                                    c("spermatids_1", "spermatids_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

# K4 signal
olap.K4 <- findOverlaps(cur_prom, H3K4me3.normReads$windows, minoverlap = 500L)
cur_df$K4.spermatocytes <- rep(NA, nrow(cur_df))
cur_df$K4.spermatocytes[unique(queryHits(olap.K4))] <- ddply(data.frame(olap.K4, 
                enrichment.spermatocytes =
                  rowMeans(H3K4me3.normReads$normReads[subjectHits(olap.K4),
                                    c("spermatocytes_1", "spermatocytes_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

cur_df$K4.spermatids <- rep(NA, nrow(cur_df))
cur_df$K4.spermatids[unique(queryHits(olap.K4))] <- ddply(data.frame(olap.K4, 
                enrichment.spermatocytes =
                  rowMeans(H3K4me3.normReads$normReads[subjectHits(olap.K4),
                                    c("spermatids_1", "spermatids_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

```

Plot heatmaps

```{r, fig.height=30, fig.width=10}
cur_df <- cur_df[!is.na(cur_df$K9.spermatocytes) & !is.na(cur_df$K4.spermatocytes),]
cur_df <- cur_df[order(as.numeric(cur_df$spermatid_specific), cur_df$K9.spermatocytes),]
cur_mat <- bulk.X[rownames(cur_df),]

# Full heatmap
pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/Bulk_heatmap_X.pdf", 
    width = 7, height = 12)
pheatmap(log2(cur_mat + 1), 
         cluster_cols = FALSE, 
         cluster_rows = FALSE, show_rownames = FALSE,
         color = viridis(100),
         border_color = NA,
         labels_col = paste("P", meta$`Characteristics.[Age]`, sep = ""),
         annotation_row = data.frame(row.names = rownames(cur_mat),
        H3K4me3.spermatocytes = cur_df[rownames(cur_mat),"K4.spermatocytes"],
        H3K9me3.spermatocytes = cur_df[rownames(cur_mat),"K9.spermatocytes"],
        retro = cur_df[rownames(cur_mat),"retro"],
        Rnf8 = cur_df[rownames(cur_mat),"Rnf8"],
        Scml2 = cur_df[rownames(cur_mat),"Scml2"]),       
        cellheight = 0.5, cellwidth = 8, fontsize = 7, 
        annotation_colors = list(H3K9me3.spermatocytes = inferno(100)))
dev.off()

# Test for statistical enrichment of K4 and K9
# Collect mean K4 and K9 signal from spermtid specifc and non-specific genes  
ddply(cur_df, .(spermatid_specific), summarize, mean.K9.spermatocytes = mean(K9.spermatocytes, na.rm = TRUE),
      mean.K9.spermatids = mean(K9.spermatids, na.rm = TRUE),
      mean.K4.spermatocytes = mean(K4.spermatocytes, na.rm = TRUE),
      mean.K4.spermatids = mean(K4.spermatids, na.rm = TRUE))

# Fisher test
# Set up contigency table - K9 high > 0.5
K9.spermatocytes.table <- matrix(data = NA, ncol = 2, nrow = 2)
colnames(K9.spermatocytes.table) <- c("specific", "non-specific")
rownames(K9.spermatocytes.table) <- c("K9-high", "K9-low")
K9.spermatocytes.table[1,1] <- sum(cur_df$K9.spermatocytes > 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K9.spermatocytes.table[1,2] <- sum(cur_df$K9.spermatocytes > 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == FALSE) 
K9.spermatocytes.table[2,1] <- sum(cur_df$K9.spermatocytes < 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K9.spermatocytes.table[2,2] <- sum(cur_df$K9.spermatocytes < 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == FALSE)

fisher.test(K9.spermatocytes.table, alternative = "greater")

# Set up contigency table - K4 high > 0.5
K4.spermatocytes.table <- matrix(data = NA, ncol = 2, nrow = 2)
colnames(K4.spermatocytes.table) <- c("specific", "non-specific")
rownames(K4.spermatocytes.table) <- c("K4-high", "K4-low")
K4.spermatocytes.table[1,1] <- sum(cur_df$K4.spermatocytes > 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K4.spermatocytes.table[1,2] <- sum(cur_df$K4.spermatocytes > 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == FALSE) 
K4.spermatocytes.table[2,1] <- sum(cur_df$K4.spermatocytes < 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K4.spermatocytes.table[2,2] <- sum(cur_df$K4.spermatocytes < 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == FALSE)

fisher.test(K4.spermatocytes.table, alternative = "less")

```

# Y chromosome

# Visualize the normalized H3K9me3 and H3K4me3 normalized counts in promoter regions

```{r}
# Find spermatid specific genes first
# Select X chromosome genes
bulk.Y <- bulk[intersect(prom.Y$gene_id, rownames(bulk)),]
bulk.Y <- bulk.Y[rowMeans(bulk.Y) > 10,]

# Collect day info on libraries
df <- data.frame(library = meta$Source.Name,
                 day = meta$`Characteristics.[Age]`)
bulk.Y <- bulk.Y[,df$library]

# Order genes based on their logFC between spermatocytes (< P20)
# and spermatids (> P20)
# Perform differential expression testing
y <- DGEList(counts=bulk.Y,
             group=ifelse(df$day < 20, "Spermatocytes", "Spermatids"))
y <- calcNormFactors(y)
design <- model.matrix(~0+y$samples$group)
colnames(design) <- c("Spermatids", "Spermatocytes")
y <- estimateDisp(y,design)
  
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmTreat(fit,coef=2, lfc = 2, 
                contrast = makeContrasts(Spermatids - Spermatocytes, 
                                         levels = design))
cur_markers <- topTags(qlf, n = nrow(qlf$table))$table
cur_markers <- cur_markers[rownames(bulk.Y),]

bulk.Y <- bulk.Y[order(cur_markers$logFC, decreasing = FALSE),]
cur_markers <- cur_markers[rownames(bulk.Y),]

# Save info if gene is spermatid specifc or not
spermatid.spec <- ifelse(cur_markers$logFC > 2 & cur_markers$FDR < 0.1, TRUE, FALSE)
```

Collect H3K4 and H3K9 signal for these promoters

```{r}
cur_prom <- prom.Y[match(rownames(bulk.Y), prom.Y$gene_id)]
cur_df <- data.frame(row.names = rownames(bulk.Y),
                     genenames = prom.Y$gene_name[match(rownames(bulk.Y), 
                                                        prom.Y$gene_id)],
                     spermatid_specific = spermatid.spec)

# K9 signal
olap.K9 <- findOverlaps(cur_prom, H3K9me3.normReads$windows, minoverlap = 500L)
cur_df$K9.spermatocytes <- rep(NA, nrow(cur_df))
cur_df$K9.spermatocytes[unique(queryHits(olap.K9))] <- ddply(data.frame(olap.K9, 
                enrichment.spermatocytes =
                  rowMeans(H3K9me3.normReads$normReads[subjectHits(olap.K9),
                                    c("spermatocytes_1", "spermatocytes_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

cur_df$K9.spermatids <- rep(NA, nrow(cur_df))
cur_df$K9.spermatids[unique(queryHits(olap.K9))] <- ddply(data.frame(olap.K9, 
                enrichment.spermatocytes =
                  rowMeans(H3K9me3.normReads$normReads[subjectHits(olap.K9),
                                    c("spermatids_1", "spermatids_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

# K4 signal
olap.K4 <- findOverlaps(cur_prom, H3K4me3.normReads$windows, minoverlap = 500L)
cur_df$K4.spermatocytes <- rep(NA, nrow(cur_df))
cur_df$K4.spermatocytes[unique(queryHits(olap.K4))] <- ddply(data.frame(olap.K4, 
                enrichment.spermatocytes =
                  rowMeans(H3K4me3.normReads$normReads[subjectHits(olap.K4),
                                    c("spermatocytes_1", "spermatocytes_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

cur_df$K4.spermatids <- rep(NA, nrow(cur_df))
cur_df$K4.spermatids[unique(queryHits(olap.K4))] <- ddply(data.frame(olap.K4, 
                enrichment.spermatocytes =
                  rowMeans(H3K4me3.normReads$normReads[subjectHits(olap.K4),
                                    c("spermatids_1", "spermatids_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

```

Plot heatmaps

```{r, fig.height=30, fig.width=10}
cur_df <- cur_df[!is.na(cur_df$K9.spermatocytes) & !is.na(cur_df$K4.spermatocytes),]
cur_df <- cur_df[order(as.numeric(cur_df$spermatid_specific), cur_df$K9.spermatocytes),]
cur_mat <- bulk.Y[rownames(cur_df),]

# Full heatmap
pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/Bulk_heatmap_Y.pdf", 
    width = 7, height = 6)
pheatmap(log2(cur_mat + 1), 
         cluster_cols = FALSE, 
         cluster_rows = FALSE, labels_row = cur_df$genenames,
         color = viridis(100),
         border_color = NA,
         labels_col = paste("P", meta$`Characteristics.[Age]`, sep = ""),
         annotation_row = data.frame(row.names = rownames(cur_mat),
        H3K4me3.spermatocytes = cur_df[rownames(cur_mat),"K4.spermatocytes"],
        H3K9me3.spermatocytes = cur_df[rownames(cur_mat),"K9.spermatocytes"]),       
        cellheight = 8, cellwidth = 8, fontsize = 7, 
        annotation_colors = list(H3K9me3.spermatocytes = inferno(100)))
dev.off()

# Statistical testing
K9.spermatocytes.table <- matrix(data = NA, ncol = 2, nrow = 2)
colnames(K9.spermatocytes.table) <- c("specific", "non-specific")
rownames(K9.spermatocytes.table) <- c("K9-high", "K9-low")
K9.spermatocytes.table[1,1] <- sum(cur_df$K9.spermatocytes > 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K9.spermatocytes.table[1,2] <- sum(cur_df$K9.spermatocytes > 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == FALSE) 
K9.spermatocytes.table[2,1] <- sum(cur_df$K9.spermatocytes < 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K9.spermatocytes.table[2,2] <- sum(cur_df$K9.spermatocytes < 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == FALSE)

fisher.test(K9.spermatocytes.table, alternative = "greater")

# Set up contigency table - K4 high > 0.5
K4.spermatocytes.table <- matrix(data = NA, ncol = 2, nrow = 2)
colnames(K4.spermatocytes.table) <- c("specific", "non-specific")
rownames(K4.spermatocytes.table) <- c("K4-high", "K4-low")
K4.spermatocytes.table[1,1] <- sum(cur_df$K4.spermatocytes > 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K4.spermatocytes.table[1,2] <- sum(cur_df$K4.spermatocytes > 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == FALSE) 
K4.spermatocytes.table[2,1] <- sum(cur_df$K4.spermatocytes < 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K4.spermatocytes.table[2,2] <- sum(cur_df$K4.spermatocytes < 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == FALSE)

fisher.test(K4.spermatocytes.table, alternative = "less")
```

# Chromosome 9

# Visualize the H3K9me3 and H3K4me3 normalized counts in promoter regions

```{r}
# Find spermatid specific genes first
# Select X chromosome genes
bulk.9 <- bulk[intersect(prom.9$gene_id, rownames(bulk)),]
bulk.9 <- bulk.9[rowMeans(bulk.9) > 10,]

# Collect day info on libraries
df <- data.frame(library = meta$Source.Name,
                 day = meta$`Characteristics.[Age]`)
bulk.9 <- bulk.9[,df$library]

# Order genes based on their logFC between spermatocytes (< P20)
# and spermatids (> P20)
# Perform differential expression testing
y <- DGEList(counts=bulk.9,
             group=ifelse(df$day < 20, "Spermatocytes", "Spermatids"))
y <- calcNormFactors(y)
design <- model.matrix(~0+y$samples$group)
colnames(design) <- c("Spermatids", "Spermatocytes")
y <- estimateDisp(y,design)
  
fit <- glmQLFit(y,design, robust = TRUE)
qlf <- glmTreat(fit,coef=2, lfc = 2, 
                contrast = makeContrasts(Spermatids - Spermatocytes, 
                                         levels = design))
cur_markers <- topTags(qlf, n = nrow(qlf$table))$table
cur_markers <- cur_markers[rownames(bulk.9),]

bulk.9 <- bulk.9[order(cur_markers$logFC, decreasing = FALSE),]
cur_markers <- cur_markers[rownames(bulk.9),]

# Save info if gene is spermatid specifc or not
spermatid.spec <- ifelse(cur_markers$logFC > 2 & cur_markers$FDR < 0.1, TRUE, FALSE)
```

Collect H3K4 and H3K9 signal for these promoters

```{r}
cur_prom <- prom.9[match(rownames(bulk.9), prom.9$gene_id)]
cur_df <- data.frame(row.names = rownames(bulk.9),
                     genenames = prom.9$gene_name[match(rownames(bulk.9), 
                                                        prom.9$gene_id)],
                     spermatid_specific = spermatid.spec)


# K9 signal
olap.K9 <- findOverlaps(cur_prom, H3K9me3.normReads$windows, minoverlap = 500L)
cur_df$K9.spermatocytes <- rep(NA, nrow(cur_df))
cur_df$K9.spermatocytes[unique(queryHits(olap.K9))] <- ddply(data.frame(olap.K9, 
                enrichment.spermatocytes =
                  rowMeans(H3K9me3.normReads$normReads[subjectHits(olap.K9),
                                    c("spermatocytes_1", "spermatocytes_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

cur_df$K9.spermatids <- rep(NA, nrow(cur_df))
cur_df$K9.spermatids[unique(queryHits(olap.K9))] <- ddply(data.frame(olap.K9, 
                enrichment.spermatocytes =
                  rowMeans(H3K9me3.normReads$normReads[subjectHits(olap.K9),
                                    c("spermatids_1", "spermatids_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

# K4 signal
olap.K4 <- findOverlaps(cur_prom, H3K4me3.normReads$windows, minoverlap = 500L)
cur_df$K4.spermatocytes <- rep(NA, nrow(cur_df))
cur_df$K4.spermatocytes[unique(queryHits(olap.K4))] <- ddply(data.frame(olap.K4, 
                enrichment.spermatocytes =
                  rowMeans(H3K4me3.normReads$normReads[subjectHits(olap.K4),
                                    c("spermatocytes_1", "spermatocytes_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

cur_df$K4.spermatids <- rep(NA, nrow(cur_df))
cur_df$K4.spermatids[unique(queryHits(olap.K4))] <- ddply(data.frame(olap.K4, 
                enrichment.spermatocytes =
                  rowMeans(H3K4me3.normReads$normReads[subjectHits(olap.K4),
                                    c("spermatids_1", "spermatids_2")]
                )), .(queryHits), summarize, 
                mean = mean(enrichment.spermatocytes))$mean

# Statistical testing
K9.spermatocytes.table <- matrix(data = NA, ncol = 2, nrow = 2)
colnames(K9.spermatocytes.table) <- c("specific", "non-specific")
rownames(K9.spermatocytes.table) <- c("K9-high", "K9-low")
K9.spermatocytes.table[1,1] <- sum(cur_df$K9.spermatocytes > 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K9.spermatocytes.table[1,2] <- sum(cur_df$K9.spermatocytes > 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == FALSE) 
K9.spermatocytes.table[2,1] <- sum(cur_df$K9.spermatocytes < 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K9.spermatocytes.table[2,2] <- sum(cur_df$K9.spermatocytes < 0.5 & !is.na(cur_df$K9.spermatocytes) & cur_df$spermatid_specific == FALSE)

fisher.test(K9.spermatocytes.table, alternative = "greater")

# Set up contigency table - K4 high > 0.5
K4.spermatocytes.table <- matrix(data = NA, ncol = 2, nrow = 2)
colnames(K4.spermatocytes.table) <- c("specific", "non-specific")
rownames(K4.spermatocytes.table) <- c("K4-high", "K4-low")
K4.spermatocytes.table[1,1] <- sum(cur_df$K4.spermatocytes > 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K4.spermatocytes.table[1,2] <- sum(cur_df$K4.spermatocytes > 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == FALSE) 
K4.spermatocytes.table[2,1] <- sum(cur_df$K4.spermatocytes < 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == TRUE) 
K4.spermatocytes.table[2,2] <- sum(cur_df$K4.spermatocytes < 0.5 & !is.na(cur_df$K4.spermatocytes) & cur_df$spermatid_specific == FALSE)

fisher.test(K4.spermatocytes.table, alternative = "less")
```

# Visualize spermatid spceific genes on X for single cells

```{r}
# Collect expression of spermatic specific genes in heatmap
for.heatmap <- logcounts(sce)[rownames(cur_markers.X)[cur_markers.X$logFC > 2 &
                                                        cur_markers.X$FDR < 0.1],]
for.heatmap <- for.heatmap[Matrix::rowMeans(for.heatmap) > 0.1,]

# Include only annotated genes
for.heatmap <- for.heatmap[!grepl("Rik|-ps|Gm", rowData(sce)$Symbol[match(rownames(for.heatmap), rowData(sce)$ID)]),]

# Build mean expression matrix
df <- as.data.frame(t(as.matrix(for.heatmap)))
df$groups <- colData(sce)$AnnotatedClusters
df.melt <- melt(df, id.vars = "groups")

# Collect mean expression for each gene in each group
mat <- ddply(df.melt, .(groups, variable), summarize, mean=mean(value))
mat.for.heatmap <- dcast(data = mat,formula = variable~groups,fun.aggregate = sum,value.var = "mean")
rownames(mat.for.heatmap) <- mat.for.heatmap$variable
mat.for.heatmap <- mat.for.heatmap[,-1]

pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_7/SC_heatmap_spermatid_specific.pdf",   width = 7, height = 12)
pheatmap(mat.for.heatmap, show_colnames = FALSE, cluster_cols = FALSE, 
         color = colorRampPalette(c("#053061", "#4393c3", "#f7f7f7", "#d6604d", "#67001f"))(100),
         annotation_col = data.frame(row.names = colnames(mat.for.heatmap),
                  cell_type = colnames(mat.for.heatmap)),
         labels_row = rowData(sce)$Symbol[match(rownames(for.heatmap), rowData(sce)$ID)],
         annotation_colors = list(cell_type = metadata(sce)$color_vector), 
         scale = "row", border_color = NA, cellheight = 8, fontsize = 7)
dev.off()

```

