---
title: "Figure 3"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, 
        encoding = encoding, output_file = '/Users/nils/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/Figures/Fig_S3/Fig_3.html') })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Read in the data

```{r data, message=FALSE}
library(scater)
library(ggplot2)
library(reshape2)
library(plyr)
library(openxlsx)
library(cowplot)
library(viridis)
library(pheatmap)
library(Rtsne)
library(irlba)
library(RColorBrewer)
source("../Functions/auxiliary.R")

# Single cell data
sce <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_all.rds")
```

# Visualize spermatogonia from P10 and P15 and RA syncrhonised cells 

```{r}
# Read in only spermatogonia from P10 and P15
sce.spermatogonia <- sce[,grepl("P10|P15", colData(sce)$Sample) &
                                       colData(sce)$AnnotatedClusters == "Spermatogonia"]
sce.spermatogonia <- normalize(sce.spermatogonia)

sce_RA <- readRDS("../../../Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/10X_data/SCE_RA.rds")

# Batch correction
sce.spermatogonia <- sce.spermatogonia[match(intersect(rownames(sce_RA), 
                          rowData(sce.spermatogonia)$Symbol), rowData(sce.spermatogonia)$Symbol),]
sce_RA <- sce_RA[match(intersect(rownames(sce_RA), rowData(sce.spermatogonia)$Symbol), 
                       rownames(sce_RA)),]

rownames(sce_RA) <- rownames(sce.spermatogonia)
rowData(sce_RA)$Symbol <- rowData(sce.spermatogonia)$Symbol
rowData(sce_RA)$ID <- rowData(sce.spermatogonia)$ID

# Select specific cell types 
sce_early_RA <- sce_RA[,grepl("A1|ln|TypeBS|G1|TypeBG2M|ePL|mPL|lPL", colData(sce_RA)$cell_type)]
sce_early_RA <- normalize(sce_early_RA)

sce.single <- split.sce(sce = sce.spermatogonia, groups = unique(colData(sce.spermatogonia)$Library), 
                        colData.name = "Library")
sce.single$RA <- sce_early_RA

corrected <- batch.correction(sce.single)

# Visualize PCA 
pca <- prcomp(t(corrected))

# Plot sample distribution
sample <- ggplot() +
  geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                  PC2 = pca$x[1:ncol(sce.spermatogonia),2],
                  sample = sce.spermatogonia$Sample),
             aes(PC1, PC2, colour = sample)) +
            scale_colour_manual(values = c("dark blue", "coral"))

#Cluster spermatogonia
set.seed(1234)
g <- buildSNNGraph(corrected[,1:ncol(sce.spermatogonia)], k = 15, pc.approx = TRUE)
clusters <- igraph::cluster_louvain(g)$membership

clusters[clusters == 1] <- "Diff_In_B"
clusters[clusters == 2] <- "Diff_A"
clusters[clusters == 3] <- "pL"
clusters[clusters == 4] <- "Undiff_2"
clusters[clusters == 5] <- "Undiff_to_Diff"
clusters[clusters == 6] <- "Undiff_1"
clusters <- factor(clusters,
                   levels = c("Undiff_1", "Undiff_2", "Undiff_to_Diff",
                              "Diff_A", "Diff_In_B", "pL"))

col_vector <- c("Undiff_1" = "#0D3D21",
                "Undiff_2" = "#14713E",
                "Undiff_to_Diff" = "#41AE76",
                "Diff_A" = "#52E8A4",
                "Diff_In_B" = "#A3EFCE",
                "pL" = "#E5DFEA")

colour_vector_RA <- c("A1" = "#67001f",
                      "TypeBS" = "#49006a",
                      "ePL" = "#800026", 
                      "G1" = "#ae017e",        
                      "mPL" = "#fc4e2a",
                      "TypeBG2M" = "#e7298a",
                      "lPL" = "#fed976",
                      "ln" = "#fde0dd")

col_vector <- c(col_vector, colour_vector_RA)


# Plot cluster
cluster <- ggplot() +
  geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                  PC2 = pca$x[1:ncol(sce.spermatogonia),2],
                  cluster = clusters),
             aes(PC1, PC2, fill = cluster), shape = 21) +
            scale_fill_manual(values = c(col_vector))

# Include RA-cells
cluster_RA <- ggplot() +
   geom_point(data = data.frame(PC1 = pca$x[1:ncol(sce.spermatogonia),1],
                                 PC2 = pca$x[1:ncol(sce.spermatogonia),2],
                                 cluster = clusters),
               aes(PC1, PC2, fill = cluster), shape = 21) +
   geom_point(data = data.frame(PC1 = pca$x[(ncol(sce.spermatogonia)+1):(ncol(sce.spermatogonia)+ncol(sce_early_RA)),1],
                                 PC2 = pca$x[(ncol(sce.spermatogonia)+1):(ncol(sce.spermatogonia)+ncol(sce_early_RA)),2],
                                 cell_type = colData(sce_early_RA)$cell_type),
               aes(PC1, PC2), fill = "black", size = 2.5, shape = 21) +
   geom_point(data = data.frame(PC1 = pca$x[(ncol(sce.spermatogonia)+1):(ncol(sce.spermatogonia)+ncol(sce_early_RA)),1],
                                 PC2 = pca$x[(ncol(sce.spermatogonia)+1):(ncol(sce.spermatogonia)+ncol(sce_early_RA)),2],
                                 cell_type = colData(sce_early_RA)$cell_type),
               aes(PC1, PC2, colour = cell_type), shape = 16, size = 1.8) + scale_color_manual(values = c(col_vector, breaks = c(levels(colData(sce_early_RA)$cell_type),
                                                         levels(cluster)))) +
  scale_fill_manual(values = col_vector)

# Plot marker gene expression

# Gfra1 and Id4 for undifferentiated spermatogonia with stem-like features
cur_pca <- pca$x[1:ncol(sce.spermatogonia),]
gene <- "Gfra1"
cur_expr <- logcounts(sce.spermatogonia)[rowData(sce.spermatogonia)$Symbol == gene,] > 0
Gfra1 <- ggplot() +
  geom_point(data = data.frame(PC1 = cur_pca[!cur_expr,1],
                                 PC2 = cur_pca[!cur_expr,2],
                                gene = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == gene,!cur_expr]), 
                 aes(PC1, PC2, colour = gene), size = 0.5) +
  geom_point(data = data.frame(PC1 = cur_pca[cur_expr,1],
                                 PC2 = cur_pca[cur_expr,2],
                                gene = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == gene,cur_expr]), 
                 aes(PC1, PC2, colour = gene), size = 0.5) + scale_colour_viridis()

gene <- "Id4"
cur_expr <- logcounts(sce.spermatogonia)[rowData(sce.spermatogonia)$Symbol == gene,] > 0
Id4 <- ggplot() +
  geom_point(data = data.frame(PC1 = cur_pca[!cur_expr,1],
                                 PC2 = cur_pca[!cur_expr,2],
                                gene = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == gene,!cur_expr]), 
                 aes(PC1, PC2, colour = gene), size = 0.5) +
  geom_point(data = data.frame(PC1 = cur_pca[cur_expr,1],
                                 PC2 = cur_pca[cur_expr,2],
                                gene = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == gene,cur_expr]), 
                 aes(PC1, PC2, colour = gene), size = 0.5) + scale_colour_viridis()

# Nanos 3 and ZBTB16 for undifferentiated spermatogonia
gene <- "Nanos3"
cur_expr <- logcounts(sce.spermatogonia)[rowData(sce.spermatogonia)$Symbol == gene,] > 0
Nanos3 <- ggplot() +
  geom_point(data = data.frame(PC1 = cur_pca[!cur_expr,1],
                                 PC2 = cur_pca[!cur_expr,2],
                                gene = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == gene,!cur_expr]), 
                 aes(PC1, PC2, colour = gene), size = 0.5) +
  geom_point(data = data.frame(PC1 = cur_pca[cur_expr,1],
                                 PC2 = cur_pca[cur_expr,2],
                                gene = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == gene,cur_expr]), 
                 aes(PC1, PC2, colour = gene), size = 0.5) + scale_colour_viridis()

gene <- "Zbtb16"
cur_expr <- logcounts(sce.spermatogonia)[rowData(sce.spermatogonia)$Symbol == gene,] > 0
Zbtb16 <- ggplot() +
  geom_point(data = data.frame(PC1 = cur_pca[!cur_expr,1],
                                 PC2 = cur_pca[!cur_expr,2],
                                gene = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == gene,!cur_expr]), 
                 aes(PC1, PC2, colour = gene), size = 0.5) +
  geom_point(data = data.frame(PC1 = cur_pca[cur_expr,1],
                                 PC2 = cur_pca[cur_expr,2],
                                gene = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == gene,cur_expr]), 
                 aes(PC1, PC2, colour = gene), size = 0.5) + scale_colour_viridis()

# Stra8 as a marker for differentiating spemratogonia
gene <- "Stra8"
cur_expr <- logcounts(sce.spermatogonia)[rowData(sce.spermatogonia)$Symbol == gene,] > 0
Stra8 <- ggplot() +
  geom_point(data = data.frame(PC1 = cur_pca[!cur_expr,1],
                                 PC2 = cur_pca[!cur_expr,2],
                                gene = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == gene,!cur_expr]), 
                 aes(PC1, PC2, colour = gene), size = 0.5) +
  geom_point(data = data.frame(PC1 = cur_pca[cur_expr,1],
                                 PC2 = cur_pca[cur_expr,2],
                                gene = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == gene,cur_expr]), 
                 aes(PC1, PC2, colour = gene), size = 0.5) + scale_colour_viridis()

# Dmrtb1 medites mitosis-to-meiosis transition
gene <- "Dmrtb1"
cur_expr <- logcounts(sce.spermatogonia)[rowData(sce.spermatogonia)$Symbol == gene,] > 0
Dmrtb1 <- ggplot() +
  geom_point(data = data.frame(PC1 = cur_pca[!cur_expr,1],
                                 PC2 = cur_pca[!cur_expr,2],
                                gene = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == gene,!cur_expr]), 
                 aes(PC1, PC2, colour = gene), size = 0.5) +
  geom_point(data = data.frame(PC1 = cur_pca[cur_expr,1],
                                 PC2 = cur_pca[cur_expr,2],
                                gene = logcounts(sce.spermatogonia)[
                                    rowData(sce.spermatogonia)$Symbol == gene,cur_expr]), 
                 aes(PC1, PC2, colour = gene), size = 0.5) + scale_colour_viridis()

legend_RA <- plot_grid(get_legend(cluster_RA + theme(legend.position = "right")))

A <- plot_grid(sample + theme(legend.position=c(.8,.3)), cluster + theme(legend.position=c(.8,.2)), cluster_RA + theme(legend.position= "NULL"), legend_RA,  Gfra1 + theme(legend.position=c(.8,.3)), Zbtb16 + theme(legend.position=c(.8,.3)), Stra8 + theme(legend.position=c(.8,.3)), Dmrtb1 + theme(legend.position=c(.8,.3)), ncol = 4, nrow = 2)
```

# Find marker genes for these groups

```{r}
# Making sure to test across all detected genes
sce.spermatogonia <- sce[,grepl("P10|P15", colData(sce)$Sample) &
                                       colData(sce)$AnnotatedClusters == "Spermatogonia"]
sce.spermatogonia <- sce.spermatogonia[rowMeans(counts(sce.spermatogonia)) > 0,]
sce.spermatogonia <- normalize(sce.spermatogonia)

spermatogonia.markers <- marker.detection(sce.spermatogonia, 
                                          as.character(clusters),
                                          blocking = sce.spermatogonia$Sample)

write.xlsx(spermatogonia.markers, "../../../Dropbox (Personal)/Tc1_meiotic_silencing/Revisions/Figures/Supplemental Tables/Table_S4.xlsx")
```

# Heatmap of genes during spermatogonial differentiation

```{r spermatogonia}
# First we order the cells based on their differerentiation trajectory
prank <- PT(rd = pca$x[1:ncol(sce.spermatogonia),1:3], clusters = clusters, col_vector = col_vector)

# Find marker genes to visualize
spermatogonia.markers <- spermatogonia.markers[c("Undiff_1",
                                                 "Undiff_2",
                                                 "Undiff_to_Diff",
                                                 "Diff_A",
                                                 "Diff_In_B",
                                                 "pL")]

markers <- data.frame(symbol = unlist(lapply(spermatogonia.markers, 
                                             function(n){n[1:10,"GeneName"]})),
                      ID = unlist(lapply(spermatogonia.markers, 
                                             function(n){rownames(n)[1:10]})),
                      group = rep(names(spermatogonia.markers), each = 10))

# Collect count for heatmap
for.heatmap <- logcounts(sce.spermatogonia)[match(markers$symbol, 
                                                         rowData(sce.spermatogonia)$Symbol),
                                      order(prank[,"rank"], decreasing = TRUE)]
colnames(for.heatmap) <- paste(colData(sce.spermatogonia)$Library[
  order(prank[,"rank"], decreasing = TRUE)],
  colData(sce.spermatogonia)$Barcode[
  order(prank[,"rank"], decreasing = TRUE)], sep = "_")
rownames(for.heatmap) <- markers$symbol

# Visualize top genes in form of heatmap
pdf("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_3/Spermatogonia_heatmap_P10_P15.pdf", onefile = FALSE, width = 10, height = 15)
pheatmap(for.heatmap, 
         cluster_rows = FALSE, cluster_cols = FALSE,
         scale = "row", 
         show_colnames = FALSE, gaps_row = seq(10,60, 10),
         cellheight = 8, fontsize = 7,
         color = colorRampPalette(c("#053061", "#4393c3", 
                                    "#f7f7f7", "#d6604d", "#67001f"))(100),
         annotation_col = data.frame(row.names = colnames(for.heatmap),
                                     cell_type = clusters[order(prank[,"rank"], 
                                                                decreasing = TRUE)]),
         annotation_row = data.frame(row.names = rownames(for.heatmap),
                                     gene_list = markers$group),
         annotation_colors = list(cell_type = col_vector,
                                  gene_list = col_vector))
dev.off()
```

# Visualize spermatogonia from P5

```{r}
spermatogonia.P5 <- sce[,grepl("P5", colData(sce)$Sample) &
                                       colData(sce)$AnnotatedClusters == "Spermatogonia"]
spermatogonia.P5 <- spermatogonia.P5[rowMeans(counts(spermatogonia.P5)) > 0,]
spermatogonia.P5 <- normalize(spermatogonia.P5)

# Visualize non-batch corrected data
HVgenes <- HVG(sce = spermatogonia.P5)
pca.P5 <- prcomp(t(logcounts(spermatogonia.P5)[HVgenes,]))
ggplot(data.frame(PC1 = pca.P5$x[,1],
                  PC2 = pca.P5$x[,2],
                  libraries = colData(spermatogonia.P5)$Library)) +
  geom_point(aes(PC1, PC2, colour =  libraries)) + 
  scale_color_manual(values = c( "#FF9100", "#e31a1c"))

# We don't need to batch correct the P5 samples since there is no batch effect
set.seed(123)
g <- buildSNNGraph(logcounts(spermatogonia.P5)[HVgenes,], k = 15, 
                   pc.approx = TRUE)
clusters <- igraph::cluster_louvain(g)$membership


# Change cluster labels
clusters[clusters == 1] <- "Prospermatogonia"
clusters[clusters == 2] <- "Undiff_2"
clusters[clusters == 3] <- "Undiff_1"
clusters[clusters == 4] <- "Diff"


col_vector <- c("Prospermatogonia" = "#7A9606",
                "Undiff_1" = "#0D3D21",
                "Undiff_2" = "#076E38",
                "Diff" = "#5FA37D")

p5.cluster <- ggplot(data.frame(PC1 = pca.P5$x[,1],
                  PC2 = pca.P5$x[,2],
                  clusters = as.factor(clusters))) +
            geom_point(aes(PC1, PC2, colour = clusters)) + 
            scale_colour_manual(values = col_vector)

# Plot marker gene expression
gene <- "Gfra1"
cur_expr <- logcounts(spermatogonia.P5)[rowData(spermatogonia.P5)$Symbol == gene,] > 0
Gfra1.p5 <- ggplot() +
  geom_point(data = data.frame(PC1 = pca.P5$x[!cur_expr,1],
                                 PC2 = pca.P5$x[!cur_expr,2],
                                gene = logcounts(spermatogonia.P5)[
                                    rowData(spermatogonia.P5)$Symbol == gene,!cur_expr]), 
                 aes(PC1, PC2, colour = gene)) +
  geom_point(data = data.frame(PC1 = pca.P5$x[cur_expr,1],
                                 PC2 = pca.P5$x[cur_expr,2],
                                gene = logcounts(spermatogonia.P5)[
                                    rowData(spermatogonia.P5)$Symbol == gene,cur_expr]), 
                 aes(PC1, PC2, colour = gene)) + scale_colour_viridis()

gene <- "Stra8"
cur_expr <- logcounts(spermatogonia.P5)[rowData(spermatogonia.P5)$Symbol == gene,] > 0
Stra8.p5 <- ggplot() +
  geom_point(data = data.frame(PC1 = pca.P5$x[!cur_expr,1],
                                 PC2 = pca.P5$x[!cur_expr,2],
                                gene = logcounts(spermatogonia.P5)[
                                    rowData(spermatogonia.P5)$Symbol == gene,!cur_expr]), 
                 aes(PC1, PC2, colour = gene)) +
  geom_point(data = data.frame(PC1 = pca.P5$x[cur_expr,1],
                                 PC2 = pca.P5$x[cur_expr,2],
                                gene = logcounts(spermatogonia.P5)[
                                    rowData(spermatogonia.P5)$Symbol == gene,cur_expr]), 
                 aes(PC1, PC2, colour = gene)) + scale_colour_viridis()


gene <- "Eif2s2"
cur_expr <- logcounts(spermatogonia.P5)[rowData(spermatogonia.P5)$Symbol == gene,] > 0
Eif2s2.p5 <- ggplot() +
  geom_point(data = data.frame(PC1 = pca.P5$x[!cur_expr,1],
                                 PC2 = pca.P5$x[!cur_expr,2],
                                gene = logcounts(spermatogonia.P5)[
                                    rowData(spermatogonia.P5)$Symbol == gene,!cur_expr]), 
                 aes(PC1, PC2, colour = gene)) +
  geom_point(data = data.frame(PC1 = pca.P5$x[cur_expr,1],
                                 PC2 = pca.P5$x[cur_expr,2],
                                gene = logcounts(spermatogonia.P5)[
                                    rowData(spermatogonia.P5)$Symbol == gene,cur_expr]), 
                 aes(PC1, PC2, colour = gene)) + scale_colour_viridis()


spermatogonia.markers.P5 <- marker.detection(spermatogonia.P5, 
                                          clusters)
write.xlsx(spermatogonia.markers.P5, "../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Supplemental_tables/Table_S4_P5.xlsx")


legend_B <- plot_grid(get_legend(p5.cluster), get_legend(Gfra1.p5) , get_legend(Stra8.p5), get_legend(Eif2s2.p5), ncol = 4 , nrow = 1)
B_plot <- plot_grid(p5.cluster + theme(legend.position= "NONE"), Gfra1.p5 + theme(legend.position= "NONE"), Stra8.p5 + theme(legend.position= "NONE"), Eif2s2.p5 + theme(legend.position= "NONE"), ncol = 4, nrow = 1)
B <- plot_grid(B_plot, legend_B, ncol = 1, nrow = 2, rel_heights = c(2,1))
final <- plot_grid(A, B, ncol = 1, nrow = 2, rel_heights = c(2,1))

ggsave("../../../Dropbox (Cambridge University)/SST_spermatocytes/Revisions/Results/New_figures/Fig_3/Fig_3.pdf",
       final, width = 15, height = 12.5)
```