---
title: "Bulk_RNAseq_QC_norm"
author: "Nils Eling"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Stranded gene counting

# Load the data

```{r data, message=FALSE}
# Load reads
raw_counts <- read.table("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/bulk/raw/1st_wave_bulk_all.txt", sep = "\t")

# Load library
library(DESeq2)

raw_counts <- raw_counts[,!is.na(raw_counts[1,])]

# Filter reads
plot(log10(colSums(raw_counts)), pch = 16)

plot(as.numeric(raw_counts["__no_feature",]/colSums(raw_counts)), pch = 16)
plot(as.numeric(raw_counts["__ambiguous",]/colSums(raw_counts)), pch = 16)
plot(as.numeric(raw_counts["__too_low_aQual",]/colSums(raw_counts)), pch = 16)
plot(as.numeric(raw_counts["__not_aligned",]/colSums(raw_counts)), pch = 16)
plot(as.numeric(raw_counts["__alignment_not_unique",]/colSums(raw_counts)), pch = 16)

# Percentage of mapped counts
raw_counts <- raw_counts[!grepl("ERCC", rownames(raw_counts)),]
plot(as.numeric(colSums(raw_counts[!grepl("__", rownames(raw_counts)),])/colSums(raw_counts)), pch = 16)

# Total mapped counts
plot(log10(as.numeric(colSums(raw_counts[!grepl("__", rownames(raw_counts)),]))), pch = 16)

raw_counts <- raw_counts[rowSums(raw_counts) > 0,]
raw_counts <- raw_counts[!grepl("__", rownames(raw_counts)),]
```

# Normalize

```{r}
dds <- DESeqDataSetFromMatrix(countData = raw_counts, 
                              colData = data.frame(row.names = colnames(raw_counts),
                                                   names = colnames(raw_counts)),
                              design = ~ names)

dds <- estimateSizeFactors(dds)
norm <- counts(dds, normalized = TRUE)

saveRDS(norm, "norm/1st_wave_bulk_norm.rds")
```

# Non-stranded gene counting

# Load the data

```{r}
# Load reads
raw_counts <- read.table("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/bulk/raw/1st_wave_bulk_all_non-stranded.txt", sep = "\t")

# Load library
library(DESeq2)

raw_counts <- raw_counts[,!is.na(raw_counts[1,])]

# Filter reads
plot(log10(colSums(raw_counts)), pch = 16)

plot(as.numeric(raw_counts["__no_feature",]/colSums(raw_counts)), pch = 16)
plot(as.numeric(raw_counts["__ambiguous",]/colSums(raw_counts)), pch = 16)
plot(as.numeric(raw_counts["__too_low_aQual",]/colSums(raw_counts)), pch = 16)
plot(as.numeric(raw_counts["__not_aligned",]/colSums(raw_counts)), pch = 16)
plot(as.numeric(raw_counts["__alignment_not_unique",]/colSums(raw_counts)), pch = 16)

# Percentage of mapped counts
raw_counts <- raw_counts[!grepl("ERCC", rownames(raw_counts)),]
plot(as.numeric(colSums(raw_counts[!grepl("__", rownames(raw_counts)),])/colSums(raw_counts)), pch = 16)

# Total mapped counts
plot(log10(as.numeric(colSums(raw_counts[!grepl("__", rownames(raw_counts)),]))), pch = 16)

raw_counts <- raw_counts[rowSums(raw_counts) > 0,]
raw_counts <- raw_counts[!grepl("__", rownames(raw_counts)),]
```

# Normalize

```{r}
dds <- DESeqDataSetFromMatrix(countData = raw_counts, 
                              colData = data.frame(row.names = colnames(raw_counts),
                                                   names = colnames(raw_counts)),
                              design = ~ names)

dds <- estimateSizeFactors(dds)
norm <- counts(dds, normalized = TRUE)

saveRDS(norm, "norm/1st_wave_bulk_norm_non-stranded.rds")
```

# Reverse-stranded gene counting

# Load the data

```{r}
# Load reads
raw_counts <- read.table("/Users/eling01/Dropbox (Cambridge University)/SST_spermatocytes/Analysis/data/bulk/raw/1st_wave_bulk_all_non-stranded.txt", sep = "\t")

# Load library
library(DESeq2)

raw_counts <- raw_counts[,!is.na(raw_counts[1,])]

# Filter reads
plot(log10(colSums(raw_counts)), pch = 16)

plot(as.numeric(raw_counts["__no_feature",]/colSums(raw_counts)), pch = 16)
plot(as.numeric(raw_counts["__ambiguous",]/colSums(raw_counts)), pch = 16)
plot(as.numeric(raw_counts["__too_low_aQual",]/colSums(raw_counts)), pch = 16)
plot(as.numeric(raw_counts["__not_aligned",]/colSums(raw_counts)), pch = 16)
plot(as.numeric(raw_counts["__alignment_not_unique",]/colSums(raw_counts)), pch = 16)

# Percentage of mapped counts
raw_counts <- raw_counts[!grepl("ERCC", rownames(raw_counts)),]
plot(as.numeric(colSums(raw_counts[!grepl("__", rownames(raw_counts)),])/colSums(raw_counts)), pch = 16)

# Total mapped counts
plot(log10(as.numeric(colSums(raw_counts[!grepl("__", rownames(raw_counts)),]))), pch = 16)

raw_counts <- raw_counts[rowSums(raw_counts) > 0,]
raw_counts <- raw_counts[!grepl("__", rownames(raw_counts)),]
```

# Normalize

```{r}
dds <- DESeqDataSetFromMatrix(countData = raw_counts, 
                              colData = data.frame(row.names = colnames(raw_counts),
                                                   names = colnames(raw_counts)),
                              design = ~ names)

dds <- estimateSizeFactors(dds)
norm <- counts(dds, normalized = TRUE)

saveRDS(norm, "norm/1st_wave_bulk_norm_non-stranded.rds")
```